---
title: "Duplex Sequencing Summary Report"
author: "Matthew J. Meier"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
    theme: spacelab
    code_download: yes
  pdf_document:
    toc: yes
params:
  group_var: "sample"
  species: "mouse"
  project_title: "Mouse Test"
  genome: "mm10"
  mut_data: "../../from-annette/mutfiles/"
  regions_file: "../../inst/extdata/genic_regions_mm10.txt"
---

```{r load_libraries, warning=F, echo=F}
set.seed(3135)

library(DupSeqR)
library(dplyr)
library(ggplot2)
theme_set(theme_bw())
output_dir <- file.path(here::here(),"output")
if (!dir.exists(output_dir)) { dir.create(output_dir)}
```

# Import Data

```{r import-data}
#group_var <- "tissue"
#dupseq_species <- "human"
#project <- "Axelson_2022"
#project_genome <- "GRCh38"
#mut_data <- ../../data/Jonatan_Mutations_in_blood_and_sperm_samples_221021_MM.txt
#regions_file <- "../../inst/extdata/genic_regions_hg38.txt"

# group_var <- "sample"
# dupseq_species <- "mouse"
# project <- "Mouse Test"
# project_genome = "mm10"
# mut_data <- "../../data/BaP_BM_C4.1.consensus.variant-calls.genome.mut"
# regions_file <- "../../inst/extdata/genic_regions_mm10.txt"

group_var <- "sample"
dupseq_species <- "mouse"
project <- "Mouse Test"
project_genome <- "mm10"
mut_data <- "../../from-annette/mutfiles/"
regions_file <- "../../inst/extdata/genic_regions_mm10.txt"
 
mutation_data <-
  import_mut_data(mut_file = mut_data,
                  regions_file = regions_file,
                  rsids = T)

six_base_summary <- calculate_mut_freq(data = as.data.frame(mutation_data),
                                       subtype_resolution = "6base",
                                       summary = T)

# six_base_summary_snps <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                        subtype_resolution = "6base",
#                                        summary = T,
#                                        variant_types = "snv")
# 
# six_base_full <- calculate_mut_freq(data = as.data.frame(mutation_data), subtype_resolution = "6base", summary = F)
# six_base_summary <- calculate_mut_freq(data = as.data.frame(mutation_data), subtype_resolution = "6base", summary = T,vaf_cutoff = 0.01)
# 
# six_base_w_description <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                        subtype_resolution = "6base",
#                                        summary = T,
#                                        cols_to_group = c("sample","description"))

# summary <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                        subtype_resolution = "none",
#                                        summary = T,
#                                        cols_to_group = c("sample"))
# 
# var_types <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                 subtype_resolution = "6base",
#                                 summary = T,
#                                 variant_types = c("snv","indel"),
#                                 cols_to_group = c("sample"))
# 
summary <- calculate_mut_freq(data = as.data.frame(mutation_data),
                                       subtype_resolution = "96base",
                                       summary = T,
                                       cols_to_group = c("sample"))

```

```{r write_ref_and_muts, eval = F}
# Write a reference FASTA file...
# write_reference_fasta(ref_ranges = get_region_seqs("human"),
#                       fasta_out = "hg38_duplex_sequencing_targets.fasta")

write_VCF_from_mut(mutation_data)
```

```{r exploratory-spectra, warning = F}

# Look at overall spectrum
as.data.frame(mutation_data) |> group_by(subtype) |> tally()
as.data.frame(mutation_data) |> group_by(normalized_subtype) |> tally()
as.data.frame(mutation_data) |> group_by(ref) |> tally()
as.data.frame(mutation_data) |> group_by(short_ref) |> tally()
as.data.frame(mutation_data) |> group_by(normalized_ref) |> tally()
# generate frequency tables
trinucleotide_frequencies_global <- mutation_data |>
   group_by(normalized_context, normalized_subtype) |>
   summarise(depth_per_trinucleotide = sum(total_depth))
 
 trinucleotide_frequencies_group <- mutation_data |>
   group_by(normalized_context, normalized_subtype, !!sym(group_var)) |>
   summarise(depth_per_trinucleotide = sum(total_depth))
 
 trinucleotide_frequencies_per_individual <- mutation_data |>
   group_by(normalized_context, normalized_subtype, sample) |>
   summarise(depth_per_trinucleotide = sum(total_depth))

ggplot(six_base_summary, aes(x=normalized_subtype, y=sample_MF_unique)) +
  geom_col() +
  facet_grid(sample ~ .)

tvti_plot(mutations = mutation_data, y = NULL)
tvti_plot(mutations = mutation_data)


tvti_plot(mutations = mutation_data, type = "Frequency")
tvti_plot(mutations = mutation_data, type = "Proportion")
```

## CpG Mutations

```{r cpg-mutations}
# Create a GRanges object, containing all mutation data, but a column added
# to indicate whether the 'context' column contains a CpG site (i.e., is the
# mutation at a CpG site or not)
mutation_data_cpgs <- annotate_CpG_sites(mutation_data)
cpg_mutations <- get_CpG_mutations(get_region_seqs(dupseq_species), mutation_data)
# This only returns CpG sites, nothing to do with mutation data
cpg_sites <- get_CpG_regions(get_region_seqs(dupseq_species))
# Total number of CpG sites
nrow(as.data.frame(cpg_sites)) # could also use |> count() |> pull()
n_samples <- length(unique(as.data.frame(mutation_data)[["sample"]]))

# Testing general function for this application...
cpg_summary_subtype <- calculate_mut_freq(data = as.data.frame(cpg_mutations),
                                          subtype_resolution = "6base",
                                          summary = T,
                                          cols_to_group = c("sample"))
                                       

# Mutations at CpG sites...
cpg_summary_subtype <- as.data.frame(cpg_mutations) |>
  add_count(!!ensym(group_var), alt_depth, name = "total_muts_per_group") |>
  ungroup() |>
  group_by(subtype, !!ensym(group_var), total_muts_per_group) |>
  tally() |>
  mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) |>
  mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))

# Mutations at CpG sites, broken down by trinucleotide mutation context
cpg_summary_context <- as.data.frame(cpg_mutations) |>
  group_by(!!ensym(group_var)) |>
  add_count(alt_depth, name = "total_muts_per_group") |>
  ungroup() |>
  group_by(context_with_mutation, !!ensym(group_var), total_muts_per_group) |>
  tally() |>
  mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) |>
  mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))

cpg_12base_pct <- as.data.frame(mutation_data_cpgs) |>
  group_by(CpG_site, subtype, !!ensym(group_var)) |>
  tally() |> 
  ungroup() |>
  group_by(subtype, !!ensym(group_var)) |>
  mutate(total_mutations = sum(n)) |>
  ungroup() |>
  mutate(percent = 100*n/total_mutations) |>
  filter(CpG_site == T) |>
  dplyr::select(-1)

cpg_6base_pct <- as.data.frame(mutation_data_cpgs) |>
  group_by(CpG_site, normalized_subtype, !!ensym(group_var)) |>
  tally() |> 
  ungroup() |>
  group_by(normalized_subtype, !!ensym(group_var)) |>
  mutate(total_mutations = sum(n)) |>
  ungroup() |>
  mutate(percent = 100*n/total_mutations) |>
  filter(CpG_site == T) |>
  dplyr::select(-1)

cpg_96base_pct <- as.data.frame(mutation_data_cpgs) |>
  group_by(CpG_site, context_with_mutation, !!ensym(group_var)) |>
  tally() |> 
  ungroup() |>
  group_by(context_with_mutation, !!ensym(group_var)) |>
  mutate(total_mutations = sum(n)) |>
  ungroup() |>
  mutate(percent = 100*n/total_mutations) |>
  filter(CpG_site == T) |>
  dplyr::select(-1)

cpg_tables <- list(CpGs_12base_percentage = cpg_12base_pct,
                   CpGs_6base_percentage = cpg_6base_pct,
                   #CpGs_trinucleotide_percentage = cpg_96base_pct, # Data too sparse.
                   CpGs_by_subtype = cpg_summary_subtype,
                   CpGs_by_trinucleotide = cpg_summary_context)

write_excel_from_list(list_of_tables = cpg_tables,
                      output_path = output_dir,
                      workbook_name = "cpg_tables")

```

## Signature Decomposition

```{r cosmic-signatures-sigminer}
# Aggregate by sample
sigminer_results <- signature_analysis_sigminer(project_name = project, group = "sample",
                        project_genome = "BSgenome.Mmusculus.UCSC.mm10")

show_sig_profile(sigminer_results$fitting_results,
                 mode = "SBS", style = "cosmic", x_label_angle = 90)


show_sig_exposure(sigminer_results$fitting_results, style = "cosmic")

show_sig_bootstrap_exposure(sigminer_results$bootstrap_results)
show_sig_bootstrap_error(sigminer_results$bootstrap_results)
show_sig_bootstrap_stability(sigminer_results$bootstrap_results)

show_catalogue(t(sigminer_results$tally_results$nmf_matrix), style = "cosmic", x_label_angle = 90)

```
```{r display_sigs}
# Use sigminer to show signatures of interest for comparison
# Look at arbitrary indices... could program this to show most similar ones to data
#show_cosmic_sig_profile(sig_index = c(4, 5, 6), style = "cosmic", sig_db = "ID")
#show_cosmic_sig_profile(sig_index = c(1, 2, 3), style = "cosmic", sig_db = "DBS")
#show_cosmic_sig_profile(sig_index = c(1, 5, 6), style = "cosmic")

```

## Recurring mutations

Are any mutations observed in more than one individual?

```{r}

recurring_mutations <- mutation_data_cpgs |>
  group_by(seqnames,start,end) |>
  filter(!variation_type %in% "no_variant") |>
  tally() |>
  arrange(-n)

recurring_mutations_group <- mutation_data_cpgs |>
  group_by(seqnames,start,end,sample) |>
  tally() |>
  arrange(-n)

recurring_mutations_rsid <- mutation_data_cpgs |>
  group_by(id) |>
  tally() |>
  arrange(-n)
```

## Write data

```{r write-output-tables}

# All mutation data, cleaned and processed
write_excel_single_table(mut_data = mutation_data_cpgs,
                      output_path = output_dir,
                      workbook_name = paste0(project,"_mutation_data"))

```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
