---
title: "MutSeqR Summary Report"
author: "Matthew J. Meier & Annette E. Dodge"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
    theme: spacelab
    code_download: yes
  pdf_document:
    toc: yes
---

```{r load_libraries, warning=F, echo=F}
set.seed(3135)
library(MutSeqR)
library(here)
library(knitr)
```

```{r setup}
has_description <- !is.null(params$project_description)
description_text <- paste("Purpose of the report:", params$project_description)
```

# `r params$project_title` Summary Report {-}
Date report generated: `r format(Sys.time(), '%d %B, %Y')`

Report prepared by: `r params$user_name`

Report prepared for: `r params$researcher_name`

`r if(has_description) {description_text}`

Report saved to : `r params$outputdir`

***
## Import the Mutation Data

```{r import-data, echo=TRUE, warning=TRUE, message=FALSE}

no_depth_warning <- "Could not find an appropriate depth column. Some package functionality may be limited.\n"
correct_depth_warning <-  "The total_depth may be double-counted in some instances due to overlapping positions. Use the filter_mut() function to correct the total_depth for these instances."
warnings_list <- character(0)
mutation_data <- NULL

if (params$file_type == "table") {
  mutation_data <- withCallingHandlers(
    import_mut_data(mut_file = params$mutation_file,
                    mut_sep = "\t",
                    is_0_based_mut = params$is_0_based_mut,
                    sample_data = params$sample_data,
                    sd_sep = "\t",
                    regions = params$regions,
                    rg_sep = "\t",
                    is_0_based_rg = params$is_0_based_rg,
                    species = params$species,
                    genome = params$genome),
    warning = function(w) {
      warnings_list <<- c(warnings_list, w$message)
    }
  )
  no_depth <- any(grepl(no_depth_warning,
                        warnings_list,
                        fixed = TRUE))
  correct_depth <- any(grepl(correct_depth_warning,
                             warnings_list,
                             fixed = TRUE))
} else if (params$mut_file_type == "vcf") {
  mutation_data <- withCallingHandlers(
    import_vcf_data(vcf_file = params$mutation_file,
                    sample_data = params$sample_data,
                    sd_sep = "\t",
                    regions = params$regions,
                    rg_sep = "\t",
                    is_0_based_rg = params$is_0_based_rg,
                    species = params$species,
                    genome = params$genome),
    warning = function(w) {
      warnings_list <<- c(warnings_list, w$message)
    }
  )
  no_depth <- any(grepl(no_depth_warning, warnings_list, fixed = TRUE))
  correct_depth <- any(grepl(correct_depth_warning,
                             warnings_list,
                             fixed = TRUE))
}


```

# Filter Mutation Data
Flagging variants with a VAF > than `r params$vaf_cutoff` as germline.

Flagging SNVs that overlap with germline MNVs: `r params$snv_in_germ_mnv`

Removing variants with an abnormal VAF: `r params$rm_abnormal_vaf`

Using a custom filter: `r !is.null(params$custom_filter_col)`

```{r echo=FALSE, eval=!is.null(params$custom_filter_col), results='asis'}
cat("Applying Custom Filter to column: '", params$custom_filter_col, "'. ",
    "Records are ", ifelse(params$custom_filter_rm, " removed", "flagged"),
    " when the value matches '", params$custom_filter_val, "'", sep = "")
``` 

Removing records based on a regions file: `r !is.null(params$filtering_regions)`

```{r echo=FALSE, eval= !is.null(params$filtering_regions), results='asis'}
cat("Filtering records using regions file: ", params$filtering_regions, ".",
    " Records within the specified filtering regions are",
    ifelse(params$regions_filter == "keep_within",
           " retained in", " removed from"),
    " the dataset. Records must start and end within regions to be included: ",
    !params$allow_half_overlap,
    sep = "")
```

Flagged variants (except germline) have their alt_depth subtracted from their total_depth: `r params$rm_filtered_mut_from_depth`

Flagged variants are not included in the mutation counts, but their total_depth is retained for frequency calculations.
All filtered records are returned in *filtered_variants.xlsx.*

Correcting the total_depth for overlapping positions: `r correct_depth`
```{r filter-mutation-data, echo=TRUE, message=TRUE}
filtered_mutation_data <- filter_mut(mutation_data,
                                     correct_depth = correct_depth,
                                     vaf_cutoff = params$vaf_cutoff,
                                     snv_in_germ_mnv = params$snv_in_germ_mnv,
                                     rm_abnormal_vaf = params$rm_abnormal_vaf,
                                     custom_filter_col = params$custom_filter_col,
                                     custom_filter_val = params$custom_filter_val,
                                     custom_filter_rm = params$custom_filter_rm,
                                     regions = params$filtering_regions,
                                     rg_sep = params$filtering_rg_sep,
                                     regions_filter = params$regions_filter,
                                     allow_half_overlap = params$allow_half_overlap,
                                     is_0_based_rg = params$filtering_is_0_based_rg,
                                     rm_filtered_mut_from_depth = params$rm_filtered_mut_from_depth,
                                     return_filtered_rows = TRUE)
filtered_variants <- filtered_mutation_data$filtered_rows
write_excel(data = filtered_variants,
            output_path = params$outputdir,
            workbook_name = "filtered_variants")
filtered_mutation_data <- filtered_mutation_data$mutation_data
```

# Calculating Per-Sample Mutation Frequency
```{r calculate-global-mf-sample, echo=TRUE, message=FALSE}
mf_data <- calculate_mf(mutation_data = filtered_mutation_data,
                        cols_to_group = "sample",
                        subtype_resolution = "none",
                        variant_types = c("-ambiguous", "-uncategorized"),
                        calculate_depth = !no_depth,
                        precalc_depth_data = params$precalc_depth_data_global,
                        d_sep = params$d_sep,
                        summary = TRUE,
                        retain_metadata_cols = params$exp_variable)
plot_max <- all(mf_data$mf_min != mf_data$mf_max)
# Fix this - i think the levels are wrong!!
mf_data[[params$exp_variable]] <- factor(mf_data[[params$exp_variable]],
                                         levels = params$exp_variable_order)
print(mf_data)
```

```{r plot-mf, echo=TRUE, eval=FALSE}
plot1 <- plot_mf(mf_data = mf_data,
                 group_col = "sample",
                 plot_type = "bar",
                 mf_type = ifelse(plot_max, "both", "min"),
                 group_order = ifelse(is.null(params$exp_variable),
                                      "smart",
                                      "arranged"),
                 group_order_input = params$exp_variable,
                 fill_col = params$exp_variable,
                 labels = "count",
                 x_lab = "Sample",
                 y_lab = "Mutation Frequency (mutations/bp)")
plot1
```
```{r plot-mean-mf, eval = FALSE}
# eval = !is.null(params$exp_variable)
plot2 <- plot_mean_mf(mf_data = mf_data,
                      group_col = params$exp_variable,
                      fill_col = params$exp_variable,
                      mf_type = ifelse(plot_max, "both", "min"),
                      plot_type = "line",
                      plot_error_bars = TRUE,
                      plot_indiv_vals = TRUE,
                      group_order = "none",
                      add_labels = "mean_count")
plot2
```

## CpG Mutations

```{r cpg-mutations, eval=FALSE}
# Create a GRanges object, containing all mutation data, but a column added
# to indicate whether the 'context' column contains a CpG site (i.e., is the
# mutation at a CpG site or not)
mutation_data_cpgs <- annotate_CpG_sites(mutation_data)
cpg_mutations <- get_CpG_mutations(my_regions, mutation_data)
# Returns CpG sites of your regions, it has nothing to do with mutation data
cpg_sites <- get_CpG_regions(my_regions)
# Total number of CpG sites
nrow(as.data.frame(cpg_sites)) # could also use %>% count() %>% pull()
n_samples <- length(unique(as.data.frame(mutation_data)[["sample"]]))

# Testing general function for this application...
cpg_summary_subtype <- calculate_mf(data = as.data.frame(cpg_mutations),
                                          subtype_resolution = "base_6",
                                          vaf_cutoff = params$vaf_cutoff,
                                          summary = T,
                                          cols_to_group = c("sample"))
                                       

# Mutations at CpG sites...
cpg_summary_subtype <- as.data.frame(cpg_mutations) %>%
  add_count(dose, alt_depth, name = "total_muts_per_group") %>%
  ungroup()#%>%
  group_by(subtype, !!ensym(params$group_var), total_muts_per_group) %>%
  tally() %>%
  mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
  mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))

# # Mutations at CpG sites, broken down by trinucleotide mutation context
# cpg_summary_context <- as.data.frame(cpg_mutations) %>%
#   group_by(!!ensym(group_var)) %>%
#   add_count(alt_depth, name = "total_muts_per_group") %>%
#   ungroup() %>%
#   group_by(context_with_mutation, !!ensym(params$group_var), total_muts_per_group) %>%
#   tally() %>%
#   mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
#   mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))
# 
# cpg_base_12_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, subtype, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(subtype, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_base_6_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, normalized_subtype, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(normalized_subtype, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_base_96_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, context_with_mutation, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(context_with_mutation, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_tables <- list(CpGs_base_12_percentage = cpg_base_12_pct,
#                    CpGs_base_6_percentage = cpg_base_6_pct,
#                    #CpGs_trinucleotide_percentage = cpg_base_96_pct, # Data too sparse.
#                    CpGs_by_subtype = cpg_summary_subtype,
#                    CpGs_by_trinucleotide = cpg_summary_context)
# 
# write_excel_from_list(list_of_tables = cpg_tables,
#                       output_path = output_dir,
#                       workbook_name = "cpg_tables")

```

## Recurring mutations

Are any mutations observed in more than one sample?

```{r, eval=FALSE}

recurring_mutations <- mutation_data_cpgs %>%
  group_by(contig,start,end) %>%
  filter(!variation_type %in% "no_variant") %>%
  tally() %>%
  arrange(-n)

recurring_mutations_group <- mutation_data_cpgs %>%
  group_by(contig,start,end,sample) %>%
  tally() %>%
  arrange(-n)

recurring_mutations_rsid <- mutation_data_cpgs %>%
  group_by(id) %>%
  tally() %>%
  arrange(-n)
```


<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
