---
title: "Duplex Sequencing Summary Report"
author: "Matthew J. Meier & Annette E. Dodge"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
    theme: spacelab
    code_download: yes
  pdf_document:
    toc: yes
params:
  mut_file_type:
    label: "file_type"
    input: select
    value: "mut"
    choices: ["mut", "vcf"]
  group_var: "dose"
  vaf_cutoff: 
    label: "VAF cutoff value"
    input: numeric
    step: 0.01
    min: 0
    value: 0.1
  regions: 
    label: "Select a mutagenesis panel"
    input: select
    value: "mouse"
    choices: ["mouse", "human", "rat", "custom"]
  custom_regions_file:
    label: "If you are using a custom panel, please supply a regions file"
    input: file
    value: NULL
    accept: [.txt, .csv, .xlsx]
  is_0_based: 
    label: "The coordinates for my custom panel are 0-based. Disregard if not applicable"
    input: checkbox
    value: TRUE
  species: "mouse"
  genome: "mm10"
  project_title: "Mouse Test"
  mut_data: 
    label: "Enter the file path to your mutation data"
    value: "../../../DupSeq R Package Building/Test Data/mut files"
  sample_data:
    label: "Select your Sample Data File"
    input: file
    accept: [.txt, .csv, .xlsx]
    value: NULL
---

```{r load_libraries, warning=F, echo=F}
set.seed(3135)

library(DupSeqR)
library(stringr)
library(tidyr)
library(dplyr)
library(ggplot2)
theme_set(theme_bw())
output_dir <- file.path(here::here(),"output")
if (!dir.exists(output_dir)) { dir.create(output_dir)}
```

# Import Data

```{r import-data}

if(params$mut_file_type == "mut") {
mutation_data <-
  import_mut_data(mut_file = params$mut_data,
                  sample_data_file = sample_data,
                  regions = params$regions,
                  custom_regions_file = params$custom_regions_file,
                  is_0_based = params$is_0_based,
                  vaf_cutoff = params$vaf_cutoff)
} else if (params$mut_file_type == "vcf") {
  mutation_data <- 
    read_vcf(vcf_file = params$mut_data,
            sample_data_file = params$sample_data,
            regions = params$regions,
            custom_regions_file = params$custom_regions_file,
            is_0_based = params$is_0_based,
            species = params$species,
            genome_version = params$genome,
            vaf_cutoff = params$vaf_cutoff )
}

```

```{r write_ref_and_muts}
# Write a reference FASTA file...
my_regions <- get_seq(regions = params$regions,
                   custom_regions_file = params$custom_regions_file,
                   genome = params$genome,
                   padding = 0) # Grab target sequences
write_reference_fasta(ref_ranges = my_regions) 
# Write mutation data as a vcf
write_VCF_from_mut(mutation_data)
```

## Calculate Mutation Frequencies
```{r calculate-MF-summaries}
MF_per_sample <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = c("sample"),
                                       subtype_resolution = "none",
                                       vaf_cutoff = params$vaf_cutoff,
                                       summary = T,
                                      retain_metadata_cols = params$group_var) 

MF_per_target <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = c("sample", "label"),
                                       subtype_resolution = "none",
                                       vaf_cutoff = params$vaf_cutoff,
                                       summary = T,
                                       retain_metadata_cols = c("genic_context", "dose"))

six_base_summary <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = "sample",
                                       subtype_resolution = "base_6",
                                       vaf_cutoff = params$vaf_cutoff,
                                       summary = T,
                                       retain_metadata_cols = params$group_var)

six_base_summary_group <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = params$group_var,
                                       subtype_resolution = "base_6",
                                       vaf_cutoff = params$vaf_cutoff,
                                       variant_types = "snv",
                                       summary = T)

trinucleotide_frequencies_individual <- calculate_mut_freq(data = mutation_data,
                                                            cols_to_group = "sample",
                                                            subtype_resolution = "base_96",
                                                            vaf_cutoff = params$vaf_cutoff,
                                                            variant_types = "snv",
                                                            summary = T
                                                           )

trinucleotide_frequencies_group <- calculate_mut_freq(data = mutation_data,
                                                            cols_to_group = params$group_var,
                                                            subtype_resolution = "base_96",
                                                            vaf_cutoff = params$vaf_cutoff,
                                                            variant_types = "snv",
                                                            summary = T
                                                           )
```

## Visualize data
```{r plot_MF}
# Unique MF per sample
ggplot(MF_per_sample, aes(x = sample, y = sample_MF_unique, fill = factor(dose))) +
      geom_bar(stat = "identity", alpha = 0.7) +
      geom_text(aes(y = sample_MF_unique, label = sample_sum_unique), vjust = -0.5) +
      labs(title = "Min Mutation Frequency per Sample") +
      ylab("Mutation Frequency (mutations/bp)") +
      xlab("sample") +
      theme(axis.text.x = element_text(angle = 45, hjust = 1))
# Clonal MF per sample
ggplot(MF_per_sample, aes(x = sample, y = sample_MF_clonal, fill = factor(dose))) +
  geom_bar(stat = "identity", alpha = 0.7) +
  geom_text(aes(y = sample_MF_clonal, label = sample_sum_clonal), vjust = -0.5) +
  labs(title = "Max Mutation Frequency per Sample") +
  ylab("Mutation Frequency (mutations/bp)") +
  xlab("sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Average MF by dose for both clonal and unique 
plot_data <- MF_per_sample %>%
  group_by(dose) %>%
  summarize(unique = mean(sample_MF_unique),
            std_error_unique = sd(sample_MF_unique)/sqrt(n()),
            clonal = mean(sample_MF_clonal),
            std_error_clonal = sd(sample_MF_clonal)/sqrt(n())) %>%
  ungroup()
plot_data_long <- plot_data %>%
  pivot_longer(cols = c(unique, clonal), names_to = "MF_type", values_to = "avg_MF") %>%
  mutate(std_error = ifelse(grepl("unique", MF_type), std_error_unique, std_error_clonal)) %>%
  mutate(MF_type = factor(MF_type, levels = c("unique", "clonal")))  # Reorder levels
max_y_limit <- max(plot_data_long$avg_MF + plot_data_long$std_error)*1.25
ggplot(plot_data_long, aes(x = factor(dose), y = avg_MF, fill = MF_type)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.6), width = 0.5) +
  scale_y_continuous(expand=c(0,0), limits = c(0, max_y_limit)) +
 geom_errorbar(aes(ymin = pmax(avg_MF - std_error, 0), ymax = pmin(avg_MF + std_error, max_y_limit)),
                position = position_dodge(width = 0.6), width = 0.25, color = "black") +
  labs(title = "Average Mutation Frequency per Dose",
       y = "Average Mutation Frequency",
       x = "Dose") +
  scale_fill_manual(values = c("unique" = "skyblue", "clonal" = "salmon")) +
  theme_bw()


# MF per target (average per dose)
plot_data <- MF_per_target %>%
  group_by(dose, label) %>%
  summarize(avg_MF_unique = mean(sample_label_MF_unique),
            std_error = sd(sample_label_MF_unique)/sqrt(n()))
ordered_descriptions <- plot_data %>%
  filter(dose == max(plot_data$dose)) %>%
  dplyr::arrange(desc(avg_MF_unique)) %>%
  dplyr::pull(label)
plot_data$label <- factor(plot_data$label, levels = ordered_descriptions)
ggplot(plot_data, aes(x = label, y = avg_MF_unique, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  geom_errorbar(aes(ymin = avg_MF_unique - std_error, ymax = avg_MF_unique + std_error),
                position = position_dodge(width = 0.8), width = 0.25) +
  labs(title = "Average Mutation Frequency per Target and Dose with Standard Error",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Target") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Plot simple spectra averaged across dose
plot_data <- six_base_summary %>%
  group_by(dose, normalized_subtype) %>%
  summarize(avg_MF = mean(sample_MF_unique),
            std_error = sd(sample_MF_unique)/sqrt(n())) %>%
  ungroup()
  # Create subset to view low freq non-snvs
plot_data_subset <- plot_data %>%
  filter(normalized_subtype %in% c("complex", "deletion", "insertion", "mnv", "symbolic"))
# Define the custom order for normalized_subtype
custom_order <- c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G", "complex", "deletion", "insertion", "mnv", "symbolic")
# Convert normalized_subtype to a factor with custom order
plot_data$normalized_subtype <- factor(plot_data$normalized_subtype, levels = custom_order)
# Create the ggplot for all subtypes
ggplot(plot_data, aes(x = normalized_subtype, y = avg_MF, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  geom_errorbar(aes(ymin = avg_MF - std_error, ymax = avg_MF + std_error),
                position = position_dodge(width = 0.8), width = 0.25) +
  labs(title = "Average Mutation Frequency per Variant Type with Standard Error",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Variation Type") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Create subset plot for non-snv
ggplot(plot_data_subset, aes(x = normalized_subtype, y = avg_MF, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  geom_errorbar(aes(ymin = avg_MF - std_error, ymax = avg_MF + std_error),
                position = position_dodge(width = 0.8), width = 0.25) +
  labs(title = "Average Mutation Frequency per Variant Type with Standard Error",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Variation Type") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Plot simple sectra normalized proportions
ggplot(six_base_summary_group, aes(x = normalized_subtype, y = norm_prop_unique, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  labs(title = "Normalized Proportions of SNV Subtypes per Dose",
       y = "Normalized proportion",
       x = "Variation Type") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
# Plot simple sectra MF by group
ggplot(six_base_summary_group, aes(x = normalized_subtype, y = dose_MF_unique, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  labs(title = "Mutation Frequency of SNV Subtypes per Dose",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Variation Type") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))


# Plot Trinucleotide Frequency (grouped)
# Extract mutation type from within normalized_context_with_mutation and remove square brackets
trinucleotide_frequencies_group$mutation_type <- str_remove_all(str_extract(trinucleotide_frequencies_group$normalized_context_with_mutation, "\\[(.*?)\\]"), "\\[|\\]")
# Extract context from within normalized_context_with_mutation
plotdata <- trinucleotide_frequencies_group %>%
                                    dplyr::rowwise() %>%
                                    dplyr::mutate(context = get_ref_of_mut(normalized_context_with_mutation))

# Define color mapping for mutation types
mutation_colors <- c("C>A" = "skyblue", "C>G" = "black", "C>T" = "red", "T>A" = "gray", "T>C" = "green", "T>G" = "pink")

# Reorder the data frame by mutation_type and context
plotdata <- plotdata %>% 
  arrange(mutation_type, context)

# Reverse the order of levels for mutation_type
plotdata$mutation_type <- factor(plotdata$mutation_type, levels = rev(unique(plotdata$mutation_type)))

# Manually specify x-coordinates for labels
label_positions <- data.frame(
  mutation_type = unique(plotdata$mutation_type),
  x_position = c(9, 25, 40, 55, 70, 85) ,
  y_position = max(plotdata$dose_MF_unique) * 1.1  # Adjust the multiplier based on your desired position above the max y-value
)

# Calculate upper limit for y-axis
y_max_limit <- max(plotdata$dose_MF_unique) * 1.5  # Adjust the multiplier based on your desired position above the max y-value

# Create the bar plot with facets arranged vertically, increased x-axis spacing, and color mapping
ggplot(plotdata, aes(x = reorder(normalized_context_with_mutation, -as.numeric(factor(mutation_type))), y = dose_MF_unique, fill = factor(mutation_type))) +
  geom_bar(stat = "identity", position = "dodge") +
  geom_text(data = label_positions, aes(x = x_position, y = y_position, label = mutation_type), vjust = 0, size = 5, color = "black") +  # Add labels over the bars
  scale_fill_manual(values = mutation_colors) +  # Apply color mapping
  facet_wrap(~ dose, scales = "free_y", ncol = 1) +  # Facet by 'dose' with 1 column
  scale_y_continuous(limits = c(0, y_max_limit)) +  # Set limits for the y-axis
  theme_minimal() +
  labs(x = "Normalized Context with Mutation", y = "Dose_MF_Unique") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1, margin = margin(b = 10)))  # Adjust the 'margin' parameter for increased spacing

```

```{r exploratory-spectra, warning=FALSE}
# Look at overall spectrum (sum across all groups)
mutation_data %>% group_by(subtype) %>% tally()
mutation_data %>% group_by(normalized_subtype) %>% tally()
mutation_data %>% group_by(ref) %>% tally()
mutation_data %>% group_by(short_ref) %>% tally()
mutation_data %>% group_by(normalized_ref) %>% tally()



# tvti_plot(mutations = mutation_data, y = NULL)
# tvti_plot(mutations = mutation_data)
# 
# 
# tvti_plot(mutations = mutation_data, type = "Frequency")
# tvti_plot(mutations = mutation_data, type = "Proportion")
```

## CpG Mutations

```{r cpg-mutations}
# Create a GRanges object, containing all mutation data, but a column added
# to indicate whether the 'context' column contains a CpG site (i.e., is the
# mutation at a CpG site or not)
mutation_data_cpgs <- annotate_CpG_sites(mutation_data)
cpg_mutations <- get_CpG_mutations(my_regions, mutation_data)
# Returns CpG sites of your regions, it has nothing to do with mutation data
cpg_sites <- get_CpG_regions(my_regions)
# Total number of CpG sites
nrow(as.data.frame(cpg_sites)) # could also use %>% count() %>% pull()
n_samples <- length(unique(as.data.frame(mutation_data)[["sample"]]))

# Testing general function for this application...
cpg_summary_subtype <- calculate_mut_freq(data = as.data.frame(cpg_mutations),
                                          subtype_resolution = "base_6",
                                          vaf_cutoff = params$vaf_cutoff,
                                          summary = T,
                                          cols_to_group = c("sample"))
                                       

# Mutations at CpG sites...
cpg_summary_subtype <- as.data.frame(cpg_mutations) %>%
  add_count(dose, alt_depth, name = "total_muts_per_group") %>%
  ungroup()#%>%
  group_by(subtype, !!ensym(params$group_var), total_muts_per_group) %>%
  tally() %>%
  mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
  mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))

# # Mutations at CpG sites, broken down by trinucleotide mutation context
# cpg_summary_context <- as.data.frame(cpg_mutations) %>%
#   group_by(!!ensym(group_var)) %>%
#   add_count(alt_depth, name = "total_muts_per_group") %>%
#   ungroup() %>%
#   group_by(context_with_mutation, !!ensym(params$group_var), total_muts_per_group) %>%
#   tally() %>%
#   mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
#   mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))
# 
# cpg_base_12_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, subtype, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(subtype, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_base_6_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, normalized_subtype, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(normalized_subtype, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_base_96_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, context_with_mutation, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(context_with_mutation, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_tables <- list(CpGs_base_12_percentage = cpg_base_12_pct,
#                    CpGs_base_6_percentage = cpg_base_6_pct,
#                    #CpGs_trinucleotide_percentage = cpg_base_96_pct, # Data too sparse.
#                    CpGs_by_subtype = cpg_summary_subtype,
#                    CpGs_by_trinucleotide = cpg_summary_context)
# 
# write_excel_from_list(list_of_tables = cpg_tables,
#                       output_path = output_dir,
#                       workbook_name = "cpg_tables")

```

## Signature Decomposition

```{r cosmic-signatures - SigProfiler, eval=FALSE}
# Need to have a think about how this will work in Rmd - may be problematic. 
# May grab the key outputs to put in the report
signature_fitting(mutations = mutation_data,
                              project_name = "HESI GTTC",
                              project_genome = "rn6",
                              env_name = "DupSeqR",
                              group = "sample",
                              output_path = NULL,
                              python_version = "3.9.13",
                              python_path = "doesn't matter", 
                              python_home = "not implemented")
```


```{r cosmic-signatures-sigminer, eval=FALSE}
# Aggregate by sample

sigminer_results <- signature_analysis_sigminer(
  mutation_data, project_name = params$project, group = "sample",
  project_genome = "BSgenome.Mmusculus.UCSC.mm10",
  vaf_cutoff = params$vaf_cutoff,
  run_bootstrapping = F)



# Can only handle 10 signatures at a time??
sigminer::show_sig_fit(sigminer_results$fitting_results, plot_fun = "violin")

sigminer::show_sig_profile(sigminer_results$fitting_results,
                           mode = "SBS", style = "cosmic", x_label_angle = 90)

sigminer::show_sig_exposure(sigminer_results$fitting_results, style = "cosmic")

sigminer::show_sig_bootstrap_exposure(sigminer_results$bootstrap_results) + facet_grid(sample~.)
sigminer::show_sig_bootstrap_error(sigminer_results$bootstrap_results) + facet_grid(sample~.)
sigminer::show_sig_bootstrap_stability(sigminer_results$bootstrap_results) + facet_grid(sample~.)

sigminer::show_catalogue(t(sigminer_results$fitting_results$nmf_matrix), style = "cosmic", x_label_angle = 90)


```
```{r display_sigs}
# Use sigminer to show signatures of interest for comparison
# Look at arbitrary indices... could program this to show most similar ones to data
#show_cosmic_sig_profile(sig_index = c(4, 5, 6), style = "cosmic", sig_db = "ID")
#show_cosmic_sig_profile(sig_index = c(1, 2, 3), style = "cosmic", sig_db = "DBS")
#show_cosmic_sig_profile(sig_index = c(1, 5, 6), style = "cosmic")

```

## Recurring mutations

Are any mutations observed in more than one individual?

```{r, eval=FALSE}

recurring_mutations <- mutation_data_cpgs %>%
  group_by(contig,start,end) %>%
  filter(!variation_type %in% "no_variant") %>%
  tally() %>%
  arrange(-n)

recurring_mutations_group <- mutation_data_cpgs %>%
  group_by(contig,start,end,sample) %>%
  tally() %>%
  arrange(-n)

recurring_mutations_rsid <- mutation_data_cpgs %>%
  group_by(id) %>%
  tally() %>%
  arrange(-n)
```

## Write data

```{r write-output-tables, eval=FALSE}

# All mutation data, cleaned and processed
write_excel_single_table(mut_data = mutation_data_cpgs,
                      output_path = output_dir,
                      workbook_name = paste0(project,"_mutation_data"))

```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
