---
title: "Duplex Sequencing Summary Report"
author: "Matthew J. Meier & Annette E. Dodge"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
    theme: spacelab
    code_download: yes
  pdf_document:
    toc: yes
params:
  group_var: "dose"
  vaf_cutoff: 0.01
  species: "mouse"
  project_title: "Mouse Test"
  genome: "mm10"
  mut_data: "../Test Data/prj00125_PRC_BM_variant-calls.genome.mut"
  sample_data: "../Test Data/PRC_BM_sample_data.txt"
---

```{r load_libraries, warning=F, echo=F}
set.seed(3135)

library(DupSeqR)
library(tidyr)
library(dplyr)
library(ggplot2)
theme_set(theme_bw())
output_dir <- file.path(here::here(),"output")
if (!dir.exists(output_dir)) { dir.create(output_dir)}
```

# Import Data

```{r import-data}

group_var <- "dose"
species <- "mouse"
# project <- "Mouse Test"
# project_genome <- "mm10"
mut_data <- "../../../Test Data/mut files"
sample_data <- "../../../Test Data/PRC_ST_sample_data.txt"
vaf_cutoff <- 0.01
 
mutation_data <-
  import_mut_data(mut_file = mut_data,
                  sample_data_file = sample_data,
                  regions = species,
                  vaf_cutoff = vaf_cutoff)


MF_per_sample <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = c("sample"),
                                       subtype_resolution = "none",
                                       vaf_cutoff = vaf_cutoff,
                                       summary = T) 

MF_per_target <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = c("sample", "description"),
                                       subtype_resolution = "none",
                                       vaf_cutoff = vaf_cutoff,
                                       summary = T)

six_base_summary <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = "sample",
                                       subtype_resolution = "base_6",
                                       vaf_cutoff = vaf_cutoff,
                                       summary = T)
six_base_summary_group <- calculate_mut_freq(data = mutation_data,
                                       cols_to_group = group_var,
                                       subtype_resolution = "base_6",
                                       vaf_cutoff = vaf_cutoff,
                                       variant_types = "snv",
                                       summary = T)

trinucleotide_frequencies_individual <- calculate_mut_freq(data = mutation_data,
                                                            cols_to_group = "sample",
                                                            subtype_resolution = "base_96",
                                                            vaf_cutoff = vaf_cutoff,
                                                            variant_types = "snv",
                                                            summary = T
                                                           )

trinucleotide_frequencies_group <- calculate_mut_freq(data = mutation_data,
                                                            cols_to_group = group_var,
                                                            subtype_resolution = "base_96",
                                                            vaf_cutoff = vaf_cutoff,
                                                            variant_types = "snv",
                                                            summary = T
                                                           )
# six_base_summary_snps <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                        subtype_resolution = "base_6",
#                                        summary = T,
#                                        variant_types = "snv")
# 
# six_base_full <- calculate_mut_freq(data = as.data.frame(mutation_data), subtype_resolution = "base_6", summary = F)
# six_base_summary <- calculate_mut_freq(data = as.data.frame(mutation_data), subtype_resolution = "base_6", summary = T,vaf_cutoff = 0.01)
# 
# six_base_w_description <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                        subtype_resolution = "base_6",
#                                        summary = T,
#                                        cols_to_group = c("sample","description"))

# summary <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                        subtype_resolution = "none",
#                                        summary = T,
#                                        cols_to_group = c("sample"))
# 
# var_types <- calculate_mut_freq(data = as.data.frame(mutation_data),
#                                 subtype_resolution = "base_6",
#                                 summary = T,
#                                 variant_types = c("snv","indel"),
#                                 cols_to_group = c("sample"))
# 


```

```{r write_ref_and_muts, eval = F}
# Write a reference FASTA file...
# write_reference_fasta(ref_ranges = get_region_seqs("human"),
#                       fasta_out = "hg38_duplex_sequencing_targets.fasta")

write_VCF_from_mut(mutation_data)
```

```{r plot_MF}
# Unique MF per sample
ggplot(MF_per_sample, aes(x = sample, y = sample_MF_unique, fill = dose)) +
  geom_bar(stat = "identity", alpha = 0.7) +
  geom_text(aes(y = sample_MF_unique, label = sample_sum_unique), vjust = -0.5) +
  labs(title = "Min Mutation Frequency per Sample") +
  ylab("Mutation Frequency (mutations/bp)") +
  xlab("sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
# Clonal MF per sample
ggplot(MF_per_sample, aes(x = sample, y = sample_MF_clonal, fill = dose)) +
  geom_bar(stat = "identity", alpha = 0.7) +
  geom_text(aes(y = sample_MF_clonal, label = sample_sum_clonal), vjust = -0.5) +
  labs(title = "Max Mutation Frequency per Sample") +
  ylab("Mutation Frequency (mutations/bp)") +
  xlab("sample") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Average MF by dose for both clonal and unique 
plot_data <- MF_per_sample %>%
  group_by(dose) %>%
  summarize(unique = mean(sample_MF_unique),
            std_error_unique = sd(sample_MF_unique)/sqrt(n()),
            clonal = mean(sample_MF_clonal),
            std_error_clonal = sd(sample_MF_clonal)/sqrt(n())) %>%
  ungroup()
plot_data_long <- plot_data %>%
  pivot_longer(cols = c(unique, clonal), names_to = "MF_type", values_to = "avg_MF") %>%
  mutate(std_error = ifelse(grepl("unique", MF_type), std_error_unique, std_error_clonal)) %>%
  mutate(MF_type = factor(MF_type, levels = c("unique", "clonal")))  # Reorder levels
max_y_limit <- max(plot_data_long$avg_MF + plot_data_long$std_error)*1.25
ggplot(plot_data_long, aes(x = factor(dose), y = avg_MF, fill = MF_type)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.6), width = 0.5) +
  scale_y_continuous(expand=c(0,0), limits = c(0, max_y_limit)) +
 geom_errorbar(aes(ymin = pmax(avg_MF - std_error, 0), ymax = pmin(avg_MF + std_error, max_y_limit)),
                position = position_dodge(width = 0.6), width = 0.25, color = "black") +
  labs(title = "Average Mutation Frequency per Dose",
       y = "Average Mutation Frequency",
       x = "Dose") +
  scale_fill_manual(values = c("unique" = "skyblue", "clonal" = "salmon")) +
  theme_bw()


# MF per target (average per dose)
plot_data <- MF_per_target %>%
  group_by(dose, description) %>%
  summarize(avg_MF_unique = mean(sample_description_MF_unique),
            std_error = sd(sample_description_MF_unique)/sqrt(n()))
ordered_descriptions <- plot_data %>%
  filter(dose == max(plot_data$dose)) %>%
  dplyr::arrange(desc(avg_MF_unique)) %>%
  dplyr::pull(description)
plot_data$description <- factor(plot_data$description, levels = ordered_descriptions)
ggplot(plot_data, aes(x = description, y = avg_MF_unique, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  geom_errorbar(aes(ymin = avg_MF_unique - std_error, ymax = avg_MF_unique + std_error),
                position = position_dodge(width = 0.8), width = 0.25) +
  labs(title = "Average Mutation Frequency per Target and Dose with Standard Error",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Target") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
# Plot MF per target for each sample
ggplot(MF_per_target, aes(x=description, y=sample_description_MF_unique, fill = factor(dose))) +
  geom_bar(stat = "identity", alpha = 0.7) +
  geom_text(aes(y = sample_description_MF_unique, label = sample_description_sum_unique), vjust = -0.5) +
  labs(title = "Min Mutation Frequency per Target") +
  ylab("Mutation Frequency (mutations/bp)") +
  xlab("Target") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        strip.text.y.left = element_text(angle = 0)) +
  facet_grid(sample ~ ., switch = "y")

# Plot simple Spectra for each sample
ggplot(six_base_summary, aes(x=normalized_subtype, y=sample_MF_unique)) +
  geom_col() +
  facet_grid(sample ~ .)
# Plot simple spectra averaged across dose
plot_data <- six_base_summary %>%
  group_by(dose, normalized_subtype) %>%
  summarize(avg_MF = mean(sample_MF_unique),
            std_error = sd(sample_MF_unique)/sqrt(n())) %>%
  ungroup()
  # Create subset to view low freq non-snvs
plot_data_subset <- plot_data %>%
  filter(normalized_subtype %in% c("complex", "deletion", "insertion", "mnv", "symbolic"))
# Define the custom order for normalized_subtype
custom_order <- c("C>A", "C>G", "C>T", "T>A", "T>C", "T>G", "complex", "deletion", "insertion", "mnv", "symbolic")
# Convert normalized_subtype to a factor with custom order
plot_data$normalized_subtype <- factor(plot_data$normalized_subtype, levels = custom_order)
# Create the ggplot for all subtypes
ggplot(plot_data, aes(x = normalized_subtype, y = avg_MF, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  geom_errorbar(aes(ymin = avg_MF - std_error, ymax = avg_MF + std_error),
                position = position_dodge(width = 0.8), width = 0.25) +
  labs(title = "Average Mutation Frequency per Variant Type with Standard Error",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Variation Type") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
# Create subset plot for non-snv
ggplot(plot_data_subset, aes(x = normalized_subtype, y = avg_MF, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  geom_errorbar(aes(ymin = avg_MF - std_error, ymax = avg_MF + std_error),
                position = position_dodge(width = 0.8), width = 0.25) +
  labs(title = "Average Mutation Frequency per Variant Type with Standard Error",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Variation Type") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Plot simple sectra normalized proportions
ggplot(six_base_summary_group, aes(x = normalized_subtype, y = norm_prop_unique, fill = factor(dose))) +
  geom_bar(stat = "identity", position = "dodge", alpha = 0.7) +
  labs(title = "Normalized Proportions of SNV Subtypes per Dose",
       y = "Average Mutation Frequency (mutations/bp)",
       x = "Variation Type") +
  scale_fill_discrete(name = "Dose") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

# Plot Trinucleotide Frequency
```

```{r exploratory-spectra, warning = F}
# Look at overall spectrum (sum across all groups)
mutation_data %>% group_by(subtype) %>% tally()
mutation_data %>% group_by(normalized_subtype) %>% tally()
mutation_data %>% group_by(ref) %>% tally()
mutation_data %>% group_by(short_ref) %>% tally()
mutation_data %>% group_by(normalized_ref) %>% tally()



# tvti_plot(mutations = mutation_data, y = NULL)
# tvti_plot(mutations = mutation_data)
# 
# 
# tvti_plot(mutations = mutation_data, type = "Frequency")
# tvti_plot(mutations = mutation_data, type = "Proportion")
```

## CpG Mutations

```{r cpg-mutations}
# Create a GRanges object, containing all mutation data, but a column added
# to indicate whether the 'context' column contains a CpG site (i.e., is the
# mutation at a CpG site or not)
mutation_data_cpgs <- annotate_CpG_sites(mutation_data)
cpg_mutations <- get_CpG_mutations(get_seq(species = dupseq_species,
                                           genome_version = "GRCm38",
                                           regions = "mouse"), mutation_data)
# cpg_mutations <- get_CpG_mutations(get_region_seqs(dupseq_species), mutation_data)
# This only returns CpG sites, nothing to do with mutation data
cpg_sites <- get_CpG_regions(get_seq(species = dupseq_species,
                                           genome_version = "GRCm38",
                                           regions = "mouse"))
# Total number of CpG sites
nrow(as.data.frame(cpg_sites)) # could also use %>% count() %>% pull()
n_samples <- length(unique(as.data.frame(mutation_data)[["sample"]]))

# Testing general function for this application...
cpg_summary_subtype <- calculate_mut_freq(data = as.data.frame(cpg_mutations),
                                          subtype_resolution = "base_6",
                                          summary = T,
                                          cols_to_group = c("sample"))
                                       

# Mutations at CpG sites...
cpg_summary_subtype <- as.data.frame(cpg_mutations) %>%
  add_count(!!ensym(group_var), alt_depth, name = "total_muts_per_group") %>%
  ungroup() %>%
  group_by(subtype, !!ensym(group_var), total_muts_per_group) %>%
  tally() %>%
  mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
  mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))

# Mutations at CpG sites, broken down by trinucleotide mutation context
cpg_summary_context <- as.data.frame(cpg_mutations) %>%
  group_by(!!ensym(group_var)) %>%
  add_count(alt_depth, name = "total_muts_per_group") %>%
  ungroup() %>%
  group_by(context_with_mutation, !!ensym(group_var), total_muts_per_group) %>%
  tally() %>%
  mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
  mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))

cpg_base_12_pct <- as.data.frame(mutation_data_cpgs) %>%
  group_by(CpG_site, subtype, !!ensym(group_var)) %>%
  tally() %>% 
  ungroup() %>%
  group_by(subtype, !!ensym(group_var)) %>%
  mutate(total_mutations = sum(n)) %>%
  ungroup() %>%
  mutate(percent = 100*n/total_mutations) %>%
  filter(CpG_site == T) %>%
  dplyr::select(-1)

cpg_base_6_pct <- as.data.frame(mutation_data_cpgs) %>%
  group_by(CpG_site, normalized_subtype, !!ensym(group_var)) %>%
  tally() %>% 
  ungroup() %>%
  group_by(normalized_subtype, !!ensym(group_var)) %>%
  mutate(total_mutations = sum(n)) %>%
  ungroup() %>%
  mutate(percent = 100*n/total_mutations) %>%
  filter(CpG_site == T) %>%
  dplyr::select(-1)

cpg_base_96_pct <- as.data.frame(mutation_data_cpgs) %>%
  group_by(CpG_site, context_with_mutation, !!ensym(group_var)) %>%
  tally() %>% 
  ungroup() %>%
  group_by(context_with_mutation, !!ensym(group_var)) %>%
  mutate(total_mutations = sum(n)) %>%
  ungroup() %>%
  mutate(percent = 100*n/total_mutations) %>%
  filter(CpG_site == T) %>%
  dplyr::select(-1)

cpg_tables <- list(CpGs_base_12_percentage = cpg_base_12_pct,
                   CpGs_base_6_percentage = cpg_base_6_pct,
                   #CpGs_trinucleotide_percentage = cpg_base_96_pct, # Data too sparse.
                   CpGs_by_subtype = cpg_summary_subtype,
                   CpGs_by_trinucleotide = cpg_summary_context)

write_excel_from_list(list_of_tables = cpg_tables,
                      output_path = output_dir,
                      workbook_name = "cpg_tables")

```

## Signature Decomposition

```{r cosmic-signatures-sigminer}
# Aggregate by sample

sigminer_results <- signature_analysis_sigminer(
  mutation_data, project_name = project, group = "sample",
  project_genome = "BSgenome.Mmusculus.UCSC.mm10",
  vaf_cutoff = 0.1,
  run_bootstrapping = F)



# Can only handle 10 signatures at a time??
sigminer::show_sig_fit(sigminer_results$fitting_results, plot_fun = "violin")

sigminer::show_sig_profile(sigminer_results$fitting_results,
                           mode = "SBS", style = "cosmic", x_label_angle = 90)

sigminer::show_sig_exposure(sigminer_results$fitting_results, style = "cosmic")

sigminer::show_sig_bootstrap_exposure(sigminer_results$bootstrap_results) + facet_grid(sample~.)
sigminer::show_sig_bootstrap_error(sigminer_results$bootstrap_results) + facet_grid(sample~.)
sigminer::show_sig_bootstrap_stability(sigminer_results$bootstrap_results) + facet_grid(sample~.)

sigminer::show_catalogue(t(sigminer_results$fitting_results$nmf_matrix), style = "cosmic", x_label_angle = 90)


```
```{r display_sigs}
# Use sigminer to show signatures of interest for comparison
# Look at arbitrary indices... could program this to show most similar ones to data
#show_cosmic_sig_profile(sig_index = c(4, 5, 6), style = "cosmic", sig_db = "ID")
#show_cosmic_sig_profile(sig_index = c(1, 2, 3), style = "cosmic", sig_db = "DBS")
#show_cosmic_sig_profile(sig_index = c(1, 5, 6), style = "cosmic")

```

## Recurring mutations

Are any mutations observed in more than one individual?

```{r}

recurring_mutations <- mutation_data_cpgs %>%
  group_by(seqnames,start,end) %>%
  filter(!variation_type %in% "no_variant") %>%
  tally() %>%
  arrange(-n)

recurring_mutations_group <- mutation_data_cpgs %>%
  group_by(seqnames,start,end,sample) %>%
  tally() %>%
  arrange(-n)

recurring_mutations_rsid <- mutation_data_cpgs %>%
  group_by(id) %>%
  tally() %>%
  arrange(-n)
```

## Write data

```{r write-output-tables}

# All mutation data, cleaned and processed
write_excel_single_table(mut_data = mutation_data_cpgs,
                      output_path = output_dir,
                      workbook_name = paste0(project,"_mutation_data"))

```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
