---
title: "MutSeqR Summary Report"
author: "Matthew J. Meier & Annette E. Dodge"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
    theme: spacelab
    code_download: yes
  pdf_document:
    toc: yes
params:
  projectdir:
    label: "The Project Directory"
    input: text
    value: "C:/Users/adodge/OneDrive - HC-SC PHAC-ASPC/Documents/DupSeq R Package Building/Paper/Project_1_BM_BaP_Mouse"
  output_path:
    label: "The output directory"
    input: text
    value: "example_output"
  project_title:
    label: "Title of the project"
    input: text
    value: "MutSeqR Analysis"
  researcher_name:
    label: "Name of the lead researcher."
    input: text
    value: "FirstName LastName"
  user_name:
    label: "The name of the person running the analysis." 
    input: text
    value: "FirstName LastName"
  project_description:
    label: "Description of the project, optional."
    input: text
    value: "This is a test of the Rmd report generation"
  ecs_profile:
    label: "Use a preset profile for a given ECS technology."
    input: select
    choices: ["Duplex Sequencing Human Mutagenesis Panel", 
              "Duplex Sequencing Mouse Mutagenesis Panel",
              "Duplex Sequencing Rat Mutagenesis Panel",
              "Duplex Sequencing Custom Panel",
              "Non-targeted Sequencing",
              "Custom Parameters"]
    value: Duplex Sequencing Mouse Mutagenesis Panel"
  analysis_profile:
    label: "Use preset parameters for a given analysis profile."
    input: select
    choices: ["dose-response analysis", "one factor analysis"]
    value: "dose-response analysis"
  species: 
    label: "Model Species"
    input: text
    value: "mouse"
  genome:
    label: "The reference genome"
    input: text
    value: "mm10"
  file_type:
    label: "Type of mutation data file"
    input: select
    choices: ["table", "vcf"]
    value: "table"
  mutation_file:
    label: "The file path to the mutation data."
    input: text
    value: "mut"
  mut_sep:
    label: "The delimiter to import the mutation data"
    input: text
    value: "\t"
  sample_data:
    label: "The file path to the sample metadata."
    input: text
    value: "sample_data.txt"
  sd_sep:
    label: "The delimiter to import the sample data file."
    input: text
    value: "\t"
  custom_regions:
    label: "Provide the file path to your regions file (for Custom panels)."
    input: text
    value: NULL
  rg_sep:
    label: "The delimiter to import the (custom) regions file."
    input: text
    value: "\t"
  is_0_based_rg:
    label: "The regions file is 0-based"
    input: checkbox
    value: "TRUE"
  custom_filtering_regions:
    label: "Provide the file path to your regions file that will be used for filtering (for Custom panels)."
    input: text
    value: NULL
  custom_regions_filter:
    label: "How would you like to use your regions for filtering?"
    input: select
    choices: ["keep_within", "remove_within"]
    value: "keep_within"
  rg_sep_filter:
    label: "The delimiter to import the (custom) regions file for filtering."
    input: text
    value: "\t"
  is_0_based_rg_filter:
    label: "The custom_filtering_regions file is 0-based"
    input: checkbox
    value: "TRUE"
---

```{r load_libraries, warning=F, echo=F}
set.seed(3135)
library(MutSeqR)
library(here)
library(knitr)

if (!rmarkdown::pandoc_available()) {
  stop("Please install pandoc to use this report template.")
}
```

```{r Set Directories}
if (is.null(params$projectdir)) {
  project_dir <- file.path(here::here())
} else {
  project_dir <- file.path(params$projectdir)
}
if (!dir.exists(project_dir)) {
  dir.create(project_dir)
}
knitr::opts_knit$set(root.dir = project_dir)

if (is.null(params$output_path)) {
  output_dir <- file.path(project_dir, "output")
} else {
  output_dir <- file.path(params$output_path)
}
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}
```

```{r setup}
has_description <- !is.null(params$project_description)
description_text <- paste("Purpose or report:", params$project_description)
```

# `r params$project_title` Summary Report {-}
Date report generated: `r format(Sys.time(), '%d %B, %Y')`

Report prepared by: `r params$user_name`

Report prepared for: `r params$researcher_name`

`r if(has_description) {description_text}`

***

```{r retrieve-parameter-profiles}
ecs_params <- get_params(profile = params$ecs_profile,
                         custom_regions = params$custom_regions,
                         custom_regions_filter = params$custom_regions_filter)
```
## Import the Mutation Data

```{r import-data, echo = TRUE}

no_depth_warning <- "Could not find an appropriate depth column. Some package functionality may be limited.\n"
correct_depth_warning <-  "The total_depth may be double-counted in some instances due to overlapping positions. Use the filter_mut() function to correct the total_depth for these instances."
warnings_list <- character(0)
mutation_data <- NULL

if (params$file_type == "table") {
  mutation_data <- withCallingHandlers(
    import_mut_data(mut_file = params$mutation_file,
                    mut_sep = "\t",
                    is_0_based_mut = ecs_params$is_0_based_mut,
                    sample_data = params$sample_data,
                    sd_sep = "\t",
                    regions = ecs_params$regions,
                    rg_sep = "\t",
                    is_0_based_rg = params$is_0_based_rg,
                    species = params$species,
                    genome = params$genome),
    warning = function(w) {
      warnings_list <<- c(warnings_list, w$message)
    }
  )
  no_depth <- any(grepl(no_depth_warning,
                        warnings_list,
                        fixed = TRUE))
  correct_depth <- any(grepl(correct_depth_warning,
                             warnings_list,
                             fixed = TRUE))
} else if (params$mut_file_type == "vcf") {
  mutation_data <- withCallingHandlers(
    import_vcf_data(vcf_file = params$mutation_file,
                    sample_data = params$sample_data,
                    sd_sep = "\t",
                    regions = ecs_params$regions,
                    rg_sep = "\t",
                    is_0_based_rg = params$is_0_based_rg,
                    species = params$species,
                    genome = params$genome),
    warning = function(w) {
      warnings_list <<- c(warnings_list, w$message)
    }
  )
  no_depth <- any(grepl(no_depth_warning, warnings_list, fixed = TRUE))
  correct_depth <- any(grepl(correct_depth_warning,
                             warnings_list,
                             fixed = TRUE))
}


```

# Filter Mutation Data
Flagging variants with a VAF > than `r ecs_params$vaf_cutoff`

Flagging SNVs that overlap with germline MNVs: `r ecs_params$snv_in_germ_mnv`

Removing variants with an abnormal VAF: `r ecs_params$rm_abnormal_vaf`

Using a custom filter: `r !is.null(ecs_params$custom_filter_col)`
```{r echo=FALSE, eval=!is.null(ecs_params$custom_filter_col)}
cat("Filtering on column: ", ecs_params$custom_filter_col, " with value: ", ecs_params$custom_filter_val)
cat("Variants are ", ifelse(ecs_params$custom_filter_rm, " removed", "flagged"), " when value matched: ", ecs_params$custom_filter_val)
``` 

Removing records based on a regions file: `r !is.null(ecs_params$filtering_regions)`
```{r echo=FALSE, eval=!is.null(ecs_params$filtering_regions)}
cat("Filtering on regions file: ", ecs_params$filtering_regions)
cat("Records within the specified filtering regions are", ifelse(ecs_params$regions_filter == "keep_within", " retained", " removed"), " from the dataset.")
cat("Records must start and end within regions to be included: ", !ecs_params$allow_half_overlap)

```

Flagged variants (except germline) have their alt_depth subtracted from their total_depth: `r ecs_params$rm_filtered_mut_from_depth`

Flagged variants are not included in the mutation counts, but their total_depth is retained for frequency calculations.
All filtered records are returned in *filtered_variants.xlsx.*

Correcting the total_depth for overlapping positions: `r correct_depth`
```{r filter-mutation-data, echo=TRUE, message=TRUE}
filtered_mutation_data <- filter_mut(mutation_data,
                                     correct_depth = correct_depth,
                                     vaf_cutoff = ecs_params$vaf_cutoff,
                                     snv_in_germ_mnv = ecs_params$snv_in_germ_mnv,
                                     rm_abnormal_vaf = ecs_params$rm_abnormal_vaf,
                                     custom_filter_col = ecs_params$custom_filter_col,
                                     custom_filter_val = ecs_params$custom_filter_val,
                                     custom_filter_rm = ecs_params$custom_filter_rm,
                                     regions = ecs_params$filtering_regions,
                                     rg_sep = "\t",
                                     regions_filter = ecs_params$regions_filter,
                                     allow_half_overlap = ecs_params$allow_half_overlap,
                                     is_0_based_rg = params$is_0_based_rg_filter,
                                     rm_filtered_mut_from_depth = ecs_params$rm_filtered_mut_from_depth,
                                     return_filtered_data = TRUE)
filtered_variants <- filtered_mutation_data$filtered_rows
write_excel(data = filtered_variants,
            output_path = output_dir,
            workbook_name = "filtered_variants")
filtered_mutation_data <- filtered_mutation_data$mutation_data
```


## CpG Mutations

```{r cpg-mutations, eval=FALSE}
# Create a GRanges object, containing all mutation data, but a column added
# to indicate whether the 'context' column contains a CpG site (i.e., is the
# mutation at a CpG site or not)
mutation_data_cpgs <- annotate_CpG_sites(mutation_data)
cpg_mutations <- get_CpG_mutations(my_regions, mutation_data)
# Returns CpG sites of your regions, it has nothing to do with mutation data
cpg_sites <- get_CpG_regions(my_regions)
# Total number of CpG sites
nrow(as.data.frame(cpg_sites)) # could also use %>% count() %>% pull()
n_samples <- length(unique(as.data.frame(mutation_data)[["sample"]]))

# Testing general function for this application...
cpg_summary_subtype <- calculate_mf(data = as.data.frame(cpg_mutations),
                                          subtype_resolution = "base_6",
                                          vaf_cutoff = params$vaf_cutoff,
                                          summary = T,
                                          cols_to_group = c("sample"))
                                       

# Mutations at CpG sites...
cpg_summary_subtype <- as.data.frame(cpg_mutations) %>%
  add_count(dose, alt_depth, name = "total_muts_per_group") %>%
  ungroup()#%>%
  group_by(subtype, !!ensym(params$group_var), total_muts_per_group) %>%
  tally() %>%
  mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
  mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))

# # Mutations at CpG sites, broken down by trinucleotide mutation context
# cpg_summary_context <- as.data.frame(cpg_mutations) %>%
#   group_by(!!ensym(group_var)) %>%
#   add_count(alt_depth, name = "total_muts_per_group") %>%
#   ungroup() %>%
#   group_by(context_with_mutation, !!ensym(params$group_var), total_muts_per_group) %>%
#   tally() %>%
#   mutate(numeric_pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples)) %>%
#   mutate(proportional_pct_cpg = 100*(100*n/total_muts_per_group)/(nrow(as.data.frame(cpg_sites))*n_samples))
# 
# cpg_base_12_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, subtype, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(subtype, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_base_6_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, normalized_subtype, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(normalized_subtype, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_base_96_pct <- as.data.frame(mutation_data_cpgs) %>%
#   group_by(CpG_site, context_with_mutation, !!ensym(params$group_var)) %>%
#   tally() %>% 
#   ungroup() %>%
#   group_by(context_with_mutation, !!ensym(params$group_var)) %>%
#   mutate(total_mutations = sum(n)) %>%
#   ungroup() %>%
#   mutate(percent = 100*n/total_mutations) %>%
#   filter(CpG_site == T) %>%
#   dplyr::select(-1)
# 
# cpg_tables <- list(CpGs_base_12_percentage = cpg_base_12_pct,
#                    CpGs_base_6_percentage = cpg_base_6_pct,
#                    #CpGs_trinucleotide_percentage = cpg_base_96_pct, # Data too sparse.
#                    CpGs_by_subtype = cpg_summary_subtype,
#                    CpGs_by_trinucleotide = cpg_summary_context)
# 
# write_excel_from_list(list_of_tables = cpg_tables,
#                       output_path = output_dir,
#                       workbook_name = "cpg_tables")

```

## Recurring mutations

Are any mutations observed in more than one sample?

```{r, eval=FALSE}

recurring_mutations <- mutation_data_cpgs %>%
  group_by(contig,start,end) %>%
  filter(!variation_type %in% "no_variant") %>%
  tally() %>%
  arrange(-n)

recurring_mutations_group <- mutation_data_cpgs %>%
  group_by(contig,start,end,sample) %>%
  tally() %>%
  arrange(-n)

recurring_mutations_rsid <- mutation_data_cpgs %>%
  group_by(id) %>%
  tally() %>%
  arrange(-n)
```


<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
