<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Filter your mutation data — filter_mut • MutSeqR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Filter your mutation data — filter_mut"><meta name="description" content="This function creates a filter_mut`` column that will be read by the \code{calculate_mf} function and other downstream functions. Variants with filter_mut == TRUE“ will be excluded from group mutation
counts. This function may also remove records upon on user specification.
Running this function again on the same data will not overide the previous
filters. To reset previous filters, set the filter_mut column values to
FALSE."><meta property="og:description" content="This function creates a filter_mut`` column that will be read by the \code{calculate_mf} function and other downstream functions. Variants with filter_mut == TRUE“ will be excluded from group mutation
counts. This function may also remove records upon on user specification.
Running this function again on the same data will not overide the previous
filters. To reset previous filters, set the filter_mut column values to
FALSE."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MutSeqR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/MutSeqR_introduction.html">Overview of MutSeqR workflow</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Filter your mutation data</h1>
      <small class="dont-index">Source: <a href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/blob/main/R/filter_mut.r" class="external-link"><code>R/filter_mut.r</code></a></small>
      <div class="d-none name"><code>filter_mut.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function creates a <code>filter_mut`` column that will be read by the \code{calculate_mf} function and other downstream functions. Variants with </code>filter_mut == TRUE“ will be excluded from group mutation
counts. This function may also remove records upon on user specification.
Running this function again on the same data will not overide the previous
filters. To reset previous filters, set the filter_mut column values to
FALSE.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">filter_mut</span><span class="op">(</span></span>
<span>  <span class="va">mutation_data</span>,</span>
<span>  vaf_cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  snv_in_germ_mnv <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rm_abnormal_vaf <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  custom_filter_col <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_filter_val <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  custom_filter_rm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  regions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">regions_filter</span>,</span>
<span>  allow_half_overlap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rg_sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  is_0_based_rg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm_filtered_mut_from_depth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  return_filtered_rows <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mutation-data">mutation_data<a class="anchor" aria-label="anchor" href="#arg-mutation-data"></a></dt>
<dd><p>Your mutation data. This can be a data frame or a
GRanges object.</p></dd>


<dt id="arg-vaf-cutoff">vaf_cutoff<a class="anchor" aria-label="anchor" href="#arg-vaf-cutoff"></a></dt>
<dd><p>Filter out ostensibly germline variants using a cutoff for
variant allele fraction (VAF). Any variant with a <code>vaf</code> larger than
the cutoff will be filtered. The default is 1 (no filtering). It is
recommended to use a value of 0.01 (i.e. 1%) as a conservative approach
to retain only somatic variants.</p></dd>


<dt id="arg-snv-in-germ-mnv">snv_in_germ_mnv<a class="anchor" aria-label="anchor" href="#arg-snv-in-germ-mnv"></a></dt>
<dd><p>Filter out snv variants that overlap with
germline mnv variants within the same samples. mnv variants will be
considered germline if their vaf &gt; vaf_cutoff. Default is FALSE.</p></dd>


<dt id="arg-rm-abnormal-vaf">rm_abnormal_vaf<a class="anchor" aria-label="anchor" href="#arg-rm-abnormal-vaf"></a></dt>
<dd><p>A logical value. If TRUE, rows in
<code>mutation_data</code> with a variant allele fraction (VAF) between 0.05 and
0.45 or between 0.55 and 0.95 will be removed. We expect variants to have a
VAF ~0. 0.5, or 1, reflecting rare somatic mutations, heterozygous germline
mutations, and homozygous germline mutations, respectively. Default is
FALSE.</p></dd>


<dt id="arg-custom-filter-col">custom_filter_col<a class="anchor" aria-label="anchor" href="#arg-custom-filter-col"></a></dt>
<dd><p>The name of the column in mutation_data to apply a
custom filter to. This column will be checked for specific values, as defined
by <code>custom_filter_val</code>. If any row in this column contains one of the
specified values, that row will either be flagged in the
<code>filter_mut column</code> or, if specified by <code>custom_filter_rm</code>,
removed from mutation_data.</p></dd>


<dt id="arg-custom-filter-val">custom_filter_val<a class="anchor" aria-label="anchor" href="#arg-custom-filter-val"></a></dt>
<dd><p>A set of values used to filter rows in
<code>mutation_data</code> based on <code>custom_filter_col</code>. If a row in
<code>custom_filter_col</code> matches any value in <code>custom_filter_val</code>,
it will either be set to TRUE in the <code>filter_mut</code> column or removed,
depending on <code>custom_filter_rm</code>.</p></dd>


<dt id="arg-custom-filter-rm">custom_filter_rm<a class="anchor" aria-label="anchor" href="#arg-custom-filter-rm"></a></dt>
<dd><p>A logical value. If TRUE, rows in custom_filter_col
that match any value in custom_filter_val will be removed from the
mutation_data. If FALSE, <code>filter_mut</code> will be set to TRUE for those
rows.</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>Remove rows that are within/outside of specified regions.
<code>regions</code> can be either a file path, a data frame, or a GRanges object
containing the genomic ranges by which to filter. File paths will be read
using the rg_sep. Users can also choose from the built-in TwinStrand's
Mutagenesis Panels by inputting "TSpanel_human",  "TSpanel_mouse", or
"TSpanel_rat". Required columns for the regions file are "contig", "start",
and "end". In a GRanges object, the required columns are "seqnames",
"start", and "end".</p></dd>


<dt id="arg-regions-filter">regions_filter<a class="anchor" aria-label="anchor" href="#arg-regions-filter"></a></dt>
<dd><p>Specifies how the provided <code>regions</code> should be
applied to <code>mutation_data</code>. Acceptable values are "remove_within" or
"keep_within". If set to "remove_within", records that fall within the
specified regions wil be removed from mutation_data. If set to
"keep_within", only records within the specified regions will be kept in
mutation_data, and all other records will be removed.</p></dd>


<dt id="arg-allow-half-overlap">allow_half_overlap<a class="anchor" aria-label="anchor" href="#arg-allow-half-overlap"></a></dt>
<dd><p>A logical value. If TRUE, records that start or
end in your <code>regions</code>, but extend outside of them in either direction
will be included in the filter. If FALSE, only records that start and end
within the <code>regions</code> will be included in the filter. Default is FALSE.</p></dd>


<dt id="arg-rg-sep">rg_sep<a class="anchor" aria-label="anchor" href="#arg-rg-sep"></a></dt>
<dd><p>The delimiter for importing the custom_regions. The default is
tab-delimited "\t".</p></dd>


<dt id="arg-is-based-rg">is_0_based_rg<a class="anchor" aria-label="anchor" href="#arg-is-based-rg"></a></dt>
<dd><p>A logical variable. Indicates whether the position
coordinates in <code>regions</code> are 0 based (TRUE) or 1 based (FALSE).
If TRUE, positions will be converted to 1-based (start + 1).
Need not be supplied for TSpanels. Default is TRUE.</p></dd>


<dt id="arg-rm-filtered-mut-from-depth">rm_filtered_mut_from_depth<a class="anchor" aria-label="anchor" href="#arg-rm-filtered-mut-from-depth"></a></dt>
<dd><p>A logical value. If TRUE, the function will
subtract the <code>alt_depth</code> of records that were flagged by the
<code>filter_mut</code> column from their <code>total_depth</code>. This will treat
flagged variants as No-calls. This will not apply to variants flagged as
germline by the <code>vaf_cutoff</code>. However, if the germline variant
has additional filters applied, then the subtraction will still occur.
If FALSE, the <code>alt_depth</code> will be retained in the
<code>total_depth</code> for all variants.  Default is FALSE.</p></dd>


<dt id="arg-return-filtered-rows">return_filtered_rows<a class="anchor" aria-label="anchor" href="#arg-return-filtered-rows"></a></dt>
<dd><p>A logical value. If TRUE, the function will
return both the filtered mutation data and the records that were
removed/flagged in a seperate data frame. The two dataframes will be
returned inside a list, with names <code>mutation_data</code> and
<code>filtered_rows</code>. Default is FALSE.</p></dd>

</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Load example data</span></span></span>
<span class="r-in"><span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Example_files"</span>,</span></span>
<span class="r-in"><span>                            <span class="st">"example_mutation_data.rds"</span>,</span></span>
<span class="r-in"><span>                            package <span class="op">=</span> <span class="st">"MutSeqR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">example_file</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Filter the data</span></span></span>
<span class="r-in"><span><span class="co"># Basic Usage: Filter out putative germline variants</span></span></span>
<span class="r-in"><span><span class="va">filter_example_1</span> <span class="op">&lt;-</span> <span class="fu">filter_mut</span><span class="op">(</span>mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span></span>
<span class="r-in"><span>                               vaf_cutoff <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flagging germline mutations...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 612 germline mutations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering complete.</span>
<span class="r-in"><span><span class="co"># Remove rows outside of the TwinStand Mouse Mutagenesis Panel regions</span></span></span>
<span class="r-in"><span><span class="va">filter_example_2</span> <span class="op">&lt;-</span> <span class="fu">filter_mut</span><span class="op">(</span>mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span></span>
<span class="r-in"><span>                               vaf_cutoff <span class="op">=</span> <span class="fl">0.01</span>,</span></span>
<span class="r-in"><span>                               regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span>,</span></span>
<span class="r-in"><span>                               regions_filter <span class="op">=</span> <span class="st">"keep_within"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flagging germline mutations...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 612 germline mutations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Applying region filter...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 22 rows based on regions.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering complete.</span>
<span class="r-in"><span><span class="co"># Apply a custom filter to flag rows with "EndRepairFillInArtifact"</span></span></span>
<span class="r-in"><span><span class="co"># in the column 'filter'</span></span></span>
<span class="r-in"><span><span class="va">filter_example_3</span> <span class="op">&lt;-</span> <span class="fu">filter_mut</span><span class="op">(</span>mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span></span>
<span class="r-in"><span>                               vaf_cutoff <span class="op">=</span> <span class="fl">0.01</span>,</span></span>
<span class="r-in"><span>                               regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span>,</span></span>
<span class="r-in"><span>                               regions_filter <span class="op">=</span> <span class="st">"keep_within"</span>,</span></span>
<span class="r-in"><span>                               custom_filter_col <span class="op">=</span> <span class="st">"filter"</span>,</span></span>
<span class="r-in"><span>                               custom_filter_val <span class="op">=</span> <span class="st">"EndRepairFillInArtifact"</span>,</span></span>
<span class="r-in"><span>                               custom_filter_rm <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flagging germline mutations...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 612 germline mutations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Applying custom filter...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flagged 2021 rows with values in &lt;filter&gt; column that matched EndRepairFillInArtifact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Applying region filter...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 22 rows based on regions.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering complete.</span>
<span class="r-in"><span><span class="co"># Flag snv variants that overlap with germline mnv variants.</span></span></span>
<span class="r-in"><span><span class="co"># Subtract the alt_depth of these variants from their total_depth</span></span></span>
<span class="r-in"><span><span class="co"># (treat them as No-calls).</span></span></span>
<span class="r-in"><span><span class="co"># Return all the flagged/removed rows in a seperate data frame</span></span></span>
<span class="r-in"><span><span class="va">filter_example_4</span> <span class="op">&lt;-</span> <span class="fu">filter_mut</span><span class="op">(</span>mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span></span>
<span class="r-in"><span>                               vaf_cutoff <span class="op">=</span> <span class="fl">0.01</span>,</span></span>
<span class="r-in"><span>                               regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span>,</span></span>
<span class="r-in"><span>                               regions_filter <span class="op">=</span> <span class="st">"keep_within"</span>,</span></span>
<span class="r-in"><span>                               custom_filter_col <span class="op">=</span> <span class="st">"filter"</span>,</span></span>
<span class="r-in"><span>                               custom_filter_val <span class="op">=</span> <span class="st">"EndRepairFillInArtifact"</span>,</span></span>
<span class="r-in"><span>                               custom_filter_rm <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                               snv_in_germ_mnv <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                               rm_filtered_mut_from_depth <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>                               return_filtered_rows <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flagging germline mutations...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 612 germline mutations.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flagging SNVs overlapping with germline MNVs...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Found 20 SNVs overlapping with germline MNVs.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Applying custom filter...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Flagged 2021 rows with values in &lt;filter&gt; column that matched EndRepairFillInArtifact</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Applying region filter...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removed 22 rows based on regions.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Removing filtered mutations from the total_depth...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Filtering complete.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Returning a list: mutation_data and filtered_rows.</span>
<span class="r-in"><span><span class="co"># Flagging germline mutations...</span></span></span>
<span class="r-in"><span><span class="co"># Found 612 germline mutations.</span></span></span>
<span class="r-in"><span><span class="co"># Flagging SNVs overlapping with germline MNVs...</span></span></span>
<span class="r-in"><span><span class="co"># Found 20 SNVs overlapping with germline MNVs.</span></span></span>
<span class="r-in"><span><span class="co"># Applying custom filter...</span></span></span>
<span class="r-in"><span><span class="co"># Flagged 2021 rows with values in &lt;filter&gt; column that matched EndRepairFillInArtifact</span></span></span>
<span class="r-in"><span><span class="co"># Applying region filter...</span></span></span>
<span class="r-in"><span><span class="co"># Removed 22 rows based on regions.</span></span></span>
<span class="r-in"><span><span class="co"># Correcting depth...</span></span></span>
<span class="r-in"><span><span class="co"># 909 rows had their total_depth corrected.</span></span></span>
<span class="r-in"><span><span class="co"># Removing filtered mutations from the total_depth...</span></span></span>
<span class="r-in"><span><span class="co"># Filtering complete.</span></span></span>
<span class="r-in"><span><span class="co"># Returning a list: mutation_data and filtered_rows.</span></span></span>
<span class="r-in"><span><span class="va">filtered_rows</span> <span class="op">&lt;-</span> <span class="va">filter_example_4</span><span class="op">$</span><span class="va">filtered_rows</span></span></span>
<span class="r-in"><span><span class="va">filtered_example_mutation_data</span> <span class="op">&lt;-</span> <span class="va">filter_example_4</span><span class="op">$</span><span class="va">mutation_data</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Annette E. Dodge, Andrew Williams, Danielle P.M. LeBlanc, David M. Schuster, Elena Esina, Clint C. Valentine, Jesse J. Salk, Alexander Y. Maslov, Christopher Bradley, Carole L. Yauk, Francesco Marchetti, Matthew J. Meier.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

