<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get sequence of genomic target regions — get_seq • MutSeqR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Get sequence of genomic target regions — get_seq"><meta name="description" content="Create a GRanges object from the genomic target ranges and import raw
nucleotide sequences."><meta property="og:description" content="Create a GRanges object from the genomic target ranges and import raw
nucleotide sequences."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MutSeqR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/MutSeqR_introduction.html">MutSeqR: Error-Corrected Sequencing (ECS) Analysis For Mutagenicity Assessment</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Get sequence of genomic target regions</h1>
      <small class="dont-index">Source: <a href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/blob/main/R/get_seq.R" class="external-link"><code>R/get_seq.R</code></a></small>
      <div class="d-none name"><code>get_seq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Create a GRanges object from the genomic target ranges and import raw
nucleotide sequences.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_seq</span><span class="op">(</span></span>
<span>  <span class="va">regions</span>,</span>
<span>  rg_sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  is_0_based_rg <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  species <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  masked <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  padding <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  ucsc <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>The regions metadata file to import. Can be either a file
path, a data frame, or a GRanges object. File paths will be read using
the rg_sep. Users can also choose from the built-in TwinStrand's Mutagenesis
Panels by inputting "TSpanel_human",  "TSpanel_mouse", or "TSpanel_rat".
Required columns for the regions file are "contig", "start", and "end".
In a GRanges object, the required columns are "seqnames", "start", and
"end".</p></dd>


<dt id="arg-rg-sep">rg_sep<a class="anchor" aria-label="anchor" href="#arg-rg-sep"></a></dt>
<dd><p>The delimiter for importing the regions file. The default is
tab-delimited ("\t").</p></dd>


<dt id="arg-is-based-rg">is_0_based_rg<a class="anchor" aria-label="anchor" href="#arg-is-based-rg"></a></dt>
<dd><p>A logical variable. Indicates whether the position
coordinates in <code>regions</code> are 0 based (TRUE) or 1 based (FALSE).
If TRUE, positions will be converted to 1-based (start + 1). Need not be
supplied for TSpanels. Default is TRUE.</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>The species for which to retrieve the sequences.
Species may be given as the scientific name or the common name.
Ex. "Human", "Homo sapien". Used to choose the appropriate
BS genome. Need not be supplied for TSpanels.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>The genome assembly version for which to retrieve the
sequences. Used to choose the appropriate genome (BS genome or UCSC).
Ex. hg38, hg19, mm10, mm39, rn6, rn7. Need not be supplied for TSpanels.</p></dd>


<dt id="arg-masked">masked<a class="anchor" aria-label="anchor" href="#arg-masked"></a></dt>
<dd><p>A logical value indicating whether to use the masked version
of the BS genome when retrieving sequences. Default is FALSE.</p></dd>


<dt id="arg-padding">padding<a class="anchor" aria-label="anchor" href="#arg-padding"></a></dt>
<dd><p>An integer value by which the function will extend the range
of the target sequence on both sides. Start and end coordinates will be
adjusted accordingly. Default is 0.</p></dd>


<dt id="arg-ucsc">ucsc<a class="anchor" aria-label="anchor" href="#arg-ucsc"></a></dt>
<dd><p>A logical value. If TRUE, the function will retrieve the
sequences from the UCSC genome browser using an API. If FALSE, the function
will retrieve sequences using the appropriate BSgenome package, which will
be installed as needed. Default is FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a GRanges object with sequences of targeted regions.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Consult
<code>available.genomes(splitNameParts=FALSE, type=getOption("pkgType"))</code>
for a full list of the available BS genomes and their associated
species/genome/masked values. The BSgenome package will be installed if
not already available. If using the UCSC API, the function will retrieve
the sequences from the UCSC genome browser using the DAS API. See the
UCSC website for available genomes: <a href="https://genome.ucsc.edu" class="external-link">https://genome.ucsc.edu</a>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example 1: Retrieve the sequences for TwinStrand Mouse Mutagenesis Panel</span></span></span>
<span class="r-in"><span><span class="va">regions_seq</span> <span class="op">&lt;-</span> <span class="fu">get_seq</span><span class="op">(</span>regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'getOption("repos")' replaces Bioconductor standard repositories, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'help("repositories", package = "BiocManager")' for details.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacement repositories:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     CRAN: https://cran.rstudio.com</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reference genome already installed.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading reference genome: BSgenome.Mmusculus.UCSC.mm10.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example 2: Retrieve the sequences for custom regions</span></span></span>
<span class="r-in"><span><span class="co"># We will load the TSpanel_human regions file as an example</span></span></span>
<span class="r-in"><span><span class="co"># and supply it to the function as a GRanges object.</span></span></span>
<span class="r-in"><span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu"><a href="load_regions_file.html">load_regions_file</a></span><span class="op">(</span><span class="st">"TSpanel_human"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">regions_seq</span> <span class="op">&lt;-</span> <span class="fu">get_seq</span><span class="op">(</span>regions <span class="op">=</span> <span class="va">human</span>,</span></span>
<span class="r-in"><span>                       is_0_based_rg <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                       species <span class="op">=</span> <span class="st">"human"</span>,</span></span>
<span class="r-in"><span>                       genome <span class="op">=</span> <span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                       masked <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>                       padding <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'getOption("repos")' replaces Bioconductor standard repositories, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'help("repositories", package = "BiocManager")' for details.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacement repositories:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     CRAN: https://cran.rstudio.com</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reference genome already installed.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading reference genome: BSgenome.Hsapiens.UCSC.hg38.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matthew J. Meier, Annette E. Dodge, Andrew Williams, Elena Esina.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

