<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Get mutations at CpG sites. — get_cpg_mutations • MutSeqR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Get mutations at CpG sites. — get_cpg_mutations"><meta name="description" content="Needs to be reworked for variants &amp;gt;1bp. Subset the mutation data and return only mutations that are found
at positions with a specific motif. The default is CpG sites, but can be
customizable."><meta property="og:description" content="Needs to be reworked for variants &amp;gt;1bp. Subset the mutation data and return only mutations that are found
at positions with a specific motif. The default is CpG sites, but can be
customizable."><meta property="og:image" content="https://ehsrb-bsrse-bioinformatics.github.io/MutSeqR/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MutSeqR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/MutSeqR_introduction.html">Overview of MutSeqR workflow</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Get mutations at CpG sites.</h1>
      <small class="dont-index">Source: <a href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/blob/Try-quarto-vignette/R/CpG_sites_analysis.R" class="external-link"><code>R/CpG_sites_analysis.R</code></a></small>
      <div class="d-none name"><code>get_CpG_mutations.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><em>Needs to be reworked for variants &gt;1bp</em>. Subset the mutation data and return only mutations that are found
at positions with a specific motif. The default is CpG sites, but can be
customizable.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_cpg_mutations</span><span class="op">(</span></span>
<span>  <span class="va">mutation_data</span>,</span>
<span>  <span class="va">regions</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-no_variant"</span><span class="op">)</span>,</span>
<span>  motif <span class="op">=</span> <span class="st">"CG"</span>,</span>
<span>  filter_mut <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mutation-data">mutation_data<a class="anchor" aria-label="anchor" href="#arg-mutation-data"></a></dt>
<dd><p>A dataframe or GRanges object containing the mutation
data to be interrogated. If supplying a data frame, the genomic coordinates
must be 1-based (true for mutation data imported using import_mut_data or
import_vcf_data).</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>A GRanges object containing the genomic regions of interest
in which to look for CpG sites. Must have the metadata column "sequence"
populated with the raw nucleotide sequence to search for CpGs. This object
can be obtained using the get_seq.R function.</p></dd>


<dt id="arg-variant-types">variant_types<a class="anchor" aria-label="anchor" href="#arg-variant-types"></a></dt>
<dd><p>Use this parameter to choose which variation_types
to include in the output. Provide a character vector of the variation _types
that you want to include. Options are "ambiguous", "complex", "deletion",
"insertion", "mnv", "no_variant", "snv", "sv", "uncategorized".
Alternatively, provide a character vector of the variation_types that you
want to exclude preceded by "-". All variation_types except those excluded
will be returned. Ex. inclusion: variant_types = "snv", will return only
rows with variation_type == "snv". Ex. exclusion:
variant_types = "-no_variant" will return all rows, except those with
variation_type == "no_variant" (default).</p></dd>


<dt id="arg-motif">motif<a class="anchor" aria-label="anchor" href="#arg-motif"></a></dt>
<dd><p>Default "CG", which returns CpG sites. You could in theory use
an arbitrary string to look at different motifs. Use with caution.</p></dd>


<dt id="arg-filter-mut">filter_mut<a class="anchor" aria-label="anchor" href="#arg-filter-mut"></a></dt>
<dd><p>A logical value indicating whether the function should
exclude rows flagged in the filter_mut column from the output. Default
is TRUE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A GRanges object where each range is a mutation at a CpG site
(a subset of mutations from the larger object provided to the function).</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Annette E. Dodge, Andrew Williams, Danielle P.M. LeBlanc, David M. Schuster, Elena Esina, Clint C. Valentine, Jesse J. Salk, Alexander Y. Maslov, Christopher Bradley, Carole L. Yauk, Francesco Marchetti, Matthew J. Meier.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

