<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Import a VCF file — import_vcf_data • MutSeqR</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Import a VCF file — import_vcf_data"><meta name="description" content="The function reads VCF file(s) and extracts the
data into a dataframe."><meta property="og:description" content="The function reads VCF file(s) and extracts the
data into a dataframe."><meta property="og:image" content="https://ehsrb-bsrse-bioinformatics.github.io/MutSeqR/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MutSeqR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/MutSeqR_introduction.html">Overview of MutSeqR workflow</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Import a VCF file</h1>
      <small class="dont-index">Source: <a href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/blob/main/R/import_vcf_data.R" class="external-link"><code>R/import_vcf_data.R</code></a></small>
      <div class="d-none name"><code>import_vcf_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function reads VCF file(s) and extracts the
data into a dataframe.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">import_vcf_data</span><span class="op">(</span></span>
<span>  <span class="va">vcf_file</span>,</span>
<span>  sample_data <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sd_sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  regions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rg_sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  is_0_based_rg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  padding <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  genome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  species <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  masked_BS_genome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  output_granges <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-vcf-file">vcf_file<a class="anchor" aria-label="anchor" href="#arg-vcf-file"></a></dt>
<dd><p>The path to the .vcf (.gvcf, gzip, bgzip) to be
imported. If you specify a directory, the function will
attempt to read all files in the directory and combine them into
a single table. VCF files should follow the VCF specifications,
version 4.5. Multisample VCF files are not supported; VCF files
must contain one sample each. Required fields are listed in details.</p></dd>


<dt id="arg-sample-data">sample_data<a class="anchor" aria-label="anchor" href="#arg-sample-data"></a></dt>
<dd><p>An optional file containing additional sample
metadata (dose, timepoint, etc.). This can be a data frame or a file path.
Metadata will be joined with the mutation data based on the sample column.
Required columns are <code>sample</code> and any additional columns you wish to
include.</p></dd>


<dt id="arg-sd-sep">sd_sep<a class="anchor" aria-label="anchor" href="#arg-sd-sep"></a></dt>
<dd><p>The delimiter for importing sample metadata tables.
Default is tab-delimited.</p></dd>


<dt id="arg-regions">regions<a class="anchor" aria-label="anchor" href="#arg-regions"></a></dt>
<dd><p>An optional file containing metadata of genomic regions.
Region metadata will be joined with mutation data and variants will be
checked for overlap with the regions. <code>regions</code> can be either a file path,
a data frame, or a GRanges object. File paths will be read using the rg_sep.
Users can also choose from the built-in TwinStrand's Mutagenesis Panels by
inputting "TSpanel_human",  "TSpanel_mouse", or "TSpanel_rat". Required
columns for the regions file are "contig", "start", and "end". For a GRanges
object, the required columns are "seqnames", "start", and "end". Default is
NULL.</p></dd>


<dt id="arg-rg-sep">rg_sep<a class="anchor" aria-label="anchor" href="#arg-rg-sep"></a></dt>
<dd><p>The delimiter for importing the custom_regions. The default is
tab-delimited "\t".</p></dd>


<dt id="arg-is-based-rg">is_0_based_rg<a class="anchor" aria-label="anchor" href="#arg-is-based-rg"></a></dt>
<dd><p>A logical variable. Indicates whether the position
coordinates in <code>regions</code> are 0 based (TRUE) or 1 based (FALSE).
If TRUE, positions will be converted to 1-based (start + 1).
Need not be supplied for TSpanels. Default is TRUE.</p></dd>


<dt id="arg-padding">padding<a class="anchor" aria-label="anchor" href="#arg-padding"></a></dt>
<dd><p>Extend the range of your regions
in both directions by the given amount. Ex. Structural variants and
indels may start outside of the regions. Adjust the <code>padding</code> to
include these variants in your region's ranges.</p></dd>


<dt id="arg-genome">genome<a class="anchor" aria-label="anchor" href="#arg-genome"></a></dt>
<dd><p>The genome assembly version of the reference genome. This is
required if your data does not include a context column. The
function will install a BS genome for the given species/genome/masked
arguments to populate the context column.
Ex.Human GRCh38 = hg38 | Human GRCh37 = hg19 | Mouse GRCm38 = mm10 |
Mouse GRCm39 = mm39 | Rat RGSC 6.0 = rn6 | Rat mRatBN7.2 = rn7</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>The species. Required if your data does not include a
context column. The function will install a BS genome for the given
species/genome/masked to populate the context column. The species can
be the common name of the species or the scientific name.
Ex. "human" or "Homo sapiens".</p></dd>


<dt id="arg-masked-bs-genome">masked_BS_genome<a class="anchor" aria-label="anchor" href="#arg-masked-bs-genome"></a></dt>
<dd><p>A logical value. Required when using a BS genome
to poulate the context column. Whether to use the masked version of the
BS genome (TRUE) or not (FALSE). Default is FALSE.</p></dd>


<dt id="arg-output-granges">output_granges<a class="anchor" aria-label="anchor" href="#arg-output-granges"></a></dt>
<dd><p><code>TRUE</code> or <code>FALSE</code>; whether you want the mutation
data to output as a GRanges object. Default output is as a dataframe.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A table where each row is a mutation, and columns indicate the
location, type, and other data. If <code>output_granges</code> is set to TRUE, the
mutation data will be returned as a GRanges object, otherwise mutation
data is returned as a dataframe.</p>
<p>Output Column Definitions:</p><ul><li><p><code>short_ref</code>: The reference base at the start position.</p></li>
<li><p><code>normalized_ref</code>: The short_ref in C/T-base notation for
this position (e.g. A -&gt; T, G -&gt; C).</p></li>
<li><p><code>context</code> The trinucleotide context at this position. Consists
of the reference base and the two flanking bases (e.g. TAC).</p></li>
<li><p><code>normalized_context</code>: The trinucleotide context in C/T base
notation for this position (e.g. TAG -&gt; CTA).</p></li>
<li><p><code>variation_type</code> The type of variant (snv, mnv, insertion,
deletion, complex, sv, no_variant, ambiguous, uncategorized).</p></li>
<li><p><code>subtype</code> The substitution type for the snv variant (12-base spectrum;
e.g. A&gt;C).</p></li>
<li><p><code>normalized_subtype</code> The C/T-based substitution type for the snv
variant (6-base spectrum; e.g. A&gt;C -&gt; T&gt;G).</p></li>
<li><p><code>context_with_mutation</code>: The substitution type for the snv variant
including the two flanking nucleotides (192-trinucleotide spectrum;
e.g. <code>T[A&gt;C]G</code>)</p></li>
<li><p><code>normalized_context_with_mutation</code>: The C/T-based substitution
type for the snv variant including the two flanking nucleotides
(96-base spectrum e.g. <code>T[A&gt;C]G</code> -&gt; <code>C[T&gt;G]A</code>).</p></li>
<li><p><code>nchar_ref</code>: The length (in bp) of the reference allele.</p></li>
<li><p><code>nchar_alt</code>: The length (in bp) of the alternate allele.</p></li>
<li><p><code>varlen</code>: The length (in bp) of the variant.</p></li>
<li><p><code>ref_depth</code>: The depth of the reference allele. Calculated as
<code>total_depth</code> - <code>alt_depth</code>, if applicable.</p></li>
<li><p><code>vaf</code> : The variant allele fraction. Calculated as
<code>alt_depth</code>/<code>total_depth</code>.</p></li>
<li><p><code>gc_content</code>: % GC of the trinucleotide context at this position.</p></li>
<li><p><code>is_known</code>: TRUE or FALSE. Flags known variants (ID != ".").</p></li>
<li><p><code>row_has_duplicate</code>: TRUE or FALSE. Flags rows whose position is
the same as that of at least one other row for the same sample.</p></li>
<li><p><code>filter_mut</code> : A logical value, initially set to FALSE that indicates
to calculte_mf() if the variant should be excluded from mutation counts.
See the filter_mut function for more detail.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The required fields are:</p>
<p><strong>FIXED FIELDS</strong></p><ul><li><p><code>CHROM</code>: The name of the reference sequence. Equivalent to <code>contig</code>.</p></li>
<li><p><code>POS</code>: The 1-based start position of the feature. Equivalent to  <code>start</code>.</p></li>
<li><p><code>REF</code>: The reference allele at this position.</p></li>
<li><p><code>ALT</code>: The left-aligned, normalized, alternate allele at this position.
Multiple alt alleles called for a single position should be represented as
separate rows in the table.</p></li>
</ul><p><strong>INFO FIELDS</strong></p><ul><li><p><code>END</code>: The half-open end position of the feature.</p></li>
<li><p><code>sample</code>: An identifying field for your samples; either in the INFO
field or as the header to the FORMAT field.</p></li>
</ul><p><strong>SUGGESTED FIELDS</strong></p>
<p>The following <strong>FORMAT</strong> fields are not required, but are recommended for
full package functionality:</p><ul><li><p><code>AD</code>: The allelic depths for the reference and alternate allele
in the order listed. The sum of AD is equivalent to the <code>total_depth</code>
(read depth at this position excluding N-calls).</p></li>
<li><p><code>DP</code>: The read depth at this position (including N-calls).
Equivalent to <code>depth</code>. Note that in many VCF files, the DP field
is defined as <code>total_depth</code>. However, in most cases, the DP field
includes N-calls.</p></li>
<li><p><code>VD</code>: The read depth supporting the alternate allele. If
not included, the function will add this column, assuming a value of 1.
Equivalent to <code>alt_depth</code>.</p></li>
</ul><p>We recommend that files include a record for every sequenced
position, regardless of whether a variant was called, along with the
<code>AD</code> for each record. This enables site-specific depth calculations
required for some downstream analyses. AD is used to calculate the
<code>total_depth</code> (the read depth excluding No-calls). If AD is not available,
the <code>DP</code> field will be used as the <code>total_depth</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example: Import a single bg-zipped vcf file. This library was sequenced</span></span></span>
<span class="r-in"><span><span class="co"># with Duplex Sequencing using the TwinStrand Mouse Mutagenesis Panel which</span></span></span>
<span class="r-in"><span><span class="co"># consists of 20 2.4kb targets = 48kb of sequence.</span></span></span>
<span class="r-in"><span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Example_files"</span>,</span></span>
<span class="r-in"><span>                            <span class="st">"example_import_vcf_data_cleaned.vcf.bgz"</span>,</span></span>
<span class="r-in"><span>                            package <span class="op">=</span> <span class="st">"MutSeqR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># We will create an example metadata table for this data.</span></span></span>
<span class="r-in"><span><span class="va">sample_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample <span class="op">=</span> <span class="st">"dna00996.1"</span>,</span></span>
<span class="r-in"><span>                          dose <span class="op">=</span> <span class="st">"50"</span>,</span></span>
<span class="r-in"><span>                          dose_group <span class="op">=</span> <span class="st">"High"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Import the data</span></span></span>
<span class="r-in"><span><span class="va">imported_example_data</span> <span class="op">&lt;-</span> <span class="fu">import_vcf_data</span><span class="op">(</span>vcf_file <span class="op">=</span> <span class="va">example_file</span>,</span></span>
<span class="r-in"><span>                                         sample_data <span class="op">=</span> <span class="va">sample_meta</span>,</span></span>
<span class="r-in"><span>                                         regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span>,</span></span>
<span class="r-in"><span>                                         genome <span class="op">=</span> <span class="st">"mm10"</span>,</span></span>
<span class="r-in"><span>                                         species <span class="op">=</span> <span class="st">"mouse"</span>,</span></span>
<span class="r-in"><span>                                         masked_BS_genome <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample)`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Expected 'alt' but found 'alt.value', renaming it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Expected 'alt_depth' but found 'VD', renaming it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Expected 'depth' but found 'DP', renaming it.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'getOption("repos")' replaces Bioconductor standard repositories, see</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 'help("repositories", package = "BiocManager")' for details.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Replacement repositories:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     CRAN: https://cran.rstudio.com</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Reference genome already installed.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading reference genome: BSgenome.Mmusculus.UCSC.mm10.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Retrieving context sequences from the reference genome</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>150 rows were found whose position was the same as that of at least one other row for the same sample.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The total_depth may be double-counted in some instances due to overlapping positions. Set the correct_depth parameter in calculate_mf() to correct the total_depth for these instances.</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Annette E. Dodge, Andrew Williams, Danielle P.M. LeBlanc, David M. Schuster, Elena Esina, Clint C. Valentine, Jesse J. Salk, Alexander Y. Maslov, Christopher Bradley, Carole L. Yauk, Francesco Marchetti, Matthew J. Meier.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

