<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Create a heatmap plot of mutation subtype proportions. — plot_trinucleotide_heatmap • MutSeqR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Create a heatmap plot of mutation subtype proportions. — plot_trinucleotide_heatmap"><meta name="description" content="This function creates a heatmap plot of subtype proportions for
a given grouping variable. The groups may be facetted by a second variable.
Mutation sums for each facet group and normalized subtype are calculated
and displayed."><meta property="og:description" content="This function creates a heatmap plot of subtype proportions for
a given grouping variable. The groups may be facetted by a second variable.
Mutation sums for each facet group and normalized subtype are calculated
and displayed."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MutSeqR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/MutSeqR_introduction.html">MutSeqR: Error-Corrected Sequencing (ECS) Analysis For Mutagenicity Assessment</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Create a heatmap plot of mutation subtype proportions.</h1>
      <small class="dont-index">Source: <a href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/blob/main/R/plot_trinucleotide_heatmap.R" class="external-link"><code>R/plot_trinucleotide_heatmap.R</code></a></small>
      <div class="d-none name"><code>plot_trinucleotide_heatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function creates a heatmap plot of subtype proportions for
a given grouping variable. The groups may be facetted by a second variable.
Mutation sums for each facet group and normalized subtype are calculated
and displayed.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_trinucleotide_heatmap</span><span class="op">(</span></span>
<span>  <span class="va">mf_data</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  facet_col <span class="op">=</span> <span class="st">"dose"</span>,</span>
<span>  mf_type <span class="op">=</span> <span class="st">"min"</span>,</span>
<span>  mut_proportion_scale <span class="op">=</span> <span class="st">"turbo"</span>,</span>
<span>  max <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  rescale_data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  condensed <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-mf-data">mf_data<a class="anchor" aria-label="anchor" href="#arg-mf-data"></a></dt>
<dd><p>A data frame containing the mutation frequency data at the
desired base resolution. This is obtained using the 'calculate_mf' with
subtype_resolution set to the desired resolution. cols_to_group
should be the same as 'group_col'.</p></dd>


<dt id="arg-group-col">group_col<a class="anchor" aria-label="anchor" href="#arg-group-col"></a></dt>
<dd><p>The variable to group by.</p></dd>


<dt id="arg-facet-col">facet_col<a class="anchor" aria-label="anchor" href="#arg-facet-col"></a></dt>
<dd><p>The variable to facet by.</p></dd>


<dt id="arg-mf-type">mf_type<a class="anchor" aria-label="anchor" href="#arg-mf-type"></a></dt>
<dd><p>The type of mutation frequency to plot. Options are "min" or
"max". (Default: "min")</p></dd>


<dt id="arg-mut-proportion-scale">mut_proportion_scale<a class="anchor" aria-label="anchor" href="#arg-mut-proportion-scale"></a></dt>
<dd><p>The scale option for the mutation proportion.
Options are passed to viridis::scale_fill_viridis_c.
One of # inferno, magma, plasma, viridis, cividis, turbo, mako, or rocket.
We highly reccomend the default for its ability to disciminate hard to see
patterns. (Default: "turbo")</p></dd>


<dt id="arg-max">max<a class="anchor" aria-label="anchor" href="#arg-max"></a></dt>
<dd><p>Maximum value used for plotting the proportions.
Proportions that are higher will have the maximum colour. (Default: 0.2)</p></dd>


<dt id="arg-rescale-data">rescale_data<a class="anchor" aria-label="anchor" href="#arg-rescale-data"></a></dt>
<dd><p>Logical value indicating whether to rescale the mutation
proportions to increase the dynamic range of colors shown on the plot.
(Default: TRUE)</p></dd>


<dt id="arg-condensed">condensed<a class="anchor" aria-label="anchor" href="#arg-condensed"></a></dt>
<dd><p>More condensed plotting format. Default = FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot object representing the heatmap plot.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Plot the trinucleotide proportions per sample, facetted by dose group.</span></span></span>
<span class="r-in"><span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Example_files"</span>,</span></span>
<span class="r-in"><span>                            <span class="st">"example_mutation_data_filtered.rds"</span>,</span></span>
<span class="r-in"><span>                            package <span class="op">=</span> <span class="st">"MutSeqR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">example_file</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># define dose_group order</span></span></span>
<span class="r-in"><span><span class="va">example_data</span><span class="op">$</span><span class="va">dose_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">example_data</span><span class="op">$</span><span class="va">dose_group</span>,</span></span>
<span class="r-in"><span>                                  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low"</span>,</span></span>
<span class="r-in"><span>                                             <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mf_96</span> <span class="op">&lt;-</span> <span class="fu"><a href="calculate_mf.html">calculate_mf</a></span><span class="op">(</span><span class="va">example_data</span>,</span></span>
<span class="r-in"><span>                      cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span></span>
<span class="r-in"><span>                      variant_types <span class="op">=</span> <span class="st">"snv"</span>,</span></span>
<span class="r-in"><span>                      subtype_resolution <span class="op">=</span> <span class="st">"base_96"</span>,</span></span>
<span class="r-in"><span>                      retain_metadata_cols <span class="op">=</span> <span class="st">"dose_group"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Performing internal depth correction to prevent double-counting...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Internal depth correction complete.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(normalized_context_with_mutation, sample)`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample, normalized_context)`</span>
<span class="r-in"><span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu">plot_trinucleotide_heatmap</span><span class="op">(</span><span class="va">mf_96</span>,</span></span>
<span class="r-in"><span>                                  group_col <span class="op">=</span> <span class="st">"sample"</span>,</span></span>
<span class="r-in"><span>                                  facet_col <span class="op">=</span> <span class="st">"dose_group"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Cutting off at maximum mutation proportion value of 0.2</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Matthew J. Meier, Annette E. Dodge, Andrew Williams, Elena Esina.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

