---
title: "Duplex Sequencing Signatures"
author: "Matthew J. Meier"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
    theme: spacelab
    code_download: yes
  pdf_document:
    toc: yes
---

```{r load_libraries, warning=F, echo=F}
# Script to analyze distances between two dose groups of a chemical exposure using trinucleotide mutational patterns
library(ggplot2)
theme_set(theme_bw())
library(tidyverse)
#library(spgs)
#library(MutationalPatterns)
#library(SomaticSignatures)
#library(deconstructSigs)
#library(phyloseq)
#library(vegan)
#library(IRanges)
#library(GenomicRanges)
library(fuzzyjoin)
#library(pamr)
#library(heatmap3)
#library(ggpubr)
set.seed(3135)


```

# Import Data

```{r import-data}

mutation_data <-
  import_mut_data(mut_file = "../../data/Jonatan_Mutations_in_blood_and_sperm_samples_221021_MM.txt",
                 rsids = T, grouping_variable = "tissue")

# generate frequency tables

trinucleotide_frequencies_global <- mutation_data %>%
   group_by(context, subtype) %>%
   summarise(depth_per_trinucleotide = sum(total_depth))
 
 trinucleotide_frequencies_dose_group <- mutation_data %>%
   group_by(context, subtype, tissue) %>%
   summarise(depth_per_trinucleotide = sum(total_depth))
 
 trinucleotide_frequencies_per_individual <- mutation_data %>%
   group_by(context, subtype, sample) %>%
   summarise(depth_per_trinucleotide = sum(total_depth))

```

## GC Content

```{r gc-content}

gc_content_data <- annotate_CpG_sites(get_regions("human"), mutation_data)
# TODO instead of manually specifying human or mouse, it should be a config parameter.
as.data.frame(gc_content_data) %>% dplyr::group_by(subtype) %>% tally()

```

## Split by genic and intergenic regions

```{r genic_intergenic}
#################################################################
# From this point on, we split the data into genic vs. intergenic
#################################################################

trinucleotide_frequencies_intergenic <- trinucleotide_frequencies_depth %>%
  filter(location_relative_to_genes=="intergenic")

trinucleotide_frequencies_genic <- trinucleotide_frequencies_depth %>%
  filter(location_relative_to_genes=="genic")

tables_to_analyze <- list()
tables_to_analyze[[1]] <- trinucleotide_frequencies_depth
tables_to_analyze[[2]] <- trinucleotide_frequencies_intergenic
tables_to_analyze[[3]] <- trinucleotide_frequencies_genic

```

## Convert to wide format

```{r convert_to_wide, eval=T}
convert_to_wide <- function(x) {
# Convert table above to wide format
trinucleotide_frequencies_wide <- x %>%
  dplyr::select(context_with_mutation, frequency, sample) %>%
  distinct() %>%
  pivot_wider(names_from = context_with_mutation, values_from = frequency)

# Revert to dataframe
trinucleotide_frequencies_wide <- as.data.frame(trinucleotide_frequencies_wide)
# Convert sample names to row names
row.names(trinucleotide_frequencies_wide) <- trinucleotide_frequencies_wide$sample
# Remove sample columns
trinucleotide_frequencies_wide <- trinucleotide_frequencies_wide %>% dplyr::select(-sample)
# Reorder columns according to COSMIC

actual_colnames <- colnames(signatures.cosmic)[colnames(signatures.cosmic) %in% colnames(trinucleotide_frequencies_wide)]
empty_matrix <- matrix(data = 0, nrow=nrow(trinucleotide_frequencies_wide), ncol=96)
colnames(empty_matrix) <- colnames(signatures.cosmic)
rownames(empty_matrix) <- rownames(trinucleotide_frequencies_wide)
for(i in actual_colnames) {
  #print(i)
  empty_matrix[,i] <- trinucleotide_frequencies_wide[,i]
  }
trinucleotide_frequencies_mut_matrix <- empty_matrix
# Remove extra sample
trinucleotide_frequencies_mut_matrix <- trinucleotide_frequencies_mut_matrix[1:24,]

# Set NA to 0
trinucleotide_frequencies_mut_matrix[is.na(trinucleotide_frequencies_mut_matrix)] <- 0

# Convert to proportion of row sums
trinucleotide_frequencies_proportions <- trinucleotide_frequencies_mut_matrix/rowSums(trinucleotide_frequencies_mut_matrix)
return(trinucleotide_frequencies_proportions)
}
```

## Code chunk to run main analysis


```{r main_analysis, eval=F, fig.height=5}

#################################
# Per mouse re-analysis as above#
#NMDS
#COSMIC
#################################

# Vegan package
dist <- vegdist(trinucleotide_frequencies_proportions, method="euclidean") # Could change distance here to others...
mds <- metaMDS(trinucleotide_frequencies_proportions, k=2)
stressplot(mds)
ordiplot(mds, type="n")
orditorp(mds, display="sites",col="red",air=0.01)
orditorp(mds, display="species",col="black",air=0.01)

# Vegan package
dist <- vegdist(trinucleotide_frequencies_proportions, method="binomial") # Could change distance here to others...
nmds <- metaMDS(trinucleotide_frequencies_proportions, k=2)
stressplot(nmds)
ordiplot(nmds, type="n")
orditorp(nmds, display="sites",col="red",air=0.01)
orditorp(nmds, display="species",col="black",air=0.01)

trinucleotide_frequencies_proportions.transposed <- t(trinucleotide_frequencies_proportions)
dist_transposed <- vegdist(trinucleotide_frequencies_proportions.transposed, method="euclidean") # Could change distance here to others...
df <- trinucleotide_frequencies_proportions.transposed
df <- df[rowSums(df>0),]
mds_transposed <- metaMDS(df, k=2) # Stress is nearly zero
stressplot(mds_transposed)
ordiplot(mds_transposed, type="n")
orditorp(mds_transposed, display="sites",cex=1.25,air=0.01)
orditorp(mds_transposed, display="species",col="red",air=0.01)

# Do more distance metrics in phyloseq

# Create phyloseq object
otu = otu_table(trinucleotide_frequencies_proportions, taxa_are_rows = F)
#tax = as.matrix(colnames(trinucleotide_frequencies_proportions))
tax <- x %>% dplyr::select(context_with_mutation,gc_content) %>% distinct()
tax <- as.matrix(tax)
rownames(tax) <- tax[,1]
tax = tax_table(tax)
sampledata_ps <- sample_data(sampledata)
# Put it together
ps <- phyloseq(otu, tax, sampledata_ps)

# Get distances metrics available
dist_methods <- unlist(distanceMethodList)
dist_methods <- dist_methods[-(1:3)]
dist_methods <- dist_methods[!dist_methods %in% c("ANY")]
print(dist_methods)

# Loop over various distance methods, ordinate, plot
plist <- vector("list", length(dist_methods))
for( i in dist_methods ){
  # Calculate distance matrix
  iDist <- phyloseq::distance(ps, method=i)
  # Calculate ordination
  iMDS  <- ordinate(ps, "MDS", distance=iDist)
  ## Make plot
  # Don't carry over previous plot (if error, p will be blank)
  p <- NULL
  # Create plot, store as temp variable, p
  p <- plot_ordination(ps, iMDS, type="samples", color="Dose")
  # Add title to each plot
  p <- p + ggtitle(paste("MDS using distance method ", i, sep=""))
  # Save the plot
  plist[[i]] = p
}

# Create dataframe from ordination plot data
df = plyr::ldply(plist, function(x) x$data)
names(df)[1] <- "distance"

# Plot as faceted ggplot
ggplot(df, aes(Axis.1, Axis.2, color=Dose)) +
  geom_point(size=3, alpha=0.5) +
  facet_wrap(~distance, scales="free", ncol = 6) +
  #theme(text = element_text(size=9)) +
  ggtitle("MDS on various distance metrics for Duplex Sequencing PRC Bone Marrow dataset")

# Binomial distance split NMDS plot
myord <- ordinate(ps, method="NMDS", distance="binomial")
plot_ordination(ps,
                myord,
                type="biplot",
                color="Dose",
                label="Group",
                title="Binomial distance, NMDS biplot")

ord_plot_data <- as.data.frame(myord$points) %>% 
  tibble::rownames_to_column(var="Sample") %>%
  dplyr::left_join(psmelt(ps))

gc_content_data <- ord_plot_data %>%
  dplyr::select(context_with_mutation,gc_content) %>%
  mutate(Reference=substr(context_with_mutation, 3, 3)) %>%
  distinct()

ord_plot_mutations <- as.data.frame(myord$species) %>%
  tibble::rownames_to_column(var="Mutation Type") %>%
  dplyr::left_join(gc_content_data, by=c("Mutation Type"="context_with_mutation"))


ord_plot_data$Group <- factor(ord_plot_data$Group, levels=c("C","L","M","H"))

#Looking for outliers
ggplot(ord_plot_mutations, aes(x=MDS1,y=MDS2)) +
  geom_point() +
  geom_point(data=ord_plot_data, aes(color=Group), size=4) +
  geom_text(data=ord_plot_mutations %>% dplyr::filter(!between(MDS1, -20, 20)),
            aes(label=`Mutation Type`),
            nudge_y=-1) +
  theme_bw()

#PRC
#A[T>C]T
#A[T>C]A
#G[T>C]T
#G[T>C]G
#G[T>C]A

#BaP
#C[C>A]C
#G[C>A]C
#C[C>A]A

#LABEL: CHANGE HERE for mut types of interest - IDed by NSC
mut_types_of_interest <- ord_plot_mutations %>%
              dplyr::filter(`Mutation Type` %in% c("G[T>C]A", "G[T>C]T", "G[T>C]G", "A[T>C]A", "A[T>C]T"))

ggplot(ord_plot_mutations, aes(x=MDS1,y=MDS2)) +
  geom_point(data=ord_plot_mutations, aes(shape=gc_content)) +
  geom_point(data=ord_plot_data, aes(color=Group), size=4) +
  geom_text(data=mut_types_of_interest,
            aes(label=`Mutation Type`),
            nudge_y=-2, nudge_x=1,
            size=3) +
  geom_point(data=mut_types_of_interest, shape=1, size=4) +
  theme_bw()

ggplot(ord_plot_mutations, aes(x=MDS1,y=MDS2)) +
  geom_point(data=ord_plot_mutations, aes(shape=gc_content)) +
  geom_point(data=ord_plot_data, aes(color=Group), size=4) +
  geom_text(data=mut_types_of_interest,
            aes(label=`Mutation Type`),
            nudge_y=-2, nudge_x=1,
            size=3) +
  geom_point(data=mut_types_of_interest, shape=1, size=4) +
  theme_bw() +
  facet_wrap(~gc_content)

ggplot(ord_plot_mutations, aes(x=MDS1,y=MDS2)) +
  geom_point(data=ord_plot_mutations, aes(shape=gc_content)) +
  geom_point(data=ord_plot_data, aes(color=Group), size=4) +
  geom_text(data=mut_types_of_interest,
            aes(label=`Mutation Type`),
            nudge_y=-2, nudge_x=1,
            size=3) +
  geom_point(data=mut_types_of_interest, shape=1, size=4) +
  theme_bw() +
  facet_grid(gc_content~Reference)


cbp1 <- c("#999999", "#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
cbp2 <- c("#000000", "#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

ord_plot_data <- ord_plot_data %>% dplyr::select(sample, MDS1, MDS2, Group, Dose, sample) %>% distinct()
#joined_nmds_data <- ord_plot_mutations %>% full_join(ord_plot_data)

nmdsplot <- ggplot(ord_plot_mutations, aes(x=MDS1,y=MDS2)) +
  geom_point(data=ord_plot_data, aes(color=Group), size=2) +
  scale_color_manual(values=cbp1, name = "Dose group") +
  #scale_color_brewer(palette="Dark2") +
  # geom_text(data=mut_types_of_interest,
  #           aes(label=`Mutation Type`),
  #           nudge_y=-2, nudge_x=1,
  #           size=3) +
  geom_point(data=ord_plot_mutations, aes(shape=gc_content), color="black") +
  geom_point(data=mut_types_of_interest, shape=1, size=4, color="red") +
  #theme_Publication() +
  theme_bw() +
  scale_shape(name = "Cs or Gs in\ntrinucleotide", labels=c("0","1","2","3")) +
  theme(panel.background = element_rect(colour = NA),
        panel.border = element_rect(colour = NA),
        plot.background = element_rect(colour = NA),
        legend.key = element_rect(colour = NA),
        legend.key.size= unit(0.2, "cm"),
        legend.position="right",
        legend.direction="vertical",
        strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
        strip.text = element_text(face="bold")) +
  facet_grid(~Reference) +
  coord_fixed(ylim = c(-20,20), xlim=c(-40,30))

  # theme(panel.background = element_rect(colour = NA),
  #       plot.background = element_rect(colour = NA),
  #       panel.border = element_rect(colour = NA),
  #       axis.title = element_text(face = "bold",size = rel(1)),
  #       axis.title.y = element_text(angle=90,vjust =2),
  #       axis.title.x = element_text(vjust = -0.2),
  #       axis.text = element_text(), 
  #       axis.line = element_line(colour="black"),
  #       axis.ticks = element_line(),
  #       panel.grid.major = element_line(colour="#f0f0f0"),
  #       panel.grid.minor = element_blank(),
  #       legend.key = element_rect(colour = NA),
  #       legend.key.size= unit(0.2, "cm"),
  #       legend.position="right",
  #       legend.direction="vertical",
  #       strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
  #       strip.text = element_text(face="bold")) +

print(nmdsplot)

# Manhattan distance, but with different ordination methods
dist = "binomial"
ord_meths = c("DCA", "CCA", "RDA",  "NMDS", "MDS", "PCoA")
plist = plyr::llply(as.list(ord_meths), function(i, ps, dist){
  ordi = ordinate(ps, method=i, distance=dist)
  plot_ordination(ps, ordi, "samples", color="Dose")
}, ps, dist)

names(plist) <- ord_meths

pdataframe = plyr::ldply(plist, function(x){
  df = x$data[, 1:2]
  colnames(df) = c("Axis_1", "Axis_2")
  return(cbind(df, x$data))
})

names(pdataframe)[1] = "method"

# Plot for binomial distance but various ordination methods
ggplot(pdataframe, aes(Axis_1, Axis_2, color=Dose)) +
  geom_point(size=4) +
  facet_wrap(~method, scales="free") +
  scale_colour_brewer(palette="Set1")

# Cosine similarity
pairwise_cosine_similarity <- MutationalPatterns::cos_sim_matrix(t(trinucleotide_frequencies_proportions),
                                                                 t(trinucleotide_frequencies_proportions))

pheatmap::pheatmap(t(pairwise_cosine_similarity),
                   cluster_rows=F,
                   annotation=sampledata %>% dplyr::select(Dose))

pheatmap::pheatmap(t(pairwise_cosine_similarity),
                   cluster_rows=F,
                   cluster_cols=F,
                   annotation=sampledata %>% dplyr::select(Dose))

# Cosine similarity between known signatures
signatures <- get_known_signatures(muttype = "snv", genome = "mm10")

pairwise_cosine_similarity_sigs <- MutationalPatterns::cos_sim_matrix(t(trinucleotide_frequencies_proportions),
                                                                 signatures)


pheatmap::pheatmap(t(pairwise_cosine_similarity_sigs),
                   annotation=sampledata %>% dplyr::select(Dose))

#sorted <- dendextend::sort_dist_mat(pairwise_cosine_similarity_sigs) # Doesn't look better
sorted <- dendsort::dendsort(hclust(as.dist(pairwise_cosine_similarity_sigs), method = "average"))
pheatmap::pheatmap(t(pairwise_cosine_similarity_sigs),
                   cluster_cols = sorted,
                   cluster_rows = sorted,
                   annotation=sampledata %>% dplyr::select(Dose))

pheatmap::pheatmap(t(pairwise_cosine_similarity_sigs),
                   cluster_cols=F,
                   annotation=sampledata %>% dplyr::select(Dose))

```

## Andrew's code to run shrunken centroid analysis

```{r shrunken_centroid, eval=F}
################################################################
# Shrunken Centroid Analysis
################################################################
#Choose Delta Value before you Knit
#x <- tables_to_analyze[[1]]
#tnp_all <- convert_to_wide(x)
#dist_all <- vegdist(tnp_all, method="binomial")
#trinucleotide_frequencies_proportions <- tnp_all
################################################################
# Training Set
################################################################
allDat <- as.matrix(t(trinucleotide_frequencies_proportions))
trainSet <- allDat[,sampledata$Group %in% c("C", "H")]
labels <- sampledata$Group[sampledata$Group %in% c("C", "H")]
p <- rownames(allDat)

################################################################
# Shrinkage Estimation
################################################################
train.data <- list(x = as.matrix(trainSet), y = as.factor(labels),
                   geneid=as.character(p), genenames=as.character(p))
my.train <- pamr.train(train.data)

################################################################
# 6-fold Cross Validation (CV) (6 fold because min class size is 6)
################################################################
results <- pamr.cv(my.train, train.data)
results

# Call: 
# threshold (delta)

pamr.cv(fit = my.train, data = train.data)
#Ex. BaP data
#   threshold nonzero errors
#1  0.000     96      1    
#2  0.146     75      1    
#3  0.292     61      1    
#4  0.438     50      1    
#5  0.585     38      1    
#6  0.731     30      1    
#7  0.877     26      1    
#8  1.023     23      1     
#9  1.169     23      1    
#10 1.315     21      1    
#11 1.461     15      1    
#12 1.607     13      1    
#13 1.754     10      1    
#14 1.900      9      1    
#15 2.046      8      1    
#16 2.192      6      1    
#17 2.338      5      1    
#18 2.484      5      1    
#19 2.630      5      1    
#20 2.777      5      1    
#21 2.923      5      1    
#22 3.069      3      1    
#23 3.215      3      1    
#24 3.361      3      1    
#25 3.507      3      1    
#26 3.653      3      1    
#27 3.799      3      2    
#28 3.946      2      4    
#29 4.092      1      4    
#30 4.238      0      4

################################################################
################################################################
# Choosing the largest delta with the lowest number of errors based on the CV ^Example
################################################################
################################################################
#Ex. BaP data
#cut <-3.653

#Value for PRC data
cut <- 2.850
#Obce you've chosen the delta value, you can knit the Rmd

pamr.confusion(my.train, threshold = cut)

############################################################
# Writing out the centroids
################################################################
gs <- pamr.listgenes(my.train, train.data,  threshold=cut)[,1]
flag <- p %in% gs
x <- trainSet[flag,]
pp <- p[flag]
td <- list(x = as.matrix(x), y = as.factor(labels), geneid=as.character(pp), genenames=as.character(pp))
#ISSUE: contrasts can be applied only to factors with 2 or more levels
out <- pamr.train(td)
classifier <- data.frame(ID = pp)
classifier <- cbind.data.frame(classifier, out$centroids)
names(classifier) <- c("ID", "Control.score", "PRC.score")
classifier$std.dev <- out$sd

write.table(classifier,
            file="PRC Shunken Centroid Signature.txt", row.name = FALSE, col.name = TRUE, quote = FALSE, sep="\t")

################################################################
# Let's have a look at where the L and M groups get classified
################################################################
testDat <- cbind.data.frame(rownames(allDat), allDat)
names(testDat)[1] <- "ID"
testDat <- merge(classifier, testDat)

################################################################
#Predictions
################################################################
probPRC <- rep(0, ncol(testDat)-4)

for(k in 1:length(probPRC)){
  #Gausian Linear Discriminant Analysis
  PRC <- sum(((testDat[,k+4] - testDat$PRC.score)/testDat$std.dev)^2) - 2*log(0.5)
  Control <- sum(((testDat[,k+4] - testDat$Control.score)/testDat$std.dev)^2) - 2*log(0.5)
  probPRC[k] <- 1/(1 + exp(PRC/2 - Control/2))
}             
probPRC
################################################################
#Heatmap
################################################################
heatDat <- as.matrix(testDat[,-c(1:4)])
rownames(heatDat) <- as.character(testDat$ID)
colnames(heatDat) <- as.character(sampledata$Group)
reOrder <- c(grep("C", sampledata$Group), grep("H", sampledata$Group),
             grep("L", sampledata$Group), grep("M", sampledata$Group))
heatDat <- heatDat[,reOrder]
probPRC <- probPRC[reOrder]

##############################################################
#Colours
my.col <- matrix(rep("white", 2*ncol(heatDat)), ncol = 2)
my.col[grep("C", colnames(heatDat)),1] <- "blue"
my.col[grep("H", colnames(heatDat)),1] <- "red"

flag <- probPRC > 0.9
my.col[flag,2] <- "red"
flag <- probPRC < 0.1
my.col[flag,2] <- "blue"

##############################################################
#Row center based on the mean of the controls
mn <- apply(heatDat[,1:6], 1, mean)
heatDat <- heatDat - mn

end.breaks <- c(0, 12, ncol(heatDat))
y <- NULL
my.col2 <- NULL
g <- NULL
col.blank <- rep(NA, nrow(heatDat))
pp <- NULL

for(k in 2:length(end.breaks)){
  y <- cbind(y, heatDat[, (end.breaks[k-1]+1):end.breaks[k]], col.blank)
  my.col2 <- rbind(my.col2, my.col[(end.breaks[k-1]+1):end.breaks[k],], c("white", "white"))
  g <- c(g, colnames(heatDat)[(end.breaks[k-1]+1):end.breaks[k]], " ")
}

y <- y[,-ncol(y)]
my.col2 <- my.col2[-nrow(my.col2),]
g <- g[-length(g)]
colnames(y) <- g

colnames(my.col) <- c("Class", "Prediction")
colnames(my.col2) <- c("Class", "Prediction")

NSCheatmap<-heatmap3(y, Rowv = NA, Colv = NA, scale = "none", margins = c(10, 10), main = "", cexRow=1, cexCol=1, ColSideColors = my.col2)
NSCheatmap
```

# Print results from main analysis

## All data

```{r}
x <- tables_to_analyze[[1]]
```

```{r combine_results}
tnp_all <- convert_to_wide(x)
dist_all <- vegdist(tnp_all, method="binomial")
trinucleotide_frequencies_proportions <- tnp_all
<<main_analysis>>
pdf("./NMDS_binomial_all.pdf", width=7.3, height=4.3)
print(nmdsplot)
dev.off()
ord_plot_mutations_list <- list()
ord_plot_mutations_list[[1]] <- ord_plot_mutations
<<shrunken_centroid>>
print(NSCheatmap)
```

## Intergenic only

```{r}
x <- as.data.frame(tables_to_analyze[[2]])
```

```{r combine_results2}
tnp_intergenic <- convert_to_wide(x)
dist_intergenic <- vegdist(tnp_intergenic, method="binomial")
trinucleotide_frequencies_proportions <- tnp_intergenic 
<<main_analysis>>
pdf("./NMDS_binomial_intergenic.pdf", width=7.3, height=4.3)
print(nmdsplot)
dev.off()
ord_plot_mutations_list[[2]] <- ord_plot_mutations
```

## Genic only

```{r}
x <- as.data.frame(tables_to_analyze[[3]])
```

```{r combine_results3}
tnp_genic <- convert_to_wide(x)
dist_genic <- vegdist(tnp_genic, method="binomial")
trinucleotide_frequencies_proportions <- tnp_genic
<<main_analysis>>
pdf("./NMDS_binomial_genic.pdf", width=7.3, height=4.3)
print(nmdsplot)
dev.off()
ord_plot_mutations_list[[3]] <- ord_plot_mutations
```


## Mantel statistic

```{r mantel}
mantel(dist_all, tnp_genic)
mantel(dist_all, tnp_intergenic)
mantel(tnp_intergenic, tnp_genic)
```

<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
