---
title: "Duplex Sequencing Signatures"
author: "Matthew J. Meier"
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
    code_folding: hide
    theme: spacelab
    code_download: yes
  pdf_document:
    toc: yes
---

```{r load_libraries, warning=F, echo=F}
# Script to analyze distances between two dose groups of a chemical exposure using trinucleotide mutational patterns
library(ggplot2)
theme_set(theme_bw())
library(tidyverse)
#library(spgs)
#library(MutationalPatterns)
#library(SomaticSignatures)
#library(deconstructSigs)
#library(phyloseq)
#library(vegan)
#library(IRanges)
#library(GenomicRanges)
library(fuzzyjoin)
#library(pamr)
#library(heatmap3)
#library(ggpubr)
set.seed(3135)

source(file.path(here::here(),"R","CpG_sites_analysis.R"))
source(file.path(here::here(),"R","process_mut_file.R"))
output_path <- file.path(here::here(),"output")
if (!dir.exists(output_path)) { dir.create(output_path)}
```

# Import Data

```{r import-data}
group_var <- "tissue"
dupseq_species <- "human"

mutation_data <-
  import_mut_data(mut_file = "../../data/Jonatan_Mutations_in_blood_and_sperm_samples_221021_MM.txt",
                 rsids = T, grouping_variable = group_var) %>%
  dplyr::mutate(sample_and_group = stringr::str_c(!!ensym(group_var), sample, sep = "_"))

```

```{r exploratory-spectra}

# Look at overall spectrum
as.data.frame(mutation_data) %>% dplyr::group_by(subtype) %>% tally()
as.data.frame(mutation_data) %>% dplyr::group_by(normalized_subtype) %>% tally()

# generate frequency tables
trinucleotide_frequencies_global <- mutation_data %>%
   group_by(normalized_context, normalized_subtype) %>%
   summarise(depth_per_trinucleotide = sum(total_depth))
 
 trinucleotide_frequencies_group <- mutation_data %>%
   group_by(normalized_context, normalized_subtype, !!sym(group_var)) %>%
   summarise(depth_per_trinucleotide = sum(total_depth))
 
 trinucleotide_frequencies_per_individual <- mutation_data %>%
   group_by(normalized_context, normalized_subtype, sample) %>%
   summarise(depth_per_trinucleotide = sum(total_depth))

```

## GC Content

```{r gc-content}
cpg_mutations <- get_CpG_mutations(get_region_seqs(dupseq_species), mutation_data)
cpg_sites <- get_CpG_regions(get_region_seqs(dupseq_species))
# Total number of CpG sites
nrow(as.data.frame(cpg_sites)) # could also use %>% count() %>% pull()
n_samples <- length(unique(as.data.frame(mutation_data)[["sample"]]))

# Mutations at CpG sites...
cpg_summary_subtype <- as.data.frame(cpg_mutations) %>%
  dplyr::group_by(subtype) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples))

# Mutations at CpG sites, broken down by trinucleotide mutation context
cpg_summary_context <- as.data.frame(cpg_mutations) %>%
  dplyr::group_by(context_with_mutation) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct_cpg = 100*n/(nrow(as.data.frame(cpg_sites))*n_samples))
cpg_tables <- list(CpGs_by_subtype = cpg_summary_subtype,
                   CpGs_by_trinucleotide = cpg_summary_context)
write_excel_from_list(list_of_tables = cpg_tables,
                      output_path = output_path,
                      workbook_name = "cpg_tables")

```


<div class="tocify-extend-page" data-unique="tocify-extend-page" style="height: 0;"></div>
