<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>MutSeqR: Error-Corrected Sequencing (ECS) Analysis For Mutagenicity Assessment • MutSeqR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="MutSeqR: Error-Corrected Sequencing (ECS) Analysis For Mutagenicity Assessment">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MutSeqR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item"><a class="nav-link" href="../articles/index.html">Articles</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<link href="MutSeqR_introduction_files/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
<script src="MutSeqR_introduction_files/htmlwidgets-1.6.4/htmlwidgets.js"></script><link href="MutSeqR_introduction_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="MutSeqR_introduction_files/datatables-binding-0.33/datatables.js"></script><link href="MutSeqR_introduction_files/dt-core-1.13.6/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="MutSeqR_introduction_files/dt-core-1.13.6/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="MutSeqR_introduction_files/dt-core-1.13.6/js/jquery.dataTables.min.js"></script><link href="MutSeqR_introduction_files/crosstalk-1.2.1/css/crosstalk.min.css" rel="stylesheet">
<script src="MutSeqR_introduction_files/crosstalk-1.2.1/js/crosstalk.min.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>MutSeqR: Error-Corrected Sequencing (ECS) Analysis For Mutagenicity Assessment</h1>
                        <h4 data-toc-skip class="author">Annette E.
Dodge</h4>
            <address class="author_afil">
      Environmental Health Science and Research Bureau, Health Canada,
Ottawa, ON,
Canada.<br><h4 data-toc-skip class="author">Matthew
J. Meier</h4>
                  <a class="author_email" href="mailto:#"></a><a href="mailto:matthew.meier@hc-sc.gc.ca" class="email">matthew.meier@hc-sc.gc.ca</a>
      </address>
                  
      
      <small class="dont-index">Source: <a href="https://github.com/EHSRB-BSRSE-Bioinformatics/MutSeqR/blob/main/vignettes/MutSeqR_introduction.rmd" class="external-link"><code>vignettes/MutSeqR_introduction.rmd</code></a></small>
      <div class="d-none name"><code>MutSeqR_introduction.rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="what-is-ecs">What is ECS?<a class="anchor" aria-label="anchor" href="#what-is-ecs"></a>
</h3>
<p>Error-corrected next-generation sequencing (ECS) uses various methods
to combine multiple independent raw sequence reads derived from an
original starting molecule, thereby subtracting out artifacts introduced
during sequencing or library preparation. This results in a highly
accurate representation of the original molecule. ECS is particularly
useful for detecting rare somatic mutations (or mutations induced in
germ cells), such as those that arise from mutagen exposure or other
sources of DNA damage. ECS is a powerful tool for assessing the
mutagenicity of chemicals, drugs, or other agents, and can be used to
identify the mutational signatures of these agents. ECS can also be used
to detect rare mutations in cancer or other diseases, and to track the
clonal evolution of these diseases over time.</p>
<p>For more background on how ECS works and its context in regulatory
toxicology testing and genetic toxicology, see the following
articles:</p>
<ul>
<li><span class="citation">Menon and Brash (<a href="#ref-menon-2023">2023</a>)</span></li>
<li><span class="citation">Marchetti, Cardoso, Chen, Douglas, Elloway,
Escobar, Harper, et al. (<a href="#ref-marchetti-2023a">2023</a>)</span></li>
<li><span class="citation">Marchetti, Cardoso, Chen, Douglas, Elloway,
Escobar, Harper Jr, et al. (<a href="#ref-marchetti-2023b">2023</a>)</span></li>
<li><span class="citation">Kennedy et al. (<a href="#ref-kennedy-2014">2014</a>)</span></li>
</ul>
<p>This R package is meant to facilitate the import, cleaning, and
analysis of ECS data, beginning with a table of variant calls or a
variant call file (VCF). The package is designed to be flexible and
enable users to perform common statistical analyses and
visualisations.</p>
</div>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the package from github:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span></span>
<span>  <span class="st">"EHSRB-BSRSE-Bioinformatics/MutSeqR"</span>,</span>
<span>  auth_token <span class="op">=</span> <span class="st">"your personal_access_token from GitHub"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Install the package from Bioconductor (currently not
implemented):</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"MutSeqR"</span><span class="op">)</span></span></code></pre></div>
<p>Load the package</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ehsrb-bsrse-bioinformatics.github.io/MutSeqR/">MutSeqR</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-import">Data import<a class="anchor" aria-label="anchor" href="#data-import"></a>
</h2>
<p>The main goal of MutSeqR is to generate summary statistics,
visualizations, exploratory analyses, and other post-processing tasks
such as mutational signature analysis or generalized linear modeling.
Mutation data should be supplied as a table of variants with their
genomic positions. Mutation data can be imported as either VCF files or
as tabular data using the functions <code>import_vcf_data</code> and
<code>import_mut_data</code>, respectively. It is reccomended that files
include a record for every sequenced position, regardless of whether a
variant was called or not, along with the <code>total_depth</code> for
each record. This enables site-specific depth calculations that are
required for the calculation of mutation subtype frequencies ad other
site-specific frequemcies. The data set can be pared down later to
include only mutations of interest (SNVs, indels, SVs, or any
combination).</p>
<table class="table">
<caption>(#tab:required-columns) Required columns for mutation data
import.</caption>
<colgroup>
<col width="23%">
<col width="45%">
<col width="31%">
</colgroup>
<thead><tr class="header">
<th><strong>Column</strong></th>
<th><strong>VCF Specification</strong></th>
<th><strong>Definition</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>contig</td>
<td>CHROM</td>
<td>The name of the reference sequence.</td>
</tr>
<tr class="even">
<td>start</td>
<td>POS</td>
<td>The start position of the feature. 0-based coordinates are accepted
but will be changed to 1-based during import.</td>
</tr>
<tr class="odd">
<td>end</td>
<td>INFO(END)</td>
<td>The half-open end position of the feature in contig.</td>
</tr>
<tr class="even">
<td>ref</td>
<td>REF</td>
<td>The reference allele at this position.</td>
</tr>
<tr class="odd">
<td>alt</td>
<td>ALT</td>
<td>The left-aligned, normalized, alternate allele at this
position.</td>
</tr>
<tr class="even">
<td>sample</td>
<td>INFO(sample) or Genotype Header</td>
<td>A unique identifier for the sample library. For VCF files, this
field may be provided in either the INFO field, or as the header to the
GENOTYPE field.</td>
</tr>
<tr class="odd">
<td><em>SUGGESTED FIELDS</em></td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>alt_depth</td>
<td>VD</td>
<td>The read depth supporting the alternate allele. If not included, the
function will assume an alt_depth of 1 at variant sites.</td>
</tr>
<tr class="odd">
<td>total_depth or depth</td>
<td>AD or DP</td>
<td>The total read depth at this position. This column can be
“total_depth” which excludes N-calls, or “depth”, which includes
N-calls, if “total_depth” is not available. For VCF files, the
total_depth is calculated as the sum of AD. DP is equivalent to
“depth”.</td>
</tr>
</tbody>
</table>
<p>VCF files should follow the VCF specification (version 4.5; <span class="citation">Danecek et al. (<a href="#ref-danecek-2011">2011</a>)</span>). VCF files may be
bg/g-zipped. Each individual VCF file should contain the mutation data
for a single sample library. Multiple alt alleles called for a single
position should be represented as separate rows in the data. All extra
columns, INFO fields, and FORMAT fields will be retained upon
import.</p>
<p>Upon import, records are categorized within the
<code>variation_type</code> column based on their REF and ALT.
Categories are listed below.</p>
<table class="table">
<caption>(#tab:variation-types) Definitions of Variation
types.</caption>
<colgroup>
<col width="60%">
<col width="40%">
</colgroup>
<thead><tr class="header">
<th>variation_type</th>
<th>Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>no_variant</td>
<td>No variation, the null-case.</td>
</tr>
<tr class="even">
<td>snv</td>
<td>Single nucleotide variant.</td>
</tr>
<tr class="odd">
<td>mnv</td>
<td>Multiple nucleotide variant.</td>
</tr>
<tr class="even">
<td>insertion</td>
<td>Insertion.</td>
</tr>
<tr class="odd">
<td>deletion</td>
<td>Deletion.</td>
</tr>
<tr class="even">
<td>complex</td>
<td>REF and ALT are of different lengths and nucleotide
compositions.</td>
</tr>
<tr class="odd">
<td>symbolic</td>
<td>Structural variant</td>
</tr>
<tr class="even">
<td>ambiguous</td>
<td>ALT contains IUPAC ambiguity codes.</td>
</tr>
<tr class="odd">
<td>uncategorized</td>
<td>The record does not fall into any of the preceding categories.</td>
</tr>
</tbody>
</table>
<p>Additional columns are created to further characterise variants.</p>
<table class="table">
<caption>(#tab:mut-columns) Definitions of Mutation data
columns.</caption>
<colgroup>
<col width="28%">
<col width="71%">
</colgroup>
<thead><tr class="header">
<th>Column Name</th>
<th>Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>short_ref</td>
<td>The reference base at the start position.</td>
</tr>
<tr class="even">
<td>normalized_ref</td>
<td>The short_ref in C/T (pyrimidine) notation for this position. Ex.
<code>A</code> -&gt; <code>T</code>, <code>G</code> -&gt;
<code>C</code>
</td>
</tr>
<tr class="odd">
<td>context</td>
<td>The trinucleotide context at this position. Consists of the
reference base and the two flanking bases. Sequences are retrieved from
the appropriate BS genome. Ex. <code>TAC</code>
</td>
</tr>
<tr class="even">
<td>normalized_context</td>
<td>The trinucleotide context in C/T (pyrimidine) notation for this
position (Ex. <code>TAG</code> -&gt; <code>CTA</code>)</td>
</tr>
<tr class="odd">
<td>variation_type</td>
<td>The type of variant (no_variant, snv, mnv, insertion, deletion,
complex, sv, ambiguous, uncategorized)</td>
</tr>
<tr class="even">
<td>subtype</td>
<td>The substitution type of the snv variant (12-base spectrum; Ex.
<code>A&gt;C</code>)</td>
</tr>
<tr class="odd">
<td>normalized_subtype</td>
<td>The snv subtype in C/T (pyrimidine) notation (6-base spectrum; Ex.
<code>A&gt;C</code> -&gt; <code>T&gt;G</code>)</td>
</tr>
<tr class="even">
<td>context_with_mutation</td>
<td>The snv subtype including the two flanking nucleotides (192-base
spectrum; Ex. <code>T[A&gt;C]G</code>)</td>
</tr>
<tr class="odd">
<td>normalized_context_with_mutation</td>
<td>The snv subtype in C/T (pyrimidine) notation including the two
flanking nucleotides (96-base spectrum; Ex. <code>T[A&gt;C]G</code>
-&gt; <code>C[T&gt;G]A</code>)</td>
</tr>
<tr class="even">
<td>nchar_ref</td>
<td>The length (in bp) of the reference allele.</td>
</tr>
<tr class="odd">
<td>nchar_alt</td>
<td>The length (in bp) of the alternate allele.</td>
</tr>
<tr class="even">
<td>varlen</td>
<td>The length (in bp) of the variant.</td>
</tr>
<tr class="odd">
<td>ref_depth</td>
<td>The depth of the reference allele. Calculated as
<code>total_depth</code> - <code>alt_depth</code>, if applicable.</td>
</tr>
<tr class="even">
<td>vaf</td>
<td>The variant allele fraction. Calculated as
<code>alt_depth</code>/<code>total_depth</code>
</td>
</tr>
<tr class="odd">
<td>gc_content</td>
<td>% GC of the trinucleotide context at this position.</td>
</tr>
<tr class="even">
<td>is_known</td>
<td>A logical value indicating if the record is a known variant; i.e. ID
field is not NULL.</td>
</tr>
<tr class="odd">
<td>row_has_duplicate</td>
<td>A logical value that flags rows whose position is the same as that
of at least one other row for the same sample.</td>
</tr>
</tbody>
</table>
<div class="section level3">
<h3 id="general-usage">General Usage<a class="anchor" aria-label="anchor" href="#general-usage"></a>
</h3>
<p>Indicate the filepath to your mutation data file(s) using the
<code>vcf_file</code>/ <code>mut_file</code> parameter. This can be
either a single file or a directory containing multiple files. Should
you provide a directory, then all files within will be bound into a
single data frame.</p>
<p>An important component of importing your data for proper use is to
assign each mutation to a biological sample, and also make sure that
some additional information about each sample is present (ex., a
chemical treatment, a dose, etc.). This is done by providing
<code>sample_data</code>. This parameter can take a data frame, or it
can read in a file if provided with a filepath. If using a filepath,
specify the proper delimiter using the <code>sd_sep</code> parameter.
Sample metadata will be joined with the mutation data using the “sample”
column to capture that information and associate it with each
mutation.</p>
<p>Specify the appropriate BS genome with which to populate the context
column by supplying the species, genome, and masked_BS_genome
parameters. The function will browse <a href="https://www.rdocumentation.org/packages/BSgenome/versions/1.40.1/topics/available.genomes" class="external-link">BSgenome::available.genomes</a>
for the appropriate reference genome and install the corresponding
package. Context information will be extracted from the installed
BSgenome object. BSgenome offers genomes with masked sequences. If you
wish to use the masked version of the genome, set
<code>masked_BS_genome</code> to <code>TRUE</code>.</p>
</div>
<div class="section level3">
<h3 class="tabset" id="examples">Examples<a class="anchor" aria-label="anchor" href="#examples"></a>
</h3>
<p>We provide an example data set taken from <span class="citation">LeBlanc et al. (<a href="#ref-leblanc-2022">2022</a>)</span>. This data consists of 24
mouse bone marrow samples sequenced with Duplex Sequencing using
Twinstrand’s Mouse Mutagenesis Panel of twenty 2.4kb targeted genomic
loci. Mice were exposed to three doses of benzo[a]pyrene (BaP) alongside
vehicle controls, n = 6.</p>
<div class="section level4">
<h4 id="vcf">VCF<a class="anchor" aria-label="anchor" href="#vcf"></a>
</h4>
<p><em>Example 1.1. Import the example .vcf.bgz file. Provided is the
genomic</em> <em>vcf.gz file for sample dna00996.1. It is comprised of a
record for all</em> <em>48K positions sequenced for the Mouse
Mutagenesis Panel with the alt_depth</em> <em>and the tota_depth values
for each record.</em></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>,</span>
<span>  <span class="st">"Example_files"</span>,</span>
<span>  <span class="st">"example_import_vcf_data_cleaned.vcf.bgz"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"MutSeqR"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  sample <span class="op">=</span> <span class="st">"dna00996.1"</span>,</span>
<span>  dose <span class="op">=</span> <span class="st">"50"</span>,</span>
<span>  dose_group <span class="op">=</span> <span class="st">"High"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Import the data</span></span>
<span><span class="va">imported_example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_vcf_data.html">import_vcf_data</a></span><span class="op">(</span></span>
<span>  vcf_file <span class="op">=</span> <span class="va">example_file</span>,</span>
<span>  sample_data <span class="op">=</span> <span class="va">sample_metadata</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"mm10"</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>  masked_BS_genome <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Data Table 1.</strong> Mutation data imported from VCF for
sample dna00996.1. Displays the first 6 rows.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ac96cb3ee4656e2e9ec3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ac96cb3ee4656e2e9ec3">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["chr1","chr1","chr1","chr1","chr1","chr1"],[69304218,69304219,69304220,69304221,69304222,69304223],[69304218,69304219,69304220,69304221,69304222,69304223],[1,1,1,1,1,1],["*","*","*","*","*","*"],["C","A","A","T","C","T"],[1,2,3,4,5,6],[null,null,null,null,null,null],["","","","","",""],[6014,4484,4350,6466,5430,6541],[null,null,null,null,null,null],[6661,6726,6790,6857,6919,7018],[0,0,0,0,0,0],["dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1"],["50","50","50","50","50","50"],["High","High","High","High","High","High"],["ACA","CAA","AAT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[1,1,1,1,1,1],[null,null,null,null,null,null],[null,null,null,null,null,null],["C","A","A","T","C","T"],["C","T","T","T","C","T"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["ACA","TTG","ATT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[0.3333333333333333,0.3333333333333333,0,0.3333333333333333,0.3333333333333333,0.3333333333333333],[false,false,false,false,false,false],[6014,4484,4350,6466,5430,6541],[false,false,false,false,false,false],[0,0,0,0,0,0],[6014,4484,4350,6466,5430,6541]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>contig<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>ref<\/th>\n      <th>alt.group<\/th>\n      <th>alt.group_name<\/th>\n      <th>alt<\/th>\n      <th>AD_1<\/th>\n      <th>AD_2<\/th>\n      <th>depth<\/th>\n      <th>alt_depth<\/th>\n      <th>sample<\/th>\n      <th>dose<\/th>\n      <th>dose_group<\/th>\n      <th>context<\/th>\n      <th>variation_type<\/th>\n      <th>nchar_ref<\/th>\n      <th>nchar_alt<\/th>\n      <th>varlen<\/th>\n      <th>short_ref<\/th>\n      <th>normalized_ref<\/th>\n      <th>subtype<\/th>\n      <th>normalized_subtype<\/th>\n      <th>normalized_context<\/th>\n      <th>context_with_mutation<\/th>\n      <th>normalized_context_with_mutation<\/th>\n      <th>gc_content<\/th>\n      <th>filter_mut<\/th>\n      <th>total_depth<\/th>\n      <th>row_has_duplicate<\/th>\n      <th>vaf<\/th>\n      <th>ref_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[2,3,4,7,10,11,12,13,19,20,21,29,31,33,34]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"contig","targets":1},{"name":"start","targets":2},{"name":"end","targets":3},{"name":"width","targets":4},{"name":"strand","targets":5},{"name":"ref","targets":6},{"name":"alt.group","targets":7},{"name":"alt.group_name","targets":8},{"name":"alt","targets":9},{"name":"AD_1","targets":10},{"name":"AD_2","targets":11},{"name":"depth","targets":12},{"name":"alt_depth","targets":13},{"name":"sample","targets":14},{"name":"dose","targets":15},{"name":"dose_group","targets":16},{"name":"context","targets":17},{"name":"variation_type","targets":18},{"name":"nchar_ref","targets":19},{"name":"nchar_alt","targets":20},{"name":"varlen","targets":21},{"name":"short_ref","targets":22},{"name":"normalized_ref","targets":23},{"name":"subtype","targets":24},{"name":"normalized_subtype","targets":25},{"name":"normalized_context","targets":26},{"name":"context_with_mutation","targets":27},{"name":"normalized_context_with_mutation","targets":28},{"name":"gc_content","targets":29},{"name":"filter_mut","targets":30},{"name":"total_depth","targets":31},{"name":"row_has_duplicate","targets":32},{"name":"vaf","targets":33},{"name":"ref_depth","targets":34}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level4">
<h4 id="tabular">Tabular<a class="anchor" aria-label="anchor" href="#tabular"></a>
</h4>
<p><em>Example 1.2. Import example tabular data. This is the equivalent
file to</em> <em>the example vcf file. It is stored as an .rds file. We
will load the dataframe</em> <em>and supply it to
<code>import_mut_data</code>. The mut_file parameter can accept
file</em> <em>paths or data frames as input.</em></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>,</span>
<span>  <span class="st">"Example_files"</span>,</span>
<span>  <span class="st">"example_import_mut_data.rds"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"MutSeqR"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">example_file</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  sample <span class="op">=</span> <span class="st">"dna00996.1"</span>,</span>
<span>  dose <span class="op">=</span> <span class="st">"50"</span>,</span>
<span>  dose_group <span class="op">=</span> <span class="st">"High"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Import the data</span></span>
<span><span class="va">imported_example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_mut_data.html">import_mut_data</a></span><span class="op">(</span></span>
<span>  mut_file <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  sample_data <span class="op">=</span> <span class="va">sample_metadata</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"mm10"</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>  masked_BS_genome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  is_0_based_mut <span class="op">=</span> <span class="cn">TRUE</span> <span class="co"># indicates that the genomic coordinates are 0-based.</span></span>
<span><span class="co"># Coordinates will be changed to 1-based upon import.</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Data Table 2.</strong> Mutation data imported from a
data.frame object for sample dna00996.1. Displays the first 6 rows.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-e5c8c404fe174e4c81bd" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-e5c8c404fe174e4c81bd">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["chr1","chr1","chr1","chr1","chr1","chr1"],[69304218,69304219,69304220,69304221,69304222,69304223],[69304218,69304219,69304220,69304221,69304222,69304223],[1,1,1,1,1,1],["*","*","*","*","*","*"],["dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1"],["C","A","A","T","C","T"],[".",".",".",".",".","."],[0,0,0,0,0,0],[6661,6726,6790,6857,6919,7018],[647,2242,2440,391,1489,477],["50","50","50","50","50","50"],["High","High","High","High","High","High"],["ACA","CAA","AAT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[1,1,1,1,1,1],[null,null,null,null,null,null],[null,null,null,null,null,null],["C","A","A","T","C","T"],["C","T","T","T","C","T"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["ACA","TTG","ATT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[0.3333333333333333,0.3333333333333333,0,0.3333333333333333,0.3333333333333333,0.3333333333333333],[false,false,false,false,false,false],[6014,4484,4350,6466,5430,6541],[false,false,false,false,false,false],[0,0,0,0,0,0],[6014,4484,4350,6466,5430,6541]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>contig<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>sample<\/th>\n      <th>ref<\/th>\n      <th>alt<\/th>\n      <th>alt_depth<\/th>\n      <th>depth<\/th>\n      <th>no_calls<\/th>\n      <th>dose<\/th>\n      <th>dose_group<\/th>\n      <th>context<\/th>\n      <th>variation_type<\/th>\n      <th>nchar_ref<\/th>\n      <th>nchar_alt<\/th>\n      <th>varlen<\/th>\n      <th>short_ref<\/th>\n      <th>normalized_ref<\/th>\n      <th>subtype<\/th>\n      <th>normalized_subtype<\/th>\n      <th>normalized_context<\/th>\n      <th>context_with_mutation<\/th>\n      <th>normalized_context_with_mutation<\/th>\n      <th>gc_content<\/th>\n      <th>filter_mut<\/th>\n      <th>total_depth<\/th>\n      <th>row_has_duplicate<\/th>\n      <th>vaf<\/th>\n      <th>ref_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[2,3,4,9,10,11,16,17,18,26,28,30,31]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"contig","targets":1},{"name":"start","targets":2},{"name":"end","targets":3},{"name":"width","targets":4},{"name":"strand","targets":5},{"name":"sample","targets":6},{"name":"ref","targets":7},{"name":"alt","targets":8},{"name":"alt_depth","targets":9},{"name":"depth","targets":10},{"name":"no_calls","targets":11},{"name":"dose","targets":12},{"name":"dose_group","targets":13},{"name":"context","targets":14},{"name":"variation_type","targets":15},{"name":"nchar_ref","targets":16},{"name":"nchar_alt","targets":17},{"name":"varlen","targets":18},{"name":"short_ref","targets":19},{"name":"normalized_ref","targets":20},{"name":"subtype","targets":21},{"name":"normalized_subtype","targets":22},{"name":"normalized_context","targets":23},{"name":"context_with_mutation","targets":24},{"name":"normalized_context_with_mutation","targets":25},{"name":"gc_content","targets":26},{"name":"filter_mut","targets":27},{"name":"total_depth","targets":28},{"name":"row_has_duplicate","targets":29},{"name":"vaf","targets":30},{"name":"ref_depth","targets":31}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div class="section level3">
<h3 id="region-metadata">Region Metadata<a class="anchor" aria-label="anchor" href="#region-metadata"></a>
</h3>
<p>Similar to sample metadata, you may supply a file containing the
metadata of genomic regions to the <code>regions</code> parameter. In
the file, the genomic regions are defined by the name of their reference
sequence plus their start and end coordinates. Additional columns are
added to describe features of the genomic regions (Ex. transcription
status, GC content, etc.). Region metadata will be joined with mutation
data by checking for overlap between the target region ranges and the
position of the record. The <code>regions</code> parameter can be either
a filepath, a data frame, or a GRanges object (see <em><a href="https://bioconductor.org/packages/3.21/GenomicRanges" class="external-link">GenomicRanges</a></em>).
Filepaths will be read using the <code>rg_sep</code>. Users can also
choose from the built-in TwinStrand DuplexSeq™ Mutagenesis Panels by
inputting “TSpanel_human”, “TSpanel_mouse”, or “TSpanel_rat”. Required
columns for the regions file are “contig”, “start”, and “end”. In a
GRanges object, the required columns are “seqnames”, “start”, and “end”.
Users must indicate whether the region coordinates are 0-based or
1-based with <code>is_0_based_rg</code>. Mutation data and region
coordinates will be converted to 1-based. If you do not wish to specify
regions, then set the <code>regions</code> parameter to NULL
(default).</p>
<p><em>Example 1.3. Add the metadata for TwinStrand’s Mouse Mutagenesis
panel to our</em> <em>example tabular file.</em></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">imported_example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_mut_data.html">import_mut_data</a></span><span class="op">(</span></span>
<span>  mut_file <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  sample_data <span class="op">=</span> <span class="va">sample_metadata</span>,</span>
<span>  genome <span class="op">=</span> <span class="st">"mm10"</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"mouse"</span>,</span>
<span>  masked_BS_genome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  is_0_based_mut <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Data Table 3.</strong> Mutation data imported from a
data.frame object for sample dna00996.1 with added metadata for the
Mouse Mutagenesis Panels sequencing targets. Displays the first 6 rows.
Added columns are ‘target_size’, ‘label’, ‘genic_context’,
‘region_GC_content’, ‘genome’, and ‘in_regions’.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-36aa3d2a04d42bbc2145" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-36aa3d2a04d42bbc2145">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["chr1","chr1","chr1","chr1","chr1","chr1"],[69304218,69304219,69304220,69304221,69304222,69304223],[69304218,69304219,69304220,69304221,69304222,69304223],[1,1,1,1,1,1],["*","*","*","*","*","*"],["dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1"],["C","A","A","T","C","T"],[".",".",".",".",".","."],[0,0,0,0,0,0],[6661,6726,6790,6857,6919,7018],[647,2242,2440,391,1489,477],["50","50","50","50","50","50"],["High","High","High","High","High","High"],[2400,2400,2400,2400,2400,2400],["chr1","chr1","chr1","chr1","chr1","chr1"],["intergenic","intergenic","intergenic","intergenic","intergenic","intergenic"],[37.3,37.3,37.3,37.3,37.3,37.3],["mm10","mm10","mm10","mm10","mm10","mm10"],[true,true,true,true,true,true],["ACA","CAA","AAT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[1,1,1,1,1,1],[null,null,null,null,null,null],[null,null,null,null,null,null],["C","A","A","T","C","T"],["C","T","T","T","C","T"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["ACA","TTG","ATT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[0.3333333333333333,0.3333333333333333,0,0.3333333333333333,0.3333333333333333,0.3333333333333333],[false,false,false,false,false,false],[6014,4484,4350,6466,5430,6541],[false,false,false,false,false,false],[0,0,0,0,0,0],[6014,4484,4350,6466,5430,6541]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>contig<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>sample<\/th>\n      <th>ref<\/th>\n      <th>alt<\/th>\n      <th>alt_depth<\/th>\n      <th>depth<\/th>\n      <th>no_calls<\/th>\n      <th>dose<\/th>\n      <th>dose_group<\/th>\n      <th>target_size<\/th>\n      <th>label<\/th>\n      <th>genic_context<\/th>\n      <th>region_GC_content<\/th>\n      <th>genome<\/th>\n      <th>in_regions<\/th>\n      <th>context<\/th>\n      <th>variation_type<\/th>\n      <th>nchar_ref<\/th>\n      <th>nchar_alt<\/th>\n      <th>varlen<\/th>\n      <th>short_ref<\/th>\n      <th>normalized_ref<\/th>\n      <th>subtype<\/th>\n      <th>normalized_subtype<\/th>\n      <th>normalized_context<\/th>\n      <th>context_with_mutation<\/th>\n      <th>normalized_context_with_mutation<\/th>\n      <th>gc_content<\/th>\n      <th>filter_mut<\/th>\n      <th>total_depth<\/th>\n      <th>row_has_duplicate<\/th>\n      <th>vaf<\/th>\n      <th>ref_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[2,3,4,9,10,11,14,17,22,23,24,32,34,36,37]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"contig","targets":1},{"name":"start","targets":2},{"name":"end","targets":3},{"name":"width","targets":4},{"name":"strand","targets":5},{"name":"sample","targets":6},{"name":"ref","targets":7},{"name":"alt","targets":8},{"name":"alt_depth","targets":9},{"name":"depth","targets":10},{"name":"no_calls","targets":11},{"name":"dose","targets":12},{"name":"dose_group","targets":13},{"name":"target_size","targets":14},{"name":"label","targets":15},{"name":"genic_context","targets":16},{"name":"region_GC_content","targets":17},{"name":"genome","targets":18},{"name":"in_regions","targets":19},{"name":"context","targets":20},{"name":"variation_type","targets":21},{"name":"nchar_ref","targets":22},{"name":"nchar_alt","targets":23},{"name":"varlen","targets":24},{"name":"short_ref","targets":25},{"name":"normalized_ref","targets":26},{"name":"subtype","targets":27},{"name":"normalized_subtype","targets":28},{"name":"normalized_context","targets":29},{"name":"context_with_mutation","targets":30},{"name":"normalized_context_with_mutation","targets":31},{"name":"gc_content","targets":32},{"name":"filter_mut","targets":33},{"name":"total_depth","targets":34},{"name":"row_has_duplicate","targets":35},{"name":"vaf","targets":36},{"name":"ref_depth","targets":37}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p>Users can load the example TwinStrand Mutagenesis Panels with
load_regions(). This function will output a GRanges object. Custom
panels may also be loaded with this function by providing their file
path to the regions parameter.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_example</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_regions_file.html">load_regions_file</a></span><span class="op">(</span><span class="st">"TSpanel_mouse"</span><span class="op">)</span></span>
<span><span class="va">region_example</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 20 ranges and 5 metadata columns:</span></span>
<span><span class="co">##        seqnames              ranges strand | target_size       label</span></span>
<span><span class="co">##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1   69304218-69306617      * |        2400        chr1</span></span>
<span><span class="co">##    [2]     chr1 155235939-155238338      * |        2400      chr1.2</span></span>
<span><span class="co">##    [3]     chr2   50833176-50835575      * |        2400        chr2</span></span>
<span><span class="co">##    [4]     chr3 109633161-109635560      * |        2400        chr3</span></span>
<span><span class="co">##    [5]     chr4   96825281-96827680      * |        2400        chr4</span></span>
<span><span class="co">##    ...      ...                 ...    ... .         ...         ...</span></span>
<span><span class="co">##   [16]    chr15   66779763-66782162      * |        2400       chr15</span></span>
<span><span class="co">##   [17]    chr16   72381581-72383980      * |        2400       chr16</span></span>
<span><span class="co">##   [18]    chr17   94009029-94011428      * |        2400       chr17</span></span>
<span><span class="co">##   [19]    chr18   81262079-81264478      * |        2400       chr18</span></span>
<span><span class="co">##   [20]    chr19     4618814-4621213      * |        2400       chr19</span></span>
<span><span class="co">##        genic_context region_GC_content      genome</span></span>
<span><span class="co">##          &lt;character&gt;         &lt;numeric&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]    intergenic              37.3        mm10</span></span>
<span><span class="co">##    [2]         genic              54.0        mm10</span></span>
<span><span class="co">##    [3]    intergenic              45.3        mm10</span></span>
<span><span class="co">##    [4]         genic              39.2        mm10</span></span>
<span><span class="co">##    [5]    intergenic              39.4        mm10</span></span>
<span><span class="co">##    ...           ...               ...         ...</span></span>
<span><span class="co">##   [16]         genic              44.0        mm10</span></span>
<span><span class="co">##   [17]    intergenic              38.3        mm10</span></span>
<span><span class="co">##   [18]    intergenic              35.2        mm10</span></span>
<span><span class="co">##   [19]    intergenic              47.3        mm10</span></span>
<span><span class="co">##   [20]         genic              56.1        mm10</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 19 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="custom-column-names">Custom Column Names<a class="anchor" aria-label="anchor" href="#custom-column-names"></a>
</h3>
<p>We recognize that column names may differ from those specified in
Table @ref(tab:required-columns). Therefore, we have implemented some
default column name synonyms. If your column name matches one of our
listed synonyms, it will automatically be changed to match our set
values. For example, your <code>contig</code> column may be named
<code>chr</code> or <code>chromosome</code>. After importing your data,
this synonymous column name will be changed to <code>contig</code>.
Column names are case-insensitive. A list of column name synonyms are
listed below.</p>
<table class="table">
<caption>(#tab:name-sym) Predefined column name synonyms. Synonyms will
be automatically changed to the default Column value upon
import.</caption>
<thead><tr class="header">
<th>Column</th>
<th>Synonyms</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>alt</td>
<td>alternate</td>
</tr>
<tr class="even">
<td>alt_depth</td>
<td>alt_read_depth, var_depth, variant_depth, vd</td>
</tr>
<tr class="odd">
<td>context</td>
<td>sequence_context, trinucleotide_context, flanking_sequence</td>
</tr>
<tr class="even">
<td>contig</td>
<td>chr, chromosome, seqnames</td>
</tr>
<tr class="odd">
<td>depth</td>
<td>dp</td>
</tr>
<tr class="even">
<td>end</td>
<td>end_position</td>
</tr>
<tr class="odd">
<td>no_calls</td>
<td>n_calls, n_depth, no_depth</td>
</tr>
<tr class="even">
<td>ref</td>
<td>reference, ref_value</td>
</tr>
<tr class="odd">
<td>sample</td>
<td>sample_name, sample_id</td>
</tr>
<tr class="even">
<td>start</td>
<td>pos, position</td>
</tr>
<tr class="odd">
<td>total_depth</td>
<td>informative_somatic_depth</td>
</tr>
</tbody>
</table>
<p>If your data contains a column that is synonymous to one of the
required columns, but the name is not included in our synonyms list,
your column name may be substituted using the
<code>custom_column_names</code> parameter. Provide this parameter with
a list of names to specify the meaning of column headers.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mut_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"Example_files"</span>,</span>
<span>  <span class="st">"example_import_mut_data_custom_col_names.txt"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"MutSeqR"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">imported_example_data_custom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/import_mut_data.html">import_mut_data</a></span><span class="op">(</span></span>
<span>  mut_file <span class="op">=</span> <span class="va">mut_data</span>,</span>
<span>  custom_column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>my_contig_name <span class="op">=</span> <span class="st">"contig"</span>,</span>
<span>                             my_sample_name <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Expected 'contig' but found 'my_contig_name', renaming it.</span></span></code></pre>
<pre><code><span><span class="co">## Expected 'sample' but found 'my_sample_name', renaming it.</span></span></code></pre>
<p><strong>Data Table 4.</strong> Mutation data imported from tabular
file with custom column names. Displays the first 6 rows. Custom column
names are changed to default values.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-febe03efa1a2d8d52a86" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-febe03efa1a2d8d52a86">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["chr1","chr1","chr1","chr1","chr1","chr1"],[69304218,69304219,69304220,69304221,69304222,69304223],[69304218,69304219,69304220,69304221,69304222,69304223],[1,1,1,1,1,1],["*","*","*","*","*","*"],["dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1"],["C","A","A","T","C","T"],[".",".",".",".",".","."],[0,0,0,0,0,0],[6661,6726,6790,6857,6919,7018],[647,2242,2440,391,1489,477],["ACA","CAA","AAT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[1,1,1,1,1,1],[null,null,null,null,null,null],[null,null,null,null,null,null],["C","A","A","T","C","T"],["C","T","T","T","C","T"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["ACA","TTG","ATT","ATC","TCT","CTT"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],["no_variant","no_variant","no_variant","no_variant","no_variant","no_variant"],[0.3333333333333333,0.3333333333333333,0,0.3333333333333333,0.3333333333333333,0.3333333333333333],[false,false,false,false,false,false],[6014,4484,4350,6466,5430,6541],[false,false,false,false,false,false],[0,0,0,0,0,0],[6014,4484,4350,6466,5430,6541]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>contig<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>sample<\/th>\n      <th>ref<\/th>\n      <th>alt<\/th>\n      <th>alt_depth<\/th>\n      <th>depth<\/th>\n      <th>no_calls<\/th>\n      <th>context<\/th>\n      <th>variation_type<\/th>\n      <th>nchar_ref<\/th>\n      <th>nchar_alt<\/th>\n      <th>varlen<\/th>\n      <th>short_ref<\/th>\n      <th>normalized_ref<\/th>\n      <th>subtype<\/th>\n      <th>normalized_subtype<\/th>\n      <th>normalized_context<\/th>\n      <th>context_with_mutation<\/th>\n      <th>normalized_context_with_mutation<\/th>\n      <th>gc_content<\/th>\n      <th>filter_mut<\/th>\n      <th>total_depth<\/th>\n      <th>row_has_duplicate<\/th>\n      <th>vaf<\/th>\n      <th>ref_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[2,3,4,9,10,11,14,15,16,24,26,28,29]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"contig","targets":1},{"name":"start","targets":2},{"name":"end","targets":3},{"name":"width","targets":4},{"name":"strand","targets":5},{"name":"sample","targets":6},{"name":"ref","targets":7},{"name":"alt","targets":8},{"name":"alt_depth","targets":9},{"name":"depth","targets":10},{"name":"no_calls","targets":11},{"name":"context","targets":12},{"name":"variation_type","targets":13},{"name":"nchar_ref","targets":14},{"name":"nchar_alt","targets":15},{"name":"varlen","targets":16},{"name":"short_ref","targets":17},{"name":"normalized_ref","targets":18},{"name":"subtype","targets":19},{"name":"normalized_subtype","targets":20},{"name":"normalized_context","targets":21},{"name":"context_with_mutation","targets":22},{"name":"normalized_context_with_mutation","targets":23},{"name":"gc_content","targets":24},{"name":"filter_mut","targets":25},{"name":"total_depth","targets":26},{"name":"row_has_duplicate","targets":27},{"name":"vaf","targets":28},{"name":"ref_depth","targets":29}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="output">Output<a class="anchor" aria-label="anchor" href="#output"></a>
</h3>
<p>Mutation data can be output as either a data frame or a
<em>GRanges</em> object (see <em><a href="https://bioconductor.org/packages/3.21/GenomicRanges" class="external-link">GenomicRanges</a></em>)
for downstream analysis. Use the <code>output_granges</code> parameter
to specify the output. <em>GRanges</em> may faciliate use in other
packages and makes doing genomic based analyses on the ranges
significantly easier. Most downstream analyses provided by MutSeqR will
use a data frame.</p>
</div>
</div>
<div class="section level2">
<h2 id="variant-filtering">Variant Filtering<a class="anchor" aria-label="anchor" href="#variant-filtering"></a>
</h2>
<p>Following data import, the mutation data may be filtered using the
<code><a href="../reference/filter_mut.html">filter_mut()</a></code> function. This function will flag variants
based on various parameters in the <code>filter_mut</code> column.
Variants flagged as TRUE in the <code>filter_mut</code> column will be
automatically excluded from downstream analyses such as
<code><a href="../reference/calculate_mf.html">calculate_mf()</a></code>, <code>plot_bubbles</code>, and
<code>signature_fitting</code>. When specified, this function may also
remove records from the mutation data.</p>
<p>Flagging the variants in the <code>filter_mut</code> column will not
remove them from the mutation data, however, they will be excluded from
mutation counts in all downstream analyses. Filtered variants are
retained in the data so that their <code>total_depth</code> values may
still be used in frequency calculations.</p>
<p>By default, all filtering parameters are disabled. Users should be
mindful of the filters that they use, ensuring first that they are
applicable to their data.</p>
<div class="section level3">
<h3 id="germline-variants">Germline Variants<a class="anchor" aria-label="anchor" href="#germline-variants"></a>
</h3>
<p>Germline variants may be identified and flagged for filtering by
setting the <code>vaf_cutoff</code> parameter. The variant allele
fraction (vaf) is the fraction of haploid genomes in the original sample
that harbor a specific mutation at a specific base-pair coordinate of
the reference genome. Specifically, is it calculated by dividing the
number of variant reads by the total sequencing depth at a specific base
pair coordinate (<code>alt_depth</code> / <code>total_depth</code>). In
a typical diploid cell, a homozygous germline variant will appear on
both alleles, in every cell. As such, we expect this variant to occur on
every read, giving us a vaf = 1. A heterozygous germline variant occurs
on one of the two alleles in every cell, as such we expect this variant
to occur on about half of the reads, giving a vaf = 0.5. Somatic
variants occur in only a small portion of the cells, thus we expect them
to appear in only a small percentage of the reads. Typical vaf values
for somatic variants are less than 0.01 - 0.1. Setting the
<code>vaf_cutoff</code> parameter will flag all variants that have a vaf
<strong>greater than this value as germline</strong> within the
<code>is_germline</code> column. It will also flag these variants in the
<code>filter_mut</code> so as to exclude them from downstream
analyses.</p>
<p>vaf germline filtering is only applicable to users whose data was
sequenced to a sufficient depth. High coverage/low depth sequencing
cannot be used to calculate the vaf, thus it is recommended to filter
germline mutations by contrasting against a database of known
polymorphisms, or by using conventional whole-genome sequencing to
identify germline variants for each sample.</p>
</div>
<div class="section level3">
<h3 id="quality-assurance">Quality Assurance<a class="anchor" aria-label="anchor" href="#quality-assurance"></a>
</h3>
<p>The <code><a href="../reference/filter_mut.html">filter_mut()</a></code> function offers some filtering options
to ensure the quality of the mutation data.</p>
<p><code>snv_in_germ_mnv</code> = TRUE will flag all snvs that overlap
with germline mnvs. Germline mnvs are defined as having a vaf &gt;
vaf_cutoff. These snvs are often artifacts from variant calling.
No-calls in reads supporting the germline mnv will create false
minor-haplotypes from the original mnv that can appear as sub-clonal
snvs, thus such variants are excluded from downstream analyses.</p>
<p><code>rm_abnormal_vaf</code> = TRUE This parameter identifies rows
with abnormal vaf values and <strong>removes</strong> them from the
mutation data. Abnormal vaf values are defined as being between
0.05-0.45 or between 0.55-0.95. In a typical diploid organism, we expect
variants to have a vaf ~0, 0.5, or 1, reflecting rare somatic variants,
heterozygous variants, and homozygous variants respectively. Users
should be aware of the ploidy of their model system when using this
filter. Non-diploid organisms may exhibit different vafs.</p>
<p><code>rm_filtered_mut_from_depth = TRUE</code> Variants flagged in
the <code>filter_mut</code> column will have their
<code>alt_depth</code> subtracted from the <code>total_depth</code>.
When TRUE, this parameter treats flagged variants as No-calls. This does
not apply to variants that were idenfied as germline variants.</p>
</div>
<div class="section level3">
<h3 id="custom-filtering">Custom Filtering<a class="anchor" aria-label="anchor" href="#custom-filtering"></a>
</h3>
<p>Users may use the <code><a href="../reference/filter_mut.html">filter_mut()</a></code> function to flag or
remove variants based on their own custom column. Any record that
contains the <code>custom_filter_val</code> value within the
<code>custom_filter_col</code> column of the mutation data will be
either flagged in the <code>filter_mut</code> column or, if specified by
the <code>custom_filter_rm</code> parameter, removed from the mutation
data.</p>
</div>
<div class="section level3">
<h3 id="filtering-by-regions">Filtering by Regions<a class="anchor" aria-label="anchor" href="#filtering-by-regions"></a>
</h3>
<p>Users may remove rows that are either within or outside of specified
genomic regions. Provide the region ranges to the <code>regions</code>
parameter. This may be provided as either a filepath, data frame, or a
GRanges object. <code>regions</code> must contain <code>contig</code>
(or <code>seqnames</code> for GRanges), <code>start</code>, and
<code>end</code>. The function will check whether each record falls
within the given regions. Users can define how this filter should be
used with <code>regions_filter</code>.
<code>region_filter = "remove_within"</code> will remove all rows whose
positions overlap with the provided regions.
<code>region_filter = "keep_within"</code> will remove all rows whose
positions are outside of the provided regions. By default, records that
are &gt; 1bp must start and end within the regions to count as being
within the region. <code>allow_half_overlap = TRUE</code> allow records
that only start or end within the regions but extend outside of them to
be counted as being within the region. Twinstrand Mutagenesis Panels may
be used by setting <code>regions</code> to one of “TSpanel_human”,
“TSpanel_mouse”, or “TSpanel_rat”.</p>
</div>
<div class="section level3">
<h3 id="output-1">Output<a class="anchor" aria-label="anchor" href="#output-1"></a>
</h3>
<p>The function will return the mutation data as a data.frame object
with rows removed or flagged depending on input parameters. Added
columns are ‘filter_mut’, ‘filter_reason’, ‘is_germline’, and
‘snv_mnv_overlaps’. If <code>return_filtered_rows = TRUE</code> The
function will return both the data.frame containing the processed
mutation data and a data.frame containing the rows that were
removed/flagged during the filtering process. The two dataframes will be
returned inside of a list, with names <code>mutation_data</code> and
<code>filtered_rows</code>. The function will also print the number of
mutations/rows that were filtered according to each parameter.</p>
</div>
<div class="section level3">
<h3 id="example">Example<a class="anchor" aria-label="anchor" href="#example"></a>
</h3>
<p><em>Example 2. The example file “example_mutation_data.rds” is the
output of</em> <em>import_mut_data() run on all 24 mouse libraries from
<span class="citation">LeBlanc et al. (<a href="#ref-leblanc-2022">2022</a>)</span>.</em></p>
<p><em>Filters used:</em></p>
<ul>
<li>Filter germline variants: vaf &lt; 0.01</li>
<li>Filter snvs overlapping with germline variants and have their
<code>alt_depth</code> removed from their <code>total_depth</code>.</li>
<li>Remove records outside of the TwinStrand Mouse Mutagenesis
Panel.</li>
<li>Filter variants that contain “EndRepairFillInArtifact” in the
“filter” column. Their <code>alt_depth</code> will be removed from their
<code>total_depth</code>.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the example data</span></span>
<span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"Example_files"</span>,</span>
<span>  <span class="st">"example_mutation_data.rds"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"MutSeqR"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">example_file</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Filter</span></span>
<span><span class="va">filtered_example_mutation_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_mut.html">filter_mut</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  vaf_cutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span>,</span>
<span>  regions_filter <span class="op">=</span> <span class="st">"keep_within"</span>,</span>
<span>  custom_filter_col <span class="op">=</span> <span class="st">"filter"</span>,</span>
<span>  custom_filter_val <span class="op">=</span> <span class="st">"EndRepairFillInArtifact"</span>,</span>
<span>  custom_filter_rm <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  snv_in_germ_mnv <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm_filtered_mut_from_depth <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  return_filtered_rows <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Flagging germline mutations...</span></span></code></pre>
<pre><code><span><span class="co">## Found 612 germline mutations.</span></span></code></pre>
<pre><code><span><span class="co">## Flagging SNVs overlapping with germline MNVs...</span></span></code></pre>
<pre><code><span><span class="co">## Found 20 SNVs overlapping with germline MNVs.</span></span></code></pre>
<pre><code><span><span class="co">## Applying custom filter...</span></span></code></pre>
<pre><code><span><span class="co">## Flagged 2021 rows with values in &lt;filter&gt; column that matched EndRepairFillInArtifact</span></span></code></pre>
<pre><code><span><span class="co">## Applying region filter...</span></span></code></pre>
<pre><code><span><span class="co">## Removed 22 rows based on regions.</span></span></code></pre>
<pre><code><span><span class="co">## Removing filtered mutations from the total_depth...</span></span></code></pre>
<pre><code><span><span class="co">## Filtering complete.</span></span></code></pre>
<p><strong>Data Table 5.</strong> Filtered Mutation data for all 24
samples. Displayed are 6 variants that were flagged during
filtering.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-1fb4450895fe099f74a1" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-1fb4450895fe099f74a1">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6"],["chr1","chr1","chr1","chr1","chr1","chr2"],[155236432,155236656,155236922,155237528,155237607,50833347],[155236432,155236656,155236922,155237528,155237607,50833347],[1,1,1,1,1,1],["*","*","*","*","*","*"],["dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1"],["A","G","G","A","G","T"],["T","C","T","C","A","A"],[3,1,1,1,1,5053],[14412,14855,15824,16149,15638,11276],[798,95,94,413,132,433],[13611,14759,15729,15735,15505,5790],[13611,14759,15729,15735,15505,10843],[".",".",".",".",".","."],["N","N","N","N","N","N"],["EndRepairFillInArtifact","EndRepairFillInArtifact,v2","EndRepairFillInArtifact,v2","EndRepairFillInArtifact,v2","EndRepairFillInArtifact,v2",""],["Control","Control","Control","Control","Control","Control"],[0,0,0,0,0,0],["C1","C1","C1","C1","C1","C1"],[2400,2400,2400,2400,2400,2400],["chr1.2","chr1.2","chr1.2","chr1.2","chr1.2","chr2"],["genic","genic","genic","genic","genic","intergenic"],[54,54,54,54,54,45.3],["mm10","mm10","mm10","mm10","mm10","mm10"],[true,true,true,true,true,true],["snv","snv","snv","snv","snv","snv"],["T&gt;A","C&gt;G","C&gt;A","T&gt;G","C&gt;T","T&gt;A"],["A&gt;T","G&gt;C","G&gt;T","A&gt;C","G&gt;A","T&gt;A"],["C[T&gt;A]C","C[C&gt;G]T","T[C&gt;A]A","G[T&gt;G]G","G[C&gt;T]A","T[T&gt;A]T"],["G[A&gt;T]G","A[G&gt;C]G","T[G&gt;T]A","C[A&gt;C]C","T[G&gt;A]C","T[T&gt;A]T"],["A","G","G","A","G","T"],["T","C","C","T","C","T"],["CTC","CCT","TCA","GTG","GCA","TTT"],["GAG","AGG","TGA","CAC","TGC","TTT"],[1,1,1,1,1,1],[1,1,1,1,1,1],[1,1,1,1,1,1],[0.0002203613926840018,6.775067750677507e-05,6.357279084551812e-05,6.354855109303508e-05,6.449116471043467e-05,0.4660149405146177],[0.6666666666666666,0.6666666666666666,0.3333333333333333,0.6666666666666666,0.6666666666666666,0],[false,false,false,false,false,false],[true,true,true,true,true,true],["EndRepairFillInArtifact","EndRepairFillInArtifact,v2","EndRepairFillInArtifact,v2","EndRepairFillInArtifact,v2","EndRepairFillInArtifact,v2","germline"],[false,false,false,false,false,true],[false,false,false,false,false,false]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>contig<\/th>\n      <th>start<\/th>\n      <th>end<\/th>\n      <th>width<\/th>\n      <th>strand<\/th>\n      <th>sample<\/th>\n      <th>ref<\/th>\n      <th>alt<\/th>\n      <th>alt_depth<\/th>\n      <th>depth<\/th>\n      <th>no_calls<\/th>\n      <th>ref_depth<\/th>\n      <th>total_depth<\/th>\n      <th>id<\/th>\n      <th>is_known<\/th>\n      <th>filter<\/th>\n      <th>dose_group<\/th>\n      <th>dose<\/th>\n      <th>new_sample_id<\/th>\n      <th>target_size<\/th>\n      <th>label<\/th>\n      <th>genic_context<\/th>\n      <th>region_GC_content<\/th>\n      <th>genome<\/th>\n      <th>in_regions<\/th>\n      <th>variation_type<\/th>\n      <th>normalized_subtype<\/th>\n      <th>subtype<\/th>\n      <th>normalized_context_with_mutation<\/th>\n      <th>context_with_mutation<\/th>\n      <th>short_ref<\/th>\n      <th>normalized_ref<\/th>\n      <th>normalized_context<\/th>\n      <th>context<\/th>\n      <th>nchar_ref<\/th>\n      <th>nchar_alt<\/th>\n      <th>varlen<\/th>\n      <th>vaf<\/th>\n      <th>gc_content<\/th>\n      <th>row_has_duplicate<\/th>\n      <th>filter_mut<\/th>\n      <th>filter_reason<\/th>\n      <th>is_germline<\/th>\n      <th>snv_in_germ_mnv<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"className":"dt-right","targets":[2,3,4,9,10,11,12,13,18,20,23,35,36,37,38,39]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"contig","targets":1},{"name":"start","targets":2},{"name":"end","targets":3},{"name":"width","targets":4},{"name":"strand","targets":5},{"name":"sample","targets":6},{"name":"ref","targets":7},{"name":"alt","targets":8},{"name":"alt_depth","targets":9},{"name":"depth","targets":10},{"name":"no_calls","targets":11},{"name":"ref_depth","targets":12},{"name":"total_depth","targets":13},{"name":"id","targets":14},{"name":"is_known","targets":15},{"name":"filter","targets":16},{"name":"dose_group","targets":17},{"name":"dose","targets":18},{"name":"new_sample_id","targets":19},{"name":"target_size","targets":20},{"name":"label","targets":21},{"name":"genic_context","targets":22},{"name":"region_GC_content","targets":23},{"name":"genome","targets":24},{"name":"in_regions","targets":25},{"name":"variation_type","targets":26},{"name":"normalized_subtype","targets":27},{"name":"subtype","targets":28},{"name":"normalized_context_with_mutation","targets":29},{"name":"context_with_mutation","targets":30},{"name":"short_ref","targets":31},{"name":"normalized_ref","targets":32},{"name":"normalized_context","targets":33},{"name":"context","targets":34},{"name":"nchar_ref","targets":35},{"name":"nchar_alt","targets":36},{"name":"varlen","targets":37},{"name":"vaf","targets":38},{"name":"gc_content","targets":39},{"name":"row_has_duplicate","targets":40},{"name":"filter_mut","targets":41},{"name":"filter_reason","targets":42},{"name":"is_germline","targets":43},{"name":"snv_in_germ_mnv","targets":44}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
<div class="section level2">
<h2 id="calculating-mf">Calculating MF<a class="anchor" aria-label="anchor" href="#calculating-mf"></a>
</h2>
<p>Mutation Frequency (MF) is calculated by dividing the sum of
mutations by the sum of the <code>total_depth</code> for a given group
(units mutations/bp). The function <code><a href="../reference/calculate_mf.html">calculate_mf()</a></code> sums
mutation counts and <code>total_depth</code> across user-defined
groupings within the mutation data to calculate the MF. Mutations can be
summarised across samples, experimental groups, and mutation subtypes
for later statistical analyses. Variants flagged as <code>TRUE</code> in
the <code>filter_mut</code> column will be excluded from the mutation
sums; however, the <em>total_depth</em> of these variants will be
counted in the <em>total_depth</em> sum.</p>
<p>If the total_depth is not available, the function will sum the
mutations across groups, without calculating the mutation
frequencies.</p>
<div class="section level3">
<h3 id="mutation-counting">Mutation Counting<a class="anchor" aria-label="anchor" href="#mutation-counting"></a>
</h3>
<p>Mutations are counted based on two opposing assumptions <span class="citation">(<a href="#ref-dodge-2023">Dodge et al.
2023</a>)</span>:</p>
<p>The <strong>Minimum Independent Mutation Counting Method</strong>
(min) Each mutation is counted once, regardless of the number of reads
that contain the non-reference allele. This method assumes that multiple
instances of the same mutation within a sample are the result of clonal
expansion of a single mutational event. When summing mutations across
groups using the Min method (<code>sum_min</code>), the
<code>alt_depth</code> of each variant is set to 1. <em>Ex. For 3
variants with</em> <em><code>alt_depth</code> values of 1, 2, and 10,
the <code>sum_min</code>= 3.</em></p>
<p>The <strong>Maximum Independent Mutation Counting Method</strong>
(max) Multiple identical mutations at the same position within a sample
are counted as independent mutation events. When summing mutations
across groups using the Max method (<code>sum_max</code>), the
<code>alt_depth</code> of each variant is summed unchanged. <em>Ex. for
3</em> <em>variants with <code>alt_depth</code> values of 1, 2, and 10,
the <code>sum_max</code> = 13.</em></p>
<p>The Min and Max mutation counting methods undercount and overcount
the mutations, respectively. We expect some recurrent mutations to be a
result of clonal expansion. We also expect some recurrent mutations to
arise independently of each other. As we expect the true number of
independent mutations to be somewhere in the middle of the two counting
methods, we calculate frequencies for both methods. However, the Min
mutation counting method is generally recommended for statistical
analyses to ensure independance of values and because the Max method
tends to increase the sample variance by a significant degree.</p>
</div>
<div class="section level3">
<h3 id="grouping-mutations">Grouping Mutations<a class="anchor" aria-label="anchor" href="#grouping-mutations"></a>
</h3>
<p>Mutation counts and <code>total_depth</code> are summed across groups
that can be designated using the <code>cols_to_group</code> parameter.
This parameter can be set to one or more columns in the mutation data
that represent experimental variables of interest.</p>
<p><em>Example 3.1. Calculate mutation sums and frequencies per
sample.</em> <em>The file “example_mutation_data_filtered.rds” is the
output of filter_mut()</em> <em>from Example 2</em></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load example data:</span></span>
<span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"extdata"</span>, <span class="st">"Example_files"</span>,</span>
<span>  <span class="st">"example_mutation_data_filtered.rds"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"MutSeqR"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">example_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">example_file</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mf_data_global</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  retain_metadata_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dose_group"</span>, <span class="st">"dose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 6.</strong> Global Mutation Frequency per Sample</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-10b3b7155e8045a1b2ad" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-10b3b7155e8045a1b2ad">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],["Control","Control","Control","Control","Control","Control","Low","Low","Low","Low","Low","Low","Medium","Medium","Medium","Medium","Medium","Medium","High","High","High","High","High","High"],[0,0,0,0,0,0,12.5,12.5,12.5,12.5,12.5,12.5,25,25,25,25,25,25,50,50,50,50,50,50],[76,112,138,123,97,124,212,230,248,273,228,221,230,160,379,471,318,467,681,723,760,770,659,686],[79,128,174,154,167,129,325,357,412,411,306,308,336,197,590,833,498,817,1516,1923,1835,1718,2078,2632],[1.344209454660627e-07,1.482316200007564e-07,2.156555910825569e-07,1.822000847769588e-07,1.621791470952174e-07,2.028508249233072e-07,3.268922378844466e-07,3.224762505985825e-07,3.702959207607104e-07,3.98573762918471e-07,3.180300522246773e-07,3.192194373518508e-07,7.726900431161044e-07,9.25584825420905e-07,5.637779512015613e-07,6.357123503490611e-07,5.69854191784267e-07,6.36475951881235e-07,9.682578640570764e-07,8.172871372514138e-07,1.022451113390202e-06,9.631359984291226e-07,9.72415634724636e-07,9.783737583203342e-07],[1.397270354186704e-07,1.694075657151502e-07,2.719135713649631e-07,2.281204313467614e-07,2.792156449989826e-07,2.110302936702147e-07,5.011319684549298e-07,5.005392237551912e-07,6.151690296508575e-07,6.000506101080277e-07,4.268298069331196e-07,4.448850077120816e-07,1.128799367334831e-06,1.139626316299489e-06,8.776490533216917e-07,1.124306555925197e-06,8.924131682659275e-07,1.113492189907857e-06,2.155475656256282e-06,2.173780311112682e-06,2.468681306672397e-06,2.148919019871731e-06,3.066281773835802e-06,3.753760542126997e-06],[565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>dose_group<\/th>\n      <th>dose<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":[6,7],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[3,4,5,6,7,8]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"dose_group","targets":2},{"name":"dose","targets":3},{"name":"sum_min","targets":4},{"name":"sum_max","targets":5},{"name":"mf_min","targets":6},{"name":"mf_max","targets":7},{"name":"group_depth","targets":8}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script><p>Alternatively, you can sum mutations by experimental groups such as
<code>label</code> (the genomic target). Counts and frequencies will be
returned for every level of the designated groups.</p>
<p><em>Example 3.2. Calculate mutation sums and frequencies per sample
and genomic target.</em></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_rg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample"</span>, <span class="st">"label"</span><span class="op">)</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  retain_metadata_cols <span class="op">=</span> <span class="st">"dose_group"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 7.</strong> Mutation Frequency per Sample and genomic
target.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-4018eef1a407a0df6b52" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-4018eef1a407a0df6b52">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312","313","314","315","316","317","318","319","320","321","322","323","324","325","326","327","328","329","330","331","332","333","334","335","336","337","338","339","340","341","342","343","344","345","346","347","348","349","350","351","352","353","354","355","356","357","358","359","360","361","362","363","364","365","366","367","368","369","370","371","372","373","374","375","376","377","378","379","380","381","382","383","384","385","386","387","388","389","390","391","392","393","394","395","396","397","398","399","400","401","402","403","404","405","406","407","408","409","410","411","412","413","414","415","416","417","418","419","420","421","422","423","424","425","426","427","428","429","430","431","432","433","434","435","436","437","438","439","440","441","442","443","444","445","446","447","448","449","450","451","452","453","454","455","456","457","458","459","460","461","462","463","464","465","466","467","468","469","470","471","472","473","474","475","476","477","478","479","480"],["dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00973.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00974.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00975.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00976.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00977.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00978.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00979.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00980.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00981.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00982.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00983.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00984.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00985.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00986.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00987.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00988.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00989.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00990.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00991.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00992.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00993.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00994.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00995.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1","dna00996.1"],["chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9","chr1","chr1.2","chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18","chr19","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9"],["Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Low","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","Medium","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High","High"],[0,1,4,2,6,4,5,3,12,2,3,3,6,5,2,1,4,5,3,5,2,3,4,4,12,7,9,5,3,3,2,6,23,1,3,5,3,2,6,9,9,10,9,10,10,6,4,8,2,6,6,1,9,4,9,6,7,7,6,9,5,10,8,12,3,6,9,12,9,4,4,2,3,9,5,8,4,2,3,5,4,1,6,9,7,1,7,3,6,7,5,3,1,5,7,5,6,3,3,8,1,6,10,8,5,3,8,12,8,6,5,6,1,5,8,6,8,6,6,6,4,10,10,20,14,5,12,9,13,11,11,14,9,7,8,7,11,14,16,7,5,8,3,28,12,11,24,5,20,9,10,7,10,6,13,19,8,8,14,10,11,10,14,25,17,15,16,7,15,12,15,10,8,3,8,6,10,14,19,13,9,16,10,22,13,5,20,12,12,24,14,18,24,2,12,6,9,17,12,16,6,9,12,20,10,10,22,9,16,12,15,4,7,6,7,12,15,10,16,10,7,9,9,23,8,8,16,15,13,12,12,5,8,8,10,9,8,12,14,15,7,9,12,23,13,3,19,4,16,14,7,13,8,1,15,13,11,15,17,10,2,6,12,11,14,2,17,3,9,8,11,5,18,3,5,4,8,10,8,4,11,20,12,29,21,11,27,12,25,19,37,11,13,11,20,7,13,25,33,22,18,17,18,59,22,15,47,15,32,20,34,20,17,10,19,22,14,29,26,17,5,14,13,35,13,16,38,12,18,24,22,9,8,8,16,11,9,9,23,15,16,21,20,41,24,16,49,18,22,23,21,25,29,9,18,20,18,35,29,13,19,40,28,46,47,22,63,23,35,31,44,33,36,15,34,34,25,34,44,28,24,31,31,62,53,28,60,15,49,43,46,25,17,19,34,24,37,43,49,33,24,25,49,49,44,25,68,21,38,40,66,32,27,12,28,37,35,50,39,51,35,25,29,63,48,30,64,24,36,41,49,28,36,15,38,40,36,38,55,40,25,31,33,63,44,15,58,10,39,49,46,22,29,8,32,24,29,31,39,32,35,33,31,50,53,24,53,16,44,36,32,31,42,9,36,25,38,32,37,29],[0,1,4,2,6,4,5,3,14,2,3,3,6,5,2,1,5,5,3,5,2,3,4,4,13,7,9,5,3,3,2,6,36,1,3,5,3,2,6,11,9,11,9,10,10,8,5,11,3,6,6,1,9,4,9,6,7,7,6,37,5,10,9,38,3,6,10,12,10,6,4,2,3,9,5,8,4,2,3,5,5,1,7,9,67,1,7,5,8,7,6,3,1,5,8,7,6,3,3,8,1,7,10,9,5,3,8,12,8,8,5,6,1,6,8,6,8,6,6,6,8,10,21,36,21,6,24,11,17,15,13,25,12,7,17,8,20,24,22,8,6,8,3,66,21,13,32,6,35,10,14,10,14,7,24,28,17,8,19,16,13,12,28,62,33,27,27,8,24,20,17,13,18,5,13,6,10,25,33,18,14,25,20,32,27,5,35,13,13,34,20,32,31,2,14,7,18,24,25,20,9,17,18,27,16,11,29,10,18,16,19,6,8,8,13,14,20,13,23,11,10,9,14,33,17,17,24,15,19,14,22,5,11,8,13,13,10,14,23,17,9,10,19,34,19,8,29,4,23,17,13,22,14,1,19,18,20,20,26,11,2,10,16,14,15,2,22,3,9,8,17,7,19,3,5,4,12,12,10,7,12,35,20,43,29,11,56,20,38,25,60,17,26,23,29,10,15,35,54,32,22,34,28,106,33,24,95,16,53,33,58,24,31,24,36,34,25,71,43,43,6,29,19,55,26,23,57,12,24,38,50,18,14,12,22,11,16,20,28,18,30,31,37,85,37,28,78,31,37,32,37,48,47,13,25,29,29,94,44,25,39,79,67,106,144,54,119,45,74,59,105,88,61,25,89,63,65,68,98,68,102,63,73,189,141,39,182,28,118,89,139,83,31,34,95,63,72,78,146,158,56,67,121,123,100,89,218,33,80,122,169,95,41,21,51,72,58,126,72,121,91,61,87,136,81,39,147,72,100,78,95,56,64,23,85,91,94,84,123,111,46,77,125,216,101,116,188,73,88,122,105,103,70,9,109,50,123,161,105,91,105,185,96,210,369,57,174,99,177,130,99,91,103,22,135,68,119,63,158,172],[0,2.83891056239386e-08,1.205246014063112e-07,8.420176933177897e-08,1.831810152661226e-07,1.195120335620863e-07,1.818890127068028e-07,9.464917007348467e-08,5.36882628598825e-07,9.12329263280435e-08,9.461935076743232e-08,8.960254729289649e-08,2.159414297741195e-07,1.814911347932843e-07,7.980926224796833e-08,4.04360706733556e-08,1.362484599666259e-07,1.614369177172174e-07,1.396871296981496e-07,1.953598671365358e-07,5.768270795856398e-08,6.628637876562699e-08,9.200502126604061e-08,1.235393632237646e-07,2.793762561891444e-07,1.590909452479421e-07,2.495165893322343e-07,1.155327523684157e-07,9.858803206358849e-08,1.017324390057228e-07,4.797356656482278e-08,1.337091898660468e-07,5.982805469106184e-07,2.700263197354109e-08,8.89319583069193e-08,1.491391879949874e-07,7.591655768212173e-08,4.802298149383984e-08,2.089691437207228e-07,2.616680471594436e-07,3.087516371984384e-07,2.571236426603606e-07,2.413721459386414e-07,3.755493771006589e-07,2.721479439767129e-07,1.581739471586186e-07,1.315017376803994e-07,2.146576681098907e-07,7.925537671447707e-08,2.503528619275176e-07,1.691610067978196e-07,2.612941318276451e-08,2.763564696308498e-07,1.27276924518579e-07,3.202283783400087e-07,2.134530415937799e-07,2.054567131029168e-07,1.972397703001989e-07,2.50802348061743e-07,3.108153269668455e-07,1.629079034704206e-07,2.447561784230443e-07,2.055197729931347e-07,4.257465092511524e-07,7.682793242358476e-08,1.500780143037855e-07,2.781077560637766e-07,3.102922464251037e-07,3.362577168810901e-07,1.555959620980684e-07,1.067346780034905e-07,4.908557500609643e-08,8.714417725265084e-08,2.744521880959901e-07,1.651326940276128e-07,2.704862389111631e-07,1.13028594115025e-07,5.334404624032629e-08,1.18102857116762e-07,1.649711590820846e-07,1.40062338945818e-07,3.004229685015831e-08,1.778869608412986e-07,3.332814154950529e-07,2.179557148860266e-07,2.941161937788072e-08,2.430777279580328e-07,8.959810233608535e-08,2.377870113997075e-07,2.95332630610434e-07,1.513363315885566e-07,9.168451127403177e-08,3.306135847601155e-08,1.649793186875473e-07,2.539604312959212e-07,1.788978852659881e-07,1.90935636169854e-07,9.193036764945847e-08,1.278947320628811e-07,2.772627718063271e-07,3.478373387382618e-08,1.71550156376545e-07,2.921294484011755e-07,2.962700599365442e-07,1.512192383776523e-07,8.685879678098983e-08,2.729860759357178e-07,3.503423779278696e-07,3.107128187059059e-07,2.410345395264105e-07,1.482539553488144e-07,1.706225213879596e-07,3.229466166013292e-08,1.63375008773238e-07,2.869409583110655e-07,2.125465897696721e-07,2.504229879037245e-07,1.80242496452302e-07,2.462557632057116e-07,2.115790067219003e-07,1.310627412414538e-07,2.606664448568999e-07,2.715831988747548e-07,7.042879940659511e-07,3.854333292018551e-07,1.3474849799218e-07,3.808630897169029e-07,2.48727097623396e-07,4.887588472870726e-07,4.231833048492498e-07,3.100518936673085e-07,3.720758558488836e-07,2.736293291238197e-07,2.184984059136656e-07,2.740805360302675e-07,2.360723868085179e-07,3.264758369890712e-07,4.009580261959053e-07,6.341773890869328e-07,2.33286458307644e-07,1.491491043596283e-07,1.916834539441566e-07,7.444903618525917e-08,8.919469360380393e-07,3.01226443464566e-07,2.704102191856739e-07,7.003211234971656e-07,1.249609559493136e-07,6.737679336389792e-07,3.15981102363974e-07,2.568957828527768e-07,1.728755197718596e-07,2.747929387118042e-07,1.697091990447069e-07,4.03423551562821e-07,5.768247364791371e-07,2.140302729768855e-07,2.065120583165271e-07,5.040314051887945e-07,3.021193764896562e-07,3.536215897939408e-07,2.498328118822884e-07,3.659109420858115e-07,8.650801005347648e-07,4.439250115603295e-07,3.826868977751962e-07,5.006491855407759e-07,1.830776322915048e-07,5.578139583348673e-07,4.63809289659453e-07,4.075550047102491e-07,2.497792451031778e-07,2.450123674586457e-07,9.259097796597066e-08,2.662611775194223e-07,2.004674433020436e-07,2.837523248537356e-07,3.710918298275714e-07,7.557854681468446e-07,4.191214691716332e-07,2.873132707158535e-07,3.815498880771097e-07,2.533139673116564e-07,7.645449108971559e-07,3.291259975587459e-07,1.231293784653071e-07,6.281809907532386e-07,2.996594071178698e-07,4.472213244668227e-07,9.3043638862281e-07,3.713546074430339e-07,4.234556208863933e-07,6.979411782154411e-07,6.001470840473583e-08,3.983472175596234e-07,2.006733191878711e-07,2.492311841048326e-07,4.378046927512419e-07,4.842909339042155e-07,5.136065283498605e-07,1.816609900130357e-07,2.092056142046707e-07,2.910148509971224e-07,6.491231401202079e-07,2.420904217408819e-07,2.350549533800232e-07,6.390566872620113e-07,2.146920394171722e-07,5.574918525178997e-07,4.50609855378269e-07,3.833272380043299e-07,9.456922065032417e-08,1.976434854293129e-07,1.720696093839194e-07,2.192103279881296e-07,3.773958978575487e-07,3.963004559833046e-07,2.481056882595694e-07,6.008627788641711e-07,3.024465902761183e-07,2.177862414350894e-07,2.159278662732813e-07,2.283668814431813e-07,7.64462466554767e-07,2.041359890677012e-07,1.965575258891445e-07,4.928692751423938e-07,3.786892952539199e-07,4.587818234452969e-07,4.53564792486548e-07,3.179122659109327e-07,1.207508354569179e-07,2.339906383855441e-07,2.347768903472186e-07,3.240396598620499e-07,2.886588812159082e-07,2.195158659894053e-07,3.141561921756259e-07,5.433336742105313e-07,4.63350547115059e-07,5.298882321891507e-07,4.994068156822618e-07,6.794604540426539e-07,1.860588829445323e-06,7.456016385800442e-07,1.681273571455472e-07,1.300777741858056e-06,2.322403618072596e-07,1.342648404186848e-06,1.326411486571884e-06,4.164570499515244e-07,7.605798707739696e-07,5.163072082101107e-07,6.918729353214349e-08,1.130380838869185e-06,1.022772340868622e-06,7.047450677615586e-07,8.829856263479143e-07,1.612048870494252e-06,7.217681298229899e-07,2.794618068631069e-07,5.479360481920713e-07,1.102307754771798e-06,1.620874744841161e-06,1.306941821484817e-06,1.86757719770415e-07,1.960426598053273e-06,2.974735865765639e-07,1.360649326047711e-06,1.431579791247035e-06,1.0657739485704e-06,4.829213885110297e-07,1.926637703593425e-06,3.632034691258286e-07,6.651450807798746e-07,5.805746121108444e-07,9.001266703256816e-07,1.048158699612391e-06,1.425147257793997e-06,5.013781632261679e-07,3.495070917371916e-07,5.203968233937106e-07,3.145353220283397e-07,9.81344405966344e-07,5.511993888721137e-07,2.836540364549589e-07,8.350272272998234e-07,3.164547364901188e-07,8.975172124953495e-07,7.148053186331237e-07,1.027329205894793e-06,2.913111445871754e-07,3.79346729914466e-07,3.31320081915577e-07,6.601673411981489e-07,2.26179933051386e-07,3.627891520238919e-07,6.700338099060478e-07,1.288474180675863e-06,6.912231221423618e-07,5.371910013941599e-07,3.932539369056444e-07,4.112000475895522e-07,1.843241726094029e-06,5.182446619857552e-07,3.396217238899837e-07,1.311200053251461e-06,3.551892317990974e-07,1.057731689763725e-06,7.212543709818018e-07,8.27745100375901e-07,4.731692678831722e-07,4.628601429687855e-07,2.705828958977495e-07,5.60899865233475e-07,6.718280569688205e-07,3.60853303621016e-07,6.948340406863666e-07,9.791639927957632e-07,4.861563407298208e-07,2.023869435163541e-07,4.147755739731053e-07,3.910129699302905e-07,1.517523187103932e-06,3.950644176844691e-07,4.705259182879397e-07,1.427472435789003e-06,3.710586268610639e-07,8.094224325761231e-07,1.190548355659257e-06,6.990296864244971e-07,2.772497026650967e-07,2.822207357028916e-07,2.891960094276453e-07,6.44365654634024e-07,4.630385185326747e-07,3.042606840632108e-07,2.843789235404607e-07,1.193507836287049e-06,5.775539657762312e-07,4.806083251815153e-07,4.813631985731845e-07,4.644306174677046e-07,1.324017545880518e-06,5.675658265210456e-07,3.603613748912582e-07,1.442425186557092e-06,4.215628261403239e-07,7.624052694125362e-07,8.804704682931483e-07,5.098926952675353e-07,5.863169065863722e-07,7.574942299836068e-07,2.445607780815283e-07,5.529014579089943e-07,6.255497996285798e-07,4.605046757726009e-07,8.322912848282692e-07,1.15972259435543e-06,3.728660446037277e-07,6.09043696256702e-07,9.553129899113693e-07,6.733254222405688e-07,1.581884859828151e-06,1.131653756905074e-06,5.119306484758161e-07,1.891023043466515e-06,5.596407028824441e-07,1.257125883377869e-06,1.246016515346e-06,1.107020391214968e-06,8.025028751489373e-07,9.674908560020472e-07,4.282153858301987e-07,1.068395103696072e-06,1.148857267155776e-06,6.768863142901206e-07,8.525830067919521e-07,1.817004444351575e-06,8.641861417443666e-07,5.769574169847224e-07,6.374919372693128e-07,6.127489858559546e-07,1.591863952528358e-06,1.083696497750553e-06,5.509790405605189e-07,1.435790686700637e-06,2.956927618157349e-07,1.317071602570795e-06,1.245163769440843e-06,9.55986188077813e-07,5.190034999935229e-07,3.643787531949318e-07,4.201792639542916e-07,8.223963885768755e-07,5.956560442745182e-07,7.785614150391594e-07,8.851920159550302e-07,1.505639032997493e-06,7.774027215361591e-07,7.191016407023034e-07,5.627977115474429e-07,1.12649302857145e-06,1.575818277408079e-06,1.003420294905225e-06,5.538025735028374e-07,1.956019477351596e-06,4.728466675343028e-07,1.309215245701287e-06,1.512888371682784e-06,1.602977652427084e-06,7.150435327440779e-07,7.188350080635982e-07,3.288661074810078e-07,8.44365139794148e-07,1.159919586850437e-06,8.839811601894084e-07,1.178355985538178e-06,1.52520524863016e-06,1.471503639547855e-06,9.930779910395424e-07,5.230233191411304e-07,6.194292087998676e-07,1.87890807287169e-06,1.033028459374499e-06,6.197993027877643e-07,1.685410405465765e-06,5.084041536788824e-07,1.141619880599879e-06,1.423265919706627e-06,1.101384314210469e-06,5.970924964729426e-07,8.586270143807146e-07,3.832347960025851e-07,1.07097367938447e-06,1.161405215528208e-06,8.438448468422775e-07,8.336104905405062e-07,1.966057834558164e-06,1.078250324277046e-06,7.883955740103192e-07,7.941775208103972e-07,8.517830839906131e-07,2.090724223220453e-06,1.150785185963224e-06,3.830943918275648e-07,1.789826533406881e-06,2.576988962601654e-07,1.376961645157055e-06,1.842652932860421e-06,1.268887983810093e-06,5.803610975092643e-07,8.650114185981452e-07,2.377810062493901e-07,1.047831202252051e-06,7.698391148518164e-07,8.015420342047836e-07,8.24522512363715e-07,1.523879841840028e-06,9.955057271755579e-07,1.060547891161121e-06,8.189649246256946e-07,7.717212842447896e-07,1.601215642916102e-06,1.346365234289016e-06,5.949682198964859e-07,1.570933206646956e-06,4.047371447261612e-07,1.498315739871547e-06,1.294673518942045e-06,8.544319049145454e-07,7.927870089726611e-07,1.169554426496178e-06,2.588260845158043e-07,1.142179304764341e-06,7.807946015486468e-07,1.027190100395939e-06,8.331370688040521e-07,1.375200839719933e-06,8.724063847031181e-07],[0,2.83891056239386e-08,1.205246014063112e-07,8.420176933177897e-08,1.831810152661226e-07,1.195120335620863e-07,1.818890127068028e-07,9.464917007348467e-08,6.263630666986291e-07,9.12329263280435e-08,9.461935076743232e-08,8.960254729289649e-08,2.159414297741195e-07,1.814911347932843e-07,7.980926224796833e-08,4.04360706733556e-08,1.703105749582824e-07,1.614369177172174e-07,1.396871296981496e-07,1.953598671365358e-07,5.768270795856398e-08,6.628637876562699e-08,9.200502126604061e-08,1.235393632237646e-07,3.026576108715731e-07,1.590909452479421e-07,2.495165893322343e-07,1.155327523684157e-07,9.858803206358849e-08,1.017324390057228e-07,4.797356656482278e-08,1.337091898660468e-07,9.364391169035767e-07,2.700263197354109e-08,8.89319583069193e-08,1.491391879949874e-07,7.591655768212173e-08,4.802298149383984e-08,2.089691437207228e-07,3.198165020837644e-07,3.087516371984384e-07,2.828360069263967e-07,2.413721459386414e-07,3.755493771006589e-07,2.721479439767129e-07,2.108985962114914e-07,1.643771721004993e-07,2.951542936510997e-07,1.188830650717156e-07,2.503528619275176e-07,1.691610067978196e-07,2.612941318276451e-08,2.763564696308498e-07,1.27276924518579e-07,3.202283783400087e-07,2.134530415937799e-07,2.054567131029168e-07,1.972397703001989e-07,2.50802348061743e-07,1.277796344197031e-06,1.629079034704206e-07,2.447561784230443e-07,2.312097446172766e-07,1.348197279295316e-06,7.682793242358476e-08,1.500780143037855e-07,3.090086178486406e-07,3.102922464251037e-07,3.736196854234335e-07,2.333939431471026e-07,1.067346780034905e-07,4.908557500609643e-08,8.714417725265084e-08,2.744521880959901e-07,1.651326940276128e-07,2.704862389111631e-07,1.13028594115025e-07,5.334404624032629e-08,1.18102857116762e-07,1.649711590820846e-07,1.750779236822725e-07,3.004229685015831e-08,2.075347876481817e-07,3.332814154950529e-07,2.086147556766255e-06,2.941161937788072e-08,2.430777279580328e-07,1.493301705601422e-07,3.170493485329433e-07,2.95332630610434e-07,1.816035979062679e-07,9.168451127403177e-08,3.306135847601155e-08,1.649793186875473e-07,2.902404929096243e-07,2.504570393723833e-07,1.90935636169854e-07,9.193036764945847e-08,1.278947320628811e-07,2.772627718063271e-07,3.478373387382618e-08,2.001418491059692e-07,2.921294484011755e-07,3.333038174286121e-07,1.512192383776523e-07,8.685879678098983e-08,2.729860759357178e-07,3.503423779278696e-07,3.107128187059059e-07,3.21379386035214e-07,1.482539553488144e-07,1.706225213879596e-07,3.229466166013292e-08,1.960500105278856e-07,2.869409583110655e-07,2.125465897696721e-07,2.504229879037245e-07,1.80242496452302e-07,2.462557632057116e-07,2.115790067219003e-07,2.621254824829076e-07,2.606664448568999e-07,5.70324717636985e-07,1.267718389318712e-06,5.781499938027827e-07,1.61698197590616e-07,7.617261794338057e-07,3.039997859841507e-07,6.391461849138641e-07,5.770681429762497e-07,3.664249652431827e-07,6.644211711587207e-07,3.648391054984263e-07,2.184984059136656e-07,5.824211390643185e-07,2.69797013495449e-07,5.935924308892205e-07,6.873566163358376e-07,8.719939099945326e-07,2.666130952087361e-07,1.78978925231554e-07,1.916834539441566e-07,7.444903618525917e-08,2.102446349232521e-06,5.271462760629905e-07,3.195757135830691e-07,9.337614979962208e-07,1.499531471391764e-07,1.179093883868214e-06,3.510901137377489e-07,3.596540959938875e-07,2.469650282455138e-07,3.847101141965259e-07,1.979940655521581e-07,7.447819413467464e-07,8.500575063903073e-07,4.548143300758818e-07,2.065120583165271e-07,6.840426213276497e-07,4.8339100238345e-07,4.179164243019301e-07,2.99799374258746e-07,7.318218841716231e-07,2.145398649326217e-06,8.617367871465221e-07,6.88836415995353e-07,8.448455006000593e-07,2.092315797617197e-07,8.925023333357877e-07,7.73015482765755e-07,4.618956720049489e-07,3.247130186341312e-07,5.512778267819527e-07,1.543182966099511e-07,4.326744134690613e-07,2.004674433020436e-07,2.837523248537356e-07,6.626639818349489e-07,1.312680023623467e-06,5.80322034237646e-07,4.469317544468832e-07,5.961717001204839e-07,5.066279346233128e-07,1.112065324941318e-06,6.835693795450876e-07,1.231293784653071e-07,1.099316733818168e-06,3.246310243776923e-07,4.844897681723913e-07,1.318118217215647e-06,5.30506582061477e-07,7.528099926869214e-07,9.015073551949448e-07,6.001470840473583e-08,4.647384204862272e-07,2.341188723858496e-07,4.984623682096653e-07,6.18077213295871e-07,1.008939445633782e-06,6.420081604373257e-07,2.724914850195536e-07,3.95166160164378e-07,4.365222764956836e-07,8.763162391622806e-07,3.873446747854111e-07,2.585604487180256e-07,8.423929059362876e-07,2.385467104635247e-07,6.271783340826372e-07,6.008131405043586e-07,4.855478348054845e-07,1.418538309754862e-07,2.258782690620718e-07,2.294261458452259e-07,4.071048948350978e-07,4.402952141671401e-07,5.284006079777395e-07,3.225373947374402e-07,8.637402446172459e-07,3.326912493037302e-07,3.111232020501277e-07,2.159278662732813e-07,3.552373711338375e-07,1.096837452013361e-06,4.337889767688652e-07,4.176847425144322e-07,7.393039127135907e-07,3.786892952539199e-07,6.705272804200493e-07,5.291589245676394e-07,5.828391541700432e-07,1.207508354569179e-07,3.217371277801232e-07,2.347768903472186e-07,4.212515578206648e-07,4.169517173118674e-07,2.743948324867567e-07,3.665155575382302e-07,8.926196076315872e-07,5.251306200637336e-07,6.812848699574795e-07,5.548964618691798e-07,1.075812385567535e-06,2.750435660919173e-06,1.089725471770834e-06,4.483396190547925e-07,1.985397605993874e-06,2.322403618072596e-07,1.930057081018594e-06,1.610642519408716e-06,7.734202356242596e-07,1.287135165925179e-06,9.035376143676937e-07,6.918729353214349e-08,1.431815729234301e-06,1.416146318125784e-06,1.281354668657379e-06,1.177314168463886e-06,2.465486507814739e-06,7.939449428052889e-07,2.794618068631069e-07,9.132267469867855e-07,1.469743673029064e-06,2.062931493434205e-06,1.400294808733732e-06,1.86757719770415e-07,2.537022656304236e-06,2.974735865765639e-07,1.360649326047711e-06,1.431579791247035e-06,1.647105193245163e-06,6.760899439154417e-07,2.033673131570838e-06,3.632034691258286e-07,6.651450807798746e-07,5.805746121108444e-07,1.350190005488522e-06,1.257790439534869e-06,1.781434072242496e-06,8.774117856457938e-07,3.812804637133e-07,9.106944409389936e-07,5.242255367138996e-07,1.455096877812165e-06,7.611801084424429e-07,2.836540364549589e-07,1.731908323288523e-06,5.274245608168646e-07,1.364226162992931e-06,9.405333139909522e-07,1.665939252802367e-06,4.502081325438166e-07,7.586934598289321e-07,6.927601712780247e-07,9.572426447373159e-07,3.231141900734086e-07,4.186028677198752e-07,9.38047333868467e-07,2.108412295651411e-06,1.00541545038889e-06,6.565667794817509e-07,7.865078738112888e-07,6.396445184726368e-07,3.311586829931646e-06,7.773669929786328e-07,5.43394758223974e-07,2.650297979976357e-06,3.788685139190372e-07,1.75186811117117e-06,1.190069712119973e-06,1.412035759464772e-06,5.678031214598067e-07,8.440390842371971e-07,6.493989501545988e-07,1.062757639389742e-06,1.038279724406359e-06,6.443808993232428e-07,1.701145409956277e-06,1.619386603469916e-06,1.22968956772837e-06,2.42864332219625e-07,8.591779746585752e-07,5.714804945135015e-07,2.384679294020464e-06,7.901288353689382e-07,6.763810075389133e-07,2.141208653683504e-06,3.710586268610639e-07,1.079229910101497e-06,1.88503489646049e-06,1.588703832782948e-06,5.544994053301934e-07,4.938862874800602e-07,4.33794014141468e-07,8.860027751217831e-07,4.630385185326747e-07,5.409078827790414e-07,6.31953163423246e-07,1.452966061566842e-06,6.930647589314774e-07,9.011406097153411e-07,7.1058376932232e-07,8.591966423152536e-07,2.74491442438644e-06,8.74997315886612e-07,6.306324060597017e-07,2.296105399009248e-06,7.26024867241669e-07,1.282227044011993e-06,1.225002390668728e-06,8.983823678523242e-07,1.125728460645835e-06,1.227663062387225e-06,3.532544572288742e-07,7.679186915402698e-07,9.070472094614407e-07,7.41924199855857e-07,2.23529659353878e-06,1.759579108677204e-06,7.170500857763995e-07,1.250142323895336e-06,1.886743155074955e-06,1.611171546075647e-06,3.645212937864869e-06,3.467194489241079e-06,1.256557046258821e-06,3.571932415436751e-06,1.094949201291738e-06,2.657923296284636e-06,2.371450787271419e-06,2.641753206308446e-06,2.140007667063833e-06,1.639359506003469e-06,7.136923097169979e-07,2.796681300851483e-06,2.128764936200408e-06,1.759904417154314e-06,1.705166013583904e-06,4.046964444237599e-06,2.098737772807747e-06,2.452069022185071e-06,1.295548130579571e-06,1.442925031209184e-06,4.852617532707413e-06,2.883041626091093e-06,7.674350922092941e-07,4.355231749658599e-06,5.519598220560386e-07,3.171723451088853e-06,2.577199429772908e-06,2.888740872669914e-06,1.723091619978496e-06,6.644553734731111e-07,7.518997354971533e-07,2.297872262200093e-06,1.56359711622061e-06,1.515038429265391e-06,1.605697145220753e-06,4.486189771788449e-06,3.722110000082216e-06,1.677903828305375e-06,1.508297866947147e-06,2.781748090962151e-06,3.955625471861095e-06,2.280500670239147e-06,1.971537161670101e-06,6.270768324450704e-06,7.430447632681901e-07,2.756242622529024e-06,4.614309533632492e-06,4.104594291820867e-06,2.122785487833981e-06,1.091564271503983e-06,5.755156880917637e-07,1.537950790339341e-06,2.257140817654904e-06,1.464883065456734e-06,2.96945708355621e-06,2.815763535932604e-06,3.491214517358636e-06,2.58200277670281e-06,1.276176898704358e-06,1.858287626399603e-06,4.05605552238968e-06,1.743235525194467e-06,8.057390936240936e-07,3.871177025054179e-06,1.525212461036647e-06,3.171166334999665e-06,2.707676627734558e-06,2.135336935714176e-06,1.194184992945885e-06,1.526448025565715e-06,5.876266872039639e-07,2.395599019675787e-06,2.642196865326674e-06,2.203372655643725e-06,1.842717926457961e-06,4.396820248193711e-06,2.992144649868803e-06,1.450647856178987e-06,1.972634487174212e-06,3.226451075722019e-06,7.168197336755838e-06,2.641575085961038e-06,2.962596630133168e-06,5.801506694491269e-06,1.881201942699207e-06,3.106990378815918e-06,4.587829751203497e-06,2.896374745653472e-06,2.717145138338828e-06,2.087958596616213e-06,2.675036320305639e-07,3.569175032671049e-06,1.603831489274618e-06,3.399643800247875e-06,4.282197564211553e-06,4.102753420338536e-06,2.830969411655493e-06,3.181643673483364e-06,4.591167001689499e-06,2.389846557661284e-06,6.725105700247628e-06,9.373750404766925e-06,1.413049522254154e-06,5.15740335767114e-06,2.504311082993122e-06,6.027315589937816e-06,4.675209929512939e-06,2.643398705829375e-06,2.327213477952005e-06,2.868192998312055e-06,6.32685984371966e-07,4.283172392866278e-06,2.123761316212319e-06,3.216726893345176e-06,1.640238604207978e-06,5.872479261506742e-06,5.174272350652976e-06],[25827681,35224780,33188245,23752470,32754486,33469433,27489291,31695999,22351254,21921910,31705988,33481191,27785312,27549555,25059748,24730395,29358130,30971850,21476567,25593793,34672436,45258167,43475888,32378344,42952827,43999990,36069746,43277771,30429657,29489119,41689625,44873505,38443503,37033427,33733655,33525729,39517071,41646727,28712373,34394723,29149643,38891795,37286821,26627657,36744720,37932922,30417849,37268643,25234881,23966173,35469167,38271047,32566634,31427535,28104942,28109227,34070437,35489800,23923221,28956101,30692188,40856987,38925695,28185786,39048298,39979207,32361557,38673219,26765185,25707608,37476105,40745168,34425708,32792597,30278680,29576366,35389275,37492469,25401587,30308328,28558712,33286403,33729285,27004206,32116616,34000168,28797373,33482852,25232665,23702088,33038993,32720903,30246791,30306829,27563349,27948905,31424202,32633395,23456791,28853495,28749070,34975194,34231400,27002391,33064576,34538816,29305524,34252208,25747248,24892698,33725913,35165346,30964870,30604436,27880300,28229105,31945949,33288487,24364912,28358201,30519734,38363204,36821129,28397474,36322754,37106165,31507385,36184236,26597984,25993464,35477932,37626736,32891211,32036847,29188501,29651922,33693152,34916373,25229534,30006028,33523500,41735475,40296022,31392002,39837140,40678936,34269993,40012498,29683811,28482716,38926291,40491563,36391037,35354595,32224197,32938948,37377890,38738658,27776047,33099499,31106698,40026768,38260676,28899058,38294756,39196534,31958506,38235146,26890686,25872703,36804848,40035352,32651413,32400565,30045687,29930047,35242002,37726511,25139409,31017261,31324693,41934228,39476702,28775288,39498551,40607693,31837958,40045464,26832352,25794348,37699815,42507406,34386852,33325164,30124473,29899341,36111051,38830100,24778494,31152252,33028555,43019878,41235009,30810795,41306880,42543243,34425741,41920511,28699971,26630576,39131057,42297060,35417307,34869609,31932802,31796848,37850070,40305404,26628376,33063689,32141608,41680586,39410268,30086500,39189562,40700553,32462969,39610309,28335909,26457080,37746263,41407581,34189402,34074904,30860420,31178670,36443835,38197560,25766855,32372898,13210333,18021380,17661072,12361678,17435584,17843616,14606646,17223535,11916746,10554794,16808456,17092222,15494651,14453521,13269864,12710551,15608481,16987819,10545586,13854865,7156613,10950183,10886252,6786459,10712030,10709062,8671582,10084929,6614489,5588232,10321138,10353652,9342701,8259833,7517157,6889726,8887638,9540540,5613455,7978010,31472895,38432210,38151518,29551297,38098736,38779635,32334275,37920115,27854619,26580664,36015719,37760313,34269440,33200523,30295349,30948811,35833486,37311550,25611689,31827639,33507635,43229065,43774314,32008824,42450992,44166786,35845026,42231010,30253419,27729468,41075447,42268172,36728157,36957251,33874139,32746474,38796929,41736585,26553264,34968175,24705151,33753193,33246979,23063898,32906026,34004503,26620479,32339903,22238079,20158778,31472197,32461712,28346606,27662899,24830622,23756123,29579898,31647915,19270925,25971599,33291142,43626102,43063483,30966357,42285844,44399875,33970566,42698262,28856044,26122398,41185136,42639057,38284120,36800668,32555530,31971875,39087551,42052585,25005980,34865068,31196448,41871094,41584647,29079234,41532138,42974571,33315300,41097797,27841285,24879285,39746332,41121348,37209654,35029101,31823433,29594625,36933824,39878815,24215681,32400427,41597524,48628066,50591679,38948052,48906682,50818630,41788821,50728330,37203748,34533610,48117850,48169232,46654751,45218795,41342594,40291709,47523547,48577031,32544321,42449041,33374976,44420934,43497828,31094956,43850020,45142441,34764480,44411860,29025021,26439492,41173375,44752520,37560775,36489014,33161009,31898763,39593604,42431999,25570329,34658426,35243959,47799016,46817295,33530113,46465322,48402765,37972947,47206538,31534139,28806985,44489466,46893907,41927402,39140496,35481731,34441037,42661871,45584839,27974762,37097137,31709970,39034094,38742258,30133099,38234764,39154841,32405375,38804978,28323229,26592094,36252215,37907434,33525569,33644403,30539270,31175345,36180261,37597518,25592569,32144466,33001810,40294766,40169943,31226275,39365247,40338289,33737908,39531830,29366307,27806238,37451785,39102558,35911112,34772384,31518694,32018664,36994126,38409046,26905161,33241389]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>label<\/th>\n      <th>dose_group<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":[6,7],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[4,5,6,7,8]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"label","targets":2},{"name":"dose_group","targets":3},{"name":"sum_min","targets":4},{"name":"sum_max","targets":5},{"name":"mf_min","targets":6},{"name":"mf_max","targets":7},{"name":"group_depth","targets":8}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script><p><strong>calculate_mf() does not calculate the mean MF for any given
group</strong> If you want to calculate the mean MF for a given
experimental variable, you may group by “sample” and retain the
experimental variable in the summary table for averaging using the
<code>retain_metadata_cols</code> parameter.</p>
<p><em>Example 3.3. Calculate the mean MF per dose</em></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_global</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  retain_metadata_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dose_group"</span>, <span class="st">"dose"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mean_mf</span> <span class="op">&lt;-</span> <span class="va">mf_data_global</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">dose_group</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mean_mf_min <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">mf_min</span><span class="op">)</span>,</span>
<span>                   SE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/sd.html" class="external-link">sd</a></span><span class="op">(</span><span class="va">mf_min</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 8.</strong> Mean Mutation Frequency per Dose Group.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-5b1b2f4ad92281566982" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-5b1b2f4ad92281566982">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4"],["Control","High","Low","Medium"],[1.742563688908099e-07,9.536535843621309e-07,3.425812769564564e-07,6.840158856255222e-07],[1.291594484410011e-08,2.863285443579539e-08,1.378665560460376e-08,5.724275468458596e-08]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>dose_group<\/th>\n      <th>mean_mf_min<\/th>\n      <th>SE<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":[2,3],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[2,3]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"dose_group","targets":1},{"name":"mean_mf_min","targets":2},{"name":"SE","targets":3}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="mutation-subtypes">Mutation Subtypes<a class="anchor" aria-label="anchor" href="#mutation-subtypes"></a>
</h3>
<p>Mutations can also be grouped by mutation subtype at varying degrees
of resolution using the <code>subtype_resolution</code> parameter.</p>
<table style="width:100%;" class="table">
<caption>(#tab:subtypes) Definitions of mutation subtype
resolutions.</caption>
<colgroup>
<col width="47%">
<col width="30%">
<col width="22%">
</colgroup>
<thead><tr class="header">
<th>Subtype resolutions</th>
<th>Definition</th>
<th>Example</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>type</code></td>
<td>The variation type</td>
<td>snv, mnv, insertion, deletion, complex, and symbolic variants</td>
</tr>
<tr class="even">
<td><code>base_6</code></td>
<td>The simple spectrum; snv subtypes in their pyrimidine context</td>
<td>C&gt;A, C&gt;G, C&gt;T, T&gt;A, T&gt;C, T&gt;G</td>
</tr>
<tr class="odd">
<td><code>base_12</code></td>
<td>The snv subtypes</td>
<td>A&gt;C, A&gt;G, A&gt;T, C&gt;A, C&gt;G, C&gt;T, G&gt;A, G&gt;C,
G&gt;T, T&gt;A, T&gt;C, T&gt;G</td>
</tr>
<tr class="even">
<td><code>base_96</code></td>
<td>The trinculeotide spectrum; the snv subtypes in their pyrimidine
context alongside their two flanking nucleotides</td>
<td>A[C&gt;T]A</td>
</tr>
<tr class="odd">
<td><code>base_192</code></td>
<td>The snv subtypes reported alongside their two flanking
nucleotides</td>
<td>A[G&gt;A]A</td>
</tr>
</tbody>
</table>
<p>Mutations and <code>total_depth</code> will be summed across groups
for each mutation subtype to calculate frequencies. For SNV subtypes,
the <code>total_depth</code> is summed based on the sequence context in
which the SNV subtype occurs (<code>subtype_depth</code>). In the
simplest example, for the <code>base_6</code> SNV subtypes, the two
possible reference bases are C or T; hence, the <code>total_depth</code>
is summed separately for C:G positions and T:A positions. Thus, the MF
for C&gt;T mutations is calculated as the total number of C&gt;T
mutations divided by total_depth for C:G positions within the group:
<code>sum</code> / <code>subtype_depth</code>. Non-snv mutation
subtypes, such as mnvs, insertions, deletions, complex variants, and
structural variants, will be calculated as their <code>sum</code> /
<code>group_depth</code>, since they can occur in the context of any
nucleotide.<br>
Upon import of mutation data, columns are created that facilitate the
grouping of SNV subtypes and their associated sequence context for the
various resolutions. Below the columns associated with each
subtype_resolution are defined:</p>
<table class="table">
<caption>(#tab:subtypes-cols) Subtype Resolutions and their associated
subtype/context columns.</caption>
<colgroup>
<col width="37%">
<col width="31%">
<col width="31%">
</colgroup>
<thead><tr class="header">
<th>Subtype resolution</th>
<th>Subtype Column</th>
<th>Context Column</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>type</code></td>
<td><code>variation_type</code></td>
<td>NA</td>
</tr>
<tr class="even">
<td><code>base_6</code></td>
<td><code>normalized_subtype</code></td>
<td><code>normalized_ref</code></td>
</tr>
<tr class="odd">
<td><code>base_12</code></td>
<td><code>subtype</code></td>
<td><code>short_ref</code></td>
</tr>
<tr class="even">
<td><code>base_96</code></td>
<td><code>normalized_context_with_mutation</code></td>
<td><code>normalized_context</code></td>
</tr>
<tr class="odd">
<td><code>base_192</code></td>
<td><code>context_with_mutation</code></td>
<td><code>context</code></td>
</tr>
</tbody>
</table>
<p>The function will also calculate the the proportion of mutations for
eachsubtype, normalized to the <code>total_depth</code>:<br><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>P</mi><mi>s</mi></msub><mo>=</mo><mfrac><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><msub><mi>M</mi><mi>s</mi></msub><msub><mi>D</mi><mi>s</mi></msub></mfrac><mo stretchy="true" form="postfix">)</mo></mrow><mrow><munder><mo>∑</mo><mi>s</mi></munder><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><msub><mi>M</mi><mi>s</mi></msub><msub><mi>D</mi><mi>s</mi></msub></mfrac><mo stretchy="true" form="postfix">)</mo></mrow></mrow></mfrac></mrow><annotation encoding="application/x-tex">P_s = \frac{\left(\frac{M_s}{D_s}\right)}{\sum_s \left(\frac{M_s}{D_s}\right)}</annotation></semantics></math><br>
Where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>P</mi><mi>s</mi></msub><annotation encoding="application/x-tex">P_s</annotation></semantics></math>
is the normalized mutation proportion for subtype
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>M</mi><mi>s</mi></msub><annotation encoding="application/x-tex">M_s</annotation></semantics></math>
is the group mutation sum for subtype
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>D</mi><mi>s</mi></msub><annotation encoding="application/x-tex">D_s</annotation></semantics></math>
is the group sum of the <code>subtype_depth</code> for subtype
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.<br>
If total_depth is not available for the mutation data,
<code><a href="../reference/calculate_mf.html">calculate_mf()</a></code> will return the subtype mutation counts per
group. It will also calculate subtype proportions, without normalizing
to the total_depth:<br><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><msub><mi>′</mi><mi>s</mi></msub><mo>=</mo><mfrac><msub><mi>M</mi><mi>s</mi></msub><msub><mi>M</mi><mrow><mi>t</mi><mi>o</mi><mi>t</mi><mi>a</mi><mi>l</mi></mrow></msub></mfrac></mrow><annotation encoding="application/x-tex">P'_s = \frac{M_s}{M_{total}}</annotation></semantics></math><br>
Where,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><msub><mi>′</mi><mi>s</mi></msub></mrow><annotation encoding="application/x-tex">P'_s</annotation></semantics></math>
is the non-normalized mutation proportion of subtype
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>M</mi><mi>s</mi></msub><annotation encoding="application/x-tex">M_s</annotation></semantics></math>
is the group mutation sum for subtype
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>s</mi><annotation encoding="application/x-tex">s</annotation></semantics></math>.
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>M</mi><mrow><mi>t</mi><mi>o</mi><mi>t</mi><mi>a</mi><mi>l</mi></mrow></msub><annotation encoding="application/x-tex">M_{total}</annotation></semantics></math>
is the total mutation sum for the group.</p>
<p><em>Example 3.4. The following code will return the base_6 mutation
spectra for</em> <em>all samples with mutation proportions normalized to
depth.</em></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 9.</strong> Frequency and Proportions of 6-base
Mutation Subtypes per Sample.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-25c3e940e6859592f801" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-25c3e940e6859592f801">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312"],["ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],["N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T"],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,2,0,4,0,0,0,0,4,3,1,2,1,4,3,3,4,3,18,10,18,26,9,14,20,19,23,21,26,21,19,10,55,39,32,32,41,51,53,57,43,55,13,15,37,33,37,55,14,22,24,31,26,14,5,0,12,6,7,12,10,14,25,15,10,10,2,2,2,0,2,1,8,4,4,11,7,8,2,7,3,16,8,14,19,13,12,12,14,22,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,2,2,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,17,33,33,21,15,19,97,107,102,113,85,99,118,90,187,245,175,230,415,399,452,444,382,400,17,16,18,12,16,15,35,31,28,34,26,26,35,16,52,63,40,68,88,103,105,92,95,85,3,17,19,13,6,7,26,27,42,34,34,24,33,20,45,62,30,69,76,93,68,82,74,72,0,5,4,5,1,3,1,7,8,9,10,14,8,6,11,19,9,18,17,26,19,27,18,17,2,9,5,6,7,5,5,5,8,7,3,12,8,5,4,9,10,7,7,9,15,21,12,13,2,4,2,1,2,2,2,1,6,6,9,2,2,5,5,7,5,10,5,8,5,15,5,6],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,0,0,0,2,0,4,0,0,0,0,5,3,3,3,2,19,3,5,6,12,18,11,18,49,9,15,21,41,51,37,32,22,24,11,69,44,40,34,67,125,96,109,83,115,16,17,43,36,47,58,17,23,27,32,28,19,6,0,12,7,9,22,19,21,27,15,13,16,2,3,2,0,2,1,11,5,9,19,7,8,4,9,3,38,17,23,43,35,22,27,58,57,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,3,3,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,17,35,34,21,15,19,154,176,189,173,131,162,188,111,316,459,282,432,954,1152,1156,1038,1266,1673,17,17,18,12,17,15,55,51,42,61,34,31,42,22,84,111,69,123,172,265,303,210,347,311,3,23,48,17,65,8,53,40,63,56,46,37,45,24,63,117,44,128,205,226,158,197,247,309,0,5,4,5,1,3,2,7,14,11,12,14,10,8,24,26,14,24,24,44,31,43,31,108,2,9,5,6,7,5,6,6,8,9,4,12,12,5,6,18,14,12,19,15,17,43,17,16,2,7,2,1,2,2,2,1,6,6,10,2,5,6,7,8,5,11,9,18,19,28,7,12],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4.443904506755093e-09,0,0,0,2.804141309552892e-09,0,5.83990861419005e-09,0,0,0,0,5.950163073367402e-09,4.049123250630963e-09,1.791994313787003e-09,2.725807074437837e-09,1.421817715208629e-09,4.521643912870893e-09,4.035991237066589e-09,3.752477915957621e-09,5.902371075718579e-09,4.278602441634115e-09,3.183653971564643e-08,1.323496607149611e-08,2.812899014120308e-08,3.851383905854413e-08,1.504754973048409e-08,2.290251249134113e-08,3.083889036645722e-08,2.663934244075247e-08,3.43419603931304e-08,3.065952022449777e-08,3.626658490281408e-08,3.033306870764193e-08,6.383091660524341e-08,5.784905158880656e-08,8.181474225880177e-08,5.263860225820251e-08,5.734381804118409e-08,4.36129131910054e-08,5.82945263235538e-08,5.765095988910388e-08,7.130251185484307e-08,7.12970804031948e-08,6.345048906397472e-08,7.844104476329212e-08,2.29930564613002e-08,1.985244910724416e-08,5.782070195691743e-08,4.888294957430602e-08,6.186214889199014e-08,8.997415621598302e-08,2.158722325652006e-08,3.084555440508181e-08,3.58350891058752e-08,4.525929175997289e-08,3.626658490281408e-08,2.022204580509462e-08,1.679760963295879e-08,0,1.78504892201022e-08,8.098246501261925e-09,1.254396019650902e-08,1.635484244662703e-08,1.421817715208629e-08,1.582575369504812e-08,3.363326030888824e-08,1.87623895797881e-08,1.475592768929645e-08,1.426200813878038e-08,3.537393301738492e-09,2.646993214299222e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635893749381509e-09,1.233555614658289e-08,5.608282619105783e-09,5.9725148509792e-09,1.605974868902264e-08,9.764080550757637e-09,1.155545474576835e-08,6.719043853183517e-09,4.049433611216459e-08,4.462622305025551e-09,2.159532400336514e-08,1.433595451029602e-08,1.908064952106486e-08,2.701453658896395e-08,1.46953427168304e-08,1.614396494826636e-08,1.500991166383048e-08,2.065829876501503e-08,3.137641790531685e-08,3.537393301738492e-09,1.323496607149611e-09,0,4.443904506755093e-09,3.343899940107575e-09,4.907681248144528e-09,6.167778073291445e-09,7.010353273882228e-09,4.479386138234399e-09,4.379931460642538e-09,2.789737300216468e-09,1.444431843221044e-09,0,5.784905158880655e-09,1.48754076834185e-09,2.699415500420642e-09,1.791994313787003e-09,6.814517686094594e-09,2.843635430417258e-09,3.391232934653169e-09,4.035991237066589e-09,2.501651943971747e-09,2.95118553785929e-09,4.278602441634115e-09,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6.768438796963514e-08,9.866274665821205e-08,1.163233345261175e-07,7.019875349677959e-08,5.705222777729199e-08,7.074183179904539e-08,3.388872528747683e-07,3.405785922278743e-07,3.442097111568596e-07,3.719776368493135e-07,2.673840065766399e-07,3.231388150819526e-07,8.869240331301002e-07,1.146752212343037e-06,6.309637099805067e-07,7.442455511157748e-07,7.014340847920938e-07,7.026786844681858e-07,1.321062990438468e-06,1.023954525620063e-06,1.366773172667736e-06,1.247763696218023e-06,1.276951737829694e-06,1.293930153915547e-06,6.768438796963514e-08,4.783648322822403e-08,6.344909155970046e-08,4.011357342673119e-08,6.085570962911145e-08,5.584881457819373e-08,1.222789056764628e-07,9.867230242115984e-08,9.44889403175693e-08,1.119224748042182e-07,8.178804907050163e-08,8.486473931445219e-08,2.630706877928263e-07,2.038670599720954e-07,1.754551492993922e-07,1.913774274297706e-07,1.603277908096214e-07,2.077484806253767e-07,2.801290196592413e-07,2.643291131299913e-07,3.175026175444961e-07,2.585456307478786e-07,3.175665316592171e-07,2.749601577070538e-07,1.194430375934738e-08,5.082626342998803e-08,6.697404109079492e-08,4.345637121229212e-08,2.28208911109168e-08,2.606278013649041e-08,9.083575850251523e-08,8.594039243133277e-08,1.417334104763539e-07,1.119224748042182e-07,1.06953602630656e-07,7.833668244410972e-08,2.480380770618077e-07,2.548338249651193e-07,1.518361868937048e-07,1.88339690486441e-07,1.202458431072161e-07,2.108036053404558e-07,2.419296078875266e-07,2.386660924377591e-07,2.056207427907213e-07,2.304428447970223e-07,2.473676141345481e-07,2.329074277047985e-07,0,1.187362093087597e-08,1.122910372578574e-08,1.330029886675974e-08,2.983408985176398e-09,8.753899887932282e-09,2.760136861386272e-09,1.754124059682377e-08,2.142453621282099e-08,2.361209905153819e-08,2.506149300947495e-08,3.627470661198666e-08,4.859756213609879e-08,6.357200680423903e-08,2.926477992302863e-08,4.614913363599911e-08,2.916882357963362e-08,4.429038357128141e-08,4.368112944066087e-08,5.25286195567069e-08,4.60487855327183e-08,6.086082634756239e-08,4.755063508628221e-08,4.33642677438963e-08,6.364920043603903e-09,2.137251767557675e-08,1.403637965723217e-08,1.596035864011169e-08,2.088386289623479e-08,1.45898331465538e-08,1.380068430693136e-08,1.252945756915984e-08,2.142453621282099e-08,1.836496592897415e-08,7.518447902842486e-09,3.109260566741714e-08,4.859756213609879e-08,5.297667233686586e-08,1.064173815382859e-08,2.186011593284168e-08,3.240980397737068e-08,1.722403805549832e-08,1.798634741674271e-08,1.818298369270623e-08,3.63543043679355e-08,4.73361982703263e-08,3.170042339085481e-08,3.316091062768541e-08,6.364920043603903e-09,9.498896744700776e-09,5.614551862892868e-09,2.660059773351948e-09,5.966817970352797e-09,5.835933258621521e-09,5.520273722772544e-09,2.505891513831967e-09,1.606840215961575e-08,1.574139936769213e-08,2.255534370852746e-08,5.182100944569522e-09,1.21493905340247e-08,5.297667233686586e-08,1.330217269228574e-08,1.70023123922102e-08,1.620490198868534e-08,2.460576865071189e-08,1.284739101195908e-08,1.616265217129443e-08,1.21181014559785e-08,3.381157019309021e-08,1.32085097461895e-08,1.530503567431634e-08],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5.92520600900679e-09,0,0,0,2.804141309552892e-09,0,5.83990861419005e-09,0,0,0,0,7.437703841709251e-09,4.049123250630963e-09,5.375982941361009e-09,4.088710611656756e-09,2.843635430417258e-09,2.147780858613674e-08,4.035991237066589e-09,6.254129859929368e-09,8.853556613577869e-09,1.711440976653646e-08,3.183653971564643e-08,1.455846267864572e-08,2.812899014120308e-08,7.258377361033318e-08,1.504754973048409e-08,2.453840624072264e-08,3.238083488478009e-08,5.748489684583428e-08,7.61495643499848e-08,5.401915468125797e-08,4.463579680346348e-08,3.177750055086297e-08,8.06285262382022e-08,6.363395674768722e-08,1.026403130155877e-07,5.938714100925412e-08,7.167977255148012e-08,4.633872026544324e-08,9.526178691897815e-08,1.413013722772154e-07,1.291517195861308e-07,1.363400309464602e-07,1.224741998211605e-07,1.640130935959744e-07,2.829914641390793e-08,2.249944232154339e-08,6.719703200398513e-08,5.332685408106111e-08,7.858164859252802e-08,9.488183746412755e-08,2.621305681148864e-08,3.224762505985825e-08,4.03144752441096e-08,4.67192689135204e-08,3.905632220303055e-08,2.744420502119984e-08,2.015713155955055e-08,0,1.78504892201022e-08,9.447954251472247e-09,1.612794882408303e-08,2.998387781881621e-08,2.701453658896395e-08,2.373863054257218e-08,3.63239211335993e-08,1.87623895797881e-08,1.918270599608538e-08,2.281921302204861e-08,3.537393301738492e-09,3.970489821448833e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635893749381509e-09,1.696138970155147e-08,7.010353273882228e-09,1.34381584147032e-08,2.773956591740274e-08,9.764080550757637e-09,1.155545474576835e-08,1.343808770636703e-08,5.20641464299259e-08,4.462622305025551e-09,5.12888945079922e-08,3.046390333437905e-08,3.134678135603513e-08,6.113816175397105e-08,3.956438423762031e-08,2.959726907182165e-08,3.377230124361859e-08,8.55843805979194e-08,8.129344639104818e-08,3.537393301738492e-09,1.323496607149611e-09,0,4.443904506755093e-09,3.343899940107575e-09,4.907681248144528e-09,6.167778073291445e-09,7.010353273882228e-09,4.479386138234399e-09,4.379931460642538e-09,2.789737300216468e-09,1.444431843221044e-09,0,5.784905158880655e-09,1.48754076834185e-09,2.699415500420642e-09,1.791994313787003e-09,6.814517686094594e-09,2.843635430417258e-09,3.391232934653169e-09,4.035991237066589e-09,3.752477915957621e-09,4.426778306788935e-09,4.278602441634115e-09,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6.768438796963514e-08,1.046423070617401e-07,1.19848284057212e-07,7.019875349677959e-08,5.705222777729199e-08,7.074183179904539e-08,5.380271849764363e-07,5.602040395523914e-07,6.378003471435928e-07,5.694878865038162e-07,4.120859395475275e-07,5.287726064977406e-07,1.413065408715753e-06,1.414327728556412e-06,1.066227445742461e-06,1.394321256988329e-06,1.130310925207831e-06,1.319813876914158e-06,3.036853235851321e-06,2.956379983745145e-06,3.495552627442262e-06,2.917069181698891e-06,4.231991885058619e-06,5.411862868751777e-06,6.768438796963514e-08,5.082626342998803e-08,6.344909155970046e-08,4.011357342673119e-08,6.465919148093092e-08,5.584881457819373e-08,1.92152566063013e-07,1.623318523702952e-07,1.417334104763539e-07,2.008020871487444e-07,1.06953602630656e-07,1.011848814903084e-07,3.156848253513916e-07,2.803172074616313e-07,2.83427548868249e-07,3.371888007095959e-07,2.76565439146597e-07,3.757803399547255e-07,5.475249020612444e-07,6.800700483441524e-07,9.162218391998317e-07,5.901585049679838e-07,1.159953541955245e-06,1.006030694669338e-06,1.194430375934738e-08,6.876494464057203e-08,1.691975774925345e-07,5.682756235453585e-08,2.472263203682653e-07,2.978603444170332e-08,1.85165200024358e-07,1.273190998982708e-07,2.126001157145309e-07,1.843428996775359e-07,1.447019329708875e-07,1.207690521013358e-07,3.382337414479196e-07,3.058005899581431e-07,2.125706616511867e-07,3.55415222369574e-07,1.763605698905836e-07,3.910559635301208e-07,6.525732844334599e-07,5.79984267644447e-07,4.777658435431466e-07,5.536248832318705e-07,8.256729823139643e-07,9.995610438997603e-07,0,1.187362093087597e-08,1.122910372578574e-08,1.330029886675974e-08,2.983408985176398e-09,8.753899887932282e-09,5.520273722772544e-09,1.754124059682377e-08,3.749293837243674e-08,2.885923217410223e-08,3.007379161136994e-08,3.627470661198666e-08,6.074695267012349e-08,8.476267573898537e-08,6.385042892297155e-08,6.31514460282093e-08,4.537372556831895e-08,5.905384476170854e-08,6.166747685740359e-08,8.889458694211937e-08,7.51322290270667e-08,9.692650122019194e-08,8.189276042637493e-08,2.754906421376941e-07,6.364920043603903e-09,2.137251767557675e-08,1.403637965723217e-08,1.596035864011169e-08,2.088386289623479e-08,1.45898331465538e-08,1.656082116831763e-08,1.50353490829918e-08,2.142453621282099e-08,2.361209905153819e-08,1.002459720378998e-08,3.109260566741714e-08,7.289634320414819e-08,5.297667233686586e-08,1.596260723074289e-08,4.372023186568336e-08,4.537372556831895e-08,2.952692238085427e-08,4.882008584544451e-08,3.030497282117706e-08,4.12015449503269e-08,9.692650122019194e-08,4.490893313704431e-08,4.081342846484358e-08,6.364920043603903e-09,1.662306930322636e-08,5.614551862892868e-09,2.660059773351948e-09,5.966817970352797e-09,5.835933258621521e-09,5.520273722772544e-09,2.505891513831967e-09,1.606840215961575e-08,1.574139936769213e-08,2.506149300947495e-08,5.182100944569522e-09,3.037347633506175e-08,6.357200680423903e-08,1.862304176920004e-08,1.943121416252594e-08,1.620490198868534e-08,2.706634551578308e-08,2.312530382152635e-08,3.636596738541247e-08,4.60487855327183e-08,6.311493102710173e-08,1.84919136446653e-08,3.061007134863268e-08],[565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532],[565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,251165749,334472748,283692005,299150611,262916990,268582245,286230890,314171244,296330977,303781703,317894855,306369880,133044089,78482517,296372037,329192428,249488874,327318880,314140963,389665742,330705935,355836607,299149912,309135697,251165749,334472748,283692005,299150611,262916990,268582245,286230890,314171244,296330977,303781703,317894855,306369880,133044089,78482517,296372037,329192428,249488874,327318880,314140963,389665742,330705935,355836607,299149912,309135697,251165749,334472748,283692005,299150611,262916990,268582245,286230890,314171244,296330977,303781703,317894855,306369880,133044089,78482517,296372037,329192428,249488874,327318880,314140963,389665742,330705935,355836607,299149912,309135697,314222329,421101535,356217210,375931402,335187031,342704399,362300875,399059574,373403649,381160522,399018526,385943852,164617311,94381164,375878446,411708704,308548611,406408763,389184076,494968271,412605887,443635120,378543840,392027835,314222329,421101535,356217210,375931402,335187031,342704399,362300875,399059574,373403649,381160522,399018526,385943852,164617311,94381164,375878446,411708704,308548611,406408763,389184076,494968271,412605887,443635120,378543840,392027835,314222329,421101535,356217210,375931402,335187031,342704399,362300875,399059574,373403649,381160522,399018526,385943852,164617311,94381164,375878446,411708704,308548611,406408763,389184076,494968271,412605887,443635120,378543840,392027835],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.01575997893739194,0,0,0,0.004446250224791244,0,0.007727410024949062,0,0,0,0,0.005298315103751796,0.003119377399242391,0.001561516140402668,0.002103344078543636,0.0007033725098726454,0.002644377054605577,0.001909561688936986,0.001885587438325342,0.002890395146052891,0.002111571401022107,0.1434418861934289,0.04775751768352558,0.07670811154435228,0.1365864841240635,0.05920493016060689,0.07602409572135183,0.04781130953130446,0.04223937713551682,0.04769859240526619,0.04056890263098258,0.06081447263352983,0.04873215417758,0.0400262874746608,0.03078539941748785,0.07285183267658719,0.04055190619015108,0.04996851649288536,0.03365350525669817,0.02883827290477846,0.0337158074462211,0.03373558983788676,0.03582616132818149,0.03107174782006858,0.03871214235207195,0.1035969178063653,0.07163627652528835,0.1576777848411686,0.1733597683113113,0.2433980462158283,0.2986660903338822,0.03346791667191312,0.04890875247270369,0.04977244424897341,0.05988742769335522,0.06081447263352983,0.03248810278505333,0.01053323354596337,0,0.01589494531125539,0.006238754798484782,0.01093061298281867,0.01262006447126181,0.007033725098726453,0.009255319691119517,0.01591301407447488,0.009427937191626707,0.007225987865132227,0.007038571336740354,0.01593798735482543,0.009551503536705114,0.008523123504928032,0,0.01315665114680153,0.005430292551525131,0.01912452381252178,0.008892500449582489,0.008295407374828902,0.02125037756860992,0.0163731272474888,0.01856463016288762,0.004213293418385347,0.0215497795922415,0.003973736327813848,0.01663667946262608,0.01249212912322134,0.01472340854980545,0.01336407768758026,0.008594225427468124,0.007638246755747946,0.007542349753301366,0.01011638301118512,0.01548485694082878,0.01593798735482543,0.004775751768352557,0,0.01575997893739194,0.01315665114680153,0.01629087765457539,0.00956226190626089,0.01111562556197811,0.006221555531121676,0.005795557518711797,0.004678036356425372,0.002320578770360952,0,0.003078539941748786,0.001324578775937949,0.002079584932828261,0.001561516140402668,0.005258360196359089,0.001406745019745291,0.001983282790954182,0.001909561688936986,0.001257058292216894,0.001445197573026446,0.002111571401022107,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.3049570199188703,0.3560181297617881,0.317215203078698,0.2489546917263193,0.224473301072981,0.2348250566058077,0.5253964442583268,0.5400218730395296,0.478083327988745,0.4922035446003132,0.448369136332351,0.5191446572422447,0.5561611552251279,0.6102645405633442,0.5618408291147965,0.5733544295732366,0.6112188171606081,0.5422155749595021,0.6535291981139947,0.5988357121182594,0.6466658460562222,0.6269903792323005,0.6253241379469532,0.638579055941615,0.3049570199188703,0.1726148507935942,0.1730264744065625,0.142259823843611,0.2394381878111797,0.1853882025835323,0.189576036588056,0.1564549351796768,0.1312385606243614,0.1480966417381473,0.1371482064075427,0.1363410210939228,0.1649630545159277,0.1084914738779279,0.1562338134436867,0.1474339961759751,0.139707158208139,0.1603072134662876,0.1385796853832085,0.1545866625267687,0.1502210483095207,0.1299169254265127,0.1555125473951847,0.1356980493875932,0.05381594469156534,0.1834032789681939,0.1826390563180382,0.1541148091639119,0.08978932042919238,0.08651449453898176,0.1408279128939845,0.1362672016081056,0.196857840936542,0.1480966417381473,0.1793476545329404,0.1258532502405442,0.1555365942578747,0.1356143423474098,0.1352023385570366,0.1450937740144517,0.1047803686561043,0.1626646724878507,0.1196824555582255,0.1395782486892183,0.0972860122385467,0.1157955204888483,0.1211360895499333,0.1149442300694907,0,0.04284519193404229,0.03062190774772811,0.04716852706499961,0.01173829120166087,0.02905826700310311,0.004279199292204054,0.02781341463821615,0.02975718941460412,0.03124370310822958,0.04202495175281454,0.05827780276370322,0.0304739473609156,0.03383097159743997,0.02605878587902373,0.03555252718877355,0.02541726191078171,0.03417626907446988,0.02160903280364689,0.03072012722418632,0.02178721199089259,0.03058203464943713,0.02328557847040985,0.02140108805231855,0.02867761838062455,0.07712134548127612,0.03827738468466015,0.05660223247799955,0.08216803841162607,0.04843044500517186,0.02139599646102027,0.01986672474158296,0.02975718941460412,0.02430065797306745,0.01260748552584436,0.04995240236888848,0.0304739473609156,0.02819247633119997,0.009475922137826812,0.01684067077362958,0.02824140212309079,0.01329077130673829,0.008897837036795779,0.01063389019298757,0.01720043051912573,0.02378602694956222,0.01552371898027324,0.01636553792236125,0.02867761838062455,0.03427615354723383,0.01531095387386406,0.009433705412999925,0.02347658240332174,0.01937217800206874,0.008558398584408109,0.003973344948316594,0.02231789206095309,0.02082913540548639,0.03782245657753309,0.008325400394814747,0.0076184868402289,0.02819247633119997,0.01184490267228352,0.01309829949060078,0.0141207010615454,0.01898681615248326,0.006355597883425556,0.009452346838211176,0.005733476839708576,0.0169900192496873,0.006468216241780515,0.007553325194935959],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.01766923098129314,0,0,0,0.002828195557424438,0,0.00502771437185138,0,0,0,0,0.004144628801828324,0.001730320088155684,0.002960948117172106,0.001761714695966153,0.0006236499345253106,0.004667593919110283,0.0007680188762324304,0.001384667416909661,0.001340750448797079,0.002124987434421812,0.140092692842879,0.04600489932179704,0.05837992064666763,0.2164480795208409,0.03015004323055594,0.07876988051983851,0.03160125162682196,0.05797800892720097,0.06337867647813776,0.04650635793962526,0.05409433823168548,0.03548586002436011,0.03437987556031626,0.02739017913892541,0.05719587746523087,0.02537802795961669,0.03947930822896141,0.0199660998876164,0.0208922728065979,0.03070785473098871,0.02457660403943777,0.0301857496886306,0.01854704787502626,0.02036446291320904,0.1245268380825591,0.07109848077004997,0.1394631437670394,0.1590230788316382,0.1574502257595699,0.3045768713433756,0.02558196560266539,0.03252424891038103,0.03355341695901411,0.04022171497481104,0.04733254595272479,0.03064687911194736,0.008594968890079066,0,0.009947109124387978,0.004037413539029928,0.008882844351516319,0.01291924110375179,0.005924674377990451,0.005158919594806103,0.006912169886091874,0.004154002250728981,0.002904959305727004,0.002833316579229083,0.01556585476031989,0.01254679072412647,0.006486657849629737,0,0.00670000960679021,0.005251325367989234,0.01655303656643055,0.007070488893561096,0.01118447231967137,0.02388164326629405,0.0118331364881812,0.01290394909976731,0.00572997926005271,0.0224101465682117,0.002486777281096995,0.02191738778330533,0.0167787059973086,0.0135064793357405,0.01340847359229418,0.008598199324676838,0.00563213842570449,0.007477204051312167,0.0129605876717051,0.01009369031350361,0.01556585476031989,0.004182263574708822,0,0.01325192323596985,0.00670000960679021,0.0157539761039677,0.006019286024156564,0.007070488893561096,0.003728157439890457,0.003770785778888535,0.003380896139480343,0.001612993637470914,0,0.002490016285356856,0.0008289257603656649,0.001153546725437122,0.0009869827057240352,0.002936191159943588,0.0006236499345253106,0.0007369885135437291,0.0007680188762324304,0.0008308004501457963,0.0006703752243985394,0.0005312468586054531,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.2978366448985746,0.3306708206380307,0.2487374512123195,0.2093358422049822,0.1143127728230834,0.2270858825915834,0.5250739369448015,0.5650095344752502,0.5308361538809747,0.4902854857374528,0.4994089450796171,0.590479124337087,0.6025288465116098,0.608773865813673,0.5941507049719387,0.5958381434631403,0.6225451312473658,0.5686720640660807,0.6660253636746077,0.6424855300933795,0.6651774602689089,0.6458405468439375,0.640877476346194,0.6719565997185339,0.2978366448985746,0.1606115414527577,0.1316845329947574,0.1196204812599898,0.1295544758661612,0.1792783283617763,0.1875264060517148,0.1637243537399872,0.1179635897513277,0.1728752290750556,0.1296175887992899,0.1129929188546278,0.134607508263232,0.1206578833144217,0.1579387949925407,0.1440915771773607,0.1523248725392491,0.1619135737965924,0.1200800446038077,0.1477939804468277,0.1743501474580272,0.1306613823094671,0.1756591503097388,0.1249124342573007,0.05255940792327786,0.2172979678478487,0.3511587546526864,0.1694623484516523,0.4953553489000282,0.09561510845961405,0.1807072640134706,0.1284112578352841,0.1769453846269916,0.1587051283311986,0.1753649730813923,0.1348625160522977,0.1442223302820342,0.1316267817975509,0.1184540962444055,0.1518803110788397,0.09713470132937622,0.1684954263899498,0.1431186578126777,0.1260431682301248,0.09091525841045642,0.1225728205474525,0.1250369167910821,0.1241091388601476,0,0.03752077039778108,0.02330528686432726,0.03966208979734731,0.005977711420702094,0.02810058817555655,0.005387370633987689,0.01769171138367499,0.03120507427199614,0.02484559022204034,0.036446573643521,0.04050787944122092,0.0259024041602504,0.03648468508709916,0.0355803796918088,0.02698663609255718,0.0249906386898961,0.02544470275626821,0.01352455996752637,0.01931872294088128,0.01429709994815168,0.02145957488579714,0.01240154212449575,0.03420592125753864,0.02800803091674752,0.06753738671600595,0.02913160858040908,0.04759450775681677,0.04184397994491465,0.04683431362592757,0.01616211190196307,0.01516432404314999,0.01783147101256922,0.02032821018166937,0.01214885788117367,0.0347210395210465,0.03108288499230049,0.02280292817943697,0.0088950949229522,0.01868305575638575,0.0249906386898961,0.01272235137813411,0.01070694330762504,0.006585928275300437,0.007840345132857371,0.02145957488579714,0.006800845681175086,0.005067543890005724,0.02800803091674752,0.05252907855689352,0.01165264343216363,0.007932417959469463,0.01195542284140419,0.01873372545037103,0.005387370633987689,0.002527387340524999,0.01337360325942692,0.01355214012111291,0.03037214470293417,0.005786839920174417,0.0129512020801252,0.02736351381532437,0.01037761074344423,0.008303580336171441,0.00892522810353432,0.01166215542995627,0.005071709987822386,0.007903113930360525,0.008762738677899416,0.01397367666982139,0.00280034822166033,0.003800657917504293]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>normalized_subtype<\/th>\n      <th>sample<\/th>\n      <th>normalized_ref<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n      <th>subtype_depth<\/th>\n      <th>proportion_min<\/th>\n      <th>proportion_max<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":[6,7],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[4,5,6,7,8,9,10,11]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"normalized_subtype","targets":1},{"name":"sample","targets":2},{"name":"normalized_ref","targets":3},{"name":"sum_min","targets":4},{"name":"sum_max","targets":5},{"name":"mf_min","targets":6},{"name":"mf_max","targets":7},{"name":"group_depth","targets":8},{"name":"subtype_depth","targets":9},{"name":"proportion_min","targets":10},{"name":"proportion_max","targets":11}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script><div class="section level4">
<h4 id="selecting-variation-types">Selecting Variation Types<a class="anchor" aria-label="anchor" href="#selecting-variation-types"></a>
</h4>
<p><code><a href="../reference/calculate_mf.html">calculate_mf()</a></code> can be used on a user-defined subset of
<code>variation_type</code> values. The <code>variant_types</code>
parameter can be set to a character string of
<code>variation_type</code> values that the user wants to include in the
mutation counts. When calculating group mutation sums, only variants of
the specified <code>variation_type</code>s will be counted. The
<code>total_depth</code> for records of excluded
<code>variation_types</code> will still be included in the
<code>group_depth</code> and the <code>subtype_depth</code>, if
applicable.</p>
<p>By default the function will calculate summary values based on all
mutation types.</p>
<p><em>Example 3.5. Calculate global mutation frequencies per sample
including only</em> <em>insertion and deletion mutations in the
count.</em></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_global_indels</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"insertion"</span>, <span class="st">"deletion"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 10.</strong> Indel Mutation Frequency per Sample.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-3f27c09be0c60bb52829" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-3f27c09be0c60bb52829">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],[31,25,55,59,46,69,34,41,47,52,52,35,24,10,67,45,39,44,51,65,78,72,53,65],[34,28,61,85,56,73,38,64,78,69,60,41,30,11,81,51,49,56,86,146,123,124,96,131],[5.482959617694662e-08,3.308741517874028e-08,8.594969209812051e-08,8.739678863285015e-08,7.690969862247424e-08,1.128766687073242e-07,5.242611362297728e-08,5.748489684583428e-08,7.01770494990056e-08,7.591881198447066e-08,7.253316980562816e-08,5.055511451273655e-08,8.06285262382022e-08,5.784905158880656e-08,9.966523147890397e-08,6.073684875946445e-08,6.988777823769311e-08,5.996775563763242e-08,7.251270347564009e-08,7.347671358415201e-08,1.049357721637313e-07,9.005946998298291e-08,7.820641675327117e-08,9.27030529020725e-08],[6.013568612955437e-08,3.705790500018911e-08,9.53260221451882e-08,1.259106276913943e-07,9.362919832301211e-08,1.194202437048502e-07,5.859389169626872e-08,8.973252190569253e-08,1.164640395940944e-07,1.007384235947784e-07,8.369211900649403e-08,5.922170557206281e-08,1.007856577977527e-07,6.363395674768722e-08,1.204908022356899e-07,6.883509526072637e-08,8.780772137556315e-08,7.632259808425945e-08,1.222763235079421e-07,1.650400028197876e-07,1.654756407197301e-07,1.551024205262483e-07,1.416569058172459e-07,1.86832306618023e-07],[565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":[4,5],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[2,3,4,5,6]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"sum_min","targets":2},{"name":"sum_max","targets":3},{"name":"mf_min","targets":4},{"name":"mf_max","targets":5},{"name":"group_depth","targets":6}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script><p>Users may also supply a list of variation_types to exclude to the
<code>variant_types</code> parameter, so long as the value is preceeded
by a “-”.</p>
<p><em>Example 3.6. Calculate mutation frequencies at the “type” subtype
resolution,</em> <em>excluding ambiguous and uncategorized
mutations.</em></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"type"</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-ambiguous"</span>, <span class="st">"-uncategorized"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 11.</strong> Frequency and Proportions of Variation
Types per Sample.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-416566eb193bf50d04e6" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-416566eb193bf50d04e6">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144"],["complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],[0,0,0,3,0,0,0,2,0,4,0,0,0,0,4,3,1,2,1,4,3,3,4,3,18,10,18,26,9,14,20,19,23,21,26,21,19,10,55,39,32,32,41,51,53,57,43,55,13,15,37,33,37,55,14,22,24,31,26,14,5,0,12,6,7,12,10,14,25,15,10,10,2,2,2,0,2,1,8,4,4,11,7,8,2,7,3,16,8,14,19,13,12,12,14,22,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,2,2,3,41,84,81,58,47,51,166,178,194,203,167,177,204,142,304,405,269,402,608,638,664,681,586,593],[0,0,0,4,0,0,0,2,0,4,0,0,0,0,5,3,3,3,2,19,3,5,6,12,18,11,18,49,9,15,21,41,51,37,32,22,24,11,69,44,40,34,67,125,96,109,83,115,16,17,43,36,47,58,17,23,27,32,28,19,6,0,12,7,9,22,19,21,27,15,13,16,2,3,2,0,2,1,11,5,9,19,7,8,4,9,3,38,17,23,43,35,22,27,58,57,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,3,3,3,41,96,111,62,107,52,272,281,322,316,237,258,302,176,500,739,428,730,1383,1720,1684,1559,1915,2429],[0,0,0,4.443904506755093e-09,0,0,0,2.804141309552892e-09,0,5.83990861419005e-09,0,0,0,0,5.950163073367402e-09,4.049123250630963e-09,1.791994313787003e-09,2.725807074437837e-09,1.421817715208629e-09,4.521643912870893e-09,4.035991237066589e-09,3.752477915957621e-09,5.902371075718579e-09,4.278602441634115e-09,3.183653971564643e-08,1.323496607149611e-08,2.812899014120308e-08,3.851383905854413e-08,1.504754973048409e-08,2.290251249134113e-08,3.083889036645722e-08,2.663934244075247e-08,3.43419603931304e-08,3.065952022449777e-08,3.626658490281408e-08,3.033306870764193e-08,6.383091660524341e-08,5.784905158880656e-08,8.181474225880177e-08,5.263860225820251e-08,5.734381804118409e-08,4.36129131910054e-08,5.82945263235538e-08,5.765095988910388e-08,7.130251185484307e-08,7.12970804031948e-08,6.345048906397472e-08,7.844104476329212e-08,2.29930564613002e-08,1.985244910724416e-08,5.782070195691743e-08,4.888294957430602e-08,6.186214889199014e-08,8.997415621598302e-08,2.158722325652006e-08,3.084555440508181e-08,3.58350891058752e-08,4.525929175997289e-08,3.626658490281408e-08,2.022204580509462e-08,1.679760963295879e-08,0,1.78504892201022e-08,8.098246501261925e-09,1.254396019650902e-08,1.635484244662703e-08,1.421817715208629e-08,1.582575369504812e-08,3.363326030888824e-08,1.87623895797881e-08,1.475592768929645e-08,1.426200813878038e-08,3.537393301738492e-09,2.646993214299222e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635893749381509e-09,1.233555614658289e-08,5.608282619105783e-09,5.9725148509792e-09,1.605974868902264e-08,9.764080550757637e-09,1.155545474576835e-08,6.719043853183517e-09,4.049433611216459e-08,4.462622305025551e-09,2.159532400336514e-08,1.433595451029602e-08,1.908064952106486e-08,2.701453658896395e-08,1.46953427168304e-08,1.614396494826636e-08,1.500991166383048e-08,2.065829876501503e-08,3.137641790531685e-08,3.537393301738492e-09,1.323496607149611e-09,0,4.443904506755093e-09,3.343899940107575e-09,4.907681248144528e-09,6.167778073291445e-09,7.010353273882228e-09,4.479386138234399e-09,4.379931460642538e-09,2.789737300216468e-09,1.444431843221044e-09,0,5.784905158880655e-09,1.48754076834185e-09,2.699415500420642e-09,1.791994313787003e-09,6.814517686094594e-09,2.843635430417258e-09,3.391232934653169e-09,4.035991237066589e-09,2.501651943971747e-09,2.95118553785929e-09,4.278602441634115e-09,7.251656268563909e-08,1.111737150005673e-07,1.265804556354139e-07,8.591548713059846e-08,7.858164859252802e-08,8.343058121845699e-08,2.55962790041595e-07,2.495685765502074e-07,2.896669702724912e-07,2.963753621701451e-07,2.329430645680751e-07,2.556644362501248e-07,6.853424730247187e-07,8.214565325610531e-07,4.522123935759225e-07,5.4663163883518e-07,4.820464704087038e-07,5.478872219620053e-07,8.644651708468465e-07,7.212022041029073e-07,8.932993938040717e-07,8.518124869223799e-07,8.646973625927719e-07,8.457370826296767e-07],[0,0,0,5.92520600900679e-09,0,0,0,2.804141309552892e-09,0,5.83990861419005e-09,0,0,0,0,7.437703841709251e-09,4.049123250630963e-09,5.375982941361009e-09,4.088710611656756e-09,2.843635430417258e-09,2.147780858613674e-08,4.035991237066589e-09,6.254129859929368e-09,8.853556613577869e-09,1.711440976653646e-08,3.183653971564643e-08,1.455846267864572e-08,2.812899014120308e-08,7.258377361033318e-08,1.504754973048409e-08,2.453840624072264e-08,3.238083488478009e-08,5.748489684583428e-08,7.61495643499848e-08,5.401915468125797e-08,4.463579680346348e-08,3.177750055086297e-08,8.06285262382022e-08,6.363395674768722e-08,1.026403130155877e-07,5.938714100925412e-08,7.167977255148012e-08,4.633872026544324e-08,9.526178691897815e-08,1.413013722772154e-07,1.291517195861308e-07,1.363400309464602e-07,1.224741998211605e-07,1.640130935959744e-07,2.829914641390793e-08,2.249944232154339e-08,6.719703200398513e-08,5.332685408106111e-08,7.858164859252802e-08,9.488183746412755e-08,2.621305681148864e-08,3.224762505985825e-08,4.03144752441096e-08,4.67192689135204e-08,3.905632220303055e-08,2.744420502119984e-08,2.015713155955055e-08,0,1.78504892201022e-08,9.447954251472247e-09,1.612794882408303e-08,2.998387781881621e-08,2.701453658896395e-08,2.373863054257218e-08,3.63239211335993e-08,1.87623895797881e-08,1.918270599608538e-08,2.281921302204861e-08,3.537393301738492e-09,3.970489821448833e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635893749381509e-09,1.696138970155147e-08,7.010353273882228e-09,1.34381584147032e-08,2.773956591740274e-08,9.764080550757637e-09,1.155545474576835e-08,1.343808770636703e-08,5.20641464299259e-08,4.462622305025551e-09,5.12888945079922e-08,3.046390333437905e-08,3.134678135603513e-08,6.113816175397105e-08,3.956438423762031e-08,2.959726907182165e-08,3.377230124361859e-08,8.55843805979194e-08,8.129344639104818e-08,3.537393301738492e-09,1.323496607149611e-09,0,4.443904506755093e-09,3.343899940107575e-09,4.907681248144528e-09,6.167778073291445e-09,7.010353273882228e-09,4.479386138234399e-09,4.379931460642538e-09,2.789737300216468e-09,1.444431843221044e-09,0,5.784905158880655e-09,1.48754076834185e-09,2.699415500420642e-09,1.791994313787003e-09,6.814517686094594e-09,2.843635430417258e-09,3.391232934653169e-09,4.035991237066589e-09,3.752477915957621e-09,4.426778306788935e-09,4.278602441634115e-09,7.251656268563909e-08,1.270556742863627e-07,1.734621058707523e-07,9.184069313960525e-08,1.788986467957553e-07,8.506647496783849e-08,4.194089089838183e-07,3.939818539921812e-07,4.807874455038256e-07,4.61352780521014e-07,3.305838700756514e-07,3.726634155510294e-07,1.014575621830711e-06,1.018143307962995e-06,7.437703841709251e-07,9.974340274054271e-07,7.669735663008373e-07,9.949195821698107e-07,1.966373900133534e-06,1.944306882534484e-06,2.265536414406712e-06,1.950037690325977e-06,2.82576015250027e-06,3.464241776909755e-06],[565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532],[0,0,0,0.02439024390243902,0,0,0,0.008695652173913044,0,0.01465201465201465,0,0,0,0,0.01055408970976253,0.006369426751592356,0.003144654088050315,0.004282655246252677,0.001468428781204112,0.005532503457814661,0.003947368421052632,0.003896103896103896,0.006069802731411229,0.004373177842565598,0.2368421052631579,0.0892857142857143,0.1304347826086956,0.2113821138211382,0.09278350515463918,0.1129032258064516,0.09433962264150943,0.08260869565217391,0.09274193548387097,0.07692307692307693,0.1140350877192982,0.09502262443438914,0.08260869565217391,0.0625,0.1451187335092348,0.08280254777070063,0.1006289308176101,0.06852248394004283,0.06020558002936857,0.07053941908713693,0.06973684210526315,0.07402597402597402,0.06525037936267071,0.08017492711370262,0.1710526315789474,0.1339285714285714,0.2681159420289855,0.2682926829268293,0.3814432989690721,0.4435483870967742,0.06603773584905659,0.09565217391304347,0.09677419354838708,0.1135531135531135,0.1140350877192982,0.06334841628959276,0.02173913043478261,0,0.0316622691292876,0.01273885350318471,0.0220125786163522,0.02569593147751606,0.01468428781204112,0.01936376210235131,0.03289473684210526,0.01948051948051948,0.01517450682852807,0.01457725947521866,0.02631578947368421,0.01785714285714286,0.0144927536231884,0,0.02061855670103093,0.00806451612903226,0.03773584905660377,0.01739130434782609,0.01612903225806452,0.0402930402930403,0.03070175438596491,0.03619909502262444,0.008695652173913044,0.04375,0.0079155672823219,0.03397027600849257,0.02515723270440252,0.02997858672376874,0.02790014684287812,0.01798063623789765,0.01578947368421053,0.01558441558441558,0.0212443095599393,0.03206997084548105,0.02631578947368421,0.008928571428571428,0,0.02439024390243902,0.02061855670103093,0.02419354838709677,0.01886792452830189,0.02173913043478261,0.01209677419354839,0.01098901098901099,0.008771929824561403,0.004524886877828055,0,0.00625,0.002638522427440633,0.004246284501061571,0.003144654088050315,0.01070663811563169,0.002936857562408223,0.004149377593360995,0.003947368421052632,0.002597402597402597,0.003034901365705615,0.004373177842565598,0.5394736842105263,0.75,0.5869565217391305,0.4715447154471544,0.4845360824742268,0.4112903225806452,0.7830188679245282,0.7739130434782608,0.782258064516129,0.7435897435897436,0.7324561403508772,0.8009049773755657,0.8869565217391304,0.8875,0.8021108179419525,0.8598726114649681,0.8459119496855346,0.860813704496788,0.8928046989720998,0.8824343015214384,0.8736842105263158,0.8844155844155844,0.8892261001517451,0.8644314868804664],[0,0,0,0.02597402597402598,0,0,0,0.005602240896358543,0,0.009732360097323601,0,0,0,0,0.008474576271186441,0.003601440576230492,0.006024096385542168,0.003671970624235006,0.001319261213720317,0.00988039521580863,0.001634877384196185,0.002910360884749709,0.002887391722810394,0.004559270516717325,0.2278481012658228,0.0859375,0.103448275862069,0.3181818181818182,0.05389221556886227,0.1162790697674419,0.06461538461538463,0.1148459383753501,0.1237864077669903,0.09002433090024331,0.1045751633986928,0.07142857142857142,0.07142857142857142,0.05583756345177664,0.1169491525423729,0.05282112845138055,0.08032128514056223,0.0416156670746634,0.0441952506596306,0.06500260010400416,0.05231607629427792,0.06344586728754366,0.03994225216554378,0.04369300911854104,0.2025316455696203,0.1328125,0.2471264367816092,0.2337662337662337,0.281437125748503,0.4496124031007752,0.05230769230769231,0.06442577030812324,0.06553398058252427,0.07785888077858881,0.09150326797385619,0.06168831168831168,0.01785714285714286,0,0.02033898305084746,0.008403361344537813,0.01807228915662651,0.02692778457772338,0.01253298153034301,0.0109204368174727,0.01471389645776567,0.008731082654249126,0.006256015399422521,0.0060790273556231,0.02531645569620253,0.0234375,0.01149425287356322,0,0.01197604790419162,0.007751937984496125,0.03384615384615385,0.01400560224089636,0.02184466019417476,0.0462287104622871,0.02287581699346405,0.02597402597402598,0.0119047619047619,0.04568527918781726,0.005084745762711864,0.04561824729891956,0.03413654618473895,0.02815177478580171,0.0283641160949868,0.01820072802912116,0.01198910081743869,0.01571594877764843,0.02791145332050048,0.02165653495440729,0.02531645569620253,0.0078125,0,0.01948051948051948,0.01197604790419162,0.02325581395348837,0.01230769230769231,0.01400560224089636,0.007281553398058253,0.007299270072992701,0.0065359477124183,0.003246753246753247,0,0.005076142131979694,0.001694915254237288,0.002400960384153661,0.002008032128514056,0.006119951040391677,0.001319261213720317,0.0015600624024961,0.001634877384196185,0.001746216530849825,0.001443695861405197,0.001139817629179331,0.5189873417721519,0.75,0.6379310344827587,0.4025974025974026,0.6407185628742514,0.4031007751937984,0.836923076923077,0.7871148459383753,0.7815533980582524,0.7688564476885644,0.7745098039215687,0.8376623376623377,0.8988095238095238,0.8934010152284263,0.847457627118644,0.8871548619447779,0.8594377510040161,0.8935128518971848,0.912269129287599,0.8944357774310973,0.9177111716621253,0.9074505238649592,0.9215591915303176,0.9228723404255319]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>variation_type<\/th>\n      <th>sample<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n      <th>proportion_min<\/th>\n      <th>proportion_max<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":[5,6],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[3,4,5,6,7,8,9]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"variation_type","targets":1},{"name":"sample","targets":2},{"name":"sum_min","targets":3},{"name":"sum_max","targets":4},{"name":"mf_min","targets":5},{"name":"mf_max","targets":6},{"name":"group_depth","targets":7},{"name":"proportion_min","targets":8},{"name":"proportion_max","targets":9}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script><p><em>Example 3.7. Include only snv mutations at the base_96
resolution</em></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_96</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_96"</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="st">"snv"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 12.</strong> Frequency and Proportions of 96-base
Trinucleotide Subtypes per Sample.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-72cbf064100ce560a04c" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-72cbf064100ce560a04c">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144"],["complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv","snv"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],[0,0,0,3,0,0,0,2,0,4,0,0,0,0,4,3,1,2,1,4,3,3,4,3,18,10,18,26,9,14,20,19,23,21,26,21,19,10,55,39,32,32,41,51,53,57,43,55,13,15,37,33,37,55,14,22,24,31,26,14,5,0,12,6,7,12,10,14,25,15,10,10,2,2,2,0,2,1,8,4,4,11,7,8,2,7,3,16,8,14,19,13,12,12,14,22,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,2,2,3,41,84,81,58,47,51,166,178,194,203,167,177,204,142,304,405,269,402,608,638,664,681,586,593],[0,0,0,4,0,0,0,2,0,4,0,0,0,0,5,3,3,3,2,19,3,5,6,12,18,11,18,49,9,15,21,41,51,37,32,22,24,11,69,44,40,34,67,125,96,109,83,115,16,17,43,36,47,58,17,23,27,32,28,19,6,0,12,7,9,22,19,21,27,15,13,16,2,3,2,0,2,1,11,5,9,19,7,8,4,9,3,38,17,23,43,35,22,27,58,57,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,3,3,3,41,96,111,62,107,52,272,281,322,316,237,258,302,176,500,739,428,730,1383,1720,1684,1559,1915,2429],[0,0,0,4.443904506755093e-09,0,0,0,2.804141309552892e-09,0,5.83990861419005e-09,0,0,0,0,5.950163073367402e-09,4.049123250630963e-09,1.791994313787003e-09,2.725807074437837e-09,1.421817715208629e-09,4.521643912870893e-09,4.035991237066589e-09,3.752477915957621e-09,5.902371075718579e-09,4.278602441634115e-09,3.183653971564643e-08,1.323496607149611e-08,2.812899014120308e-08,3.851383905854413e-08,1.504754973048409e-08,2.290251249134113e-08,3.083889036645722e-08,2.663934244075247e-08,3.43419603931304e-08,3.065952022449777e-08,3.626658490281408e-08,3.033306870764193e-08,6.383091660524341e-08,5.784905158880656e-08,8.181474225880177e-08,5.263860225820251e-08,5.734381804118409e-08,4.36129131910054e-08,5.82945263235538e-08,5.765095988910388e-08,7.130251185484307e-08,7.12970804031948e-08,6.345048906397472e-08,7.844104476329212e-08,2.29930564613002e-08,1.985244910724416e-08,5.782070195691743e-08,4.888294957430602e-08,6.186214889199014e-08,8.997415621598302e-08,2.158722325652006e-08,3.084555440508181e-08,3.58350891058752e-08,4.525929175997289e-08,3.626658490281408e-08,2.022204580509462e-08,1.679760963295879e-08,0,1.78504892201022e-08,8.098246501261925e-09,1.254396019650902e-08,1.635484244662703e-08,1.421817715208629e-08,1.582575369504812e-08,3.363326030888824e-08,1.87623895797881e-08,1.475592768929645e-08,1.426200813878038e-08,3.537393301738492e-09,2.646993214299222e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635893749381509e-09,1.233555614658289e-08,5.608282619105783e-09,5.9725148509792e-09,1.605974868902264e-08,9.764080550757637e-09,1.155545474576835e-08,6.719043853183517e-09,4.049433611216459e-08,4.462622305025551e-09,2.159532400336514e-08,1.433595451029602e-08,1.908064952106486e-08,2.701453658896395e-08,1.46953427168304e-08,1.614396494826636e-08,1.500991166383048e-08,2.065829876501503e-08,3.137641790531685e-08,3.537393301738492e-09,1.323496607149611e-09,0,4.443904506755093e-09,3.343899940107575e-09,4.907681248144528e-09,6.167778073291445e-09,7.010353273882228e-09,4.479386138234399e-09,4.379931460642538e-09,2.789737300216468e-09,1.444431843221044e-09,0,5.784905158880655e-09,1.48754076834185e-09,2.699415500420642e-09,1.791994313787003e-09,6.814517686094594e-09,2.843635430417258e-09,3.391232934653169e-09,4.035991237066589e-09,2.501651943971747e-09,2.95118553785929e-09,4.278602441634115e-09,7.251656268563909e-08,1.111737150005673e-07,1.265804556354139e-07,8.591548713059846e-08,7.858164859252802e-08,8.343058121845699e-08,2.55962790041595e-07,2.495685765502074e-07,2.896669702724912e-07,2.963753621701451e-07,2.329430645680751e-07,2.556644362501248e-07,6.853424730247187e-07,8.214565325610531e-07,4.522123935759225e-07,5.4663163883518e-07,4.820464704087038e-07,5.478872219620053e-07,8.644651708468465e-07,7.212022041029073e-07,8.932993938040717e-07,8.518124869223799e-07,8.646973625927719e-07,8.457370826296767e-07],[0,0,0,5.92520600900679e-09,0,0,0,2.804141309552892e-09,0,5.83990861419005e-09,0,0,0,0,7.437703841709251e-09,4.049123250630963e-09,5.375982941361009e-09,4.088710611656756e-09,2.843635430417258e-09,2.147780858613674e-08,4.035991237066589e-09,6.254129859929368e-09,8.853556613577869e-09,1.711440976653646e-08,3.183653971564643e-08,1.455846267864572e-08,2.812899014120308e-08,7.258377361033318e-08,1.504754973048409e-08,2.453840624072264e-08,3.238083488478009e-08,5.748489684583428e-08,7.61495643499848e-08,5.401915468125797e-08,4.463579680346348e-08,3.177750055086297e-08,8.06285262382022e-08,6.363395674768722e-08,1.026403130155877e-07,5.938714100925412e-08,7.167977255148012e-08,4.633872026544324e-08,9.526178691897815e-08,1.413013722772154e-07,1.291517195861308e-07,1.363400309464602e-07,1.224741998211605e-07,1.640130935959744e-07,2.829914641390793e-08,2.249944232154339e-08,6.719703200398513e-08,5.332685408106111e-08,7.858164859252802e-08,9.488183746412755e-08,2.621305681148864e-08,3.224762505985825e-08,4.03144752441096e-08,4.67192689135204e-08,3.905632220303055e-08,2.744420502119984e-08,2.015713155955055e-08,0,1.78504892201022e-08,9.447954251472247e-09,1.612794882408303e-08,2.998387781881621e-08,2.701453658896395e-08,2.373863054257218e-08,3.63239211335993e-08,1.87623895797881e-08,1.918270599608538e-08,2.281921302204861e-08,3.537393301738492e-09,3.970489821448833e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635893749381509e-09,1.696138970155147e-08,7.010353273882228e-09,1.34381584147032e-08,2.773956591740274e-08,9.764080550757637e-09,1.155545474576835e-08,1.343808770636703e-08,5.20641464299259e-08,4.462622305025551e-09,5.12888945079922e-08,3.046390333437905e-08,3.134678135603513e-08,6.113816175397105e-08,3.956438423762031e-08,2.959726907182165e-08,3.377230124361859e-08,8.55843805979194e-08,8.129344639104818e-08,3.537393301738492e-09,1.323496607149611e-09,0,4.443904506755093e-09,3.343899940107575e-09,4.907681248144528e-09,6.167778073291445e-09,7.010353273882228e-09,4.479386138234399e-09,4.379931460642538e-09,2.789737300216468e-09,1.444431843221044e-09,0,5.784905158880655e-09,1.48754076834185e-09,2.699415500420642e-09,1.791994313787003e-09,6.814517686094594e-09,2.843635430417258e-09,3.391232934653169e-09,4.035991237066589e-09,3.752477915957621e-09,4.426778306788935e-09,4.278602441634115e-09,7.251656268563909e-08,1.270556742863627e-07,1.734621058707523e-07,9.184069313960525e-08,1.788986467957553e-07,8.506647496783849e-08,4.194089089838183e-07,3.939818539921812e-07,4.807874455038256e-07,4.61352780521014e-07,3.305838700756514e-07,3.726634155510294e-07,1.014575621830711e-06,1.018143307962995e-06,7.437703841709251e-07,9.974340274054271e-07,7.669735663008373e-07,9.949195821698107e-07,1.966373900133534e-06,1.944306882534484e-06,2.265536414406712e-06,1.950037690325977e-06,2.82576015250027e-06,3.464241776909755e-06],[565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532,565388078,755574283,639909215,675082013,598104021,611286644,648531765,713230818,669734626,684942225,716913381,692313732,297661400,172863681,672250483,740901132,558037485,733727643,703325039,884634013,743311822,799471727,677693752,701163532],[0,0,0,0.02439024390243902,0,0,0,0.008695652173913044,0,0.01465201465201465,0,0,0,0,0.01055408970976253,0.006369426751592356,0.003144654088050315,0.004282655246252677,0.001468428781204112,0.005532503457814661,0.003947368421052632,0.003896103896103896,0.006069802731411229,0.004373177842565598,0.2368421052631579,0.0892857142857143,0.1304347826086956,0.2113821138211382,0.09278350515463918,0.1129032258064516,0.09433962264150943,0.08260869565217391,0.09274193548387097,0.07692307692307693,0.1140350877192982,0.09502262443438914,0.08260869565217391,0.0625,0.1451187335092348,0.08280254777070063,0.1006289308176101,0.06852248394004283,0.06020558002936857,0.07053941908713693,0.06973684210526315,0.07402597402597402,0.06525037936267071,0.08017492711370262,0.1710526315789474,0.1339285714285714,0.2681159420289855,0.2682926829268293,0.3814432989690721,0.4435483870967742,0.06603773584905659,0.09565217391304347,0.09677419354838708,0.1135531135531135,0.1140350877192982,0.06334841628959276,0.02173913043478261,0,0.0316622691292876,0.01273885350318471,0.0220125786163522,0.02569593147751606,0.01468428781204112,0.01936376210235131,0.03289473684210526,0.01948051948051948,0.01517450682852807,0.01457725947521866,0.02631578947368421,0.01785714285714286,0.0144927536231884,0,0.02061855670103093,0.00806451612903226,0.03773584905660377,0.01739130434782609,0.01612903225806452,0.0402930402930403,0.03070175438596491,0.03619909502262444,0.008695652173913044,0.04375,0.0079155672823219,0.03397027600849257,0.02515723270440252,0.02997858672376874,0.02790014684287812,0.01798063623789765,0.01578947368421053,0.01558441558441558,0.0212443095599393,0.03206997084548105,0.02631578947368421,0.008928571428571428,0,0.02439024390243902,0.02061855670103093,0.02419354838709677,0.01886792452830189,0.02173913043478261,0.01209677419354839,0.01098901098901099,0.008771929824561403,0.004524886877828055,0,0.00625,0.002638522427440633,0.004246284501061571,0.003144654088050315,0.01070663811563169,0.002936857562408223,0.004149377593360995,0.003947368421052632,0.002597402597402597,0.003034901365705615,0.004373177842565598,0.5394736842105263,0.75,0.5869565217391305,0.4715447154471544,0.4845360824742268,0.4112903225806452,0.7830188679245282,0.7739130434782608,0.782258064516129,0.7435897435897436,0.7324561403508772,0.8009049773755657,0.8869565217391304,0.8875,0.8021108179419525,0.8598726114649681,0.8459119496855346,0.860813704496788,0.8928046989720998,0.8824343015214384,0.8736842105263158,0.8844155844155844,0.8892261001517451,0.8644314868804664],[0,0,0,0.02597402597402598,0,0,0,0.005602240896358543,0,0.009732360097323601,0,0,0,0,0.008474576271186441,0.003601440576230492,0.006024096385542168,0.003671970624235006,0.001319261213720317,0.00988039521580863,0.001634877384196185,0.002910360884749709,0.002887391722810394,0.004559270516717325,0.2278481012658228,0.0859375,0.103448275862069,0.3181818181818182,0.05389221556886227,0.1162790697674419,0.06461538461538463,0.1148459383753501,0.1237864077669903,0.09002433090024331,0.1045751633986928,0.07142857142857142,0.07142857142857142,0.05583756345177664,0.1169491525423729,0.05282112845138055,0.08032128514056223,0.0416156670746634,0.0441952506596306,0.06500260010400416,0.05231607629427792,0.06344586728754366,0.03994225216554378,0.04369300911854104,0.2025316455696203,0.1328125,0.2471264367816092,0.2337662337662337,0.281437125748503,0.4496124031007752,0.05230769230769231,0.06442577030812324,0.06553398058252427,0.07785888077858881,0.09150326797385619,0.06168831168831168,0.01785714285714286,0,0.02033898305084746,0.008403361344537813,0.01807228915662651,0.02692778457772338,0.01253298153034301,0.0109204368174727,0.01471389645776567,0.008731082654249126,0.006256015399422521,0.0060790273556231,0.02531645569620253,0.0234375,0.01149425287356322,0,0.01197604790419162,0.007751937984496125,0.03384615384615385,0.01400560224089636,0.02184466019417476,0.0462287104622871,0.02287581699346405,0.02597402597402598,0.0119047619047619,0.04568527918781726,0.005084745762711864,0.04561824729891956,0.03413654618473895,0.02815177478580171,0.0283641160949868,0.01820072802912116,0.01198910081743869,0.01571594877764843,0.02791145332050048,0.02165653495440729,0.02531645569620253,0.0078125,0,0.01948051948051948,0.01197604790419162,0.02325581395348837,0.01230769230769231,0.01400560224089636,0.007281553398058253,0.007299270072992701,0.0065359477124183,0.003246753246753247,0,0.005076142131979694,0.001694915254237288,0.002400960384153661,0.002008032128514056,0.006119951040391677,0.001319261213720317,0.0015600624024961,0.001634877384196185,0.001746216530849825,0.001443695861405197,0.001139817629179331,0.5189873417721519,0.75,0.6379310344827587,0.4025974025974026,0.6407185628742514,0.4031007751937984,0.836923076923077,0.7871148459383753,0.7815533980582524,0.7688564476885644,0.7745098039215687,0.8376623376623377,0.8988095238095238,0.8934010152284263,0.847457627118644,0.8871548619447779,0.8594377510040161,0.8935128518971848,0.912269129287599,0.8944357774310973,0.9177111716621253,0.9074505238649592,0.9215591915303176,0.9228723404255319]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>variation_type<\/th>\n      <th>sample<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n      <th>proportion_min<\/th>\n      <th>proportion_max<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":[5,6],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[3,4,5,6,7,8,9]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"variation_type","targets":1},{"name":"sample","targets":2},{"name":"sum_min","targets":3},{"name":"sum_max","targets":4},{"name":"mf_min","targets":5},{"name":"mf_max","targets":6},{"name":"group_depth","targets":7},{"name":"proportion_min","targets":8},{"name":"proportion_max","targets":9}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
</div>
<div class="section level3">
<h3 id="correcting-depth">Correcting Depth<a class="anchor" aria-label="anchor" href="#correcting-depth"></a>
</h3>
<p>There may be cases when the same genomic position is represented
multiple times within the mutation data of a single sample. For
instance, we require that multiple different alternate alleles for a
single position be reported as seperate rows. Another common example of
this that we have observed are deletion calls. Often, for each deletion
called, a no_variant call at the same start position will also be
present. For data that includes the <code>total_depth</code> of each
record, we want to prevent double-counting the depth at these positions
during when summing the depth across groups. When set to TRUE,
<code>correct_depth</code> parameter will correct the depth at these
positions. For positions with the same sample, contig, and start values,
the <code>total_depth</code> will be retained for only one row. All
other rows in the group will have their <code>total_depth</code> set to
0 before beign summed. The import functions will automatically check for
duplicated rows and return a warning advising to correct the depth
should it find any in the mutation data. This is recommended for all
users whose data contain <code>total_depth</code> values. By default,
correct_depth is set to TRUE.</p>
<p><em>NOTE: The case of deletions and no_variants: When identifying a
deletion,</em> <em>some variant callers will re-align the sequences. As
such a second</em> <em>total_depth value will be calculated,
specifically for the deletion. The</em> <em>variant caller will then
provide both the no_variant and the deletion call,</em> <em>the former
with the initial depth, and the latter with the re-aligned depth</em>
<em>value. Generally, when correcting the depth, the function will
retain the</em> <em>total_depth of the first row in the group, and set
the rest to 0. However, in</em> <em>the case of deletions, the function
may prioritize retaining the re-aligned</em> <em>total_depth over the
total_depth assigned to the no_variant. This</em> <em>prioritization can
be activated with the <code>correct_depth_by_indel_priority</code></em>
<em>parameter.</em></p>
</div>
<div class="section level3">
<h3 id="precalculated-depth">Precalculated Depth<a class="anchor" aria-label="anchor" href="#precalculated-depth"></a>
</h3>
<p>For mutation data that does not include a <code>total_depth</code>
value for each sequenced site, precalculated depths for the specified
groups can be supplied separately to <code><a href="../reference/calculate_mf.html">calculate_mf()</a></code> through
the <code>precalc_depth_data</code> parameter. Depth should be provided
at the correct subtype resolution for each level of the specified
grouping variable. This parameter will accept either a data frame or a
filepath to import the depth data. Required columns are:</p>
<ol style="list-style-type: decimal">
<li>The user-defined cols_to_group variable(s)</li>
<li>
<code>group_depth</code>: the <code>total_depth</code> summed across
the cols_to_group.</li>
<li>The context column for the specified
<code>subtype_resolution</code>. Only applicable if using the SNV
resolutions (base_6, base_12, base_96, base_192). Column names are
listed in the table above.</li>
<li>
<code>subtype_depth</code>: the <code>total_depth</code> summed
across the cols_to_group for each context. Only applicable if using the
SNV resolutions.</li>
</ol>
<p><em>Example 3.8. Use precalculated depth values to calculate the
global per sample MF.</em></p>
<p><strong>Table 13. </strong> ‘sample_depth’ - Precalculated Depth Data
per sample.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_depth</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">example_data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>,</span>
<span>  group_depth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">565395266</span>, <span class="fl">755574283</span>, <span class="fl">639909215</span>, <span class="fl">675090988</span>, <span class="fl">598104021</span>,</span>
<span>                  <span class="fl">611295330</span>, <span class="fl">648531765</span>, <span class="fl">713240735</span>, <span class="fl">669734626</span>, <span class="fl">684951248</span>,</span>
<span>                  <span class="fl">716913381</span>, <span class="fl">692323218</span>, <span class="fl">297661400</span>, <span class="fl">172863681</span>, <span class="fl">672259724</span>,</span>
<span>                  <span class="fl">740901132</span>, <span class="fl">558051386</span>, <span class="fl">733727643</span>, <span class="fl">703349287</span>, <span class="fl">884821671</span>,</span>
<span>                  <span class="fl">743311822</span>, <span class="fl">799605045</span>, <span class="fl">677693752</span>, <span class="fl">701163532</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html" class="external-link">datatable</a></span><span class="op">(</span><span class="va">sample_depth</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item" id="htmlwidget-d11fc4360aa0230696d7" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-d11fc4360aa0230696d7">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],[565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>group_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":2},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"group_depth","targets":2}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_global_precalc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  calculate_depth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  precalc_depth_data <span class="op">=</span> <span class="va">sample_depth</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 14.</strong> Mutation Frequency per Sample,
Precalculated Depth Example.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-21c7483268bafca56cec" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-21c7483268bafca56cec">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],[76,112,138,123,97,124,212,230,248,273,228,221,230,160,379,471,318,467,681,723,760,770,659,686],[79,128,174,154,167,129,325,357,412,411,306,308,336,197,590,833,498,817,1516,1923,1835,1718,2078,2632],[1.344192365416799e-07,1.482316200007564e-07,2.156555910825569e-07,1.821976625171598e-07,1.621791470952174e-07,2.02847942581207e-07,3.268922378844466e-07,3.224717668432104e-07,3.702959207607104e-07,3.985685124264494e-07,3.180300522246773e-07,3.19215063505208e-07,7.726900431161044e-07,9.25584825420905e-07,5.637702014110249e-07,6.357123503490611e-07,5.698399967776444e-07,6.36475951881235e-07,9.682244833213288e-07,8.171138023584868e-07,1.022451113390202e-06,9.629754149437615e-07,9.72415634724636e-07,9.783737583203342e-07],[1.397252590367462e-07,1.694075657151502e-07,2.719135713649631e-07,2.281173985987204e-07,2.792156449989826e-07,2.110272951046428e-07,5.011319684549298e-07,5.0053226418707e-07,6.151690296508575e-07,6.000427055211381e-07,4.268298069331196e-07,4.448789120344076e-07,1.128799367334831e-06,1.139626316299489e-06,8.776369890039701e-07,1.124306555925197e-06,8.923909383498959e-07,1.113492189907857e-06,2.155401346130888e-06,2.173319283451411e-06,2.468681306672397e-06,2.148560731004392e-06,3.066281773835802e-06,3.753760542126997e-06],[565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":[4,5],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[2,3,4,5,6]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"sum_min","targets":2},{"name":"sum_max","targets":3},{"name":"mf_min","targets":4},{"name":"mf_max","targets":5},{"name":"group_depth","targets":6}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script><p><em>Example 3.9. Use precalculated depth values to calculate the
base_6 per sample MF.</em></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mf_data_6_precalc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span>,</span>
<span>  calculate_depth <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  precalc_depth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Example_files"</span>,</span>
<span>                              <span class="st">"precalc_depth_base_6_example.txt"</span>,</span>
<span>                              package <span class="op">=</span> <span class="st">"MutSeqR"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 16.</strong> Precalculated Depth Data per sample for
base_6 subtype resolution.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-1834a22cd196f3aa03a1" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-1834a22cd196f3aa03a1">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],["C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T"],[565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532],[251172937,334472748,283692005,299159586,262916990,268590931,286230890,314181161,296330977,303790726,317894855,306379366,133044089,78482517,296381278,329192428,249502775,327318880,314149620,389791799,330705935,355948966,299149912,309135697,314222329,421101535,356217210,375931402,335187031,342704399,362300875,399059574,373403649,381160522,399018526,385943852,164617311,94381164,375878446,411708704,308548611,406408763,389199667,495029872,412605887,443656079,378543840,392027835]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>sample<\/th>\n      <th>normalized_ref<\/th>\n      <th>group_depth<\/th>\n      <th>subtype_depth<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[3,4]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"sample","targets":1},{"name":"normalized_ref","targets":2},{"name":"group_depth","targets":3},{"name":"subtype_depth","targets":4}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script><p><strong>Table 17.</strong> Frequency and Proportions of 6-base
Subtypes per Sample, Precalculated Depth Example.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-28515d92cb327f90c9eb" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-28515d92cb327f90c9eb">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251","252","253","254","255","256","257","258","259","260","261","262","263","264","265","266","267","268","269","270","271","272","273","274","275","276","277","278","279","280","281","282","283","284","285","286","287","288","289","290","291","292","293","294","295","296","297","298","299","300","301","302","303","304","305","306","307","308","309","310","311","312"],["ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","ambiguous","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","complex","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","deletion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","insertion","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","mnv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","sv","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","uncategorized","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;A","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;G","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","C&gt;T","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;A","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;C","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G","T&gt;G"],["dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1","dna00973.1","dna00974.1","dna00975.1","dna00976.1","dna00977.1","dna00978.1","dna00979.1","dna00980.1","dna00981.1","dna00982.1","dna00983.1","dna00984.1","dna00985.1","dna00986.1","dna00987.1","dna00988.1","dna00989.1","dna00990.1","dna00991.1","dna00992.1","dna00993.1","dna00994.1","dna00995.1","dna00996.1"],["N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","N","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","C","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T","T"],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,2,0,4,0,0,0,0,4,3,1,2,1,4,3,3,4,3,18,10,18,26,9,14,20,19,23,21,26,21,19,10,55,39,32,32,41,51,53,57,43,55,13,15,37,33,37,55,14,22,24,31,26,14,5,0,12,6,7,12,10,14,25,15,10,10,2,2,2,0,2,1,8,4,4,11,7,8,2,7,3,16,8,14,19,13,12,12,14,22,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,2,2,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,17,33,33,21,15,19,97,107,102,113,85,99,118,90,187,245,175,230,415,399,452,444,382,400,17,16,18,12,16,15,35,31,28,34,26,26,35,16,52,63,40,68,88,103,105,92,95,85,3,17,19,13,6,7,26,27,42,34,34,24,33,20,45,62,30,69,76,93,68,82,74,72,0,5,4,5,1,3,1,7,8,9,10,14,8,6,11,19,9,18,17,26,19,27,18,17,2,9,5,6,7,5,5,5,8,7,3,12,8,5,4,9,10,7,7,9,15,21,12,13,2,4,2,1,2,2,2,1,6,6,9,2,2,5,5,7,5,10,5,8,5,15,5,6],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4,0,0,0,2,0,4,0,0,0,0,5,3,3,3,2,19,3,5,6,12,18,11,18,49,9,15,21,41,51,37,32,22,24,11,69,44,40,34,67,125,96,109,83,115,16,17,43,36,47,58,17,23,27,32,28,19,6,0,12,7,9,22,19,21,27,15,13,16,2,3,2,0,2,1,11,5,9,19,7,8,4,9,3,38,17,23,43,35,22,27,58,57,2,1,0,3,2,3,4,5,3,3,2,1,0,1,1,2,1,5,2,3,3,3,3,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,17,35,34,21,15,19,154,176,189,173,131,162,188,111,316,459,282,432,954,1152,1156,1038,1266,1673,17,17,18,12,17,15,55,51,42,61,34,31,42,22,84,111,69,123,172,265,303,210,347,311,3,23,48,17,65,8,53,40,63,56,46,37,45,24,63,117,44,128,205,226,158,197,247,309,0,5,4,5,1,3,2,7,14,11,12,14,10,8,24,26,14,24,24,44,31,43,31,108,2,9,5,6,7,5,6,6,8,9,4,12,12,5,6,18,14,12,19,15,17,43,17,16,2,7,2,1,2,2,2,1,6,6,10,2,5,6,7,8,5,11,9,18,19,28,7,12],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,4.443845427247801e-09,0,0,0,2.804102320375743e-09,0,5.83983168390402e-09,0,0,0,0,5.950081281382848e-09,4.049123250630963e-09,1.791949675401398e-09,2.725807074437837e-09,1.421768697975519e-09,4.520684936976414e-09,4.035991237066589e-09,3.751852266014655e-09,5.902371075718579e-09,4.278602441634115e-09,3.183613497039786e-08,1.323496607149611e-08,2.812899014120308e-08,3.85133270361476e-08,1.504754973048409e-08,2.290218706562015e-08,3.083889036645722e-08,2.663897204356955e-08,3.43419603931304e-08,3.06591163404961e-08,3.626658490281408e-08,3.033265309325506e-08,6.383091660524341e-08,5.784905158880656e-08,8.181361761901416e-08,5.263860225820251e-08,5.734238961284472e-08,4.36129131910054e-08,5.82925166169963e-08,5.763873294644928e-08,7.130251185484307e-08,7.128519305427845e-08,6.345048906397472e-08,7.844104476329212e-08,2.299276414528734e-08,1.985244910724416e-08,5.782070195691743e-08,4.888229969972581e-08,6.186214889199014e-08,8.997287775779343e-08,2.158722325652006e-08,3.084512552413317e-08,3.58350891058752e-08,4.525869555025615e-08,3.626658490281408e-08,2.022176872883671e-08,1.679760963295879e-08,0,1.785024384414854e-08,8.098246501261925e-09,1.254364772780978e-08,1.635484244662703e-08,1.42176869797552e-08,1.582239727941745e-08,3.363326030888824e-08,1.875926133007327e-08,1.475592768929645e-08,1.426200813878038e-08,3.537348330044207e-09,2.646993214299222e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635870504687153e-09,1.233555614658289e-08,5.608204640751485e-09,5.9725148509792e-09,1.605953713073606e-08,9.764080550757637e-09,1.155529641647812e-08,6.719043853183517e-09,4.049433611216459e-08,4.462560961037136e-09,2.159532400336514e-08,1.433559740321118e-08,1.908064952106486e-08,2.701360526153487e-08,1.469222604517335e-08,1.614396494826636e-08,1.500740906405862e-08,2.065829876501503e-08,3.137641790531685e-08,3.537348330044207e-09,1.323496607149611e-09,0,4.443845427247801e-09,3.343899940107575e-09,4.90761151406146e-09,6.167778073291445e-09,7.010255800939356e-09,4.479386138234399e-09,4.379873762928015e-09,2.789737300216468e-09,1.444412052059765e-09,0,5.784905158880655e-09,1.487520320345712e-09,2.699415500420642e-09,1.791949675401398e-09,6.814517686094594e-09,2.843537395951039e-09,3.39051370273231e-09,4.035991237066589e-09,2.50123484400977e-09,2.95118553785929e-09,4.278602441634115e-09,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6.768245099590487e-08,9.866274665821205e-08,1.163233345261175e-07,7.019664748432965e-08,5.705222777729199e-08,7.073954406896932e-08,3.388872528747683e-07,3.405678420037413e-07,3.442097111568596e-07,3.719665886048147e-07,2.673840065766399e-07,3.231288101823411e-07,8.869240331301002e-07,1.146752212343037e-06,6.309440368902114e-07,7.442455511157748e-07,7.013950045244987e-07,7.026786844681858e-07,1.321026585994279e-06,1.023623383107658e-06,1.366773172667736e-06,1.247369826605986e-06,1.276951737829694e-06,1.293930153915547e-06,6.768245099590487e-08,4.783648322822403e-08,6.344909155970046e-08,4.011236999104551e-08,6.085570962911145e-08,5.58470084755021e-08,1.222789056764628e-07,9.86691878702428e-08,9.44889403175693e-08,1.119191505536611e-07,8.178804907050163e-08,8.486211176505927e-08,2.630706877928263e-07,2.038670599720954e-07,1.754496787074385e-07,1.913774274297706e-07,1.603188581770283e-07,2.077484806253767e-07,2.801213001626422e-07,2.642436302257862e-07,3.175026175444961e-07,2.584640181255647e-07,3.175665316592171e-07,2.749601577070538e-07,1.19439619404538e-08,5.082626342998803e-08,6.697404109079492e-08,4.345506749029931e-08,2.28208911109168e-08,2.606193728856765e-08,9.083575850251523e-08,8.59376797579534e-08,1.417334104763539e-07,1.119191505536611e-07,1.06953602630656e-07,7.833425701390087e-08,2.480380770618077e-07,2.548338249651193e-07,1.51831452727591e-07,1.88339690486441e-07,1.202391436327712e-07,2.108036053404558e-07,2.419229410495547e-07,2.38588908844642e-07,2.056207427907213e-07,2.303701031119163e-07,2.473676141345481e-07,2.329074277047985e-07,0,1.187362093087597e-08,1.122910372578574e-08,1.330029886675974e-08,2.983408985176398e-09,8.753899887932282e-09,2.760136861386272e-09,1.754124059682377e-08,2.142453621282099e-08,2.361209905153819e-08,2.506149300947495e-08,3.627470661198666e-08,4.859756213609879e-08,6.357200680423903e-08,2.926477992302863e-08,4.614913363599911e-08,2.916882357963362e-08,4.429038357128141e-08,4.367937961262438e-08,5.252208295017801e-08,4.60487855327183e-08,6.085795118790653e-08,4.755063508628221e-08,4.33642677438963e-08,6.364920043603903e-09,2.137251767557675e-08,1.403637965723217e-08,1.596035864011169e-08,2.088386289623479e-08,1.45898331465538e-08,1.380068430693136e-08,1.252945756915984e-08,2.142453621282099e-08,1.836496592897415e-08,7.518447902842486e-09,3.109260566741714e-08,4.859756213609879e-08,5.297667233686586e-08,1.064173815382859e-08,2.186011593284168e-08,3.240980397737068e-08,1.722403805549832e-08,1.798562689931592e-08,1.818072102121546e-08,3.63543043679355e-08,4.73339620350384e-08,3.170042339085481e-08,3.316091062768541e-08,6.364920043603903e-09,9.498896744700776e-09,5.614551862892868e-09,2.660059773351948e-09,5.966817970352797e-09,5.835933258621521e-09,5.520273722772544e-09,2.505891513831967e-09,1.606840215961575e-08,1.574139936769213e-08,2.255534370852746e-08,5.182100944569522e-09,1.21493905340247e-08,5.297667233686586e-08,1.330217269228574e-08,1.70023123922102e-08,1.620490198868534e-08,2.460576865071189e-08,1.284687635665423e-08,1.616064090774708e-08,1.21181014559785e-08,3.380997288217029e-08,1.32085097461895e-08,1.530503567431634e-08],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,5.925127236330401e-09,0,0,0,2.804102320375743e-09,0,5.83983168390402e-09,0,0,0,0,7.437601601728561e-09,4.049123250630963e-09,5.375849026204193e-09,4.088710611656756e-09,2.843537395951039e-09,2.147325345063796e-08,4.035991237066589e-09,6.253087110024424e-09,8.853556613577869e-09,1.711440976653646e-08,3.183613497039786e-08,1.455846267864572e-08,2.812899014120308e-08,7.258280864504742e-08,1.504754973048409e-08,2.45380575703073e-08,3.238083488478009e-08,5.748409756770272e-08,7.61495643499848e-08,5.401844307611218e-08,4.463579680346348e-08,3.177706514531483e-08,8.06285262382022e-08,6.363395674768722e-08,1.026389021038541e-07,5.938714100925412e-08,7.16779870160559e-08,4.633872026544324e-08,9.52585027643598e-08,1.412714042805129e-07,1.291517195861308e-07,1.363172989985325e-07,1.224741998211605e-07,1.640130935959744e-07,2.829878664035365e-08,2.249944232154339e-08,6.719703200398513e-08,5.332614512697361e-08,7.858164859252802e-08,9.488048927185489e-08,2.621305681148864e-08,3.224717668432104e-08,4.03144752441096e-08,4.671865347123216e-08,3.905632220303055e-08,2.744382898913553e-08,2.015713155955055e-08,0,1.785024384414854e-08,9.447954251472247e-09,1.612754707861258e-08,2.998387781881621e-08,2.701360526153487e-08,2.373359591912617e-08,3.63239211335993e-08,1.875926133007327e-08,1.918270599608538e-08,2.281921302204861e-08,3.537348330044207e-09,3.970489821448833e-09,3.125443349022564e-09,0,3.343899940107575e-09,1.635870504687153e-09,1.696138970155147e-08,7.010255800939356e-09,1.34381584147032e-08,2.77392004985441e-08,9.764080550757637e-09,1.155529641647812e-08,1.343808770636703e-08,5.20641464299259e-08,4.462560961037136e-09,5.12888945079922e-08,3.046314448182376e-08,3.134678135603513e-08,6.113605401294734e-08,3.955599319854362e-08,2.959726907182165e-08,3.37666703941319e-08,8.55843805979194e-08,8.129344639104818e-08,3.537348330044207e-09,1.323496607149611e-09,0,4.443845427247801e-09,3.343899940107575e-09,4.90761151406146e-09,6.167778073291445e-09,7.010255800939356e-09,4.479386138234399e-09,4.379873762928015e-09,2.789737300216468e-09,1.444412052059765e-09,0,5.784905158880655e-09,1.487520320345712e-09,2.699415500420642e-09,1.791949675401398e-09,6.814517686094594e-09,2.843537395951039e-09,3.39051370273231e-09,4.035991237066589e-09,3.751852266014655e-09,4.426778306788935e-09,4.278602441634115e-09,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6.768245099590487e-08,1.046423070617401e-07,1.19848284057212e-07,7.019664748432965e-08,5.705222777729199e-08,7.073954406896932e-08,5.380271849764363e-07,5.601863569407333e-07,6.378003471435928e-07,5.694709719348049e-07,4.120859395475275e-07,5.287562348438308e-07,1.413065408715753e-06,1.414327728556412e-06,1.066194201375972e-06,1.394321256988329e-06,1.130247950148049e-06,1.319813876914158e-06,3.036769549490462e-06,2.955423903107823e-06,3.495552627442262e-06,2.916148378416697e-06,4.231991885058619e-06,5.411862868751777e-06,6.768245099590487e-08,5.082626342998803e-08,6.344909155970046e-08,4.011236999104551e-08,6.465919148093092e-08,5.58470084755021e-08,1.92152566063013e-07,1.623267284316898e-07,1.417334104763539e-07,2.007961230521566e-07,1.06953602630656e-07,1.011817486429553e-07,3.156848253513916e-07,2.803172074616313e-07,2.834187117581698e-07,3.371888007095959e-07,2.765500303553738e-07,3.757803399547255e-07,5.475098139542553e-07,6.798501166003238e-07,9.162218391998317e-07,5.89972215286615e-07,1.159953541955245e-06,1.006030694669338e-06,1.19439619404538e-08,6.876494464057203e-08,1.691975774925345e-07,5.682585748731448e-08,2.472263203682653e-07,2.978507118693446e-08,1.85165200024358e-07,1.273150811228939e-07,2.126001157145309e-07,1.843374244413241e-07,1.447019329708875e-07,1.207653128964305e-07,3.382337414479196e-07,3.058005899581431e-07,2.125640338186274e-07,3.55415222369574e-07,1.763507439947311e-07,3.910559635301208e-07,6.525553015152462e-07,5.797967032138611e-07,4.777658435431466e-07,5.534501257688721e-07,8.256729823139643e-07,9.995610438997603e-07,0,1.187362093087597e-08,1.122910372578574e-08,1.330029886675974e-08,2.983408985176398e-09,8.753899887932282e-09,5.520273722772544e-09,1.754124059682377e-08,3.749293837243674e-08,2.885923217410223e-08,3.007379161136994e-08,3.627470661198666e-08,6.074695267012349e-08,8.476267573898537e-08,6.385042892297155e-08,6.31514460282093e-08,4.537372556831895e-08,5.905384476170854e-08,6.166500651194031e-08,8.888352499260894e-08,7.51322290270667e-08,9.69219222622215e-08,8.189276042637493e-08,2.754906421376941e-07,6.364920043603903e-09,2.137251767557675e-08,1.403637965723217e-08,1.596035864011169e-08,2.088386289623479e-08,1.45898331465538e-08,1.656082116831763e-08,1.50353490829918e-08,2.142453621282099e-08,2.361209905153819e-08,1.002459720378998e-08,3.109260566741714e-08,7.289634320414819e-08,5.297667233686586e-08,1.596260723074289e-08,4.372023186568336e-08,4.537372556831895e-08,2.952692238085427e-08,4.881813015528608e-08,3.030120170202577e-08,4.12015449503269e-08,9.69219222622215e-08,4.490893313704431e-08,4.081342846484358e-08,6.364920043603903e-09,1.662306930322636e-08,5.614551862892868e-09,2.660059773351948e-09,5.966817970352797e-09,5.835933258621521e-09,5.520273722772544e-09,2.505891513831967e-09,1.606840215961575e-08,1.574139936769213e-08,2.506149300947495e-08,5.182100944569522e-09,3.037347633506175e-08,6.357200680423903e-08,1.862304176920004e-08,1.943121416252594e-08,1.620490198868534e-08,2.706634551578308e-08,2.312437744197762e-08,3.636144204243092e-08,4.60487855327183e-08,6.311194938005121e-08,1.84919136446653e-08,3.061007134863268e-08],[565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532],[565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,565395266,755574283,639909215,675090988,598104021,611295330,648531765,713240735,669734626,684951248,716913381,692323218,297661400,172863681,672259724,740901132,558051386,733727643,703349287,884821671,743311822,799605045,677693752,701163532,251172937,334472748,283692005,299159586,262916990,268590931,286230890,314181161,296330977,303790726,317894855,306379366,133044089,78482517,296381278,329192428,249502775,327318880,314149620,389791799,330705935,355948966,299149912,309135697,251172937,334472748,283692005,299159586,262916990,268590931,286230890,314181161,296330977,303790726,317894855,306379366,133044089,78482517,296381278,329192428,249502775,327318880,314149620,389791799,330705935,355948966,299149912,309135697,251172937,334472748,283692005,299159586,262916990,268590931,286230890,314181161,296330977,303790726,317894855,306379366,133044089,78482517,296381278,329192428,249502775,327318880,314149620,389791799,330705935,355948966,299149912,309135697,314222329,421101535,356217210,375931402,335187031,342704399,362300875,399059574,373403649,381160522,399018526,385943852,164617311,94381164,375878446,411708704,308548611,406408763,389199667,495029872,412605887,443656079,378543840,392027835,314222329,421101535,356217210,375931402,335187031,342704399,362300875,399059574,373403649,381160522,399018526,385943852,164617311,94381164,375878446,411708704,308548611,406408763,389199667,495029872,412605887,443656079,378543840,392027835,314222329,421101535,356217210,375931402,335187031,342704399,362300875,399059574,373403649,381160522,399018526,385943852,164617311,94381164,375878446,411708704,308548611,406408763,389199667,495029872,412605887,443656079,378543840,392027835],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.01576009880063126,0,0,0,0.004446312422615971,0,0.007727502946855426,0,0,0,0,0.005298390469668754,0.003119377399242391,0.001561554667395308,0.002103344078543636,0.0007033682177480282,0.00264462820720336,0.001909561688936986,0.001885816500836074,0.002890395146052891,0.002111571401022107,0.1434432958465718,0.04775751768352558,0.07670811154435228,0.1365875229388042,0.05920493016060689,0.07602468953172063,0.04781130953130446,0.04223996801485172,0.04769859240526619,0.04056939047099099,0.06081447263352983,0.04873273356119795,0.0400262874746608,0.03078539941748785,0.07285286895794538,0.04055190619015108,0.04996974935664985,0.03365350525669817,0.02883809692766916,0.03371900964184284,0.03373558983788676,0.0358305135158854,0.03107174782006858,0.03871214235207195,0.1035979358891908,0.07163627652528835,0.1576777848411686,0.1733610868069438,0.2433980462158283,0.2986684231603311,0.03346791667191312,0.04890943664877567,0.04977244424897341,0.05988814783812956,0.06081447263352983,0.03248848904079863,0.01053323354596337,0,0.01589517140900627,0.006238754798484782,0.01093088267176715,0.01262006447126181,0.007033682177480283,0.009256198725211759,0.01591301407447488,0.009429082504180369,0.007225987865132227,0.007038571336740354,0.01593814398295242,0.009551503536705114,0.008523123504928032,0,0.01315665114680153,0.005430334966551474,0.01912452381252178,0.008892624845231942,0.008295407374828902,0.02125063310385242,0.0163731272474888,0.01856485088045636,0.004213293418385347,0.0215497795922415,0.003973792852251566,0.01663667946262608,0.01249243733916246,0.01472340854980545,0.01336399613721254,0.008595041673410918,0.007638246755747946,0.007543266003344295,0.01011638301118512,0.01548485694082878,0.01593814398295242,0.004775751768352557,0,0.01576009880063126,0.01315665114680153,0.01629100489965442,0.00956226190626089,0.01111578105653993,0.006221555531121676,0.00579562721014157,0.004678036356425372,0.002320606360057046,0,0.003078539941748786,0.001324597617417189,0.002079584932828261,0.001561554667395308,0.005258360196359089,0.001406736435496056,0.00198347115540252,0.001909561688936986,0.001257211000557383,0.001445197573026446,0.002111571401022107,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.3049551665380866,0.3560181297617881,0.317215203078698,0.2489524259873471,0.224473301072981,0.2348226333166246,0.5253964442583268,0.5400198900166523,0.478083327988745,0.4922013279077767,0.448369136332351,0.519141868802255,0.5561611552251279,0.6102645405633442,0.5618390260336896,0.5733544295732366,0.6112150681673824,0.5422155749595021,0.6535297314616761,0.5988259102900553,0.6466658460562222,0.6269731414977342,0.6253241379469532,0.638579055941615,0.3049551665380866,0.1726148507935942,0.1730264744065625,0.1422585291356269,0.2394381878111797,0.1853862894604931,0.189576036588056,0.1564543606590301,0.1312385606243614,0.1480959747687116,0.1371482064075427,0.1363402887763498,0.1649630545159277,0.1084914738779279,0.156233312052149,0.1474339961759751,0.1397063012954017,0.1603072134662876,0.1385797984786205,0.1545841322302649,0.1502210483095207,0.1299133536436747,0.1555125473951847,0.1356980493875932,0.05381561762436821,0.1834032789681939,0.1826390563180382,0.1541134065635958,0.08978932042919238,0.08651360174823011,0.1408279128939845,0.1362667012191553,0.196857840936542,0.1480959747687116,0.1793476545329404,0.1258525742550921,0.1555365942578747,0.1356143423474098,0.1352019046605135,0.1450937740144517,0.1047797259715513,0.1626646724878507,0.1196825532315359,0.1395759640525693,0.0972860122385467,0.1157923369432752,0.1211360895499333,0.1149442300694907,0,0.04284519193404229,0.03062190774772811,0.04716951290267477,0.01173829120166087,0.02905890687491837,0.004279199292204054,0.02781419044769456,0.02975718941460412,0.03124449040288511,0.04202495175281454,0.05827929416189649,0.0304739473609156,0.03383097159743997,0.0260595147710473,0.03555252718877355,0.02541852219873544,0.03417626907446988,0.02160878027081179,0.03072573824709344,0.02178721199089259,0.03058940502450535,0.02328557847040985,0.02140108805231855,0.02867826479889558,0.07712134548127612,0.03827738468466015,0.05660341548320973,0.08216803841162607,0.04843151145819728,0.02139599646102027,0.0198672788912104,0.02975718941460412,0.02430127031335509,0.01260748552584436,0.049953680710197,0.0304739473609156,0.02819247633119997,0.009476187189471746,0.01684067077362958,0.02824280244303937,0.01329077130673829,0.008897733052687208,0.01063583247014773,0.01720043051912573,0.02379175946350416,0.01552371898027324,0.01636553792236125,0.02867826479889558,0.03427615354723383,0.01531095387386406,0.009433902580534956,0.02347658240332174,0.01937260458327891,0.008558398584408109,0.00397345577824208,0.02231789206095309,0.02082966026859008,0.03782245657753309,0.0083256134516995,0.0076184868402289,0.02819247633119997,0.01184523398683968,0.01309829949060078,0.01412140122151969,0.01898681615248326,0.006355523609062292,0.009454073306797983,0.005733476839708576,0.01699411390250297,0.006468216241780515,0.007553325194935959],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.01766935574970328,0,0,0,0.002828236979208065,0,0.005027778780192548,0,0,0,0,0.004144688578965792,0.001730320088155684,0.002961023313282898,0.001761714695966153,0.0006236460273437216,0.004668056307776829,0.0007680188762324304,0.001384843462250947,0.001340750448797079,0.002124987434421812,0.1400940374368396,0.04600489932179704,0.05837992064666763,0.2164496079338651,0.03015004323055594,0.07877049256867404,0.03160125162682196,0.05797885807376533,0.06337867647813776,0.04650695371678107,0.05409433823168548,0.03548633411182708,0.03437987556031626,0.02739017913892541,0.05719670238972793,0.02537802795961669,0.03948031084377197,0.0199660998876164,0.02089214191601468,0.03071089676168966,0.02457660403943777,0.03018958747707065,0.01854704787502626,0.02036446291320904,0.1245280332771908,0.07109848077004997,0.1394631437670394,0.1590242017473295,0.1574502257595699,0.3045792379322063,0.02558196560266539,0.03252472526089274,0.03355341695901411,0.04022223024154038,0.04733254595272479,0.03064728855112338,0.008594968890079066,0,0.0099472525895179,0.004037413539029928,0.008883069939848692,0.01291924110375179,0.005924637259765356,0.005159430655963864,0.006912169886091874,0.004154530386752842,0.002904959305727004,0.002833316579229083,0.01556600415964884,0.01254679072412647,0.006486657849629737,0,0.00670000960679021,0.005251366171244935,0.01655303656643055,0.007070592448020163,0.01118447231967137,0.0238819492059146,0.0118331364881812,0.01290412149520985,0.00572997926005271,0.0224101465682117,0.002486813147379475,0.02191738778330533,0.01677913210860308,0.0135064793357405,0.01340838958789002,0.008599051093273106,0.00563213842570449,0.007478154696155116,0.0129605876717051,0.01009369031350361,0.01556600415964884,0.004182263574708822,0,0.01325201681227746,0.00670000960679021,0.01575409851373481,0.006019286024156564,0.007070592448020163,0.003728157439890457,0.00377083408514441,0.003380896139480343,0.001613015186901231,0,0.002490016285356856,0.0008289377157931583,0.001153546725437122,0.0009870077710942989,0.002936191159943588,0.0006236460273437216,0.000737061522280552,0.0007680188762324304,0.0008309060773505684,0.0006703752243985394,0.0005312468586054531,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0.2978347664518292,0.3306708206380307,0.2487374512123195,0.2093338231172408,0.1143127728230834,0.2270835299179859,0.5250739369448015,0.5650078310035894,0.5308361538809747,0.4902836628872428,0.4994089450796171,0.5904768211782569,0.6025288465116098,0.608773865813673,0.5941489160128614,0.5958381434631403,0.6225417629597438,0.5686720640660807,0.6660257987796548,0.6424776396725767,0.6651774602689089,0.6458264447220039,0.640877476346194,0.6719565997185339,0.2978347664518292,0.1606115414527577,0.1316845329947574,0.1196193274955662,0.1295544758661612,0.1792764709878836,0.1875264060517148,0.1637238601203583,0.1179635897513277,0.1728745863359642,0.1296175887992899,0.1129924781267035,0.134607508263232,0.1206578833144217,0.1579383194464568,0.1440915771773607,0.1523240483837671,0.1619135737965924,0.1200801230504199,0.1477921653760701,0.1743501474580272,0.1306585292790181,0.1756591503097388,0.1249124342573007,0.05255907643267575,0.2172979678478487,0.3511587546526864,0.1694607139520521,0.4953553489000282,0.09561411786020457,0.1807072640134706,0.128410870682634,0.1769453846269916,0.1587045382756393,0.1753649730813923,0.1348619900221945,0.1442223302820342,0.1316267817975509,0.1184537395848426,0.1518803110788397,0.09713417578095294,0.1684954263899498,0.1431187513100935,0.1260416202829881,0.09091525841045642,0.1225701441331742,0.1250369167910821,0.1241091388601476,0,0.03752077039778108,0.02330528686432726,0.03966289716331899,0.005977711420702094,0.02810120581441976,0.005387370633987689,0.01769221649175592,0.03120507427199614,0.02484623581489955,0.036446573643521,0.04050897566389176,0.0259024041602504,0.03648468508709916,0.03558138196829509,0.02698663609255718,0.02499189589651797,0.02544470275626821,0.01352439971144957,0.01932233047278919,0.01429709994815168,0.02146486815743524,0.01240154212449575,0.03420592125753864,0.0280086558151689,0.06753738671600595,0.02913160858040908,0.04759547659598279,0.04184397994491465,0.04683534302403292,0.01616211190196307,0.01516475699293364,0.01783147101256922,0.02032873839400872,0.01214885788117367,0.03472197914047865,0.03108288499230049,0.02280292817943697,0.008895345492073771,0.01868305575638575,0.02499189589651797,0.01272235137813411,0.01070681643823091,0.006587158115723588,0.007840345132857371,0.02146486815743524,0.006800845681175086,0.005067543890005724,0.0280086558151689,0.05252907855689352,0.01165264343216363,0.007932579432663799,0.01195542284140419,0.01873413720961317,0.005387370633987689,0.002527459498822274,0.01337360325942692,0.01355249226267248,0.03037214470293417,0.005786996523413109,0.0129512020801252,0.02736351381532437,0.01037790307408607,0.008303580336171441,0.008925677105899277,0.01166215542995627,0.00507164989179359,0.007904589738868305,0.008762738677899416,0.01397712345135317,0.00280034822166033,0.003800657917504293]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>normalized_subtype<\/th>\n      <th>sample<\/th>\n      <th>normalized_ref<\/th>\n      <th>sum_min<\/th>\n      <th>sum_max<\/th>\n      <th>mf_min<\/th>\n      <th>mf_max<\/th>\n      <th>group_depth<\/th>\n      <th>subtype_depth<\/th>\n      <th>proportion_min<\/th>\n      <th>proportion_max<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":[6,7],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[4,5,6,7,8,9,10,11]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"normalized_subtype","targets":1},{"name":"sample","targets":2},{"name":"normalized_ref","targets":3},{"name":"sum_min","targets":4},{"name":"sum_max","targets":5},{"name":"mf_min","targets":6},{"name":"mf_max","targets":7},{"name":"group_depth","targets":8},{"name":"subtype_depth","targets":9},{"name":"proportion_min","targets":10},{"name":"proportion_max","targets":11}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
<div class="section level3">
<h3 id="summary-table">Summary Table<a class="anchor" aria-label="anchor" href="#summary-table"></a>
</h3>
<p>The function will output the resulting <code>mf_data</code> as a data
frame with the MF and proportion calculated. If the <code>summary</code>
parameter is set to <code>TRUE</code>, the data frame will be a summary
table with the MF calculated for each group. If <code>summary</code> is
set to <code>FALSE</code>, the MF will be appended to each row of the
original <code>mutation_data</code>.</p>
<p>The summary table will include:</p>
<ul>
<li>
<code>cols_to_group</code>: all columns used to group the data.</li>
<li>Subtype column for the given resolution, if applicable.</li>
<li>Context column for the given resolution, if applicable.</li>
<li>
<code>sum_min</code> &amp; <code>sum_max</code>: the min/max
mutation counts for the group(s)/subtypes.</li>
<li>
<code>group_depth</code>: the total_depth summed across the
group(s).</li>
<li>
<code>subtype_depth</code>: the total_depth summed across the
group(s), for the given context, if applicable.</li>
<li>
<code>MF_min</code> &amp; <code>MF_max</code>: the min/max MF for
the group(s)/subtype.</li>
<li>
<code>proportion_min</code> &amp; <code>proportion_max</code>: the
min/max subtype proportion, if applicable.</li>
</ul>
<p>Additional columns from the orginal mutation data can be retained
using the <code>retain_metadata_cols</code> parameter. Retaining
higher-order experimental groups may be useful for later statistical
analyses or plotting. See above for example <em>calculating mean MF per
dose</em>.</p>
<p><strong>NOTE</strong> The summary table uses a pre-defined list of
possible subtypes for each resolution. If a particular subtype within a
given group is not recorded in the mutation data, the summary table will
have no frame of reference for populating the
<code>metadata_cols</code>. Thus, for subtypes that do not occur in the
mutation data for a given group, the corresponding
<code>metadata_col</code> will be NA.</p>
</div>
<div class="section level3">
<h3 class="tabset" id="mf-plots">MF Plots<a class="anchor" aria-label="anchor" href="#mf-plots"></a>
</h3>
<p>You can visualize the results of <code>calculate_mf</code> using the
<code><a href="../reference/plot_mf.html">plot_mf()</a></code> and <code>plot_mean_mf</code>() functions. These
functions offer variable aesthetics for easy visualization. The output
is a ggplot object that can be modified using ggplot2.</p>
<div class="section level4">
<h4 id="plot_mf">plot_mf<a class="anchor" aria-label="anchor" href="#plot_mf"></a>
</h4>
<p><em>Example 3.10. Plot the Min and Max MF per sample, coloured and
ordered by</em> <em>dose group. See example 3.1 for calculating
mf_data_global</em></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define the order for dose groups</span></span>
<span><span class="va">mf_data_global</span><span class="op">$</span><span class="va">dose_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">mf_data_global</span><span class="op">$</span><span class="va">dose_group</span>,</span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_mf.html">plot_mf</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_global</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>  mf_type <span class="op">=</span> <span class="st">"both"</span>,</span>
<span>  fill_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  group_order <span class="op">=</span> <span class="st">"arranged"</span>,</span>
<span>  group_order_input <span class="op">=</span> <span class="st">"dose_group"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-mf-1.png" alt="Mutation Frequency (MF) Minimum and Maximum (mutations/bp) per Sample. Light colored bars represent MFmin and dark coloured bars represent MFmax. Bars are coloured and grouped by dose. Data labels are the number of mutations per sample." width="1152"><p class="caption">
Mutation Frequency (MF) Minimum and Maximum (mutations/bp) per Sample.
Light colored bars represent MFmin and dark coloured bars represent
MFmax. Bars are coloured and grouped by dose. Data labels are the number
of mutations per sample.
</p>
</div>
</div>
<div class="section level4">
<h4 id="plot_mean_mf">plot_mean_mf()<a class="anchor" aria-label="anchor" href="#plot_mean_mf"></a>
</h4>
<p>Calculate and plot the mean MF for a user-defined group using
<code><a href="../reference/plot_mean_mf.html">plot_mean_mf()</a></code>.</p>
<p><em>Example 3.11. Plot the mean MF min per dose, including SEM</em>
<em>and individual values coloured by dose. See example 3.1 for
calculating</em> <em>mf_data_global</em></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_mean</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_mean_mf.html">plot_mean_mf</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_global</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  mf_type <span class="op">=</span> <span class="st">"min"</span>,</span>
<span>  fill_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  add_labels <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  group_order <span class="op">=</span> <span class="st">"arranged"</span>,</span>
<span>  group_order_input <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  plot_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  x_lab <span class="op">=</span> <span class="st">"Dose Group"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot_mean</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-mean-mf-1.png" alt="Mean Mutation Frequency (MF) Minimum per Dose. Lines are mean ± S.E.M. Points are individual samples, coloured by dose." width="700"><p class="caption">
Mean Mutation Frequency (MF) Minimum per Dose. Lines are mean ± S.E.M.
Points are individual samples, coloured by dose.
</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="modeling-mf">Modeling MF<a class="anchor" aria-label="anchor" href="#modeling-mf"></a>
</h2>
<div class="section level3">
<h3 id="generalized-linear-modeling">Generalized Linear Modeling<a class="anchor" aria-label="anchor" href="#generalized-linear-modeling"></a>
</h3>
<p>An important component of analysing mutagencity data is how MF
changes based on experimental variables.</p>
<p>The <code><a href="../reference/model_mf.html">model_mf()</a></code> function can be used to quantitatively
evaluate how MF changes based on experimental variables. The function
models the proportion of reads that carry a mutation (i.e. the MF)
across user-supplied fixed and/or random effects and interaction
parameters. Depending on the supplied effects, <code><a href="../reference/model_mf.html">model_mf()</a></code>
will automatically choose to fit either a generalized linear model (GLM)
using the <code><a href="https://rdrr.io/r/stats/glm.html" class="external-link">glm()</a></code> function from the <em><a href="https://CRAN.R-project.org/package=stats" class="external-link">stats</a></em> library
or a generalized linear mixed model (GLMM) using the
<code>glmer()</code> function from the <em><a href="https://CRAN.R-project.org/package=lme4" class="external-link">lme4</a></em>
library.</p>
<div class="section level4">
<h4 id="model-formula">Model Formula<a class="anchor" aria-label="anchor" href="#model-formula"></a>
</h4>
<ul>
<li>
<strong>Fixed effects</strong>: The experimental
variable(s)/factor(s) of interest.</li>
<li>
<strong>Random effects</strong>: Variable used to account for
variability within a larger group. Covariate.</li>
<li>
<strong>Interaction</strong>: The product of two or more interacting
fixed_effects. An interaction implies that the effect of one fixed
effect changes depending on the levels of another fixed effect,
indicating a non-additive reationship.</li>
<li>
<strong>Response</strong>: the function will model the MF, taking
the mutation count (<code>sum</code>) and the <code>group_depth</code>
as input.</li>
</ul>
<p>The model formula is built as:</p>
<p><code>cbind(sum, group_depth) ~ fixed_effect1 + fixed_effect2 + ... + (1|random_effect)</code></p>
<p>If <code>test_interaction</code> is set to TRUE, the model will use
the product of the fixed_effects:</p>
<p><code>cbind(sum, group_depth) ~ fixed_effect1 * fixed_effect2 * ... + (1|random_effect)</code></p>
</div>
<div class="section level4">
<h4 id="family">Family<a class="anchor" aria-label="anchor" href="#family"></a>
</h4>
<p>The occurence of mutations is assumed to follow a binomial
distribution as:</p>
<ol style="list-style-type: decimal">
<li>There is a finite number of sequenced bases.</li>
<li>A mutation at any given base is equally probable.</li>
<li>Mutations occur independently of other mutations.</li>
</ol>
<p>To account for over-dispersion, the function will fit a GLM with a
quasibinomial distribution. If the dispersion parameter of the model is
less than 1, the model will be refit using a binomial distribution. The
GLMM will always use a binomial distribution.</p>
<p>The <code>model_mf</code> function will fit a generalized linear
model to analyse the effect(s) of given fixed_effects(s) on MF and
perform specified pairwise comparisons between levels of your
factors.</p>
<p>Additional arguments can be passed to the model to further customize
it to your needs. Details on the arguments for the generalized linear
model can be found at <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/glm" class="external-link">stats::glm</a>
and for the general linear mixed model at <a href="https://www.rdocumentation.org/packages/lme4/versions/1.1-35.3/topics/glmer" class="external-link">lme4::glmer</a>.</p>
</div>
<div class="section level4">
<h4 id="estimates-comparisons">Estimates &amp; Comparisons<a class="anchor" aria-label="anchor" href="#estimates-comparisons"></a>
</h4>
<p><code><a href="../reference/model_mf.html">model_mf()</a></code> provides estimates of the mean for each level
of the fixed effects. Furthermore, pairwise comparisons can be performed
based on a user-supplied table. Mean estimates and comparisons are
conducted using the doBy R library <span class="citation">(<a href="#ref-doBy">Halekoh and Højsgaard 2025</a>)</span>. Estimates of
the back-transformed standard errors are approximated using the delta
method. The p-values are adjusted for multiple comparisons using the
Sidak method.</p>
</div>
<div class="section level4">
<h4 id="goodness-of-fit">Goodness of Fit<a class="anchor" aria-label="anchor" href="#goodness-of-fit"></a>
</h4>
<p>The <code>model_mf</code> function will output the Pearson residuals
appended to the <code>mf_data</code>. The row with the highest residual
will be printed to the console so that users may assess it as an
outlier. Additionally, Pearson residuals will be plotted as a histogram
and a QQ-plot to check for deviances from model assumptions. We assume
that residuals will follow a normal distribution with a mean of 0. Thus,
we expect the histogram to follow a bell curve and the QQ-plot to be
plotted along the y=x line.</p>
</div>
<div class="section level4">
<h4 id="general-usage-1">General Usage<a class="anchor" aria-label="anchor" href="#general-usage-1"></a>
</h4>
<p>Mutation data should first be summarised by sample using
<code><a href="../reference/calculate_mf.html">calculate_mf()</a></code>. The <code>mf_data</code> should be output
as a summary table. Be sure to retain the columns for experimental
variables of interest using the <code>retain_metadata_cols</code>
parameter.</p>
<p>You may specify factors and covariates for your model using the
<code>fixed_effects</code> and <code>random_effects</code> parameters
respectively. If more than one fixed_effect is supplied, then you may
specify whether you wish to test the interaction between your
fixed_effects using the <code>test_interaction</code> parameter.</p>
<p>You must specify the columns in your <code>mf_data</code> that
contain the mutation counts and the total_depth per sample using the
<code>muts</code> and <code>total_counts</code> parameters
respectively.</p>
<p>To perform pairwise comparisons between levels of your fixed effects,
supply a constrast table using the <code>contrasts</code> parameter.
This can either be a data frame or a filepath to a text file that will
be loaded into R. The table must consist of two columns, each containing
levels within the fixed_effects. The level in the first column will be
compared to the level in the second column for each row. You should also
provide the reference level for each fixed effect using the
<code>reference_level</code> parameter. If you specify multiple pairwise
comparisons, then the p-values will automatically be corrected for
multiple comparisons using the Sidak method. For multiple fixed effects,
the user must include levels for all fixed_effects in each value of the
contrasts table. Within each value, the levels of the different
fixed_effects should be seperated by a colon.</p>
</div>
<div class="section level4">
<h4 id="output-2">Output<a class="anchor" aria-label="anchor" href="#output-2"></a>
</h4>
<p>The function will output a list of results.</p>
<ul>
<li>model: the glmm or glmer model object.</li>
<li>model_formula: the formula given to glm or glmer.</li>
<li>model_data: the supplied <code>mf_data</code> with added column for
Pearson residuals.</li>
<li>summary: the summary of the model.</li>
<li>anova: the analysis of variance for models with two or more
fixed_effects. See <a href="https://www.rdocumentation.org/packages/car/versions/1.0-9/topics/Anova" class="external-link">car::Anova.</a>.</li>
<li>residuals_histogram: the Pearson residuals plotted as a histogram.
This is used to check whether the variance is normally distributed. A
symmetric bell-shaped histogram, evenly distributed around zero
indicates that the normality assumption is likely to be true.</li>
<li>residuals_qq_plot: the Pearson residuals plotted in a
quantile-quantile plot. For a normal distribution, we expect points to
roughly follow the y=x line.<br>
</li>
<li>point_estimates_matrix: the contrast matrix used to generate
point-estimates for the fixed effects.</li>
<li>point_estimates: the point estimates for the fixed effects.</li>
<li>pairwise_comparisons_matrix: the contrast matrix used to conduct the
pairwise comparisons specified in the <code>contrasts</code>.</li>
<li>pairwise_comparisons: the results of pairwise comparisons specified
in the <code>contrasts</code>.</li>
</ul>
</div>
<div class="section level4">
<h4 id="examples-1">Examples<a class="anchor" aria-label="anchor" href="#examples-1"></a>
</h4>
<div class="section level5">
<h5 class="tabset" id="example-4-1-model-by-dose">Example 4.1 Model by Dose<a class="anchor" aria-label="anchor" href="#example-4-1-model-by-dose"></a>
</h5>
<p><em>Example 4.1. Model the effect of dose on MF. Our example data
consists of 24</em> <em>mouse samples, exposed to 3 doses of BaP or a
vehicle control. Dose Groups</em> <em>are : Control, Low, Medium, and
High. We will determine if the MFmin of each</em> <em>BaP dose group is
significantly increased from that of the Control group.</em> <em>See
example 3.1 for calculating mf_data_global</em></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a contrasts table for pairwise comparisons</span></span>
<span><span class="va">contrasts_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  col1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span>,</span>
<span>  col2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Control"</span>, <span class="st">"Control"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Run the model</span></span>
<span><span class="va">model_by_dose</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/model_mf.html">model_mf</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_global</span>,</span>
<span>  fixed_effects <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  muts <span class="op">=</span> <span class="st">"sum_min"</span>,</span>
<span>  total_count <span class="op">=</span> <span class="st">"group_depth"</span>,</span>
<span>  reference_level <span class="op">=</span> <span class="st">"Control"</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="va">contrasts_table</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        sample dose_group dose sum_min sum_max       mf_min       mf_max</span></span>
<span><span class="co">## 14 dna00986.1     Medium   25     160     197 9.255848e-07 1.139626e-06</span></span>
<span><span class="co">##    group_depth residuals</span></span>
<span><span class="co">## 14   172863681  4.739692</span></span></code></pre>
<div class="section level6">
<h6 id="residuals-histogram">Residuals Histogram<a class="anchor" aria-label="anchor" href="#residuals-histogram"></a>
</h6>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/model-dose-hist-1.png" alt="GLM residuals of MFmin modelled as an effect of Dose. x is pearson's residuals, y is frequency. Plotted to validate model assumptions. n = 24." width="700"><p class="caption">
GLM residuals of MFmin modelled as an effect of Dose. x is pearson’s
residuals, y is frequency. Plotted to validate model assumptions. n =
24.
</p>
</div>
</div>
<div class="section level6">
<h6 id="residuals-qq-plot">Residuals QQ-plot<a class="anchor" aria-label="anchor" href="#residuals-qq-plot"></a>
</h6>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/model-dose-qq-1.png" alt="GLM residuals of MFmin modelled as an effect of Dose expressed as a quantile-quantile plot. Y is the pearson's residuals of the model in ascending order x is the quantiles of standard normal distribution for n of 24. Plotted to validate model assumptions." width="700"><p class="caption">
GLM residuals of MFmin modelled as an effect of Dose expressed as a
quantile-quantile plot. Y is the pearson’s residuals of the model in
ascending order x is the quantiles of standard normal distribution for n
of 24. Plotted to validate model assumptions.
</p>
</div>
</div>
<div class="section level6">
<h6 id="model-summary">Model Summary<a class="anchor" aria-label="anchor" href="#model-summary"></a>
</h6>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_by_dose</span><span class="op">$</span><span class="va">summary</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Call:</span></span>
<span><span class="co">## stats::glm(formula = model_formula, family = "quasibinomial", </span></span>
<span><span class="co">##     data = mf_data)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Coefficients:</span></span>
<span><span class="co">##                   Estimate Std. Error  t value Pr(&gt;|t|)    </span></span>
<span><span class="co">## (Intercept)      -15.56285    0.08290 -187.736  &lt; 2e-16 ***</span></span>
<span><span class="co">## dose_groupLow      0.67512    0.10066    6.707 1.58e-06 ***</span></span>
<span><span class="co">## dose_groupMedium   1.29746    0.09563   13.567 1.51e-11 ***</span></span>
<span><span class="co">## dose_groupHigh     1.69485    0.08915   19.011 2.84e-14 ***</span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## (Dispersion parameter for quasibinomial family taken to be 4.604243)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Null deviance: 2919.634  on 23  degrees of freedom</span></span>
<span><span class="co">## Residual deviance:   89.569  on 20  degrees of freedom</span></span>
<span><span class="co">## AIC: NA</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Number of Fisher Scoring iterations: 5</span></span></code></pre>
</div>
<div class="section level6">
<h6 id="model-estimates">Model Estimates<a class="anchor" aria-label="anchor" href="#model-estimates"></a>
</h6>
<p><strong>Table 18.</strong> model_by_dose$point_estimates: Estimated
Mean Mutation Frequency per Dose.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-0caf26d4e3c00206b0c5" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-0caf26d4e3c00206b0c5">{"x":{"filter":"none","vertical":false,"data":[["Control","Low","Medium","High"],[1.742366757678581e-07,3.42247727467635e-07,6.377065341569276e-07,9.488646685114914e-07],[1.444379508937598e-08,1.954350276020435e-08,3.040798182832912e-08,3.112522101531163e-08],[1.465685596335918e-07,3.03815088965098e-07,5.773291077181194e-07,8.861100699143531e-07],[2.071277718668112e-07,3.855420985039251e-07,7.043982682837411e-07,1.016063567855002e-06],["Control","Low","Medium","High"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Estimate<\/th>\n      <th>Std.Err<\/th>\n      <th>Lower<\/th>\n      <th>Upper<\/th>\n      <th>dose_group<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":[1,2,3,4],"render":"function(data, type, row, meta) {\n            return data.toExponential(2); \n          }"},{"className":"dt-right","targets":[1,2,3,4]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Estimate","targets":1},{"name":"Std.Err","targets":2},{"name":"Lower","targets":3},{"name":"Upper","targets":4},{"name":"dose_group","targets":5}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
<div class="section level6">
<h6 id="pairwise-comparisons">Pairwise Comparisons<a class="anchor" aria-label="anchor" href="#pairwise-comparisons"></a>
</h6>
<p><strong>Table 19.</strong> model_by_dose$pairwise_comparisons</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-da0b268a2927f570ebf3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-da0b268a2927f570ebf3">{"x":{"filter":"none","vertical":false,"data":[["Low vs Control","Medium vs Control","High vs Control"],[1.964269152630208,3.660001726654651,5.445837762513892],[0.1977269031176319,0.3500176367380217,0.4855053807161875],[6.70681555203679,13.56708516380477,19.01088480443303],[1.584454647041156e-06,1.509636859964303e-11,2.842170943040401e-14],[20,20,20],[1.592242454917549,2.998092588572272,4.521684205639225],[2.423219712587296,4.468045013077525,6.5588722225748],[4.753356409592513e-06,4.528910579892909e-11,8.526512829121202e-14],["***","***","***"],["Low","Medium","High"],["Control","Control","Control"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Fold.Change<\/th>\n      <th>FC.Std.Err<\/th>\n      <th>Obs.T<\/th>\n      <th>p.value<\/th>\n      <th>df<\/th>\n      <th>FC.Lower<\/th>\n      <th>FC.Upper<\/th>\n      <th>adj_p.value<\/th>\n      <th>Significance<\/th>\n      <th>dose_group_1<\/th>\n      <th>dose_group_2<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":1,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":2,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":6,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":7,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":[4,8],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[1,2,3,4,5,6,7,8]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Fold.Change","targets":1},{"name":"FC.Std.Err","targets":2},{"name":"Obs.T","targets":3},{"name":"p.value","targets":4},{"name":"df","targets":5},{"name":"FC.Lower","targets":6},{"name":"FC.Upper","targets":7},{"name":"adj_p.value","targets":8},{"name":"Significance","targets":9},{"name":"dose_group_1","targets":10},{"name":"dose_group_2","targets":11}],"rowCallback":"function(row, data, dataIndex) {\n      $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render","options.columnDefs.4.render","options.columnDefs.5.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
</div>
<div class="section level5">
<h5 class="tabset" id="example-4-2-model-by-dose-and-target">Example 4.2 Model by Dose and Target<a class="anchor" aria-label="anchor" href="#example-4-2-model-by-dose-and-target"></a>
</h5>
<p><em>Example 4.2.1. Model the effects of dose and genomic locus on MF.
Sequencing for</em> <em>the example data was done on a panel of 20
genomic targets. We will determine</em> <em>if the MF of each BaP dose
group is significantly different from the Control</em> <em>individually
for all 20 targets. In this model, dose group and target label</em>
<em>will be our fixed effects. We include the interaction between the
two fixed</em> <em>effects. Because sample will be a repeated measure,
we will use it as a random</em> <em>effect. See example 3.2 for
calculating mf_data_rg</em></p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a contrasts table for the pairwise comparisons.</span></span>
<span><span class="va">combinations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span>dose_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">mf_data_rg</span><span class="op">$</span><span class="va">dose_group</span><span class="op">)</span>,</span>
<span>                            label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">mf_data_rg</span><span class="op">$</span><span class="va">label</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">combinations</span> <span class="op">&lt;-</span> <span class="va">combinations</span><span class="op">[</span><span class="va">combinations</span><span class="op">$</span><span class="va">dose_group</span> <span class="op">!=</span> <span class="st">"Control"</span>, <span class="op">]</span></span>
<span><span class="va">combinations</span><span class="op">$</span><span class="va">col1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">combinations</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">dose_group</span>, <span class="va">label</span>, sep <span class="op">=</span> <span class="st">":"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">combinations</span><span class="op">$</span><span class="va">col2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/with.html" class="external-link">with</a></span><span class="op">(</span><span class="va">combinations</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="va">label</span>, sep <span class="op">=</span> <span class="st">":"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">contrasts2</span> <span class="op">&lt;-</span> <span class="va">combinations</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"col1"</span>, <span class="st">"col2"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Run the model</span></span>
<span><span class="co"># To improve convergence of the model, we will supply the control argument</span></span>
<span><span class="co"># to the  glmer function</span></span>
<span><span class="va">model_by_target</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/model_mf.html">model_mf</a></span><span class="op">(</span>mf_data <span class="op">=</span> <span class="va">mf_data_rg</span>,</span>
<span>  fixed_effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dose_group"</span>, <span class="st">"label"</span><span class="op">)</span>,</span>
<span>  test_interaction <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  random_effects <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  muts <span class="op">=</span> <span class="st">"sum_min"</span>,</span>
<span>  total_count <span class="op">=</span> <span class="st">"group_depth"</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="va">contrasts2</span>,</span>
<span>  reference_level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"chr1"</span><span class="op">)</span>,</span>
<span>  control <span class="op">=</span> <span class="fu">lme4</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/lme4/man/lmerControl.html" class="external-link">glmerControl</a></span><span class="op">(</span>optimizer <span class="op">=</span> <span class="st">"bobyqa"</span>,</span>
<span>                               optCtrl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxfun <span class="op">=</span> <span class="fl">2e5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        sample label dose_group sum_min sum_max       mf_min       mf_max</span></span>
<span><span class="co">## 33 dna00974.1  chr2    Control      23      36 5.982805e-07 9.364391e-07</span></span>
<span><span class="co">##    group_depth residuals</span></span>
<span><span class="co">## 33    38443503  5.481194</span></span></code></pre>
<div class="section level6">
<h6 id="residuals-histogram-1">Residuals Histogram<a class="anchor" aria-label="anchor" href="#residuals-histogram-1"></a>
</h6>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/model-rg-hist-1.png" alt="GLMM residuals of MFmin modelled as an effect of Dose and Genomic Target. x is pearson's residuals, y is frequency. Plotted to validate model assumptions. n = 24." width="700"><p class="caption">
GLMM residuals of MFmin modelled as an effect of Dose and Genomic
Target. x is pearson’s residuals, y is frequency. Plotted to validate
model assumptions. n = 24.
</p>
</div>
</div>
<div class="section level6">
<h6 id="residuals-qq-plot-1">Residuals QQ-plot<a class="anchor" aria-label="anchor" href="#residuals-qq-plot-1"></a>
</h6>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/model-rg-qq-1.png" alt="GLMM residuals of MFmin modelled as an effect of Dose and Genomic Target expressed as a quantile-quantile plot. Y is the pearson's residuals of the model in ascending order x is the quantiles of standard normal distribution for n of 24. Plotted to validate model assumptions." width="700"><p class="caption">
GLMM residuals of MFmin modelled as an effect of Dose and Genomic Target
expressed as a quantile-quantile plot. Y is the pearson’s residuals of
the model in ascending order x is the quantiles of standard normal
distribution for n of 24. Plotted to validate model assumptions.
</p>
</div>
</div>
<div class="section level6">
<h6 id="model-summary-1">Model Summary<a class="anchor" aria-label="anchor" href="#model-summary-1"></a>
</h6>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_by_target</span><span class="op">$</span><span class="va">summary</span></span></code></pre></div>
<pre><code><span><span class="co">## Generalized linear mixed model fit by maximum likelihood (Laplace</span></span>
<span><span class="co">##   Approximation) [glmerMod]</span></span>
<span><span class="co">##  Family: binomial  ( logit )</span></span>
<span><span class="co">## Formula: cbind(sum_min, group_depth) ~ dose_group * label + (1 | sample)</span></span>
<span><span class="co">##    Data: mf_data</span></span>
<span><span class="co">## Control: ..1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##       AIC       BIC    logLik -2*log(L)  df.resid </span></span>
<span><span class="co">##    2825.5    3163.6   -1331.7    2663.5       399 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Scaled residuals: </span></span>
<span><span class="co">##     Min      1Q  Median      3Q     Max </span></span>
<span><span class="co">## -2.7513 -0.7140 -0.0326  0.6138  5.4812 </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Random effects:</span></span>
<span><span class="co">##  Groups Name        Variance Std.Dev.</span></span>
<span><span class="co">##  sample (Intercept) 0.01088  0.1043  </span></span>
<span><span class="co">## Number of obs: 480, groups:  sample, 24</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Fixed effects:</span></span>
<span><span class="co">##                                Estimate Std. Error z value Pr(&gt;|z|)    </span></span>
<span><span class="co">## (Intercept)                  -1.596e+01  2.222e-01 -71.807  &lt; 2e-16 ***</span></span>
<span><span class="co">## dose_groupHigh                1.903e+00  2.395e-01   7.947 1.92e-15 ***</span></span>
<span><span class="co">## dose_groupLow                 6.208e-01  2.738e-01   2.267 0.023380 *  </span></span>
<span><span class="co">## dose_groupMedium              1.293e+00  2.612e-01   4.951 7.39e-07 ***</span></span>
<span><span class="co">## labelchr1.2                   1.382e-01  2.824e-01   0.489 0.624624    </span></span>
<span><span class="co">## labelchr10                    4.517e-01  2.682e-01   1.684 0.092109 .  </span></span>
<span><span class="co">## labelchr11                    8.366e-01  2.641e-01   3.168 0.001535 ** </span></span>
<span><span class="co">## labelchr12                    5.183e-01  2.660e-01   1.948 0.051365 .  </span></span>
<span><span class="co">## labelchr13                    1.984e-02  2.908e-01   0.068 0.945593    </span></span>
<span><span class="co">## labelchr14                    6.560e-01  2.671e-01   2.456 0.014043 *  </span></span>
<span><span class="co">## labelchr15                    5.084e-01  2.660e-01   1.911 0.056012 .  </span></span>
<span><span class="co">## labelchr16                    7.757e-01  2.693e-01   2.881 0.003970 ** </span></span>
<span><span class="co">## labelchr17                    4.593e-01  2.885e-01   1.592 0.111307    </span></span>
<span><span class="co">## labelchr18                   -7.519e-03  2.958e-01  -0.025 0.979721    </span></span>
<span><span class="co">## labelchr19                   -2.378e-01  3.084e-01  -0.771 0.440585    </span></span>
<span><span class="co">## labelchr2                     6.258e-01  2.660e-01   2.352 0.018661 *  </span></span>
<span><span class="co">## labelchr3                     2.568e-01  2.863e-01   0.897 0.369732    </span></span>
<span><span class="co">## labelchr4                     5.106e-01  2.773e-01   1.841 0.065603 .  </span></span>
<span><span class="co">## labelchr5                     4.209e-01  2.824e-01   1.490 0.136186    </span></span>
<span><span class="co">## labelchr6                     2.937e-01  2.806e-01   1.046 0.295350    </span></span>
<span><span class="co">## labelchr7                    -3.983e-04  2.958e-01  -0.001 0.998926    </span></span>
<span><span class="co">## labelchr8                     4.383e-01  2.908e-01   1.507 0.131704    </span></span>
<span><span class="co">## labelchr9                     7.000e-01  2.671e-01   2.621 0.008766 ** </span></span>
<span><span class="co">## dose_groupHigh:labelchr1.2   -2.483e-01  3.022e-01  -0.822 0.411347    </span></span>
<span><span class="co">## dose_groupLow:labelchr1.2    -2.540e-03  3.459e-01  -0.007 0.994141    </span></span>
<span><span class="co">## dose_groupMedium:labelchr1.2 -2.494e-02  3.289e-01  -0.076 0.939548    </span></span>
<span><span class="co">## dose_groupHigh:labelchr10    -4.750e-01  2.882e-01  -1.648 0.099285 .  </span></span>
<span><span class="co">## dose_groupLow:labelchr10     -3.356e-01  3.360e-01  -0.999 0.317857    </span></span>
<span><span class="co">## dose_groupMedium:labelchr10  -3.318e-01  3.167e-01  -1.048 0.294861    </span></span>
<span><span class="co">## dose_groupHigh:labelchr11    -5.554e-02  2.809e-01  -0.198 0.843268    </span></span>
<span><span class="co">## dose_groupLow:labelchr11      4.252e-01  3.174e-01   1.340 0.180390    </span></span>
<span><span class="co">## dose_groupMedium:labelchr11   4.359e-01  3.029e-01   1.439 0.150068    </span></span>
<span><span class="co">## dose_groupHigh:labelchr12    -1.676e-01  2.836e-01  -0.591 0.554373    </span></span>
<span><span class="co">## dose_groupLow:labelchr12     -1.544e-01  3.287e-01  -0.470 0.638477    </span></span>
<span><span class="co">## dose_groupMedium:labelchr12  -1.756e-01  3.115e-01  -0.564 0.572958    </span></span>
<span><span class="co">## dose_groupHigh:labelchr13    -3.978e-01  3.125e-01  -1.273 0.202970    </span></span>
<span><span class="co">## dose_groupLow:labelchr13      2.170e-03  3.561e-01   0.006 0.995138    </span></span>
<span><span class="co">## dose_groupMedium:labelchr13  -2.386e-01  3.425e-01  -0.697 0.486037    </span></span>
<span><span class="co">## dose_groupHigh:labelchr14     1.207e-01  2.832e-01   0.426 0.670094    </span></span>
<span><span class="co">## dose_groupLow:labelchr14      2.822e-01  3.228e-01   0.874 0.381936    </span></span>
<span><span class="co">## dose_groupMedium:labelchr14   4.877e-01  3.055e-01   1.596 0.110378    </span></span>
<span><span class="co">## dose_groupHigh:labelchr15    -1.145e+00  2.935e-01  -3.903 9.51e-05 ***</span></span>
<span><span class="co">## dose_groupLow:labelchr15     -4.122e-01  3.348e-01  -1.231 0.218290    </span></span>
<span><span class="co">## dose_groupMedium:labelchr15  -6.712e-01  3.214e-01  -2.088 0.036792 *  </span></span>
<span><span class="co">## dose_groupHigh:labelchr16    -2.607e-01  2.878e-01  -0.906 0.364977    </span></span>
<span><span class="co">## dose_groupLow:labelchr16      1.133e-01  3.279e-01   0.346 0.729681    </span></span>
<span><span class="co">## dose_groupMedium:labelchr16   6.542e-02  3.125e-01   0.209 0.834154    </span></span>
<span><span class="co">## dose_groupHigh:labelchr17     1.326e-01  3.059e-01   0.434 0.664586    </span></span>
<span><span class="co">## dose_groupLow:labelchr17      3.705e-01  3.457e-01   1.072 0.283845    </span></span>
<span><span class="co">## dose_groupMedium:labelchr17   3.523e-01  3.308e-01   1.065 0.286780    </span></span>
<span><span class="co">## dose_groupHigh:labelchr18     3.822e-01  3.118e-01   1.226 0.220225    </span></span>
<span><span class="co">## dose_groupLow:labelchr18      4.491e-01  3.525e-01   1.274 0.202678    </span></span>
<span><span class="co">## dose_groupMedium:labelchr18   5.986e-01  3.347e-01   1.789 0.073665 .  </span></span>
<span><span class="co">## dose_groupHigh:labelchr19     6.478e-02  3.273e-01   0.198 0.843111    </span></span>
<span><span class="co">## dose_groupLow:labelchr19      3.157e-01  3.689e-01   0.856 0.392152    </span></span>
<span><span class="co">## dose_groupMedium:labelchr19   3.342e-01  3.522e-01   0.949 0.342680    </span></span>
<span><span class="co">## dose_groupHigh:labelchr2     -6.041e-01  2.869e-01  -2.106 0.035210 *  </span></span>
<span><span class="co">## dose_groupLow:labelchr2      -2.459e-01  3.312e-01  -0.742 0.457836    </span></span>
<span><span class="co">## dose_groupMedium:labelchr2   -2.993e-01  3.138e-01  -0.954 0.340146    </span></span>
<span><span class="co">## dose_groupHigh:labelchr3     -1.072e+00  3.178e-01  -3.375 0.000739 ***</span></span>
<span><span class="co">## dose_groupLow:labelchr3      -5.817e-01  3.701e-01  -1.572 0.116009    </span></span>
<span><span class="co">## dose_groupMedium:labelchr3   -6.911e-01  3.503e-01  -1.973 0.048510 *  </span></span>
<span><span class="co">## dose_groupHigh:labelchr4     -2.788e-01  2.967e-01  -0.940 0.347321    </span></span>
<span><span class="co">## dose_groupLow:labelchr4      -1.492e-01  3.434e-01  -0.435 0.663885    </span></span>
<span><span class="co">## dose_groupMedium:labelchr4   -5.001e-02  3.234e-01  -0.155 0.877121    </span></span>
<span><span class="co">## dose_groupHigh:labelchr5     -2.603e-01  3.023e-01  -0.861 0.389180    </span></span>
<span><span class="co">## dose_groupLow:labelchr5      -4.738e-02  3.471e-01  -0.136 0.891432    </span></span>
<span><span class="co">## dose_groupMedium:labelchr5   -1.236e-01  3.312e-01  -0.373 0.709068    </span></span>
<span><span class="co">## dose_groupHigh:labelchr6     -2.353e-01  2.999e-01  -0.785 0.432557    </span></span>
<span><span class="co">## dose_groupLow:labelchr6      -4.394e-02  3.448e-01  -0.127 0.898606    </span></span>
<span><span class="co">## dose_groupMedium:labelchr6   -2.395e-01  3.307e-01  -0.724 0.468894    </span></span>
<span><span class="co">## dose_groupHigh:labelchr7      1.376e-01  3.131e-01   0.439 0.660386    </span></span>
<span><span class="co">## dose_groupLow:labelchr7       4.021e-01  3.530e-01   1.139 0.254618    </span></span>
<span><span class="co">## dose_groupMedium:labelchr7    5.092e-01  3.355e-01   1.518 0.129058    </span></span>
<span><span class="co">## dose_groupHigh:labelchr8      2.822e-01  3.074e-01   0.918 0.358623    </span></span>
<span><span class="co">## dose_groupLow:labelchr8       5.457e-01  3.454e-01   1.580 0.114098    </span></span>
<span><span class="co">## dose_groupMedium:labelchr8    6.388e-01  3.299e-01   1.936 0.052824 .  </span></span>
<span><span class="co">## dose_groupHigh:labelchr9     -4.547e-01  2.867e-01  -1.586 0.112740    </span></span>
<span><span class="co">## dose_groupLow:labelchr9      -1.701e-01  3.304e-01  -0.515 0.606713    </span></span>
<span><span class="co">## dose_groupMedium:labelchr9   -4.265e-01  3.172e-01  -1.345 0.178701    </span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre>
</div>
<div class="section level6">
<h6 id="anova">ANOVA<a class="anchor" aria-label="anchor" href="#anova"></a>
</h6>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_by_target</span><span class="op">$</span><span class="va">anova</span></span></code></pre></div>
<pre><code><span><span class="co">## Analysis of Deviance Table (Type II Wald chisquare tests)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Response: cbind(sum_min, group_depth)</span></span>
<span><span class="co">##                    Chisq Df Pr(&gt;Chisq)    </span></span>
<span><span class="co">## dose_group        600.61  3  &lt; 2.2e-16 ***</span></span>
<span><span class="co">## label            1239.68 19  &lt; 2.2e-16 ***</span></span>
<span><span class="co">## dose_group:label  129.41 57  1.493e-07 ***</span></span>
<span><span class="co">## ---</span></span>
<span><span class="co">## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span></span></code></pre>
</div>
<div class="section level6">
<h6 id="model-estimates-1">Model Estimates<a class="anchor" aria-label="anchor" href="#model-estimates-1"></a>
</h6>
<p><strong>Table 20.</strong> model_by_target$point_estimates: Estimated
Mean Mutation Frequency per Dose and Genomic Target.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-0ed12bb554391c49c2e3" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-0ed12bb554391c49c2e3">{"x":{"filter":"none","vertical":false,"data":[["Control:chr1","High:chr1","Low:chr1","Medium:chr1","Control:chr1.2","High:chr1.2","Low:chr1.2","Medium:chr1.2","Control:chr10","High:chr10","Low:chr10","Medium:chr10","Control:chr11","High:chr11","Low:chr11","Medium:chr11","Control:chr12","High:chr12","Low:chr12","Medium:chr12","Control:chr13","High:chr13","Low:chr13","Medium:chr13","Control:chr14","High:chr14","Low:chr14","Medium:chr14","Control:chr15","High:chr15","Low:chr15","Medium:chr15","Control:chr16","High:chr16","Low:chr16","Medium:chr16","Control:chr17","High:chr17","Low:chr17","Medium:chr17","Control:chr18","High:chr18","Low:chr18","Medium:chr18","Control:chr19","High:chr19","Low:chr19","Medium:chr19","Control:chr2","High:chr2","Low:chr2","Medium:chr2","Control:chr3","High:chr3","Low:chr3","Medium:chr3","Control:chr4","High:chr4","Low:chr4","Medium:chr4","Control:chr5","High:chr5","Low:chr5","Medium:chr5","Control:chr6","High:chr6","Low:chr6","Medium:chr6","Control:chr7","High:chr7","Low:chr7","Medium:chr7","Control:chr8","High:chr8","Low:chr8","Medium:chr8","Control:chr9","High:chr9","Low:chr9","Medium:chr9"],[1.174932797045947e-07,7.88125247800619e-07,2.185866844864108e-07,4.282878757116527e-07,1.349056336285138e-07,7.059666725579152e-07,2.503441710178697e-07,4.796448141675249e-07,1.845751230406704e-07,7.699388957464031e-07,2.454784427682318e-07,4.828371170278464e-07,2.712331292276511e-07,1.721091409459824e-06,7.720118403900807e-07,1.528869341976379e-06,1.973001964593873e-07,1.119189859678827e-06,3.145318852788088e-07,6.033742551643626e-07,1.198477963333967e-07,5.400498846860443e-07,2.234514480150588e-07,3.441301330519316e-07,2.264086091138454e-07,1.713482745896675e-06,5.585617562575728e-07,1.344104214358601e-06,1.953431624320503e-07,4.168726067818566e-07,2.406647624915153e-07,3.639452002640926e-07,2.552061437329376e-07,1.318988438959318e-06,5.317538840990062e-07,9.931776558017677e-07,1.859967757035631e-07,1.424566812213175e-06,5.01193358828343e-07,9.643634480972733e-07,1.16613165922487e-07,1.146356157697704e-06,3.399271726669925e-07,7.734681848478962e-07,9.262386182004394e-08,6.628834812236883e-07,2.362889741840838e-07,4.716324682181592e-07,2.196768301048894e-07,8.053871302438059e-07,3.196001711886703e-07,5.936419371045899e-07,1.518986299687973e-07,3.486999285444691e-07,1.579514242357194e-07,2.77422336960644e-07,1.957836741359204e-07,9.937020984606937e-07,3.137458254131134e-07,6.788625227551306e-07,1.789725178390153e-07,9.254011011048573e-07,3.175570724476225e-07,5.765639814025972e-07,1.575993905330554e-07,8.35461815958605e-07,2.805963435069776e-07,4.521149496517134e-07,1.174464942251494e-07,9.04010831623778e-07,3.266564216904236e-07,7.123781560454438e-07,1.821207651370085e-07,1.619959416244377e-06,5.847671342772275e-07,1.257447219484351e-06,2.366028204001955e-07,1.00724969877913e-06,3.713302861842747e-07,5.630031562202195e-07],[2.610911891531588e-08,7.041919065498308e-08,3.497361595829037e-08,5.880416591435256e-08,2.489310453382039e-08,5.997436336884642e-08,3.352074904855903e-08,5.54877728914948e-08,2.987040639582259e-08,6.34292488088923e-08,3.387204206168782e-08,5.585983014207539e-08,4.204009522005315e-08,1.194319401031207e-07,7.345980056269211e-08,1.275826419443456e-07,3.123239299620678e-08,8.126981467253787e-08,3.89259085189622e-08,6.396100148139264e-08,2.361368029220218e-08,5.053254187636061e-08,3.184920075981991e-08,4.586496215890782e-08,3.623145440645891e-08,1.155155779836323e-07,5.830725970644785e-08,1.12236917910412e-07,3.092377505526921e-08,4.369011498876823e-08,3.347085560570722e-08,4.81686863037799e-08,4.17781190248131e-08,1.018368185957717e-07,6.072088109334177e-08,9.982595925059703e-08,3.601710816309042e-08,1.101485125197214e-07,5.993872991591916e-08,1.019429371268863e-07,2.38355926310915e-08,8.381917145925248e-08,4.133945978608267e-08,7.526465834550884e-08,2.058498452953981e-08,5.801536910852348e-08,3.260547635509741e-08,5.568712531035517e-08,3.477574700557513e-08,6.814580511150003e-08,4.161124084643162e-08,6.67532346904963e-08,2.892825308024051e-08,4.217353693772907e-08,2.870721742486946e-08,4.447570350305906e-08,3.458464739819476e-08,8.17137103093762e-08,4.329089488114085e-08,7.636720196368286e-08,3.302366883458987e-08,7.877277958772254e-08,4.347904442185993e-08,7.036717588600561e-08,2.864768877220069e-08,6.895277606343368e-08,3.784657747913689e-08,5.646454992071798e-08,2.400639230890775e-08,7.117111458795793e-08,4.018900313472273e-08,7.135211371527863e-08,3.588290742337117e-08,1.21380068884964e-07,6.613952714592197e-08,1.210249277188754e-07,3.786305854266426e-08,8.124749948809588e-08,4.68012562852975e-08,6.719750903054299e-08],[7.600825148041449e-08,6.61515716617396e-07,1.597469015133391e-07,3.272392059879494e-07,9.396462331145166e-08,5.976839869443556e-07,1.925585193036707e-07,3.823388167343322e-07,1.344065359545955e-07,6.551374358130519e-07,1.873101699540694e-07,3.848791894235811e-07,2.00174963720084e-07,1.502229850727653e-06,6.406620006435932e-07,1.298190542417578e-06,1.44671604302691e-07,9.707197951508176e-07,2.467865485099908e-07,4.901790138587482e-07,8.145492028439167e-08,4.495591406624066e-07,1.689891730893513e-07,2.650185643727401e-07,1.654542179000283e-07,1.501396565002092e-06,4.552141935945901e-07,1.14118208183258e-06,1.432349034832925e-07,3.394639235930686e-07,1.832441943597479e-07,2.807877852180927e-07,1.85159721906866e-07,1.133760150700993e-06,4.251205692031841e-07,8.155887739478559e-07,1.272552122371748e-07,1.224242121095573e-06,3.964682276343288e-07,7.83899085856586e-07,7.812013795437747e-08,9.933024039572258e-07,2.678360107748346e-07,6.391660765121432e-07,5.991692090996728e-08,5.583934118962316e-07,1.802960852817622e-07,3.741964459341869e-07,1.610777456885333e-07,6.823110569071314e-07,2.476181248740774e-07,4.762220964903923e-07,1.045796624294031e-07,2.751080059042737e-07,1.106165507133981e-07,2.026183503544696e-07,1.384886123123925e-07,8.457861026252337e-07,2.394024000249388e-07,5.445374064863807e-07,1.246591774646508e-07,7.832011386727746e-07,2.428161708316745e-07,4.539018873670576e-07,1.103639031618192e-07,7.106811396166587e-07,2.154134041379454e-07,3.539504502496665e-07,7.867777661828504e-08,7.747473473433106e-07,2.566650095431906e-07,5.854015499556256e-07,1.237795352307536e-07,1.398702776104621e-06,4.684989478919595e-07,1.041272916648394e-06,1.729035459741732e-07,8.59957354942801e-07,2.900532328288923e-07,4.455693052771279e-07],[1.816206860027466e-07,9.389669672505144e-07,2.990990008702734e-07,5.6053951092998e-07,1.936849139956359e-07,8.338669826349134e-07,3.254709487238445e-07,6.017153835511707e-07,2.534696382398193e-07,9.048573181404213e-07,3.217105930697325e-07,6.057269085629564e-07,3.67515542520578e-07,1.971839155160968e-06,9.302912941672059e-07,1.800538047737079e-06,2.69074001843973e-07,1.290368186849741e-06,4.008739838307544e-07,7.427092582549003e-07,1.763367300075019e-07,6.487553061865674e-07,2.954659680690092e-07,4.468575579021642e-07,3.098189875814418e-07,1.955528065619038e-06,6.853723806147417e-07,1.58310945099609e-06,2.664081880950575e-07,5.11932957251216e-07,3.160783789492886e-07,4.717303093950099e-07,3.517513157197158e-07,1.534478435348677e-06,6.651341142687326e-07,1.209435303050112e-06,2.718537022094209e-07,1.657670952085123e-06,6.335811180443911e-07,1.18637319114849e-06,1.740733032807385e-07,1.322993315083161e-06,4.314225050735121e-07,9.359899640426762e-07,1.431845903989274e-07,7.86926386160326e-07,3.096710571042795e-07,5.944395984901332e-07,2.995938978327023e-07,9.506638108762761e-07,4.125072406382664e-07,7.400134350893139e-07,2.206279237320474e-07,4.419778325507063e-07,2.255417680011819e-07,3.79842955537158e-07,2.767826640626265e-07,1.167486504472303e-06,4.11175673067194e-07,8.46322620469943e-07,2.569498916413081e-07,1.093419245249434e-06,4.15303865126062e-07,7.323741846043281e-07,2.250515538579028e-07,9.821513573602771e-07,3.655032903108773e-07,5.775043584614396e-07,1.753186172596076e-07,1.054841409261355e-06,4.157341821602529e-07,8.668966408598928e-07,2.679600713660516e-07,1.876215987493352e-06,7.298897956322812e-07,1.518500562636683e-06,3.237694999597585e-07,1.179770077968726e-06,4.753823292810425e-07,7.11387768770974e-07],["Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium","Control","High","Low","Medium"],["chr1","chr1","chr1","chr1","chr1.2","chr1.2","chr1.2","chr1.2","chr10","chr10","chr10","chr10","chr11","chr11","chr11","chr11","chr12","chr12","chr12","chr12","chr13","chr13","chr13","chr13","chr14","chr14","chr14","chr14","chr15","chr15","chr15","chr15","chr16","chr16","chr16","chr16","chr17","chr17","chr17","chr17","chr18","chr18","chr18","chr18","chr19","chr19","chr19","chr19","chr2","chr2","chr2","chr2","chr3","chr3","chr3","chr3","chr4","chr4","chr4","chr4","chr5","chr5","chr5","chr5","chr6","chr6","chr6","chr6","chr7","chr7","chr7","chr7","chr8","chr8","chr8","chr8","chr9","chr9","chr9","chr9"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Estimate<\/th>\n      <th>Std.Err<\/th>\n      <th>Lower<\/th>\n      <th>Upper<\/th>\n      <th>dose_group<\/th>\n      <th>label<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":[1,2,3,4],"render":"function(data, type, row, meta) {\n            return data.toExponential(2); \n          }"},{"className":"dt-right","targets":[1,2,3,4]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Estimate","targets":1},{"name":"Std.Err","targets":2},{"name":"Lower","targets":3},{"name":"Upper","targets":4},{"name":"dose_group","targets":5},{"name":"label","targets":6}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
<div class="section level6">
<h6 id="pairwise-comparisons-1">Pairwise Comparisons<a class="anchor" aria-label="anchor" href="#pairwise-comparisons-1"></a>
</h6>
<p><strong>Table 21.</strong> model_by_target$pairwise_comparisons.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-ec658d41f8c4f2d124e9" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-ec658d41f8c4f2d124e9">{"x":{"filter":"none","vertical":false,"data":[["Low:chr1 vs Control:chr1","Medium:chr1 vs Control:chr1","High:chr1 vs Control:chr1","Low:chr1.2 vs Control:chr1.2","Medium:chr1.2 vs Control:chr1.2","High:chr1.2 vs Control:chr1.2","Low:chr10 vs Control:chr10","Medium:chr10 vs Control:chr10","High:chr10 vs Control:chr10","Low:chr11 vs Control:chr11","Medium:chr11 vs Control:chr11","High:chr11 vs Control:chr11","Low:chr12 vs Control:chr12","Medium:chr12 vs Control:chr12","High:chr12 vs Control:chr12","Low:chr13 vs Control:chr13","Medium:chr13 vs Control:chr13","High:chr13 vs Control:chr13","Low:chr14 vs Control:chr14","Medium:chr14 vs Control:chr14","High:chr14 vs Control:chr14","Low:chr15 vs Control:chr15","Medium:chr15 vs Control:chr15","High:chr15 vs Control:chr15","Low:chr16 vs Control:chr16","Medium:chr16 vs Control:chr16","High:chr16 vs Control:chr16","Low:chr17 vs Control:chr17","Medium:chr17 vs Control:chr17","High:chr17 vs Control:chr17","Low:chr18 vs Control:chr18","Medium:chr18 vs Control:chr18","High:chr18 vs Control:chr18","Low:chr19 vs Control:chr19","Medium:chr19 vs Control:chr19","High:chr19 vs Control:chr19","Low:chr2 vs Control:chr2","Medium:chr2 vs Control:chr2","High:chr2 vs Control:chr2","Low:chr3 vs Control:chr3","Medium:chr3 vs Control:chr3","High:chr3 vs Control:chr3","Low:chr4 vs Control:chr4","Medium:chr4 vs Control:chr4","High:chr4 vs Control:chr4","Low:chr5 vs Control:chr5","Medium:chr5 vs Control:chr5","High:chr5 vs Control:chr5","Low:chr6 vs Control:chr6","Medium:chr6 vs Control:chr6","High:chr6 vs Control:chr6","Low:chr7 vs Control:chr7","Medium:chr7 vs Control:chr7","High:chr7 vs Control:chr7","Low:chr8 vs Control:chr8","Medium:chr8 vs Control:chr8","High:chr8 vs Control:chr8","Low:chr9 vs Control:chr9","Medium:chr9 vs Control:chr9","High:chr9 vs Control:chr9"],[1.860418613183564,3.645211681795481,6.707832565250944,1.85569841884614,3.555409816971101,5.233040708306646,1.329964941776804,2.615938210272548,4.171412068227361,2.846303630343463,5.636735255497386,6.345432117236964,1.594179280726427,3.058153341923113,5.672522783874683,1.864460214132381,2.87139308005813,4.506131119705485,2.467051754099643,5.936630323464169,7.568098901376497,1.232010168644781,1.863106933116689,2.134052718261201,2.083624932852182,3.891668285388485,5.168325572677365,2.694634661985389,5.184839599769468,7.659094125822983,2.914998233500861,6.632768939332474,9.83041793462402,2.551059408893606,5.091911079398525,7.156724716483811,1.454865180984587,2.702342057745202,3.666236124489121,1.039847589594228,1.826364971281384,2.295609437794783,1.602512705912851,3.467411293363706,5.075510523777526,1.774334273674704,3.221522434641124,5.170632409257657,1.780440536971014,2.86876077453412,5.301174155133373,2.781321178171662,6.065554878801126,7.697214272660661,3.210875672729084,6.904469232480878,8.894973700696283,1.569424597543674,2.379528508020079,4.257133102113673],[0.5094243445269129,0.9523112876230735,1.606580380709246,0.4230611976941378,0.774472598685223,1.063031558349787,0.2828388996091907,0.5205184084595208,0.757506717552575,0.5176513375491637,0.9925806158235084,1.077590302875372,0.3203147761339311,0.5827703335914091,0.9879167159254402,0.4533896530669063,0.6831468806454769,0.9828753297772794,0.4713508282450781,1.071887444987505,1.314179672638959,0.2595999402047224,0.3845330712555969,0.4051567374431498,0.4158700708089288,0.747793438221904,0.9354540981313217,0.6132764586772003,1.144122353712777,1.597002042641902,0.6932911518419593,1.50179743715266,2.134015978038764,0.6673340462582219,1.281638978810363,1.709409508655244,0.2981908489978758,0.5248730027469525,0.6580825293195746,0.2737362942592773,0.454731508553622,0.5178973428936785,0.3591938102747744,0.7263310848488914,0.9889614921961082,0.4076764627554532,0.7128528838146766,1.050706609305862,0.4029932487828106,0.63282877087697,1.058296585127847,0.6635317996270325,1.380939703354937,1.685995059862849,0.7294455655532918,1.514349629895873,1.875012857303892,0.3196849118358639,0.4751597551751844,0.7629116151020234],[5.140051956285146,24.51111412590189,63.14849406427542,7.354485478378336,33.9100048216,66.37552580343807,1.797863542875751,23.35566796312175,61.85922502710306,33.08024592644858,96.44215619439875,118.3844120497217,5.38717626236594,34.40809256576594,99.31803198158944,6.562956154288858,19.6559999885871,47.63619580250386,22.33913641015492,97.31472964603051,135.8501710901952,0.9804923117829134,9.08932606573225,15.94148441850713,13.52835567648665,50.00847984261954,82.35525309729206,18.96988362739155,55.62212559038833,95.33537391583032,20.23516070859374,69.82638295305429,110.8419178652077,12.81675959194155,41.81710560106551,67.89061207661472,3.345942136822313,26.19683152162467,52.38526413967514,0.02203199265202625,5.852354482200907,13.56778550831217,4.42632033272018,35.23456985792637,69.50261784329494,6.228641012651717,27.95026095633037,65.37276968666441,6.495334204482858,22.82436408492813,69.80461361189268,18.38515466155532,62.69067052450988,86.81211525863273,26.36715207726319,77.60656626270425,107.4922216410494,4.895862314801877,18.84702893871026,65.34021196735038],[0.02337994847046232,7.38824322388254e-07,1.887379141862766e-15,0.00668956509375096,5.772111566493265e-09,3.33066907387547e-16,0.179970996870508,1.346461687368183e-06,3.663735981263017e-15,8.843262455648926e-09,0,0,0.02028527126806545,4.468614145380911e-09,0,0.01041232533482717,9.271173554981083e-06,5.131228775212548e-12,2.28502003751796e-06,0,0,0.3220772925365391,0.002571055959358692,6.533102236205846e-05,0.0002349858434343988,1.530775506353166e-12,0,1.327980781784532e-05,8.781864124784988e-14,0,6.848260289515906e-06,1.110223024625157e-16,0,0.000343528176320218,1.002227190127769e-10,2.220446049250313e-16,0.06737090107314447,3.083280164517532e-07,4.560796185160143e-13,0.882002022798234,0.01555622468476503,0.0002301006889470214,0.03538872361502587,2.922887598622026e-09,1.110223024625157e-16,0.01257001693714932,1.244741704153896e-07,6.661338147750939e-16,0.01081579622771023,1.77501179710049e-06,1.110223024625157e-16,1.804584798104081e-05,2.442490654175344e-15,0,2.823007971830549e-07,0,0,0.02692112564452687,1.416315357050113e-05,6.661338147750939e-16],[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],[1.087758789794525,2.184471645485083,4.194812278184079,1.187000482466525,2.319919903631327,3.514328761468415,0.8766337714568608,1.771151520255874,2.922195878508369,1.99285163636602,3.991520499601642,4.548926778452794,1.07524440683298,2.104998977014648,4.032110881477162,1.157610333420143,1.801271648126971,2.93859760303259,1.696481394673218,4.16726801114948,5.384904173376155,0.8151837032840944,1.243241094956744,1.470963196997571,1.409056756565326,2.670394271033437,3.624803742252551,1.724938015344216,3.364366814998677,5.089708291806662,1.828908730806758,4.255641727413929,6.423779907820893,1.527755848312825,3.109084698133124,4.481276631988958,0.9735527727486291,1.846768067676573,2.578881263520459,0.6207172528097139,1.121124741697696,1.475244558124957,1.032784366869858,2.299858506881824,3.464363087641641,1.130995234193951,2.087895519989434,3.471953469249448,1.14251663032695,1.861766301263557,3.5846188527161,1.742538719621733,3.882202701077175,5.010564938030808,2.057064139397004,4.491973670333567,5.884586322490436,1.052820739264832,1.608856992838493,2.996230644126589],[3.181916293164278,6.082737778977948,10.72634834160428,2.901108021921279,5.448868707418671,7.792303143361063,2.017725992253106,3.863663069874071,5.954658539808434,4.065252128291609,7.96007044025908,8.851430395667387,2.363562705322857,4.442900906288862,7.98031494654462,3.002920576747225,4.577265305196505,6.90983265181443,3.58762812048316,8.457238532098364,10.63642344170183,1.861971785659173,2.792030812292484,3.096053669876786,3.081134127901243,5.6714778816678,7.369113233311638,4.209459062866181,7.99037773036297,11.52555696023855,4.646057268021731,10.33771793315592,15.04365313819715,4.25978019648335,8.339289841820609,11.42949049427406,2.174132470359496,3.954287885346682,5.212061334751087,1.741989617157301,2.975234498234877,3.572168873200556,2.486527735112134,5.227687286574432,7.435943180688096,2.783621026467432,4.970654277254543,7.700402597113489,2.774549115125348,4.42041967131973,7.839730966588133,4.439354723678944,9.476825096623607,11.82443662381423,5.011862483172536,10.61263909383847,13.44538983712175,2.339518472152308,3.519365577975059,6.048660601158409],[0.7581543908772616,4.432849318247722e-05,1.13242748511766e-13,0.3315023725143706,3.463266350367533e-07,1.998401444325282e-14,0.9999932430177531,8.078449238801966e-05,2.19824158875781e-13,5.305956088941244e-07,0,0,0.7075994205606213,2.681168134177625e-07,0,0.4663498320009645,0.0005561183008185866,3.078737265127529e-10,0.0001370919609290366,0,0,0.9999999999257304,0.1431239540340262,0.003912316262629112,0.01400185667225773,9.184653038118995e-11,0,0.0007964764038632666,5.269118474870993e-12,0,0.0004108126177163607,6.661338147750939e-15,0,0.02040419055755782,6.013363140766614e-09,1.332267629550188e-14,0.9847756869981413,1.849951272103745e-05,2.736477711096086e-11,1,0.6096492344812281,0.01371274187624993,0.8848828382957761,1.753732408182884e-07,6.661338147750939e-15,0.5318569076376329,7.468422800971375e-06,3.996802888650564e-14,0.4792487138891175,0.0001064951313369811,6.661338147750939e-15,0.001082174674756686,1.465494392505207e-13,0,1.693790677381735e-05,0,0,0.8055165523130204,0.0008494342584246528,3.996802888650564e-14],["","***","***","","***","***","","***","***","***","***","***","","***","***","","***","***","***","***","***","","","**","*","***","***","***","***","***","***","***","***","*","***","***","","***","***","","","*","","***","***","","***","***","","***","***","**","***","***","***","***","***","","***","***"],["Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High","Low","Medium","High"],["Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control","Control"],["chr1","chr1","chr1","chr1.2","chr1.2","chr1.2","chr10","chr10","chr10","chr11","chr11","chr11","chr12","chr12","chr12","chr13","chr13","chr13","chr14","chr14","chr14","chr15","chr15","chr15","chr16","chr16","chr16","chr17","chr17","chr17","chr18","chr18","chr18","chr19","chr19","chr19","chr2","chr2","chr2","chr3","chr3","chr3","chr4","chr4","chr4","chr5","chr5","chr5","chr6","chr6","chr6","chr7","chr7","chr7","chr8","chr8","chr8","chr9","chr9","chr9"],["chr1","chr1","chr1","chr1.2","chr1.2","chr1.2","chr10","chr10","chr10","chr11","chr11","chr11","chr12","chr12","chr12","chr13","chr13","chr13","chr14","chr14","chr14","chr15","chr15","chr15","chr16","chr16","chr16","chr17","chr17","chr17","chr18","chr18","chr18","chr19","chr19","chr19","chr2","chr2","chr2","chr3","chr3","chr3","chr4","chr4","chr4","chr5","chr5","chr5","chr6","chr6","chr6","chr7","chr7","chr7","chr8","chr8","chr8","chr9","chr9","chr9"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Fold.Change<\/th>\n      <th>FC.Std.Err<\/th>\n      <th>Obs.T<\/th>\n      <th>p.value<\/th>\n      <th>df<\/th>\n      <th>FC.Lower<\/th>\n      <th>FC.Upper<\/th>\n      <th>adj_p.value<\/th>\n      <th>Significance<\/th>\n      <th>dose_group_1<\/th>\n      <th>dose_group_2<\/th>\n      <th>label_1<\/th>\n      <th>label_2<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"scrollX":true,"columnDefs":[{"targets":1,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":2,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":6,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":7,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":[4,8],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[1,2,3,4,5,6,7,8]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Fold.Change","targets":1},{"name":"FC.Std.Err","targets":2},{"name":"Obs.T","targets":3},{"name":"p.value","targets":4},{"name":"df","targets":5},{"name":"FC.Lower","targets":6},{"name":"FC.Upper","targets":7},{"name":"adj_p.value","targets":8},{"name":"Significance","targets":9},{"name":"dose_group_1","targets":10},{"name":"dose_group_2","targets":11},{"name":"label_1","targets":12},{"name":"label_2","targets":13}],"rowCallback":"function(row, data, dataIndex) {\n      $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render","options.columnDefs.4.render","options.columnDefs.5.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
</div>
</div>
<div class="section level4">
<h4 class="tabset" id="plot-model-results">Plot Model Results<a class="anchor" aria-label="anchor" href="#plot-model-results"></a>
</h4>
<p>Plot the results of <code><a href="../reference/model_mf.html">model_mf()</a></code> using
<code><a href="../reference/plot_model_mf.html">plot_model_mf()</a></code>. This function will create a bar or line
plot of the point estimates. Users can include the estimated standard
error as error bars with <code>plot_error_bars</code> and add
significance labels based on the pairwise comparisons with
<code>plot_signif</code>. The function can plot model results with up to
two fixed effects. Users must specify which fixed effect should be
represented on the x-axis with <code>x_effect</code>. The other fixed
effect will be represented by the fill aesthetic.</p>
<p>When adding significance labels, the function will generate a unique
symbol for each unique level or combination of levels in column 2 of the
contrasts table. If using two fixed effects, but your comparisons only
contrast levels of one fixed effect, you may specify this using the
<code>ref_effect</code> parameter. Supply it with the fixed effect that
is being contrasted to reduce the number of unique symbols
generated.</p>
<p>The output is a ggplot object that can be modified with ggplot2.</p>
<div class="section level5">
<h5 id="plot-model-by-dose">Plot Model by Dose<a class="anchor" aria-label="anchor" href="#plot-model-by-dose"></a>
</h5>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_model_mf.html">plot_model_mf</a></span><span class="op">(</span></span>
<span>  <span class="va">model_by_dose</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>  x_effect <span class="op">=</span> <span class="st">"dose"</span>,</span>
<span>  plot_error_bars <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_signif <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  x_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span>,</span>
<span>  x_label <span class="op">=</span> <span class="st">"Dose Group"</span>,</span>
<span>  y_label <span class="op">=</span> <span class="st">"Estimated Mean Mutation Frequency (mutations/bp)"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-model-dose-1.png" alt="Mean Mutation Frequency Minimum (mutations/bp) per Dose estimated using a generalized linear model. Error bars are the S.E.M. Symbols indicate significance differences (p &lt; 0.05)." width="700"><p class="caption">
Mean Mutation Frequency Minimum (mutations/bp) per Dose estimated using
a generalized linear model. Error bars are the S.E.M. Symbols indicate
significance differences (p &lt; 0.05).
</p>
</div>
</div>
<div class="section level5">
<h5 id="plot-model-by-dose-and-target">Plot Model by Dose and Target<a class="anchor" aria-label="anchor" href="#plot-model-by-dose-and-target"></a>
</h5>
<p>In this example, we only made comparisons between dose groups. For
each contrast, we held the label (target) constant. Thus, we will set
the ref_effect to dose_group so that significance labels are generated
to indicate differences in dose, and not label.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define the order of the genomic targets for the x-axis:</span></span>
<span><span class="co"># We will order them from lowest to highest MF at the High dose.</span></span>
<span><span class="va">label_order</span> <span class="op">&lt;-</span> <span class="va">model_by_target</span><span class="op">$</span><span class="va">point_estimates</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">dose_group</span> <span class="op">==</span> <span class="st">"High"</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">Estimate</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">label</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define the order of the doses for the fill</span></span>
<span><span class="va">dose_order</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_model_mf.html">plot_model_mf</a></span><span class="op">(</span></span>
<span>  model <span class="op">=</span> <span class="va">model_by_target</span>,</span>
<span>  plot_type <span class="op">=</span> <span class="st">"bar"</span>,</span>
<span>  x_effect <span class="op">=</span> <span class="st">"label"</span>,</span>
<span>  plot_error_bars <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_signif <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ref_effect <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  x_order <span class="op">=</span> <span class="va">label_order</span>,</span>
<span>  fill_order <span class="op">=</span> <span class="va">dose_order</span>,</span>
<span>  x_label <span class="op">=</span> <span class="st">"Target"</span>,</span>
<span>  y_label <span class="op">=</span> <span class="st">"Mutation Frequency (mutations/bp)"</span>,</span>
<span>  fill_label <span class="op">=</span> <span class="st">"Dose"</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="st">""</span>,</span>
<span>  custom_palette <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#ef476f"</span>,</span>
<span>                     <span class="st">"#ffd166"</span>,</span>
<span>                     <span class="st">"#06d6a0"</span>,</span>
<span>                     <span class="st">"#118ab2"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Rotate the x-axis labels for clarity using ggplot2 functions.</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="va">plot</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-model-rg-1.png" alt="Mean Mutation Frequency Minimum (mutations/bp) per Genomic Target and Dose estimated using a generalized linear mixed model. Error bars are the SEM. Symbols indicate significance differences (p &lt; 0.05) between dose levels for individual genomic regions." width="700"><p class="caption">
Mean Mutation Frequency Minimum (mutations/bp) per Genomic Target and
Dose estimated using a generalized linear mixed model. Error bars are
the SEM. Symbols indicate significance differences (p &lt; 0.05) between
dose levels for individual genomic regions.
</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="benchmark-dose-modeling">Benchmark Dose Modeling<a class="anchor" aria-label="anchor" href="#benchmark-dose-modeling"></a>
</h3>
<p>Dose-response models are essential for quatitative risk assessment of
mutagenicity, as they provide a framework to evaluate the levels at
which exposure to a substance might cause an adverse effect. The
<strong>benchmark dose</strong> (BMD) is a dose that produces a
predetermined change in the measured response, defined as the
<strong>benchmark response</strong> (BMR). The BMD is used as a point of
departure to derive human health-based guidance values to inform
regulatory risk assessment such as the reference dose (RfD), the derived
no-effect level (DNEL) or the acceptable daily intake (ADI).</p>
<p>The BMD is estimated by applying various mathmatical models to fit
the dose-response data. Some requirements must be met before modelling
the BMD. There must be a clear dose-response trend in the MF data. We
suggest using <code><a href="../reference/model_mf.html">model_mf()</a></code> to test for significant increases
in MF with dose prior to running a BMD analysis. In general, studies
with more dose groups and a graded monotonic response with dose will be
more useful for BMD analysis. A minimum of three dose groups + 1 control
group is suggested. Datasets in which a response is only observed at the
high dose are usually not suitable for BMD modeling. However, if the one
elevated response is near the BMR, adequate BMD computation may result.
For a better estimate of the BMD, it is preferable to have studies with
one or more doses near the level of the BMR.</p>
<p>Protection and safety authorities recommend the use of model
averaging to determine the BMD and its confidence intervals. Model
averaging incorporates information across multiple models to acount for
model uncertainty, allowing the BMD to be more accurately estimated.</p>
<p>Ideally, the BMR would be based on a consensus scientific definition
of what minimal level of change in MF is biologically significant.
Currently, the default provided by this package calculates the BMD at a
50% relative increase in MF from the background. This BMR was selected
based on previous recommendations for genotoxicity assessment by <span class="citation">White, Long, and Johnson (<a href="#ref-white-2020">2020</a>)</span>.</p>
<p>MutSeqR provides two functions for BMD modelling, each employing
widely-used software designed to be consistent with methods used by
regulatory authorities.</p>
<ol style="list-style-type: decimal">
<li>
<code><a href="../reference/bmd_proast.html">bmd_proast()</a></code> runs a modified version of the <a href="https://www.rivm.nl/en/proast" class="external-link">proast71.1</a> R library that is
parametirized instead of menu-based.</li>
<li>
<code>bmd_toxicr</code> uses the <a href="https://github.com/NIEHS/ToxicR" class="external-link">ToxicR</a> library, available on
Github.</li>
</ol>
<div class="section level4">
<h4 id="proast">PROAST<a class="anchor" aria-label="anchor" href="#proast"></a>
</h4>
<p><code>bmd_proast</code> will analyze the continuous, individual MF
data following a log transformation. PROAST uses four families of nested
models: exponential, Hill, inverse exponential, and log-normal. The
Akaike information criterion (AIC) is used to select best fit. BMD
confidence intervals are assessed by the Maximum Likelihood Profile
method, or by model averaging via bootstrapping (recommended). BMR
values are a user-defined relative increase in MF from the control.
Alternatively, users may set the BMR as one standard-deviation from the
control.</p>
<div class="section level5">
<h5 id="general-usage-bmd_proast">General Usage: bmd_proast<a class="anchor" aria-label="anchor" href="#general-usage-bmd_proast"></a>
</h5>
<p>Supply <code>bmd_proast</code> with per-sample mf data calculated
using <code><a href="../reference/calculate_mf.html">calculate_mf()</a></code>, with the dose column retained in the
summary table. Dose column must be numeric.</p>
<p>Specify the column that contains the numeric dose values in
<code>dose_col</code>. The function can model more than one response
variable at once. Supply all response variables to
<code>response_col</code>. If you wish to include a covariate in the
analysis, supply the covariate variable to <code>covariate_col</code>.
PROAST will assess if the BMD values differ significantly between levels
of the covariate and give a BMD estimate for each.</p>
<p>It is highly recommended to use model averaging when calculating the
BMD confidence intervals. Specify the number of bootstraps to run with
<code>num_bootstraps</code>. The recommended value is 200, but be aware
that this may take some time to run. PROAST model averaging will return
the upper and lower 90% BMD confidence intervals. MutSeqR calculates the
model-averaged BMD value as the median BMD of all bootstrap runs.</p>
<p>Users may choose to generate model plots with
<code>plot_results</code>. Plots may be automatically saved to an output
directory specified in <code>output_path</code>. Alternatively, when
output_path is NULL, plots will automcatically be displayed and then
returned within a list alongside the BMD summary results.</p>
<p>The function will output a data frame of final results, including a
BMD estimate for each model family and the model averaging results, if
applicable. Users may access the raw, unparsed PROAST results by setting
<code>raw_results = TRUE</code>.</p>
</div>
<div class="section level5">
<h5 id="example-1">Example<a class="anchor" aria-label="anchor" href="#example-1"></a>
</h5>
<p><em>Example 5.1. Calculate the BMD with model averaging for a 50%
relative</em> <em>increase in MF from control. This will be calculated
for both MFmin and MFmax.</em> * See example 3.1 for calculating
mf_data_global.*</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proast_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bmd_proast.html">bmd_proast</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_global</span>,</span>
<span>  dose_col <span class="op">=</span> <span class="st">"dose"</span>,</span>
<span>  response_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mf_min"</span>, <span class="st">"mf_max"</span><span class="op">)</span>,</span>
<span>  bmr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  model_averaging <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  num_bootstraps <span class="op">=</span> <span class="fl">10</span>, <span class="co"># recommended value 200</span></span>
<span>  plot_results <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 22.</strong> BMD Estimates by PROAST.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-6b83523733b890d61edc" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-6b83523733b890d61edc">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10"],["Expon. m3-","Expon. m5-","Hill m3-","Hill m5-","Inv.Expon. m3-","Inv.Expon. m3-","LN m3-","LN m5-","Model averaging","Model averaging"],["mf_max","mf_min","mf_max","mf_min","mf_max","mf_min","mf_max","mf_min","mf_max","mf_min"],["0.5","0.5","0.5","0.5","0.5","0.5","0.5","0.5","0.5","0.5"],["3.889","8.45","3.964","9.03","5.297","4.714","4.699","9.171","4.4125","9.107"],["2.52","6.47","2.59","7.02","3.79","2.9","3.23","7.27","3.57","8.39"],["5.64","10.3","5.72","10.7","7.09","6.96","6.49","10.7","5.45","9.87"],["-3.72","-18.46","-3.8","-18.46","-4.56","-10.18","-4.26","-18.46","N/A","N/A"],[0.2052,0.3316,0.2136,0.3316,0.3123,0.0053,0.2688,0.3316,null,null]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>Model<\/th>\n      <th>Response<\/th>\n      <th>BMR<\/th>\n      <th>BMD<\/th>\n      <th>BMDL<\/th>\n      <th>BMDU<\/th>\n      <th>AIC<\/th>\n      <th>weights<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":3,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":4,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":5,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":6,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":7,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":8,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"className":"dt-right","targets":8},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"Model","targets":1},{"name":"Response","targets":2},{"name":"BMR","targets":3},{"name":"BMD","targets":4},{"name":"BMDL","targets":5},{"name":"BMDU","targets":6},{"name":"AIC","targets":7},{"name":"weights","targets":8}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.columnDefs.2.render","options.columnDefs.3.render","options.columnDefs.4.render","options.columnDefs.5.render"],"jsHooks":[]}</script>
</div>
</div>
<div class="section level4">
<h4 id="toxicr">ToxicR<a class="anchor" aria-label="anchor" href="#toxicr"></a>
</h4>
<p><code>bmd_toxicr</code> will analyze continuous, individual or
continuous, summary MF data, assuming either a normal (default) or
log-normal distribution. This function employs Bayesian estimation using
either the Laplace Maximum <em>a posteriori</em> approach (default)
<span class="citation">(<a href="#ref-gelman-1995">Gelman et al.
1995</a>)</span> or Markov chain Monte Carlo (MCMC) simulation <span class="citation">(<a href="#ref-brooks-2011">Brooks et al.
2011</a>)</span>. The default model parameter’s prior distributions are
specified in <span class="citation">(<a href="#ref-wheeler-2020">Wheeler
et al. 2020</a>)</span>, but they are user-modifiable. The default
models are described in the European Food Safety Authority’s 2022
Guidance on the use of benchmark dose approach in risk assessment <span class="citation">(<a href="#ref-efsa-2022">Committee et al.
2022</a>)</span>. Model averaging may be applied using described
methodologies (<span class="citation">Wheeler et al. (<a href="#ref-wheeler-2020">2020</a>)</span>; <span class="citation">Wheeler et al. (<a href="#ref-wheeler-2022">2022</a>)</span>).</p>
<p>ToxicR offers several options for the BMR:</p>
<ul>
<li>
<strong>Relative deviation (<em>rel</em>)</strong>: the BMD
represents the dose that changes the mean MF a certain percentage from
the background dose.</li>
<li>
<strong>Standard deviation (<em>sd</em>)</strong>: the BMD
represents the dose associated with the mean MF changing a specified
number of standard deviations from the background mean.</li>
<li>
<strong>Absolute deviation (<em>abs</em>)</strong>: the BMD
represents the dose associated with a specified absolute deviation from
the background mean.</li>
<li>
<strong>Hybrid deviation (<em>hybrid</em>)</strong>: the BMD
represents the dose that changes the probability of an adverse event by
a specified amount.</li>
</ul>
<p>One of these options can be specified using the <code>bmr_type</code>
parameter. The <code>bmr</code> parameter is set to a numeric value
specifying the BMR, defined in relation to the calculation requested in
<code>bmr_type</code>.</p>
<p>Selecting an appropriate BMR involves making judgements about the
statistical and biological characteristics of the dataset and about the
applications for which the resuling BMDs will be used.</p>
<div class="section level5">
<h5 id="install-toxicr">Install ToxicR<a class="anchor" aria-label="anchor" href="#install-toxicr"></a>
</h5>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"NIEHS/ToxicR"</span><span class="op">)</span></span></code></pre></div>
<p><em>See the ToxicR repository on github for more information on
installing the</em> <em>package.</em></p>
</div>
<div class="section level5">
<h5 id="general-usage-bmd_toxicr">General Usage: bmd_toxicr<a class="anchor" aria-label="anchor" href="#general-usage-bmd_toxicr"></a>
</h5>
<p><em>Examples include external dependencies</em></p>
<p>Supply <code>bmd_toxicr</code> with the per-sample mf data from
<code><a href="../reference/calculate_mf.html">calculate_mf()</a></code>. Retain the dose column in the summary
table. Specify the column that contains the numeric dose values in
<code>dose_col</code>. The function can model more than one response
variable at once. Supply all response variables to
<code>response_col</code>.</p>
<p>Calculate the BMD with model averaging by setting
<code>model_averaging = TRUE</code>. The confidence level for the upper
and lower confidence intervals can be defined with the
<code>alpha</code> parameter (default 90% CI).</p>
<p>Users may choose to generate model plots with
<code>plot_results</code>. If TRUE, plots may be automatically saved to
an output directory specified in <code>output_path</code>.
Alternatively, if <code>output_path</code> is NULL, plots will be
returned within a list alongside the the BMD results.</p>
<p>The function will return the BMD with its upper and lower confidence
intervals for each response variable. If model averaging, a breakdown of
the model averaging process can be returned alongside the results
<code>ma_summary = TRUE</code>. This will return the estimate for each
model with its associated posterior probability.</p>
</div>
<div class="section level5">
<h5 id="example-2">Example<a class="anchor" aria-label="anchor" href="#example-2"></a>
</h5>
<p><em>Example 5.2. Calculate the BMD with model averaging for a 50%
relative</em> <em>increase in MF from control. This will be calculated
for both MFmin and</em> <em>MFmax. See example 3.1 for calculation of
mf_data_global.</em></p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">toxicr_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bmd_toxicr.html">bmd_toxicr</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_global</span>,</span>
<span>  dose_col <span class="op">=</span> <span class="st">"dose"</span>,</span>
<span>  response_col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mf_min"</span>, <span class="st">"mf_max"</span><span class="op">)</span>,</span>
<span>  bmr_type <span class="op">=</span> <span class="st">"rel"</span>,</span>
<span>  bmr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  model_averaging <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ma_summary <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_results <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="va">toxicr_results</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="plot-bmd-cis">Plot BMD CIs<a class="anchor" aria-label="anchor" href="#plot-bmd-cis"></a>
</h4>
<p><code><a href="../reference/plot_ci.html">plot_ci()</a></code> creates a confidence interval (CI) plot of BMD
results for easy comparison of BMDs between response variables.</p>
<p>The BMD values can be plotted on the log scale, if required.</p>
<p><em>Example 5.3. Compare the estimated BMD for MFmin (50% relative
increase)</em> <em>by PROAST versus ToxicR</em></p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  Response <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PROAST"</span>, <span class="st">"ToxicR"</span><span class="op">)</span>,</span>
<span>  BMD <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">9.111</span>, <span class="fl">9.641894</span><span class="op">)</span>,</span>
<span>  BMDL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">7.38</span>, <span class="fl">8.032936</span><span class="op">)</span>,</span>
<span>  BMDU <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10.9</span>, <span class="fl">10.97636</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_ci.html">plot_ci</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">plot_results</span>,</span>
<span>  order <span class="op">=</span> <span class="st">"asc"</span>,</span>
<span>  x_lab <span class="op">=</span> <span class="st">"Dose (mg/kg-bw/d)"</span>,</span>
<span>  y_lab <span class="op">=</span> <span class="st">"BMD Method"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-ci-1.png" alt="Benchmark dose with 90% confidence intervals representing the dose at, which a 50% increase in mutation frequency occurs from reference level. Calculated using ROAST software. Black points represent the BMD, red points the BMDL, and blue points, the BMDU" width="700"><p class="caption">
Benchmark dose with 90% confidence intervals representing the dose at,
which a 50% increase in mutation frequency occurs from reference level.
Calculated using ROAST software. Black points represent the BMD, red
points the BMDL, and blue points, the BMDU
</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="mutation-spectra-analysis">Mutation Spectra Analysis<a class="anchor" aria-label="anchor" href="#mutation-spectra-analysis"></a>
</h2>
<p>The mutation spectra is the pattern of mutation subtypes within a
sample or group. The mutation spectra can inform on the mechanisms
involved in mutagenesis.</p>
<div class="section level3">
<h3 id="group-comparison">Group Comparison<a class="anchor" aria-label="anchor" href="#group-comparison"></a>
</h3>
<p>Compare the mutation spectra between experimental groups using the
<code><a href="../reference/spectra_comparison.html">spectra_comparison()</a></code> function. This function will compare
the proportion of mutation subtypes at any resolution between
user-defined groups using a modified contingency table approach <span class="citation">(<a href="#ref-piegorsch-1994">Piegorsch and Bailer
1994</a>)</span>.</p>
<p>This approach is applied to the mutation counts of each subtype in a
given group. The contingency table is represented as
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>R</mi><mo>*</mo><mi>T</mi></mrow><annotation encoding="application/x-tex">R * T</annotation></semantics></math>
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>R</mi><annotation encoding="application/x-tex">R</annotation></semantics></math>
is the number of subtypes, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>T</mi><annotation encoding="application/x-tex">T</annotation></semantics></math>
is the number of groups. <code><a href="../reference/spectra_comparison.html">spectra_comparison()</a></code> performs
comparisons between
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>T</mi><mo>=</mo><mn>2</mn></mrow><annotation encoding="application/x-tex">T = 2</annotation></semantics></math>
specified groups. The statistical hypothesis of homogeneity is that the
proportion (count/group total) of each mutation subtype equals that of
the other group. To test the significance of the homogeneity hypothesis,
the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>G</mi><mn>2</mn></msup><annotation encoding="application/x-tex">G^{2}</annotation></semantics></math>
likelihood ratio statistic is used:</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>G</mi><mn>2</mn></msup><mo>=</mo><mn>2</mn><mspace width="0.222em"></mspace><munderover><mo>∑</mo><mrow><mi>i</mi><mo>=</mo><mn>1</mn></mrow><mi>R</mi></munderover><mspace width="0.222em"></mspace><munderover><mo>∑</mo><mrow><mi>j</mi><mo>=</mo><mn>1</mn></mrow><mi>T</mi></munderover><mspace width="0.222em"></mspace><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><mspace width="0.222em"></mspace><mi>l</mi><mi>o</mi><mi>g</mi><mrow><mo stretchy="true" form="prefix">(</mo><mfrac><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><msub><mi>E</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub></mfrac><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">G^{2} = 2\  \sum_{i=1}^{R}\  \sum_{j=1}^{T}\  Y_{ij}\  log(\frac{Y_{ij}}{E_{ij}})</annotation></semantics></math></p>
<p><math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>Y</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">Y_{ij}</annotation></semantics></math>
represents the mutation counts and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>E</mi><mrow><mi>i</mi><mi>j</mi></mrow></msub><annotation encoding="application/x-tex">E_{ij}</annotation></semantics></math>
are the <em>expected</em> counts under the null hypothesis. The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>G</mi><mn>2</mn></msup><annotation encoding="application/x-tex">G^{2}</annotation></semantics></math>
statistic possesses approximately a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>χ</mi><mn>2</mn></msup><annotation encoding="application/x-tex">\chi^{2}</annotation></semantics></math>
distribution in large sample sizes under the null hypothesis of no
spectral differences. Thus, as the column totals become large,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>G</mi><mn>2</mn></msup><annotation encoding="application/x-tex">G^{2}</annotation></semantics></math>
may be referred to a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>χ</mi><mn>2</mn></msup><annotation encoding="application/x-tex">\chi^{2}</annotation></semantics></math>
distribution with
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>R</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mrow><mo stretchy="true" form="prefix">(</mo><mi>T</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">(R -  1)(T - 1)</annotation></semantics></math>
degrees of freedom. The
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>G</mi><mn>2</mn></msup><annotation encoding="application/x-tex">G^{2}</annotation></semantics></math>
statistic may exhibit high false positive rates in small sample sizes
when referred to a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>χ</mi><mn>2</mn></msup><annotation encoding="application/x-tex">\chi^{2}</annotation></semantics></math>
distribution. In such cases, we instead switch to an F-distribution.
This has the effect of reducing the rate at which
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>G</mi><mn>2</mn></msup><annotation encoding="application/x-tex">G^{2}</annotation></semantics></math>
rejects each null hypothesis, providing greater stability in terms of
false positive error rates. Thus when
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>N</mi><mi>/</mi><mrow><mo stretchy="true" form="prefix">(</mo><mi>R</mi><mo>−</mo><mn>1</mn><mo stretchy="true" form="postfix">)</mo></mrow><mo>&lt;</mo><mn>20</mn></mrow><annotation encoding="application/x-tex">N/(R-1) &lt; 20</annotation></semantics></math>,
where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>N</mi><annotation encoding="application/x-tex">N</annotation></semantics></math>
is the total mutation counts across both groups, the function will use a
F-distribution, otherwise it will use a
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>χ</mi><mn>2</mn></msup><annotation encoding="application/x-tex">\chi^{2}</annotation></semantics></math>-distribution.</p>
<p>This comparison assumes independance among the observations. Each
tabled observation represents a sum of independent contributions to the
total mutant count. We assume independance is valid for mutants derived
from a mixed population, however, mutants that are derived clonally from
a single progenitor cell would violate this assumption. As such, it is
recommended to use the <strong>MFmin method</strong> of mutation
counting for spectral analyses to ensure that all mutation counts are
independant. In those cases where the independence may be invalid, and
where additional, extra-multinomial sources of variability are present,
more complex, hierarchical statistical models are required. This is
currently outside the scope of this package.</p>
<div class="section level5">
<h5 id="general-usage-spectra_comparison">General Usage: spectra_comparison()<a class="anchor" aria-label="anchor" href="#general-usage-spectra_comparison"></a>
</h5>
<p>The first step is to calculate the per-group mf data at the desired
subtype_resolution using <code><a href="../reference/calculate_mf.html">calculate_mf()</a></code>. This mf_data is
then supplied to <code><a href="../reference/spectra_comparison.html">spectra_comparison()</a></code> along with a
contrasts table to specify the comparisons. The contrasts table will
consist of two columns, each specifying a group to be contrasted against
the other (see Generalized Linear Modelling for more details on contrast
tables).</p>
<p>Specify the variable by which you wish to make your comparisons with
The <code>exp_variable</code> parameter.</p>
<p>The function will output the
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msup><mi>G</mi><mn>2</mn></msup><annotation encoding="application/x-tex">G^{2}</annotation></semantics></math>
statistic and p-value for each specified comparison listed in
<code>constrasts</code>. P-values are adjusted for multiple comparison
using the Sidak method.</p>
<p><em>Example 6.1. In our example data, we are studying the mutagenic
effect of</em> <em>BaP. Our samples were exposed to three doses of a BaP
(Low, Medium, High),</em> <em>or to the vehicle control (Control). We
will compare the base_6 snv subtypes,</em> <em>alongside non-snv
variants, of each of the three dose groups to the control.</em> <em>In
this way we can investigate if exposure to BaP leads to significant
spectral</em> <em>differences.</em></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the MF per dose at the base_6 resolution.</span></span>
<span><span class="va">mf_data_6_dose</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create the contrast table</span></span>
<span><span class="va">contrasts_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  col1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span>,</span>
<span>  col2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Control"</span>, <span class="st">"Control"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Run the analysis</span></span>
<span><span class="va">ex_spectra_comp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spectra_comparison.html">spectra_comparison</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_6_dose</span>,</span>
<span>  exp_variable <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  contrasts <span class="op">=</span> <span class="va">contrasts_table</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><strong>Table 23.</strong> Comparison of the 6-base subtype
Proportions between Dose Groups.</p>
<div class="datatables html-widget html-fill-item" id="htmlwidget-b3f7c917b6c8ff580948" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-b3f7c917b6c8ff580948">{"x":{"filter":"none","vertical":false,"data":[["1","2","3"],["Low vs Control","Medium vs Control","High vs Control"],[195.6280978952381,503.3807421047178,714.0200084754551],[0,0,0],[0,0,0],["***","***","***"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>contrasts<\/th>\n      <th>G2<\/th>\n      <th>p.value<\/th>\n      <th>adj_p.value<\/th>\n      <th>Significance<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"targets":2,"render":"function(data, type, row, meta) {\n    return type !== 'display' ? data : DTWidget.formatRound(data, 2, 3, \",\", \".\", null);\n  }"},{"targets":[3,4],"render":"function(data, type, row, meta) { return data.toExponential(2); }"},{"className":"dt-right","targets":[2,3,4]},{"orderable":false,"targets":0},{"name":" ","targets":0},{"name":"contrasts","targets":1},{"name":"G2","targets":2},{"name":"p.value","targets":3},{"name":"adj_p.value","targets":4},{"name":"Significance","targets":5}],"rowCallback":"function(row, data, dataIndex) {\n        $('td:eq(2)', row).css('text-align', 'right'); // Align the content of column 2 to the right\n      }","order":[],"autoWidth":false,"orderClasses":false}},"evals":["options.columnDefs.0.render","options.columnDefs.1.render","options.rowCallback"],"jsHooks":[]}</script>
</div>
</div>
<div class="section level3">
<h3 id="mutational-signatures">Mutational Signatures<a class="anchor" aria-label="anchor" href="#mutational-signatures"></a>
</h3>
<p>Mutational processes generate characteristic patterns of mutations,
known as mutational signatures. Distinct mutational signatures have been
extracted from various cancer types and normal tissues using data from
the Catalogue of Somatic Mutations in Cancer, (<a href="https://cancer.sanger.ac.uk/signatures/" class="external-link">COSMIC</a>) database.
These include signatures of single base substitutions (SBSs), doublet
base substitutions (DBSs), small insertions and deletions (IDs) and copy
number alterations (CNs). It is possible to assign mutational signatures
to individual samples or groups using the <code>signature_fitting</code>
function. Linking ECS mutational profiles of specific mutagens to
existing mutational signatures provides empirical evidence for the
contribution of environmental mutagens to the mutations found in human
cancers and informs on mutagenic mechanisms.</p>
<p>The <code>signature_fitting</code> function utilizes the <a href="https://github.com/AlexandrovLab" class="external-link">SigProfiler</a> suite of tools
(<span class="citation">Díaz-Gay et al. (<a href="#ref-diaz-gay-2023">2023</a>)</span>; <span class="citation">Khandekar et al. (<a href="#ref-khandekar-2023">2023</a>)</span>) to assign SBS signatures
from the COSMIC database to the 96-base SNV subtypes of a given group by
creating a virtual environment to run python using
<code>reticulate</code>. <code><a href="../reference/signature_fitting.html">signature_fitting()</a></code> facilitates
interoperability between these tools for users less familiar with python
and assists users by coercing the mutation data to the necessary
structure for the SigProfiler tools.</p>
<p>This function will install several python dependencies using a conda
virtual environment on first use, as well as the FASTA files for all
chromosomes for your specified reference genome. As a result ~3Gb of
storage must be available for the downloads of each genome.</p>
<p>SNVs in their 96-base trinucleotide context are summed across groups
to create a mutation count matrix by
<code>SigProfilerMatrixGeneratorR()</code> (SigProfilerMatrixGenerator;
<span class="citation">Khandekar et al. (<a href="#ref-khandekar-2023">2023</a>)</span>).
<code>Analyze.cosmic_fit</code> (SigProfilerAssignment; <span class="citation">Díaz-Gay et al. (<a href="#ref-diaz-gay-2023">2023</a>)</span>) is then run to assign
mutational signatures to each group using refitting methods, which
quantifies the contribution of a set of signatures to the mutational
profile of the group. The process is a numerical optimization approach
that finds the combination of mutational signatures that most closely
reconstructs the mutation count matrix. To quantify the number of
mutations imprinted by each signature, the tool uses a custom
implementation of the forward stagewise algorithm and it applies
nonnegative least squares, based on the Lawson-Hanson method. Cosine
similarity values, and other solution statistics, are generated to
compare the reconstructed mutational profile to the original mutational
profile of the group, with cosine values &gt; 0.9 indicating a good
reconstruction.</p>
<p>Currently, <code>signature_fitting</code> offers fitting of COSMIC
version 3.4 SBS signatures to the SBS96 matrix of any sample/group. For
advanced use, including using a custom set of reference signatures, or
fitting the DBS, ID, or CN signatures, it is suggested to use the
SigProfiler python tools directly in as described in their respective <a href="https://github.com/AlexandrovLab" class="external-link">documentation</a>.</p>
<div class="section level4">
<h4 id="general-usage-signature_fitting">General Usage: signature_fitting()<a class="anchor" aria-label="anchor" href="#general-usage-signature_fitting"></a>
</h4>
<p><em>Examples include external dependencies</em></p>
<p><code>signature_fitting</code> requires the installation of
reticulate and <a href="https://github.com/AlexandrovLab/SigProfilerMatrixGeneratorR" class="external-link">SigProfilerMatrixGeneratorR</a>
as well as a version of python 3.8 or newer.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Install reticulate</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"reticulate"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install python</span></span>
<span><span class="fu">reticulate</span><span class="fu">::</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/install_python.html" class="external-link">install_python</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Install SigProfilerMatrixGeneratorR from github using devtools.</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"AlexandrovLab/SigProfilerMatrixGeneratorR"</span><span class="op">)</span></span></code></pre></div>
<p><code>signature_fitting</code> will take the imported (and filtered,
if applicable) <code>mutation_data</code> and create mutational matrices
for the SNV mutations. Mutations are summed across levels of the
<code>group</code> parameter. This can be set to individual samples or
to an experimental group. Variants flagged in the
<code>filter_mut</code> column will not be included in the mutational
matrices.</p>
<p>The <code>project_genome</code> will be referenced for the creation
of the mutational matrices. The reference genome will be installed if it
is not already.</p>
<p>The virtual environment can be specified with the
<code>env_name</code> parameter. If no such environmnent exists, then
the function will create one in which to store the dependencies and run
the signature refitting. Specify your version of python using the
<code>python_version</code> parameter (must be 3.8 or higher).</p>
<p><em>Example 6.2. Determine the COSMIC SBS signatures associated with
each BaP</em> <em>dose group.</em></p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run Analysis</span></span>
<span><span class="fu"><a href="../reference/signature_fitting.html">signature_fitting</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>, <span class="co"># filtered mutation data</span></span>
<span>  project_name <span class="op">=</span> <span class="st">"Example"</span>,</span>
<span>  project_genome <span class="op">=</span> <span class="st">"mm10"</span>,</span>
<span>  env_name <span class="op">=</span> <span class="st">"MutSeqR"</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  python_version <span class="op">=</span> <span class="st">"3.11"</span>,</span>
<span>  output_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="output-3">Output<a class="anchor" aria-label="anchor" href="#output-3"></a>
</h4>
<p>Results from the <code>signature_fitting</code> will be stored in an
output folder. A filepath to a specific output directory can be
designated using the <code>output_path</code> parameter. If null, the
output will be stored within your working directory. Results will be
organized into subfolders based on the <code>group</code> parameter. The
output structure is divided into three folders: input, output, and
logs.</p>
<p><strong>Input</strong> folder: “mutations.txt”, a text file with the
mutation_data coerced into the required format for
SigProfilerMatrixGenerator. It consists of a list of all the snv
variants in each <code>group</code> alongside their genomic positions.
This data serves as input for matrix generation.</p>
<p><strong>Log</strong> folder: the error and the log files for
SigProfilerMatrixGeneration.</p>
<p>The output folder contains the results from matrix generation and
signature refitting desribed in detail below.</p>
<div class="section level5">
<h5 id="matrix-generation-output">Matrix Generation Output<a class="anchor" aria-label="anchor" href="#matrix-generation-output"></a>
</h5>
<p><code>signature_fitting</code> uses <a href="https://osf.io/s93d5/wiki/home/" class="external-link">SigProfilerMatrixGenerator</a> to
create mutational matrices (<span class="citation">(<a href="#ref-bergstrom-2019">Bergstrom et al. 2019</a>)</span>). Mutation
matrices are created for single-base substitutions (SBS) and
doublet-base substitutions (DBS), including matrices with extended
sequence context and transcriptional strand bias. SBS and DBS matrices
are stored in their respective folders in the output directory.
<strong>Only the SBS96 matrix is used for refitting</strong>.
<em>Matrices are stored as <code>.all</code> files which can be viewed
in a text-editor.</em></p>
<table class="table">
<caption>(#tab:mat-files) Output Files from SigProfilerMatrixGeneration
by Folder.</caption>
<colgroup>
<col width="21%">
<col width="16%">
<col width="32%">
<col width="29%">
</colgroup>
<thead><tr class="header">
<th>Folder</th>
<th>File</th>
<th>Definition</th>
<th>Plot file</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>SBS</td>
<td><em>project.SBS6.all</em></td>
<td>The 6 pyrimidine single-nucleotide variants. <em>C&gt;A, C&gt;G,
C&gt;T, T&gt;A, T&gt;C, or T&gt;G</em>
</td>
<td><em>plots/SBS_6_plots_project.pdf</em></td>
</tr>
<tr class="even">
<td>SBS</td>
<td><em>project.SBS18.all</em></td>
<td>The 6 pyrimidine single-nucleotide variants within 3 transcriptional
bias categories: Untranscribed (U), Transcribed (T), Non-Transcribed
Region (N).</td>
<td></td>
</tr>
<tr class="odd">
<td>SBS</td>
<td><em>project.SBS24.all</em></td>
<td>The 6 pyrimidine single-nucleotide variants within 4 transcriptional
bias categories: Untranscribed (U), Transcribed (T), Bidirectional (B),
Non-Transcribed Region (N).</td>
<td><em>plots/SBS_24_plots_project.pdf</em></td>
</tr>
<tr class="even">
<td>SBS</td>
<td><strong>project.SBS96.all</strong></td>
<td>The 6 pyrimidine single-nucleotide variants alongside their flanking
nucleotides (4 x 4 = 16 combinations). <em>Ex. A[C&gt;G]T</em>
</td>
<td><em>plots/SBS_96_plots_project.pdf</em></td>
</tr>
<tr class="odd">
<td>SBS</td>
<td><em>project.SBS288.all</em></td>
<td>The 96-base single-nucleotide variants within 3 transcriptional bias
categories (U, T, N).</td>
<td><em>plots/SBS_288_plots_project.pdf</em></td>
</tr>
<tr class="even">
<td>SBS</td>
<td><em>project.SBS384.all</em></td>
<td>The 96-base single-nucleotide variants within 4 transcriptional bias
categories (U, T, N, B).</td>
<td><em>plots/SBS_384_plots_project.pdf</em></td>
</tr>
<tr class="odd">
<td>SBS</td>
<td><em>project.SBS1536.all</em></td>
<td>The 6 pyrimidine single-nucleotide variants alongside their flanking
dinucleotides (16 x 16 = 256 combinations). <em>Ex.
AA[C&gt;G]TT</em>
</td>
<td><em>plots/SBS_1536_plots_project.pdf</em></td>
</tr>
<tr class="even">
<td>SBS</td>
<td><em>project.SBS4608.all</em></td>
<td>The 1536-base single-nucleotide variants within 3 transcriptional
bias categories (U, T, N).</td>
<td></td>
</tr>
<tr class="odd">
<td>SBS</td>
<td><em>project.SBS6144.all</em></td>
<td>The 1536-base single-nucleotide variants within 4 transcriptional
bias categories (U, T, N, B).</td>
<td></td>
</tr>
<tr class="even">
<td>DBS</td>
<td><em>project.DBS78.all</em></td>
<td>The 78 pyrimidine double-nucleotide variants.</td>
<td><em>plots/DBS_78_plots_project.pdf</em></td>
</tr>
<tr class="odd">
<td>DBS</td>
<td><em>project.DBS150.all</em></td>
<td>The 36 dinucleotide combinations that have only all purines or all
pyrimidines x 3 transcriptionla bias categories (U, T, N).</td>
<td></td>
</tr>
<tr class="even">
<td>DBS</td>
<td><em>project.DBS186.all</em></td>
<td>The 36 dinucleotide combinations that have only all purines or all
pyrimidines x 4 transcriptional bias categories (U, T, N, B).</td>
<td><em>plots/DBS_186_plots_project.pdf</em></td>
</tr>
<tr class="odd">
<td>DBS</td>
<td><em>project.DBS1248.all</em></td>
<td>The 78 pyrimidine double-nucleotide variants alongside their
flanking nucleotides (Possible starting nucleotides (4) x 78 x possible
ending nucleotides (4) = 1248 combinations)</td>
<td></td>
</tr>
<tr class="even">
<td>DBS</td>
<td><em>project.DBS2400.all</em></td>
<td>The 36 dinucleotide combinations that have only all purines or all
pyrimidines alongside their flanking nucleotides within transcriptional
bias categories (U, T, N).</td>
<td></td>
</tr>
<tr class="odd">
<td>DBS</td>
<td><em>project.DBS2676.all</em></td>
<td>The 36 dinucleotide combinations that have only all purines or all
pyrimidines alongside their flanking nucleotides within 4
transcriptional bias categories (U, T, N, B).</td>
<td></td>
</tr>
<tr class="even">
<td>TSB</td>
<td><em>strandBiasTes_24.txt</em></td>
<td>Transcription Strand Bias Test stats of the SBS6 variants</td>
<td></td>
</tr>
<tr class="odd">
<td>TSB</td>
<td><em>strandBiasTes_384.txt</em></td>
<td>Transcription Strand Bias Test stats of the SBS96 variants</td>
<td></td>
</tr>
<tr class="even">
<td>TSB</td>
<td><em>strandBiasTes_6144.txt</em></td>
<td>Transcription Strand Bias Test stats of the SBS1536 variants</td>
<td></td>
</tr>
<tr class="odd">
<td>TSB</td>
<td><em>significantResults_strandBiasTest.txt</em></td>
<td>Returns significant results from the three files above.</td>
<td></td>
</tr>
</tbody>
</table>
<p><strong>Doublet-base Matrices (DBS)</strong>: DBS are somatic
mutations in which a set of two adjacent DNA base-pairs are
simultaneously substituted with another set of two adjacent DNA
base-pairs. <strong>We do not recommend using the DBS matrices</strong>
<strong>generated using <code>signature_fitting</code></strong> The
<code>signature_fitting</code> function is designed to handle only SBS
mutations. All true MNVs, including doublets, are filtered out of the
<code>mutation_data</code> prior to MatrixGeneration. However, the tool
will still attempt to identify DBSs and will occasionally find two
independent SBSs occuring next to each other simply by chance. If you
wish to use DBS mutations in your signature analysis, please refer
directly to the SigProfiler tools.</p>
<p><strong>Plots</strong>: Barplots of the mutation matrices for all
groups can be found in the “plots” folder. The number of mutations are
plotted for each group at the various subtype resolutions. Files for
each matrix are listed in Table @ref(tab:mat-files).</p>
<p><strong>vcf_files</strong>: This output folder provides text-based
files containing the original mutations and their
SigProfilerMatrixGenerator classification for each chromosome. The files
are separated into dinucleotides (DBS), multinucleotide substitutions
(MNS), and single nucleotide variants (SNV) folders containing the
appropriate files. The headers are:</p>
<ol style="list-style-type: decimal">
<li>The group</li>
<li>the chromosome</li>
<li>the position</li>
<li>the SigProfilerMatrixGenerator classification</li>
<li>the strand {1, 0, -1}.</li>
</ol>
<p>The headers for each file are the same with the exception of the MNS
files which don’t contain a matrix classification or a strand
classification. As noted above the DBS and MNS matrices do no reflect
the true mutation counts for these variant types. Only SBS/SNV mutations
are included in the matrix generation.</p>
<p><strong>Transcription Strand Bias (TSB)</strong>: SBS and DBS
mutations are tested for <a href="https://osf.io/s93d5/wiki/5.%20Output%20-%20TSB/" class="external-link">transcription
strand bias</a>. Mutations are first classified within the four
transcriptional bias categories:</p>
<table class="table">
<caption>(#tab:transcript-cat) SigProfiler Transcriptional Bias
Categories.</caption>
<colgroup>
<col width="22%">
<col width="77%">
</colgroup>
<thead><tr class="header">
<th>Category</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Transcribed (T)</td>
<td>The variant is on the transcribed (template) strand.</td>
</tr>
<tr class="even">
<td>Untranscribed (U)</td>
<td>The variant is on the untranscribed (coding) strand.</td>
</tr>
<tr class="odd">
<td>Bidirectional (B)</td>
<td>The variant is on both strands and is transcribed either way.</td>
</tr>
<tr class="even">
<td>Nontranscribed (N)</td>
<td>The variant is in a non-coding region and is untranslated.</td>
</tr>
</tbody>
</table>
<p>The tool will then perform a transcription strand bias test which
compares the number of transcribed and untranscribed mutations for each
mutation type. <em>For example, it will compare the number of
transcribed T&gt;C to untranscribed</em> <em>T&gt;C mutations. Should
there be a significant difference, it would indicate that</em>
<em>T:A&gt;C:G mutations are occuring at a higher rate on one of the
strands compared</em> <em>to the other.</em> The output files contain
the following headers:</p>
<ol style="list-style-type: decimal">
<li>the <code>group</code>
</li>
<li>the mutation type</li>
<li>the enrichment value (# Transcribed / # untranscribed)</li>
<li>the p-value, corrected for multiple comparisons using the false
discrover rate method</li>
<li>the false discovery rate q-value</li>
</ol>
</div>
<div class="section level5">
<h5 id="signature-refitting-results">Signature Refitting Results<a class="anchor" aria-label="anchor" href="#signature-refitting-results"></a>
</h5>
<p>Results from the signature refitting perfomed by <a href="https://osf.io/mz79v/wiki/home/" class="external-link">SigProfilerAssignment</a> will be
stored within the “Assignment_Solution” folder. “Assignment_Solution”
consists of 3 subdirectories; “Activities”, “Signatures”, and
“Solution_Stats”.</p>
<p><strong>Activities</strong></p>
<table class="table">
<colgroup>
<col width="31%">
<col width="68%">
</colgroup>
<thead><tr class="header">
<th>File</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><em>Assignment_Solution_Activities.txt</em></td>
<td>This file contains the activity matrix for the selected signatures.
The first column lists all of the samples/groups. All of the following
columns list the calculated activity value for the respective
signatures. Signature activities correspond to the specific numbers of
mutations from the sample’s original mutation matrix caused by a
particular mutational process.</td>
</tr>
<tr class="even">
<td><em>Assignment_Solution_Activity_Plots.pdf</em></td>
<td>This file contains a stacked barplot showing the number of mutations
in each signature on the y-axis and the samples/groups on the
x-axis.</td>
</tr>
<tr class="odd">
<td><em>Assignment_Solution_TMB_plot.pdf</em></td>
<td>This file contains a tumor mutational burden plot. The y-axis is the
somatic mutations per megabase and the x-axis is the number of
samples/groups plotted over the total number of samples/groups included.
The column names are the mutational signatures and the plot is ordered
by the median somatic mutations per megabase.</td>
</tr>
<tr class="even">
<td><em>Decomposed_Mutation_Probabilities.txt</em></td>
<td>This file contains the probabilities of each of the 96 mutation
types in each sample/group. The probabilities refer to the probability
of each mutation type being caused by a specific signature. The first
column lists all the samples/groups, the second column lists all the
mutation types, and the following columns list the calculated
probability value for the respective signatures.</td>
</tr>
<tr class="odd">
<td><em>SampleReconstruction</em></td>
<td>This folder contains generated plots for each sample/group
summarizing the assignment results. Each plot consists of three panels.
(i) Original: a bar plot of the inputted 96SBS mutation matrix for the
sample/group. (ii) Reconstructed: a bar plot of the reconstruction of
the original mutation matrix. (iii) The mutational profiles for each of
the known mutational signatures assigned to that sample/group, including
the activities for each signature. Accuracy metrics for the
reconstruction are displayed at the bottom of the figure.</td>
</tr>
</tbody>
</table>
<p><strong>Signatures</strong></p>
<table class="table">
<colgroup>
<col width="35%">
<col width="65%">
</colgroup>
<thead><tr class="header">
<th>Files</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><em>Assignment_Solution_Signatures.txt</em></td>
<td>The distribution of mutation types in the input mutational
signatures. The first column lists all 96 of the mutation types. The
following columns are the signatures.</td>
</tr>
<tr class="even">
<td><em>SBS_96_plots_Assignment_Solution.pdf</em></td>
<td>Barplots for each signature identified that depicts the proportion
of the mutation types for that signature. The top right corner also
lists the total number of mutations and the percentage of total
mutations assigned to the mutational signature.</td>
</tr>
</tbody>
</table>
<p><strong>Solution_Stats</strong></p>
<table class="table">
<colgroup>
<col width="35%">
<col width="65%">
</colgroup>
<thead><tr class="header">
<th>Files</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><em>Assignment_Solution_Samples_Stats.txt</em></td>
<td>The accuracy metrics for the reconstruction. statistics for each
sample including the total number of mutations, cosine similarity, L1
norm (calculated as the sum of the absolute values of the vector), L1
norm percentage, L2 norm (calculated as the square root of the sum of
the squared vector values), and L2 norm percentage, along with the
Kullback-Leibler divergence.</td>
</tr>
<tr class="even">
<td><em>Assignment_Solution_Signature_Assignment_log.txt</em></td>
<td>The events that occur when known signatures are assigned to an input
sample. The information includes the L2 error and cosine similarity
between the reconstructed and original sample within different
composition steps.</td>
</tr>
</tbody>
</table>
<p><strong>Other Files</strong></p>
<p><em>JOB_METADATA_SPA.txt</em>: This file contains the metadata about
system and runtime.</p>
</div>
</div>
<div class="section level4">
<h4 id="use-sigprofiler-webtool">Use SigProfiler Webtool<a class="anchor" aria-label="anchor" href="#use-sigprofiler-webtool"></a>
</h4>
<p>Users may choose to use the <a href="https://cancer.sanger.ac.uk/signatures/assignment/" class="external-link">SigProfiler
Webtool</a> instead of using the
<code><a href="../reference/signature_fitting.html">signature_fitting()</a></code>function. MutSeqR offers functions to
coerce mutation data into the proper format for input files.</p>
<div class="section level5">
<h5 id="mutation-calling-file">Mutation Calling File<a class="anchor" aria-label="anchor" href="#mutation-calling-file"></a>
</h5>
<p><code><a href="../reference/write_mutation_calling_file.html">write_mutation_calling_file()</a></code> creates a simple text file
from mutation data that can be used for mutation signatures analysis
using the SigProfiler Assignment web application as a “mutation calling
file”. Signature analyses are done at the sample level when using
mutation calling files. The file will be saved to your output directory,
specified in <code>output_path</code>.</p>
<p><em>Example 6.3. Analyze the COSMIC SBS signatures contributing to
each of the</em> <em>24 samples using the SigProfiler Web Tool. Output a
mutation calling file that</em> <em>can be uploaded to the
webtool.</em></p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_mutation_calling_file.html">write_mutation_calling_file</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  project_name <span class="op">=</span> <span class="st">"Example"</span>,</span>
<span>  project_genome <span class="op">=</span> <span class="st">"mm10"</span>,</span>
<span>  output_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="mutational-matrix">Mutational Matrix<a class="anchor" aria-label="anchor" href="#mutational-matrix"></a>
</h5>
<p><code><a href="../reference/write_mutational_matrix.html">write_mutational_matrix()</a></code> will sum mutations across
user-defined groups before coercing the data into the proper format for
input as a “mutational matrix”. SNV subtypes can be resolved to either
the base_6 or base_96 resolution. The file is saved to the specified
output directory.</p>
<p><em>Example 6.4. Analyze the COSMIC SBS signatures contributing to
each dose</em> <em>group using the SigProfiler Web Tool. Output a
mutational matrix that can be</em> <em>uploaded to the webtool.</em></p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_mutational_matrix.html">write_mutational_matrix</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  group <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_96"</span>,</span>
<span>  mf_type <span class="op">=</span> <span class="st">"min"</span>,</span>
<span>  output_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="plot-spectra">Plot Spectra<a class="anchor" aria-label="anchor" href="#plot-spectra"></a>
</h3>
<div class="section level4">
<h4 class="tabset" id="plot_spectra">plot_spectra<a class="anchor" aria-label="anchor" href="#plot_spectra"></a>
</h4>
<p>The mutation spectra can be visualized with <code>plot_spectra</code>
which will create a stacked bar plot for user-defined groups at the
desired subtype resolution. Mutation subtypes are represented by colour.
The value can represent subtype count (<code>sum</code>), frequency
(<code>mf</code>), or <code>proportion</code>.</p>
<div class="section level5">
<h5 id="proportion">Proportion<a class="anchor" aria-label="anchor" href="#proportion"></a>
</h5>
<p><em>Example 6.5. Plot the base_6 proportions for each dose
group.</em></p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the mf data at the 6-base resolution for each dose</span></span>
<span><span class="co"># We will exclude ambiguous or uncategorized variants since we don't</span></span>
<span><span class="co"># have any in this data.</span></span>
<span><span class="va">mf_data_6_dose</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-ambiguous"</span>, <span class="st">"-uncategorized"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Set the desired order for the dose group:</span></span>
<span><span class="va">mf_data_6_dose</span><span class="op">$</span><span class="va">dose_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">mf_data_6_dose</span><span class="op">$</span><span class="va">dose_group</span>,</span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Plot</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_spectra.html">plot_spectra</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_6_dose</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"proportion"</span>,</span>
<span>  group_order <span class="op">=</span> <span class="st">"arranged"</span>,</span>
<span>  group_order_input <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  x_lab <span class="op">=</span> <span class="st">"Dose Group"</span>,</span>
<span>  y_lab <span class="op">=</span> <span class="st">"Subtype Proportion"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-spectra-dose-p-1.png" alt="Mutation spectrum (minimum) per Dose. Subtypes include single-nucleotide, variants at 6-base resolution, complex variants, deletions, insertions, multi-nucleotide variants (mnv) and structural variants (sv). Subtypes are represented by colour. Data is the proportion normalized to sequencing depth." width="700"><p class="caption">
Mutation spectrum (minimum) per Dose. Subtypes include
single-nucleotide, variants at 6-base resolution, complex variants,
deletions, insertions, multi-nucleotide variants (mnv) and structural
variants (sv). Subtypes are represented by colour. Data is the
proportion normalized to sequencing depth.
</p>
</div>
</div>
<div class="section level5">
<h5 id="frequency">Frequency<a class="anchor" aria-label="anchor" href="#frequency"></a>
</h5>
<p><em>Example 6.6. Plot the base_6 frequences for each dose
group.</em></p>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_spectra.html">plot_spectra</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_6_dose</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"mf"</span>,</span>
<span>  group_order <span class="op">=</span> <span class="st">"arranged"</span>,</span>
<span>  group_order_input <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  x_lab <span class="op">=</span> <span class="st">"Dose Group"</span>,</span>
<span>  y_lab <span class="op">=</span> <span class="st">"Subtype Frequency (mutations/bp)"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-spectra-dose-f-1.png" alt="Mutation spectrum (minimum) per Dose. Subtypes include single-nucleotide, variants at 6-base resolution, complex variants, deletions, insertions, multi-nucleotide variants (mnv) and structural variants (sv). Subtypes are represented by colour. Data is subtype frequency (mutations/bp)." width="700"><p class="caption">
Mutation spectrum (minimum) per Dose. Subtypes include
single-nucleotide, variants at 6-base resolution, complex variants,
deletions, insertions, multi-nucleotide variants (mnv) and structural
variants (sv). Subtypes are represented by colour. Data is subtype
frequency (mutations/bp).
</p>
</div>
</div>
<div class="section level5">
<h5 id="sum">Sum<a class="anchor" aria-label="anchor" href="#sum"></a>
</h5>
<p><em>Example 6.7. Plot the base_6 mutation sums for each dose
group.</em></p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_spectra.html">plot_spectra</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_6_dose</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"sum"</span>,</span>
<span>  group_order <span class="op">=</span> <span class="st">"arranged"</span>,</span>
<span>  group_order_input <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  x_lab <span class="op">=</span> <span class="st">"Dose Group"</span>,</span>
<span>  y_lab <span class="op">=</span> <span class="st">"Subtype Mutation Count"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="MutSeqR_introduction_files/figure-html/plot-spectra-dose-s-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 id="hierarchical-clustering">Hierarchical Clustering<a class="anchor" aria-label="anchor" href="#hierarchical-clustering"></a>
</h5>
<p><code><a href="../reference/plot_spectra.html">plot_spectra()</a></code> integrates <code><a href="../reference/cluster_spectra.html">cluster_spectra()</a></code>
which performs unsupervised hierarchical clustering of samples based on
the mutation spectra. <code><a href="../reference/cluster_spectra.html">cluster_spectra()</a></code> uses
<code><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist()</a></code> from the stats library to compute the
sample-to-sample distances using a user-defined distance measure
(default Euclidean). The resulting distance matrix is passed to
<code><a href="https://rdrr.io/r/stats/hclust.html" class="external-link">hclust()</a></code> to cluster samples using the specified linkage
method (default Ward). The function will output a dendrogram visually
representing the clusters’ relationships and hierarchy. The dendrogram
will be overlaid on the <code><a href="../reference/plot_spectra.html">plot_spectra()</a></code> bar plot and samples
will be ordered accordingly.</p>
<p><em>Example 6.8. Plot the base_6 mutation spectra per sample, with
hierarchical</em> <em>clustering. For this example we have created a new
sample column with more</em> intuitive sample names: new_sample_id.
These names correspond to their<em> </em>associated dose groups. We will
see that samples largly cluster within their<em> </em>dose groups.*</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the mf data at the 6-base resolution for each sample</span></span>
<span><span class="va">mf_data_6</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"new_sample_id"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-ambiguous"</span>, <span class="st">"-uncategorized"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Plot</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_spectra.html">plot_spectra</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_6</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"new_sample_id"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_6"</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"proportion"</span>,</span>
<span>  group_order <span class="op">=</span> <span class="st">"clustered"</span>,</span>
<span>  x_lab <span class="op">=</span> <span class="st">"Sample"</span>,</span>
<span>  y_lab <span class="op">=</span> <span class="st">"Subtype Proportion"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-spectra-clustered-1.png" alt="Mutation spectrum (minimum) per Sample. Subtypes include single-nucleotide variants at 6-base resolution, complex variants, deletions, insertions, multi-nucleotide variants (mnv) and structural variants (sv). Subtypes are represented by colour. Data is the proportion normalized to sequencing depth. Samples are clustered based on the Euclidean distance between their subtype proportions." width="700"><p class="caption">
Mutation spectrum (minimum) per Sample. Subtypes include
single-nucleotide variants at 6-base resolution, complex variants,
deletions, insertions, multi-nucleotide variants (mnv) and structural
variants (sv). Subtypes are represented by colour. Data is the
proportion normalized to sequencing depth. Samples are clustered based
on the Euclidean distance between their subtype proportions.
</p>
</div>
</div>
</div>
<div class="section level4">
<h4 class="tabset" id="trinucleotide-plots">Trinucleotide Plots<a class="anchor" aria-label="anchor" href="#trinucleotide-plots"></a>
</h4>
<p>The 96-base SNV mutation subtypes can be vizualised using
<code><a href="../reference/plot_trinucleotide.html">plot_trinucleotide()</a></code>. This function creates a bar plot of
the 96-base SNV spectrum for all levels of a user-defined group. Data
can represent subtype mutation count (<code>sum</code>), frequency
(<code>mf</code>), or <code>proportion</code>. Aesthetics are consistent
with COSMIC trinucleotide plots. Plots are automatically saved to the
specified output directory.</p>
<p><em>Example 6.9. plot the base_96 mutation spectra proportions for
each dose group.</em></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the mf data at the 96-base resolution for each dose</span></span>
<span><span class="va">mf_data_96_dose</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_96"</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="st">"snv"</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Plot</span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_trinucleotide.html">plot_trinucleotide</a></span><span class="op">(</span></span>
<span>  mf_96 <span class="op">=</span> <span class="va">mf_data_96_dose</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  response <span class="op">=</span> <span class="st">"proportion"</span>,</span>
<span>  mf_type <span class="op">=</span> <span class="st">"min"</span>,</span>
<span>  output_path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level5">
<h5 id="control-dose">Control Dose<a class="anchor" aria-label="anchor" href="#control-dose"></a>
</h5>
<pre><code><span><span class="co">## $Control</span></span></code></pre>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/trinuc-C-1.png" alt="96-base trinucleotide spectra (minimum) per Dose. Bars are the proportion of SNV subtypes within their trinucleotide context normalized to the sequencing depth.  Bars are coloured based on SNV subtype. Data labels indicate the total number of mutations for each SNV subtype." width="700"><p class="caption">
96-base trinucleotide spectra (minimum) per Dose. Bars are the
proportion of SNV subtypes within their trinucleotide context normalized
to the sequencing depth. Bars are coloured based on SNV subtype. Data
labels indicate the total number of mutations for each SNV subtype.
</p>
</div>
</div>
<div class="section level5">
<h5 id="low-dose">Low Dose<a class="anchor" aria-label="anchor" href="#low-dose"></a>
</h5>
<pre><code><span><span class="co">## $Low</span></span></code></pre>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/trinuc-L-1.png" alt="96-base trinucleotide spectra (minimum) per Dose. Bars are the proportion of SNV subtypes within their trinucleotide context normalized to the sequencing depth.  Bars are coloured based on SNV subtype. Data labels indicate the total number of mutations for each SNV subtype." width="700"><p class="caption">
96-base trinucleotide spectra (minimum) per Dose. Bars are the
proportion of SNV subtypes within their trinucleotide context normalized
to the sequencing depth. Bars are coloured based on SNV subtype. Data
labels indicate the total number of mutations for each SNV subtype.
</p>
</div>
</div>
<div class="section level5">
<h5 id="medium-dose">Medium Dose<a class="anchor" aria-label="anchor" href="#medium-dose"></a>
</h5>
<pre><code><span><span class="co">## $Medium</span></span></code></pre>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/trinuc-M-1.png" alt="96-base trinucleotide spectra (minimum) per Dose. Bars are the proportion of SNV subtypes within their trinucleotide context normalized to the sequencing depth.  Bars are coloured based on SNV subtype. Data labels indicate the total number of mutations for each SNV subtype." width="700"><p class="caption">
96-base trinucleotide spectra (minimum) per Dose. Bars are the
proportion of SNV subtypes within their trinucleotide context normalized
to the sequencing depth. Bars are coloured based on SNV subtype. Data
labels indicate the total number of mutations for each SNV subtype.
</p>
</div>
</div>
<div class="section level5">
<h5 id="high-dose">High Dose<a class="anchor" aria-label="anchor" href="#high-dose"></a>
</h5>
<pre><code><span><span class="co">## $High</span></span></code></pre>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/trinuc-H-1.png" alt="96-base trinucleotide spectra (minimum) per Dose. Bars are the proportion of SNV subtypes within their trinucleotide context normalized to the sequencing depth.  Bars are coloured based on SNV subtype. Data labels indicate the total number of mutations for each SNV subtype." width="700"><p class="caption">
96-base trinucleotide spectra (minimum) per Dose. Bars are the
proportion of SNV subtypes within their trinucleotide context normalized
to the sequencing depth. Bars are coloured based on SNV subtype. Data
labels indicate the total number of mutations for each SNV subtype.
</p>
</div>
</div>
</div>
<div class="section level4">
<h4 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h4>
<p>Another option for vizualizing the base-96 mutation spectra is
<code><a href="../reference/plot_trinucleotide_heatmap.html">plot_trinucleotide_heatmap()</a></code>. This function creates a
heatmap of the 96-base SNV proportions. Plots can be facetted by
additional grouping variables. Heatmaps are useful for making
comparisons between experimental variables when information density
becomes too high to represent using traditional plots.</p>
<p><em>Example 6.10. Plot the 96-base SNV spectrum for each sample,
facetted</em> <em>by dose group.</em></p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the mf data at the 96-base resolution for each sample</span></span>
<span><span class="va">mf_data_96</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculate_mf.html">calculate_mf</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  cols_to_group <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  subtype_resolution <span class="op">=</span> <span class="st">"base_96"</span>,</span>
<span>  variant_types <span class="op">=</span> <span class="st">"snv"</span>,</span>
<span>  retain_metadata_cols <span class="op">=</span> <span class="st">"dose_group"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">mf_data_96</span><span class="op">$</span><span class="va">dose_group</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>  <span class="va">mf_data_96</span><span class="op">$</span><span class="va">dose_group</span>,</span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span>, <span class="st">"Low"</span>, <span class="st">"Medium"</span>, <span class="st">"High"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># Plot</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_trinucleotide_heatmap.html">plot_trinucleotide_heatmap</a></span><span class="op">(</span></span>
<span>  mf_data <span class="op">=</span> <span class="va">mf_data_96</span>,</span>
<span>  group_col <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>  facet_col <span class="op">=</span> <span class="st">"dose_group"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/plot-heatmap-1.png" alt="96-base trinucleotide spectra (minimum) per Sample, facetted by Dose. Colour represents the proportion of SNV subtypes within their trinucleotide context normalized to the sequencing depth." width="700"><p class="caption">
96-base trinucleotide spectra (minimum) per Sample, facetted by Dose.
Colour represents the proportion of SNV subtypes within their
trinucleotide context normalized to the sequencing depth.
</p>
</div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="bubble-plots">Bubble Plots<a class="anchor" aria-label="anchor" href="#bubble-plots"></a>
</h2>
<p><code>plot_bubbles</code> is used to visually represent the
distribution and density of recurrent mutations. Each mutation is in a
given group is represented by a bubble whose size is scaled on either
the <code>alt_depth</code> or the <code>vaf</code>. Thus a highly
reccurent mutation is represented by a large bubble. These plots make it
easy to determine if MFmax is driven by a few highly recurrent mutations
versus serveral moderately recurrent mutations.</p>
<p>Plots can be facetted by user-defined groups, and bubbles can be
coloured by any variable of interest to help discern patterns in
mutation recurrence.</p>
<p><em>Example 7. Plot mutations per dose group, bubbles coloured by
base-6 subtype</em></p>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_bubbles.html">plot_bubbles</a></span><span class="op">(</span></span>
<span>  mutation_data <span class="op">=</span> <span class="va">example_data</span>,</span>
<span>  size_by <span class="op">=</span> <span class="st">"alt_depth"</span>,</span>
<span>  facet_col <span class="op">=</span> <span class="st">"dose_group"</span>,</span>
<span>  color_by <span class="op">=</span> <span class="st">"normalized_subtype"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<div class="figure">
<img src="MutSeqR_introduction_files/figure-html/bubble-plot-1.png" alt="Multiplet mutations plotted per Dose. Each circle represents a mutation, coloured by mutation subtype. The size of the circle is scaled by the mutation's alternative depth." width="700"><p class="caption">
Multiplet mutations plotted per Dose. Each circle represents a mutation,
coloured by mutation subtype. The size of the circle is scaled by the
mutation’s alternative depth.
</p>
</div>
</div>
<div class="section level2">
<h2 id="retrieve-sequences">Retrieve Sequences<a class="anchor" aria-label="anchor" href="#retrieve-sequences"></a>
</h2>
<p><code><a href="../reference/get_seq.html">get_seq()</a></code> will retrive raw nucleotide sequences for
specified genomic intervals. This function will install an appropriate
BS genome library to retrieve sequences based on species, genome, and
masked parameter.</p>
<p>Supply regions with a filepath, data frame or GRanges object
containing the specified genomic intervals. TwinStrand’s Mutagenesis
Panels are stored in package files and can easily be retrieved.</p>
<p>Sequences are returned within a <em>GRanges</em> object.</p>
<p><em>Example 8.1. Retrieve the sequences for our example’s target
panel,</em> <em>TwinStrand’s Mouse Mutagenesis Panel</em></p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_seq.html">get_seq</a></span><span class="op">(</span>regions <span class="op">=</span> <span class="st">"TSpanel_mouse"</span><span class="op">)</span></span>
<span><span class="va">regions_seq</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 20 ranges and 6 metadata columns:</span></span>
<span><span class="co">##        seqnames              ranges strand | target_size       label</span></span>
<span><span class="co">##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]     chr1   69304218-69306617      * |        2400        chr1</span></span>
<span><span class="co">##    [2]     chr1 155235939-155238338      * |        2400      chr1.2</span></span>
<span><span class="co">##    [3]     chr2   50833176-50835575      * |        2400        chr2</span></span>
<span><span class="co">##    [4]     chr3 109633161-109635560      * |        2400        chr3</span></span>
<span><span class="co">##    [5]     chr4   96825281-96827680      * |        2400        chr4</span></span>
<span><span class="co">##    ...      ...                 ...    ... .         ...         ...</span></span>
<span><span class="co">##   [16]    chr15   66779763-66782162      * |        2400       chr15</span></span>
<span><span class="co">##   [17]    chr16   72381581-72383980      * |        2400       chr16</span></span>
<span><span class="co">##   [18]    chr17   94009029-94011428      * |        2400       chr17</span></span>
<span><span class="co">##   [19]    chr18   81262079-81264478      * |        2400       chr18</span></span>
<span><span class="co">##   [20]    chr19     4618814-4621213      * |        2400       chr19</span></span>
<span><span class="co">##        genic_context region_GC_content      genome                sequence</span></span>
<span><span class="co">##          &lt;character&gt;         &lt;numeric&gt; &lt;character&gt;          &lt;DNAStringSet&gt;</span></span>
<span><span class="co">##    [1]    intergenic              37.3        mm10 CAATCTTTCT...CAAAATGCAA</span></span>
<span><span class="co">##    [2]         genic              54.0        mm10 AATCTCCAGG...CAAGCACTGG</span></span>
<span><span class="co">##    [3]    intergenic              45.3        mm10 TGTGCCCCAT...TGCTTGCCAC</span></span>
<span><span class="co">##    [4]         genic              39.2        mm10 AACGATGAAT...GCACTCAAGA</span></span>
<span><span class="co">##    [5]    intergenic              39.4        mm10 ATTGTTTGAA...CTCAGGGCCT</span></span>
<span><span class="co">##    ...           ...               ...         ...                     ...</span></span>
<span><span class="co">##   [16]         genic              44.0        mm10 GTGTCATTTT...CAGGTAGAGG</span></span>
<span><span class="co">##   [17]    intergenic              38.3        mm10 TCTGTAGCAA...ATAAAAACTC</span></span>
<span><span class="co">##   [18]    intergenic              35.2        mm10 TAAGGAAACT...TATCAAAGAT</span></span>
<span><span class="co">##   [19]    intergenic              47.3        mm10 AGCCATCTCC...GGGACTCAGA</span></span>
<span><span class="co">##   [20]         genic              56.1        mm10 TCCAGGCTGT...GCCCATGGAG</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 19 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p><em>Example 8.2. Retrieve sequences for a custom interval of regions.
We will use</em> <em>the Human Mutagenesis Panel as an example.</em></p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We will load the TSpanel_human regions file as an example</span></span>
<span><span class="va">human</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_regions_file.html">load_regions_file</a></span><span class="op">(</span><span class="st">"TSpanel_human"</span><span class="op">)</span></span>
<span><span class="va">regions_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_seq.html">get_seq</a></span><span class="op">(</span>regions <span class="op">=</span> <span class="va">human</span>,</span>
<span>                       is_0_based_rg <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       species <span class="op">=</span> <span class="st">"human"</span>,</span>
<span>                       genome <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>                       masked <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                       padding <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">regions_seq</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 20 ranges and 7 metadata columns:</span></span>
<span><span class="co">##        seqnames              ranges strand | target_size description</span></span>
<span><span class="co">##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |   &lt;integer&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]    chr11 108510788-108513187      * |        2400 region_1111</span></span>
<span><span class="co">##    [2]    chr13   75803913-75806312      * |        2400 region_1501</span></span>
<span><span class="co">##    [3]    chr14   74661756-74664155      * |        2400 region_1725</span></span>
<span><span class="co">##    [4]    chr18     5749265-5751664      * |        2400 region_2457</span></span>
<span><span class="co">##    [5]     chr2   40162768-40165167      * |        2400 region_2896</span></span>
<span><span class="co">##    ...      ...                 ...    ... .         ...         ...</span></span>
<span><span class="co">##   [16]    chr15   46089738-46092137      * |        2400 region_1904</span></span>
<span><span class="co">##   [17]    chr17   70672727-70675126      * |        2400 region_2378</span></span>
<span><span class="co">##   [18]    chr21   23665977-23668376      * |        2400 region_3515</span></span>
<span><span class="co">##   [19]    chr22   48262371-48264770      * |        2400 region_3703</span></span>
<span><span class="co">##   [20]    chr10 128969038-128971437      * |        2400  region_784</span></span>
<span><span class="co">##        genic_context        gene      genome       label</span></span>
<span><span class="co">##          &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##    [1]         genic       EXPH5        hg38       chr11</span></span>
<span><span class="co">##    [2]         genic        LMO7        hg38       chr13</span></span>
<span><span class="co">##    [3]         genic       AREL1        hg38       chr14</span></span>
<span><span class="co">##    [4]         genic   MIR3976HG        hg38       chr18</span></span>
<span><span class="co">##    [5]         genic  SLC8A1-AS1        hg38        chr2</span></span>
<span><span class="co">##    ...           ...         ...         ...         ...</span></span>
<span><span class="co">##   [16]    intergenic        &lt;NA&gt;        hg38       chr15</span></span>
<span><span class="co">##   [17]    intergenic        &lt;NA&gt;        hg38       chr17</span></span>
<span><span class="co">##   [18]    intergenic        &lt;NA&gt;        hg38       chr21</span></span>
<span><span class="co">##   [19]    intergenic        &lt;NA&gt;        hg38       chr22</span></span>
<span><span class="co">##   [20]    intergenic        &lt;NA&gt;        hg38       chr10</span></span>
<span><span class="co">##                       sequence</span></span>
<span><span class="co">##                 &lt;DNAStringSet&gt;</span></span>
<span><span class="co">##    [1] GTTTCCTTCA...CTTTCCTGGA</span></span>
<span><span class="co">##    [2] AGAATTATTT...TCAGACAACC</span></span>
<span><span class="co">##    [3] TTCCCTGGTT...AAGATACACT</span></span>
<span><span class="co">##    [4] TGGCAACTTG...AATGAAAACA</span></span>
<span><span class="co">##    [5] CTAGATTTTC...AGCATATCAC</span></span>
<span><span class="co">##    ...                     ...</span></span>
<span><span class="co">##   [16] TGAACAGACA...ATAAATTGCT</span></span>
<span><span class="co">##   [17] GTGGTGATCA...TAAAGATTCT</span></span>
<span><span class="co">##   [18] TGTAATAATG...TCCAGTCATT</span></span>
<span><span class="co">##   [19] GTAAAGGCAG...TCCACAGCAG</span></span>
<span><span class="co">##   [20] GTGGACTGAT...TTCTCACACT</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 20 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Sequences can be exported as FASTA files with
<code><a href="../reference/write_reference_fasta.html">write_reference_fasta()</a></code>. Supply this function with the
GRanges object containing the sequences of the regions. Each one will be
written to a single FASTA file. The FASTA file will be saved to the
<code>output_path</code>. If NULL, the file will be saved to the working
directory.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_reference_fasta.html">write_reference_fasta</a></span><span class="op">(</span><span class="va">regions_seq</span>, output_path <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exporting-results">Exporting Results<a class="anchor" aria-label="anchor" href="#exporting-results"></a>
</h2>
<p>Users can easily output data frames to an Excel workbook with
<code><a href="../reference/write_excel.html">write_excel()</a></code>. This function can write single data frames
or it can take a list of dataframes and write each one to a separate
Excel sheet in a workbook.</p>
<p>In addition to data frames, <code><a href="../reference/write_excel.html">write_excel()</a></code> will also
extract the mf_data, point_estimates, and pairwise_comparisons from
<code><a href="../reference/model_mf.html">model_mf()</a></code> output to write to an excel workbook. Set
<code>model_results</code> to TRUE if supplying the function with the
output to model_mf().</p>
<p><em>Example 9.1. Write MF data to excel workbook.</em></p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save a single data frame to an Excel file</span></span>
<span><span class="fu"><a href="../reference/write_excel.html">write_excel</a></span><span class="op">(</span><span class="va">mf_data_global</span>, workbook_name <span class="op">=</span> <span class="st">"example_mf_data"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Write multiple data frames to a list to export all at once.</span></span>
<span><span class="va">list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">mf_data_global</span>, <span class="va">mf_data_rg</span>, <span class="va">mf_data_6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">list</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mf_per_sample"</span>, <span class="st">"mf_per_region"</span>, <span class="st">"pf_6spectra"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#save a list of data frames to an Excel file</span></span>
<span><span class="fu"><a href="../reference/write_excel.html">write_excel</a></span><span class="op">(</span><span class="va">list</span>, workbook_name <span class="op">=</span> <span class="st">"example_mf_data_list"</span><span class="op">)</span></span></code></pre></div>
<p><em>Example 9.2. Export model results</em></p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_excel.html">write_excel</a></span><span class="op">(</span></span>
<span>  <span class="va">model_by_dose</span>,</span>
<span>  workbook_name <span class="op">=</span> <span class="st">"Example_model"</span>,</span>
<span>  model_results <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Mutation data can be written to a VCF file for downstream
applications with <code><a href="../reference/write_vcf_from_mut.html">write_vcf_from_mut()</a></code>.</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/write_vcf_from_mut.html">write_vcf_from_mut</a></span><span class="op">(</span><span class="va">example_data</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="summary-report">Summary Report<a class="anchor" aria-label="anchor" href="#summary-report"></a>
</h2>
<p>To facilitate reproducible analysis of error-corrected sequencing
data, the MutSeqR package includes a pre-built R Markdown workflow:
Summary_Report.Rmd. This report template allows users to efficiently
evaluate the effects of an experimental variable (e.g., dose, age,
tissue) on mutation frequency and mutational spectra.</p>
<p>The workflow guides users through a standardized analysis pipeline
that includes:</p>
<ul>
<li>Mutation data Import</li>
<li>Mutation data Filtering</li>
<li>Calculation of Global Mutation Frequencies per Sample</li>
<li>Visualization of mutation frequencies per Sample and Mean mutation
frequencies per experimental group.</li>
<li>Generalized Linear Modeling of Mutation Frequency (min) as an effect
of the experimental variable.</li>
<li>Optional Benchmark Dose Modeling of mutation frequency across a dose
or concentration variable.</li>
<li>Calculation of Mutation Subtype Frequencies and Proportions at the
base-6 and base-96 resolutions.</li>
<li>Comparison of base-6 mutation spectra between experimental
groups.</li>
<li>Visualization of base-6 subtype proportions per sample with
hierarchical clustering</li>
<li>Visualization of base-6 subtype proportions per experimental
group.</li>
<li>Visualization of base-96 subtype proportions per experimental
group.</li>
<li>Mutation signature analysis per experimental group.</li>
<li>Generalized Linear Mixed Modeling of mutation frequency as an effect
of specified genomic regions and the experimental group.</li>
<li>Visualization of multiplet mutations per experimental group.</li>
</ul>
<p>All figures will be included in the report. Data frames included in
the report will also be exported to excel workbooks within a predefined
output path.</p>
<div class="section level3">
<h3 id="general-usage-2">General Usage<a class="anchor" aria-label="anchor" href="#general-usage-2"></a>
</h3>
<p>Users must download the config file “summary_config.yaml”. This file
defines the parameters to be passed to the Summary_Report.rmd. Users
must fill out the .yaml file and save it.</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">config</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"inputs"</span>, <span class="st">"summary_config.yaml"</span>, package <span class="op">=</span> <span class="st">"MutSeqR"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="va">config</span>, to <span class="op">=</span> <span class="st">"path/to/summary_config.yaml"</span><span class="op">)</span></span></code></pre></div>
<p>Users can then render the report using MutSeqR’s render_report().
Provide the function with the .yaml file using the
<code>config_file_path</code> parameter. Provide the name and the file
path of the output file to the <code>output_file</code> parameter. Set
the <code>output_format</code> to one of “html_document” (recommended),
“pdf_document”, or “all”.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/render_report.html">render_report</a></span><span class="op">(</span>config_file_path <span class="op">=</span> <span class="st">"path/to/summary_config.yaml"</span>,</span>
<span>              output_file <span class="op">=</span> <span class="st">"path/to/output_file.html"</span>,</span>
<span>              output_format <span class="op">=</span> <span class="st">"html_document"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="parameters">Parameters<a class="anchor" aria-label="anchor" href="#parameters"></a>
</h3>
<p>Below is a description of the parameters set in summary_config.yaml.
Parameters are catagerised by their function.</p>
<p>System Set Up</p>
<ul>
<li>
<em>projectdir</em>: The working directory for the project. All
filepaths will be imported from the projectdir. If NULL, the current
working directory will be used.</li>
<li>
<em>outputdir</em>: The output directory where output files will be
saved. If NULL, the projectdir will be used.</li>
</ul>
<p>Project Details</p>
<ul>
<li>
<em>project_title</em>: The project title.</li>
<li>
<em>research_name</em>: The name of the principle investigator for
the project.</li>
<li>
<em>user_name</em>: The name of the one running the analysis.</li>
<li>
<em>project_description</em>: Optional description of the
project.</li>
</ul>
<p>Workflow Profile</p>
<ul>
<li>
<em>config_profile</em> : Choose a workflow profile to load
pre-defined parameters. Current options are “Duplex Sequencing Mouse
Mutagenesis Panel”, “Duplex Sequencing Human Mutagenesis Panel”, “Duplex
Sequencing Rat Mutagenesis Panel”, and “None”. Pre-defined parameters
are set to handle data import and filtering. More profiles coming soon.
If not using a pre-defined profile (“None”), users must fill out the
Custom_Profile_Params (see below).</li>
</ul>
<p>Data Import</p>
<ul>
<li>
<em>species</em>: The species of the model organism. This can be the
common name or the scientific name. Ex. “Mouse” or “Mus musculus”.</li>
<li>
<em>genome</em>: The reference genome. This is used to populate the
sequence context column. Ex. “mm10”.</li>
<li>
<em>file_type</em>: The format of the mutation data to be imported.
Options are table or vcf.</li>
<li>
<em>mutation_file</em>: The file path to the mutation data file(s).
This can be an individual file or a directory containing multiple files.
The file path will be read as projectdir/mutation_file.</li>
<li>
<em>mut_sep</em>: The delimiter for importing tabular mutation data.
Ex. “ is tab-delimited.</li>
<li>
<em>sample_data</em>: An optional file path to the sample metadata
that will be joined with the mutation data. Set to NULL if not importing
metadata. The file path will be read as projectdir/sample_data.</li>
<li>
<em>sd_sep</em>: The delimiter for importing the sample_data file.
Ex. “ is tab-delimited.</li>
</ul>
<p>Calculating MF</p>
<p>Precalculated-depth files: The Summary_Report will automatically
check during import whether the mutation data contains a depth column
(total_depth or depth). If a depth column does not exist, per-sample
precalculated depth can be provided for any subtype resolution to
calculate mutation frequencies. If an exp_variable is provided, the
Summary_Report will automatically sum the per-sample depths to obtain
the depth per experimental group as needed.</p>
<p><em>In order for MutSeqR to calculate the depth from the mutation
data, the</em> <em>data must have the depth-value for <strong>every
sequenced site</strong>. It is not</em> <em>sufficient to calculate the
depth from a list of variants. No-variant sites</em> <em>must be
included. The Summary Report will check for a total_depth column, but it
will</em> <em>not check whether no_variant sites are included. If it
finds a total_depth</em> <em>column, it will proceed to calculate the
depth from the mutation data</em>.</p>
<ul>
<li>
<em>precalc_depth_data_global</em>: Optional file path to the
precalculated per-sample total_depth data. This is the total number of
bases sequenced per sample, used for calculating mutation frequencies.
Columns are “sample” and “group_depth”. If using an exp_variable (see
below), please also include it in this table. The file path will be read
as projectdir/precalc_depth_data_global.</li>
<li>
<em>precalc_depth_data_base6</em>: Optional file path to the
precalculated per-sample total_depth data in the base_6 context. This is
the total number of C and T bases sequenced for each sample. Columns are
“sample”, “normalized_ref”, and “subtype_depth”. If using an
exp_variable (see below), please also include it in this table. The file
path will be read as projectdir/precalc_depth_data_base6.</li>
<li>
<em>precalc_depth_data_base96</em>: Optional file path to the
precalculated per-sample total_depth data in the base_96 context. This
is the total number bases sequenced per sample for each of the 32
possible trinucleotide contexts in their pyrimidine notation. Columns
are “sample”, “normalized_context”, and “subtype_depth”. If using an
exp_variable, please also include it in this table. The file path will
be read as projectdir/precalc_depth_data_base96.</li>
<li>
<em>precalc_depth_data_rg</em>: Optional file path to the
precalculated per-sample total_depth data for each target region. This
is the total number of bases sequenced per sample for each region.
Columns are “sample”, ‘region_col’, “group_depth”. Only applicable if
performing regions analysis. The file path will be read as
projectdir/precalc_depth_data_rg.</li>
<li>
<em>d_sep</em>: The delimiter for importing the precalc_depth_data
files.</li>
</ul>
<p>Statistical Analysis</p>
<p>The Summary Report will run several analyses to investigate the
effect of a specified experimental variable on mutation frequency and
spectra.</p>
<ul>
<li>
<em>exp_variable</em>: Optional. The experimental variable of
interest. This argument should refer to a column in your mutation data
or sample_data. The Summary Report does not support the analysis of
multiple exp_variables; supply only one column name. Ex. “dose”. If
NULL, statistical analyses will be skipped.</li>
<li>
<em>exp_variable_order</em>: A vector specifying the unique levels
of the exp_variable in the order desired for plotting.</li>
<li>
<em>reference_level</em>: The reference level of the exp_variable.
Ex. the vehicle control group for a chemical dose. This value must match
one of the levels within the exp_variable column.</li>
<li>
<em>contrasts</em>: An optional file path to the contrasts table
specifying pairwise comparisons between levels of the experimental
variable. Required 2 columns, no header. This contrasts table will be
used for Generalized Linear Modeling of MFmin by the exp_variable
(model_mf), Generalized Linear Mixed Modeling of MFmin by the
exp_variable and target regions, if using a targeted panel (model_mf),
and comparing the base_6 mutation spectra between exp_variable levels
(spectra_comparison). The file path will be read as
projectdir/contrasts. If NULL, spectra_comparison will be skipped, and
the GLM/GLMM will return only the model estimates per exp_variable.</li>
<li>
<em>cont_sep</em>: the delimiter for importing the contrasts
file.</li>
<li>
<em>bmd</em>: A logical variable indicating whether to run a BMD
analysis on the MF data. If TRUE, the exp_variable must refer to numeric
dose or concentration values.</li>
<li>
<em>bmr</em>: A numeric value indicating the benchmark response for
the bmd analysis. The Summary Report defines the bmr as a bmr-% increase
in MF from the reference level. Default is 0.5.</li>
<li>
<em>bmd_method</em>: The method for running the BMD analysis.
Optionsare “proast” or “toxicr”.</li>
<li>
<em>run_sigfitting</em>: A logical variable indicating whether to
run signature_fitting. If an exp_variable is provided,the signature
analysis will be performed per exp_variable level, otherwise it will be
performed per sample. If TRUE, this will create a virtual environment to
run python on first use.</li>
<li>
<em>python_version</em>: The python version installed on the
operating system.</li>
</ul>
<p>Custom Profile Parameters</p>
<p>Users who do not wish to use one of the pre-defined parameter
profiles may still run the Summary Report by filling out the
Custom_Profile_Params. If using one of the config_profiles, users should
skip this section.</p>
<ul>
<li>
<em>ecs_technology</em>: The technology used to generate the
data.</li>
<li>
<em>is_0_based_mut</em>: A logical variable indicating whether the
genomic coordinates of the tabular mutation data are 0-based (TRUE), or
1-based (FALSE).</li>
<li>
<em>regions</em>: An optional file path to the regions metadata
file. The file must contain columns “contig”, “start”, “end” in addition
to the metadata.</li>
<li>
<em>is_0_based_rg</em>: A logical variable indicating whether the
genomic coordinates of the regions file are 0-based (TRUE), or 1-based
(FALSE).</li>
<li>
<em>rg_sep</em>: The delimiter for importing the regions file.</li>
<li>
<em>vaf_cutoff</em>: A numeric value from 0-1. Will flag variants
with a VAF &gt; vaf_cutoff as germline variants and filter them from
mutation counts.</li>
<li>
<em>snv_in_germ_mnv</em>: A logical variable indicating whether
filter_mut should flag SNVs that overlap with germline MNVs for
filtering.</li>
<li>
<em>rm_abnormal_vaf</em>: A logical variable indicaring whether
filter_mut should remove records where the VAF is between 0.05-0.45 or
between 0.55-0.95.</li>
<li>
<em>custom_filter_col</em>: A column name used by filter_mut to
apply a custom filter to the specified column.</li>
<li>
<em>custom_filter_val</em>: The value in custom_filter_col by which
to filter.</li>
<li>
<em>custom_filter_rm</em>: A logical variable indicating whether
records that contain the custom_filter_val within the custom_filter_col
should be removed (TRUE) or flagged (FALSE).</li>
<li>
<em>filtering_regions</em>: Optional file path to regions file by
which to filter variants. Must contain the contig, start, and end of
each region. The file path will be read as
projectdir/filtering_regions.</li>
<li>
<em>filtering_rg_sep</em>: The delimiter for importing the
filtering_regions file.</li>
<li>
<em>regions_filter</em>: “keep_within” will remove any records that
fall outside of the filtering_regions. “remove_within” will remove
records that fall inside of the filtering_regions.</li>
<li>
<em>allow_half_overlap</em>: A logical variable indicating whether
to include records that half overlap with the regions. If FALSE, the
start and end position of the record must fall within the region
interval to be counted as “falling in the region”. If TRUE, records that
start/end within the region interval, but extend outside of it will be
counted as “falling inside the region”.</li>
<li>
<em>filtering_is_0_based_rg</em>: A logical variable indicating
whether the genomic coordinates of the filtering_regions are 0-based
(TRUE) or 1-based (FALSE).</li>
<li>
<em>rm_filtered_mut_from_depth</em>: A logical variable indicating
whether the alt_depth of variants flagged during the filtering process
should be removed from their total_depth values. This does not apply to
records flagged as germline variants.</li>
<li>
<em>do_regions_analysis</em>: A logical variable indicating whether
to perform Generalized Linear Mixed Modeling of the data by sequencing
target. This is applicable to data sets that used a targeted panel or
has specific regions of interest.</li>
<li>
<em>region_col</em>: The column name that uniquely identifies each
target region for the regions_analysis. This column must be present in
the mutation_data or in the regions metadata file.</li>
</ul>
</div>
<div class="section level3">
<h3 id="output-4">Output<a class="anchor" aria-label="anchor" href="#output-4"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<div class="section level3">
<h3 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h3>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] BSgenome.Hsapiens.UCSC.hg38_1.4.5  BSgenome.Mmusculus.UCSC.mm10_1.4.3</span></span>
<span><span class="co">##  [3] BSgenome_1.76.0                    rtracklayer_1.68.0                </span></span>
<span><span class="co">##  [5] BiocIO_1.18.0                      Biostrings_2.76.0                 </span></span>
<span><span class="co">##  [7] XVector_0.48.0                     GenomicRanges_1.60.0              </span></span>
<span><span class="co">##  [9] GenomeInfoDb_1.44.0                IRanges_2.42.0                    </span></span>
<span><span class="co">## [11] S4Vectors_0.46.0                   BiocGenerics_0.54.0               </span></span>
<span><span class="co">## [13] generics_0.1.4                     MutSeqR_0.99.0                    </span></span>
<span><span class="co">## [15] htmltools_0.5.8.1                  bs4Dash_2.3.4                     </span></span>
<span><span class="co">## [17] DT_0.33                            BiocStyle_2.36.0                  </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3          jsonlite_2.0.0             </span></span>
<span><span class="co">##   [3] magrittr_2.0.3              GenomicFeatures_1.60.0     </span></span>
<span><span class="co">##   [5] farver_2.1.2                nloptr_2.2.1               </span></span>
<span><span class="co">##   [7] rmarkdown_2.29              fs_1.6.6                   </span></span>
<span><span class="co">##   [9] ragg_1.4.0                  vctrs_0.6.5                </span></span>
<span><span class="co">##  [11] memoise_2.0.1               minqa_1.2.8                </span></span>
<span><span class="co">##  [13] Rsamtools_2.24.0            RCurl_1.98-1.17            </span></span>
<span><span class="co">##  [15] S4Arrays_1.8.1              curl_6.4.0                 </span></span>
<span><span class="co">##  [17] broom_1.0.8                 SparseArray_1.8.0          </span></span>
<span><span class="co">##  [19] Formula_1.2-5               sass_0.4.10                </span></span>
<span><span class="co">##  [21] bslib_0.9.0                 htmlwidgets_1.6.4          </span></span>
<span><span class="co">##  [23] desc_1.4.3                  cachem_1.1.0               </span></span>
<span><span class="co">##  [25] GenomicAlignments_1.44.0    lifecycle_1.0.4            </span></span>
<span><span class="co">##  [27] pkgconfig_2.0.3             Matrix_1.7-3               </span></span>
<span><span class="co">##  [29] R6_2.6.1                    fastmap_1.2.0              </span></span>
<span><span class="co">##  [31] GenomeInfoDbData_1.2.14     rbibutils_2.3              </span></span>
<span><span class="co">##  [33] MatrixGenerics_1.20.0       digest_0.6.37              </span></span>
<span><span class="co">##  [35] colorspace_2.1-1            patchwork_1.3.0            </span></span>
<span><span class="co">##  [37] AnnotationDbi_1.70.0        rprojroot_2.0.4            </span></span>
<span><span class="co">##  [39] textshaping_1.0.1           crosstalk_1.2.1            </span></span>
<span><span class="co">##  [41] RSQLite_2.4.1               labeling_0.4.3             </span></span>
<span><span class="co">##  [43] httr_1.4.7                  abind_1.4-8                </span></span>
<span><span class="co">##  [45] compiler_4.5.1              microbenchmark_1.5.0       </span></span>
<span><span class="co">##  [47] here_1.0.1                  bit64_4.6.0-1              </span></span>
<span><span class="co">##  [49] withr_3.0.2                 backports_1.5.0            </span></span>
<span><span class="co">##  [51] BiocParallel_1.42.1         carData_3.0-5              </span></span>
<span><span class="co">##  [53] DBI_1.2.3                   MASS_7.3-65                </span></span>
<span><span class="co">##  [55] DelayedArray_0.34.1         rjson_0.2.23               </span></span>
<span><span class="co">##  [57] tools_4.5.1                 glue_1.8.0                 </span></span>
<span><span class="co">##  [59] restfulr_0.0.15             nlme_3.1-168               </span></span>
<span><span class="co">##  [61] grid_4.5.1                  checkmate_2.3.2            </span></span>
<span><span class="co">##  [63] gtable_0.3.6                tidyr_1.3.1                </span></span>
<span><span class="co">##  [65] data.table_1.17.6           doBy_4.6.27                </span></span>
<span><span class="co">##  [67] car_3.1-3                   Deriv_4.1.6                </span></span>
<span><span class="co">##  [69] pillar_1.10.2               stringr_1.5.1              </span></span>
<span><span class="co">##  [71] splines_4.5.1               dplyr_1.1.4                </span></span>
<span><span class="co">##  [73] lattice_0.22-7              bit_4.6.0                  </span></span>
<span><span class="co">##  [75] tidyselect_1.2.1            knitr_1.50                 </span></span>
<span><span class="co">##  [77] reformulas_0.4.1            bookdown_0.43              </span></span>
<span><span class="co">##  [79] SummarizedExperiment_1.38.1 xfun_0.52                  </span></span>
<span><span class="co">##  [81] Biobase_2.68.0              matrixStats_1.5.0          </span></span>
<span><span class="co">##  [83] stringi_1.8.7               UCSC.utils_1.4.0           </span></span>
<span><span class="co">##  [85] yaml_2.3.10                 boot_1.3-31                </span></span>
<span><span class="co">##  [87] evaluate_1.0.4              codetools_0.2-20           </span></span>
<span><span class="co">##  [89] tibble_3.3.0                BiocManager_1.30.26        </span></span>
<span><span class="co">##  [91] packcircles_0.3.7           cli_3.6.5                  </span></span>
<span><span class="co">##  [93] systemfonts_1.2.3           Rdpack_2.6.4               </span></span>
<span><span class="co">##  [95] jquerylib_0.1.4             dichromat_2.0-0.1          </span></span>
<span><span class="co">##  [97] modelr_0.1.11               Rcpp_1.0.14                </span></span>
<span><span class="co">##  [99] png_0.1-8                   XML_3.99-0.18              </span></span>
<span><span class="co">## [101] parallel_4.5.1              pkgdown_2.1.3              </span></span>
<span><span class="co">## [103] ggh4x_0.3.1                 ggplot2_3.5.2              </span></span>
<span><span class="co">## [105] blob_1.2.4                  dendsort_0.3.4             </span></span>
<span><span class="co">## [107] plyranges_1.28.0            bitops_1.0-9               </span></span>
<span><span class="co">## [109] lme4_1.1-37                 viridisLite_0.4.2          </span></span>
<span><span class="co">## [111] VariantAnnotation_1.54.1    scales_1.4.0               </span></span>
<span><span class="co">## [113] purrr_1.0.4                 crayon_1.5.3               </span></span>
<span><span class="co">## [115] rlang_1.1.6                 cowplot_1.1.3              </span></span>
<span><span class="co">## [117] KEGGREST_1.48.1</span></span></code></pre>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-bergstrom-2019" class="csl-entry">
Bergstrom, Erik N., Mi Ni Huang, Uma Mahto, Mark Barnes, Michael R.
Stratton, Steven G. Rozen, and Ludmil B. Alexandrov. 2019.
<span>“<span>SigProfilerMatrixGenerator</span>: A Tool for Visualizing
and Exploring Patterns of Small Mutational Events.”</span> <em>BMC
Genomics</em> 20 (1): 685. <a href="https://doi.org/10.1186/s12864-019-6041-2" class="external-link">https://doi.org/10.1186/s12864-019-6041-2</a>.
</div>
<div id="ref-brooks-2011" class="csl-entry">
Brooks, Steve, Andrew Gelman, Galin Jones, and Xiao-Li Meng. 2011.
<em>Handbook of <span>Markov</span> <span>Chain</span>
<span>Monte</span> <span>Carlo</span></em>. CRC Press.
</div>
<div id="ref-efsa-2022" class="csl-entry">
Committee, EFSA Scientific, Simon John More, Vasileios Bampidis, Diane
Benford, Claude Bragard, Thorhallur Ingi Halldorsson, Antonio F
Hernández-Jerez, et al. 2022. <span>“Guidance on the Use of the
Benchmark Dose Approach in Risk Assessment.”</span> <em>EFSA
Journal</em> 20 (10): e07584. <a href="https://doi.org/10.2903/j.efsa.2022.7584" class="external-link">https://doi.org/10.2903/j.efsa.2022.7584</a>.
</div>
<div id="ref-danecek-2011" class="csl-entry">
Danecek, Petr, Adam Auton, Goncalo Abecasis, Cornelis A. Albers, Eric
Banks, Mark A. DePristo, Robert E. Handsaker, et al. 2011. <span>“The
Variant Call Format and <span>VCFtools</span>.”</span>
<em>Bioinformatics</em> 27 (15): 2156–58. <a href="https://doi.org/10.1093/bioinformatics/btr330" class="external-link">https://doi.org/10.1093/bioinformatics/btr330</a>.
</div>
<div id="ref-diaz-gay-2023" class="csl-entry">
Díaz-Gay, Marcos, Raviteja Vangara, Mark Barnes, Xi Wang, S M Ashiqul
Islam, Ian Vermes, Stephen Duke, et al. 2023. <span>“Assigning
Mutational Signatures to Individual Samples and Individual Somatic
Mutations with <span>SigProfilerAssignment</span>.”</span> Edited by
Christina Kendziorski. <em>Bioinformatics</em> 39 (12): btad756. <a href="https://doi.org/10.1093/bioinformatics/btad756" class="external-link">https://doi.org/10.1093/bioinformatics/btad756</a>.
</div>
<div id="ref-dodge-2023" class="csl-entry">
Dodge, Annette E., Danielle P. M. LeBlanc, Gu Zhou, Andrew Williams,
Matthew J. Meier, Phu Van, Fang Yin Lo, et al. 2023. <span>“Duplex
Sequencing Provides Detailed Characterization of Mutation Frequencies
and Spectra in the Bone Marrow of <span>MutaMouse</span> Males Exposed
to Procarbazine Hydrochloride.”</span> <em>Archives of Toxicology</em>
97 (8): 2245–59. <a href="https://doi.org/10.1007/s00204-023-03527-y" class="external-link">https://doi.org/10.1007/s00204-023-03527-y</a>.
</div>
<div id="ref-gelman-1995" class="csl-entry">
Gelman, Andrew, John B. Carlin, Hal S. Stern, and Donald B. Rubin. 1995.
<em>Bayesian <span>Data</span> <span>Analysis</span></em>. New York:
Chapman; Hall/CRC. <a href="https://doi.org/10.1201/9780429258411" class="external-link">https://doi.org/10.1201/9780429258411</a>.
</div>
<div id="ref-doBy" class="csl-entry">
Halekoh, Ulrich, and Søren Højsgaard. 2025. <em>doBy: Groupwise
Statistics, LSmeans, Linear Estimates, Utilities</em>. <a href="https://CRAN.R-project.org/package=doBy" class="external-link">https://CRAN.R-project.org/package=doBy</a>.
</div>
<div id="ref-kennedy-2014" class="csl-entry">
Kennedy, Scott R., Michael W. Schmitt, Edward J. Fox, Brendan F. Kohrn,
Jesse J. Salk, Eun Hyun Ahn, Marc J. Prindle, et al. 2014.
<span>“Detecting Ultralow-Frequency Mutations by <span>Duplex</span>
<span>Sequencing</span>.”</span> <em>Nature Protocols</em> 9 (11):
2586–606. <a href="https://doi.org/10.1038/nprot.2014.170" class="external-link">https://doi.org/10.1038/nprot.2014.170</a>.
</div>
<div id="ref-khandekar-2023" class="csl-entry">
Khandekar, Azhar, Raviteja Vangara, Mark Barnes, Marcos Díaz-Gay, Ammal
Abbasi, Erik N. Bergstrom, Christopher D. Steele, Nischalan Pillay, and
Ludmil B. Alexandrov. 2023. <span>“Visualizing and Exploring Patterns of
Large Mutational Events with
<span>SigProfilerMatrixGenerator</span>.”</span> <em>BMC Genomics</em>
24 (1): 469. <a href="https://doi.org/10.1186/s12864-023-09584-y" class="external-link">https://doi.org/10.1186/s12864-023-09584-y</a>.
</div>
<div id="ref-leblanc-2022" class="csl-entry">
LeBlanc, Danielle P. M., Matthew Meier, Fang Yin Lo, Elizabeth Schmidt,
Charles Valentine, Andrew Williams, Jesse J. Salk, Carole L. Yauk, and
Francesco Marchetti. 2022. <span>“Duplex Sequencing Identifies Genomic
Features That Determine Susceptibility to Benzo(a)pyrene-Induced in Vivo
Mutations.”</span> <em>BMC Genomics</em> 23 (1): 542. <a href="https://doi.org/10.1186/s12864-022-08752-w" class="external-link">https://doi.org/10.1186/s12864-022-08752-w</a>.
</div>
<div id="ref-marchetti-2023b" class="csl-entry">
Marchetti, Francesco, Renato Cardoso, Connie L. Chen, George R. Douglas,
Joanne Elloway, Patricia A. Escobar, Tod Harper Jr, et al. 2023.
<span>“Error-Corrected Next-Generation Sequencing to Advance Nonclinical
Genotoxicity and Carcinogenicity Testing.”</span> <em>Nature Reviews
Drug Discovery</em> 22 (3): 165–66. <a href="https://doi.org/10.1038/d41573-023-00014-y" class="external-link">https://doi.org/10.1038/d41573-023-00014-y</a>.
</div>
<div id="ref-marchetti-2023a" class="csl-entry">
Marchetti, Francesco, Renato Cardoso, Connie L. Chen, George R. Douglas,
Joanne Elloway, Patricia A. Escobar, Tod Harper, et al. 2023.
<span>“Error-Corrected Next Generation Sequencing -
<span>Promises</span> and Challenges for Genotoxicity and Cancer Risk
Assessment.”</span> <em>Mutation Research. Reviews in Mutation
Research</em> 792: 108466. <a href="https://doi.org/10.1016/j.mrrev.2023.108466" class="external-link">https://doi.org/10.1016/j.mrrev.2023.108466</a>.
</div>
<div id="ref-menon-2023" class="csl-entry">
Menon, Vijay, and Douglas E. Brash. 2023. <span>“Next-Generation
Sequencing Methodologies to Detect Low-Frequency Mutations:
<span>‘<span>Catch</span> Me If You Can’</span>.”</span> <em>Mutation
Research/Reviews in Mutation Research</em> 792 (July): 108471. <a href="https://doi.org/10.1016/j.mrrev.2023.108471" class="external-link">https://doi.org/10.1016/j.mrrev.2023.108471</a>.
</div>
<div id="ref-piegorsch-1994" class="csl-entry">
Piegorsch, W W, and A J Bailer. 1994. <span>“Statistical Approaches for
Analyzing Mutational Spectra: Some Recommendations for Categorical
Data.”</span> <em>Genetics</em> 136 (1): 403–16. <a href="https://doi.org/10.1093/genetics/136.1.403" class="external-link">https://doi.org/10.1093/genetics/136.1.403</a>.
</div>
<div id="ref-wheeler-2020" class="csl-entry">
Wheeler, Matthew W., Todd Blessinger, Kan Shao, Bruce C. Allen, Louis
Olszyk, J. Allen Davis, and Jeffrey S. Gift. 2020. <span>“Quantitative
<span>Risk</span> <span>Assessment</span>: <span>Developing</span> a
<span>Bayesian</span> <span>Approach</span> to <span>Dichotomous</span>
<span>Dose</span>-<span>Response</span>
<span>Uncertainty</span>.”</span> <em>Risk Analysis: An Official
Publication of the Society for Risk Analysis</em> 40 (9): 1706–22. <a href="https://doi.org/10.1111/risa.13537" class="external-link">https://doi.org/10.1111/risa.13537</a>.
</div>
<div id="ref-wheeler-2022" class="csl-entry">
Wheeler, Matthew W., Jose Cortinas, Marc Aerts, Jeffery S. Gift, and J.
Allen Davis. 2022. <span>“Continuous <span>Model</span>
<span>Averaging</span> for <span>Benchmark</span> <span>Dose</span>
<span>Analysis</span>: <span>Averaging</span> <span>Over</span>
<span>Distributional</span> <span>Forms</span>.”</span>
<em>Environmetrics</em> 33 (5): e2728. <a href="https://doi.org/10.1002/env.2728" class="external-link">https://doi.org/10.1002/env.2728</a>.
</div>
<div id="ref-white-2020" class="csl-entry">
White, Paul A., Alexandra S. Long, and George E. Johnson. 2020.
<span>“Quantitative <span>Interpretation</span> of <span>Genetic</span>
<span>Toxicity</span> <span>Dose</span>‐<span>Response</span>
<span>Data</span> for <span>Risk</span> <span>Assessment</span> and
<span>Regulatory</span> <span>Decision</span>‐<span>Making</span>:
<span>Current</span> <span>Status</span> and <span>Emerging</span>
<span>Priorities</span>.”</span> <em>Environmental and Molecular
Mutagenesis</em> 61 (1): 66–83. <a href="https://doi.org/10.1002/em.22351" class="external-link">https://doi.org/10.1002/em.22351</a>.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Annette E. Dodge, Andrew Williams, Danielle P.M. LeBlanc, David M. Schuster, Elena Esina, Clint C. Valentine, Jesse J. Salk, Alexander Y. Maslov, Christopher Bradley, Carole L. Yauk, Francesco Marchetti, Matthew J. Meier.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
